{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Simulations using NFBC1966 results\n",
    "\n",
    "Using chromosome 1 of the NFBC (Stampeed) data ($n=5340$ samples and $p = 24523$ SNPs), let us simulate:\n",
    "\n",
    "## Polygenic model\n",
    "\n",
    "For $r$ traits, our simulation model is:\n",
    "\n",
    "$$\\mathbf{Y}_{r \\times n} \\sim \\text{MatrixNormal}(\\mathbf{B}_{r \\times p}\\mathbf{X}_{p \\times n}, \\ \\ \\Sigma_{r \\times r} , \\ \\ \\sigma_g^2\\Phi_{n \\times n} + \\sigma_e^2 \\mathbf{I}_{n \\times n})$$\n",
    "\n",
    "+ $\\mathbf{X}_{p \\times n}$ contains *all* predictors (genetic + non-genetic)\n",
    "+ $\\mathbf{B}_{r \\times p}$ contains (true) regression coefficients. $k$ entries are non-zero\n",
    "+ $\\Phi$ is the GRM estimated from genotypes\n",
    "+ $\\Sigma_{r \\times r}$ is the symmetric PD row (trait) covariance matrix where max condition number is set to 10\n",
    "+ $\\sigma_g^2\\Phi_{n \\times n} + \\sigma_e^2 \\mathbf{I}_{n \\times n}$ is the column (sample) covariance matrix\n",
    "+ $\\sigma_g^2 = 0.1$ and $\\sigma_e^2 = 0.9$ (thus heritability is 10%)\n",
    "+ $k$ causal SNPs chosen fixed distance with each other across genome\n",
    "+ $\\beta_i:$ every causal SNP affects every phenotype. Causal SNP $i$ have effect size sums up to 0.5 across $r$ traits\n",
    "+ $r = \\{2, 3, 4\\}$  (number of traits)\n",
    "+ $n=5340$ samples\n",
    "+ $p = 24523$ SNPs\n",
    "+ q = 5 fold cross validation\n",
    "+ iterates ≥5 times\n",
    "+ init_beta=true\n",
    "+ debias=false\n",
    "\n",
    "## Sparse model\n",
    "\n",
    "With $r$ traits, each sample's phenotype $\\mathbf{y}_{i} \\in \\mathbb{R}^{n \\times 1}$ is simulated under\n",
    "\n",
    "$$\\mathbf{y}_{i}^{r \\times 1} \\sim N(\\mathbf{B}^{r \\times p}\\mathbf{x}_{i}^{p \\times 1}, \\ \\ \\Sigma_{r \\times r})$$\n",
    "\n",
    "This model assumes each sample is independent.\n",
    "\n",
    "+ $\\mathbf{X}_{p \\times n}$ contains *all* predictors (genetic + non-genetic)\n",
    "+ $\\mathbf{B}_{r \\times p}$ contains (true) regression coefficients. $k$ entries are non-zero\n",
    "+ $B_{ij} \\in \\{0.05, 0.1, ..., 0.5\\}$ for $k$ positions\n",
    "+ $k = \\{10, 20, 30\\}$ (number of causal SNPs)\n",
    "+ $\\Sigma_{r \\times r}$ is the symmetric PD row (trait) covariance matrix where max condition number is set to 10\n",
    "+ $k = \\{10, 20, 30\\}$ (number of causal SNPs)\n",
    "+ $r = \\{2, 3, 4\\}$  (number of traits)\n",
    "+ $n=5340$ samples\n",
    "+ $p = 24523$ SNPs\n",
    "+ q = 5 fold cross validation\n",
    "+ iterates ≥5 times\n",
    "+ init_beta=true\n",
    "+ debias=false"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "# Summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "set 1 summary (successful run = 100):\n",
      "\\texttt{mIHT} & $164.6 \\pm 69.3$ & $0.92 \\pm 0.16$ & $0.76 \\pm 0.2$ & $3.7 \\pm 6.4$\\\\\n",
      "\\texttt{uIHT} & $114.9 \\pm 48.6$ & $0.93 \\pm 0.16$ & $0.72 \\pm 0.2$ & $1.4 \\pm 3.7$\\\\\n",
      "\\texttt{CCA} & $152.6 \\pm 57.3$ & $0.96 \\pm 0.14$ & $0.78 \\pm 0.2$ & $77.8 \\pm 40.6$\\\\\n",
      "\\texttt{mvLMM} & $307.7 \\pm 121.4$ & $0.95 \\pm 0.15$ & $0.76 \\pm 0.2$ & $42.8 \\pm 18.5$\\\\ \n",
      "\n",
      "set 2 summary (successful run = 99):\n",
      "\\texttt{mIHT} & $214.4 \\pm 100.1$ & $0.91 \\pm 0.12$ & $0.75 \\pm 0.14$ & $5.7 \\pm 6.0$\\\\\n",
      "\\texttt{uIHT} & $169.6 \\pm 81.9$ & $0.86 \\pm 0.16$ & $0.72 \\pm 0.16$ & $2.4 \\pm 2.5$\\\\\n",
      "\\texttt{CCA} & $226.8 \\pm 101.9$ & $0.95 \\pm 0.09$ & $0.79 \\pm 0.15$ & $125.3 \\pm 55.3$\\\\\n",
      "\\texttt{mvLMM} & $449.9 \\pm 221.7$ & $0.93 \\pm 0.1$ & $0.75 \\pm 0.16$ & $66.1 \\pm 22.8$\\\\ \n",
      "\n",
      "set 3 summary (successful run = 86):\n",
      "\\texttt{mIHT} & $227.9 \\pm 41.1$ & $0.93 \\pm 0.09$ & $0.73 \\pm 0.12$ & $5.9 \\pm 4.6$\\\\\n",
      "\\texttt{uIHT} & $213.8 \\pm 45.7$ & $0.9 \\pm 0.11$ & $0.69 \\pm 0.12$ & $3.2 \\pm 3.8$\\\\\n",
      "\\texttt{CCA} & $371.5 \\pm 34.0$ & $0.96 \\pm 0.07$ & $0.75 \\pm 0.11$ & $173.2 \\pm 54.9$\\\\\n",
      "\\texttt{mvLMM} & $1135.3 \\pm 125.5$ & $0.94 \\pm 0.09$ & $0.71 \\pm 0.11$ & $93.6 \\pm 22.7$\\\\ \n",
      "\n",
      "set 4 summary (successful run = 36):\n",
      "\\texttt{mIHT} & $278.8 \\pm 53.0$ & $0.97 \\pm 0.09$ & $0.74 \\pm 0.2$ & $2.2 \\pm 2.0$\\\\\n",
      "\\texttt{uIHT} & $245.8 \\pm 34.8$ & $0.96 \\pm 0.11$ & $0.7 \\pm 0.23$ & $3.1 \\pm 6.4$\\\\\n",
      "\\texttt{CCA} & $985.1 \\pm 97.5$ & $0.99 \\pm 0.06$ & $0.78 \\pm 0.2$ & $64.6 \\pm 30.5$\\\\\n",
      "\\texttt{mvLMM} & $8067.4 \\pm 3900.8$ & $0.99 \\pm 0.06$ & $0.74 \\pm 0.18$ & $41.8 \\pm 16.4$\\\\ \n",
      "\n",
      "set 5 summary (successful run = 0):\n",
      "\\texttt{mIHT} & $1892.2 \\pm 419.0$ & $0.93 \\pm 0.12$ & $0.75 \\pm 0.14$ & $2.9 \\pm 2.5$\\\\\n",
      "\\texttt{uIHT} & $1336.9 \\pm 310.2$ & $0.92 \\pm 0.11$ & $0.72 \\pm 0.12$ & $7.6 \\pm 5.9$\\\\\n",
      "\\texttt{CCA} & $NaN \\pm NaN$ & $NaN \\pm NaN$ & $NaN \\pm NaN$ & $NaN \\pm NaN$\\\\\n",
      "\\texttt{mvLMM} & $NaN \\pm NaN$ & $NaN \\pm NaN$ & $NaN \\pm NaN$ & $NaN \\pm NaN$\\\\ \n",
      "\n",
      "set 6 summary (successful run = 0):\n",
      "\\texttt{mIHT} & $3699.3 \\pm 410.4$ & $0.91 \\pm 0.11$ & $0.71 \\pm 0.11$ & $2.8 \\pm 2.1$\\\\\n",
      "\\texttt{uIHT} & $2353.8 \\pm 212.3$ & $0.92 \\pm 0.11$ & $0.7 \\pm 0.1$ & $10.7 \\pm 4.3$\\\\\n",
      "\\texttt{CCA} & $NaN \\pm NaN$ & $NaN \\pm NaN$ & $NaN \\pm NaN$ & $NaN \\pm NaN$\\\\\n",
      "\\texttt{mvLMM} & $NaN \\pm NaN$ & $NaN \\pm NaN$ & $NaN \\pm NaN$ & $NaN \\pm NaN$\\\\ \n",
      "\n"
     ]
    }
   ],
   "source": [
    "using Statistics\n",
    "\n",
    "\"\"\"\n",
    "Find the `nsnps` most significant SNPs for mvPLINK in simulation `sim`.\n",
    "\"\"\"\n",
    "function get_top_mvPLINK_SNPs(set::Int, sim::Int, nsnps::Int)\n",
    "    n, p = 5340, 24523\n",
    "    dir = \"set$set/sim$sim/\"\n",
    "    \n",
    "    # read mvPLINK result\n",
    "    mvplink_df = CSV.read(dir * \"plink.mqfam.total\", DataFrame, delim=' ', ignorerepeated=true)\n",
    "\n",
    "    # get pvalues, possibly accounting for \"NA\"s\n",
    "    if eltype(mvplink_df[!, :P]) == Float64\n",
    "        pval = mvplink_df[!, :P]\n",
    "    else\n",
    "        mvplink_df[findall(x -> x == \"NA\", mvplink_df[!, :P]), :P] .= \"1.0\"\n",
    "        pval = parse.(Float64, mvplink_df[!, :P])\n",
    "    end\n",
    "    perm = sortperm(pval)\n",
    "    return perm[1:nsnps]\n",
    "end\n",
    "\n",
    "\"\"\"\n",
    "Find the position of the `nsnps` most significant SNPs for GEMMA in simulation `sim`.\n",
    "\"\"\"\n",
    "function get_top_GEMMA_SNP_ids(set::Int, sim::Int, nsnps::Int)\n",
    "    dir = \"set$set/sim$sim/\"\n",
    "    gemma_df = CSV.read(dir * \"gemma.sim$sim.assoc.txt\", DataFrame)\n",
    "    pval_wald = gemma_df[!, :p_wald]\n",
    "    perm = sortperm(pval_wald)\n",
    "    return perm[1:nsnps]\n",
    "end\n",
    "\n",
    "\"\"\"\n",
    "Find significant SNPs return by IHT in simulation `sim`.\n",
    "\"\"\"\n",
    "function get_IHT_SNPs(set::Int, sim::Int)\n",
    "    dir = \"set$set/sim$sim/\"\n",
    "    iht_β1 = vec(readdlm(dir * \"iht_beta1.txt\"))\n",
    "    iht_β2 = vec(readdlm(dir * \"iht_beta2.txt\"))\n",
    "    detected_snps = findall(!iszero, iht_β1) ∪ findall(!iszero, iht_β2)\n",
    "    return unique(detected_snps)\n",
    "end\n",
    "\n",
    "\"\"\"\n",
    "Get positions for the truly causal SNPs in simulation `sim`. \n",
    "\"\"\"\n",
    "function get_true_SNPs(set::Int, sim::Int)\n",
    "    dir = \"set$set/sim$sim/\"\n",
    "    trueB = readdlm(dir * \"trueb.txt\")\n",
    "    causal_snps = unique([x[1] for x in findall(!iszero, trueB)])\n",
    "    return causal_snps\n",
    "end\n",
    "\n",
    "\"\"\"\n",
    "Get causal SNPs' position in GEMMA's result for simulation `sim`. Note gemma have snp filtering.\n",
    "\"\"\"\n",
    "function get_gemma_causal_snp_pos(set::Int, sim::Int)\n",
    "    nfbc = SnpData(\"NFBC.qc.imputeBy0.chr.1\")\n",
    "    dir = \"set$set/sim$sim/\"\n",
    "    trueB = readdlm(dir * \"trueb.txt\")\n",
    "    causal_snps = unique([x[1] for x in findall(!iszero, trueB)])\n",
    "    causal_snp_rsID = nfbc.snp_info.snpid[causal_snps]\n",
    "    gemma_df = CSV.read(dir * \"gemma.sim$sim.assoc.txt\", DataFrame)\n",
    "    gemma_snps = gemma_df[!, :rs]\n",
    "    causal_snp_idx = convert(Vector{Int}, indexin(causal_snp_rsID, gemma_snps))\n",
    "    \n",
    "    # also need IHT's selected SNPs\n",
    "    iht_snps_rsID = nfbc.snp_info.snpid[get_IHT_SNPs(sim)]\n",
    "    iht_snps_idx = convert(Vector{Int}, indexin(iht_snps_rsID, gemma_snps))\n",
    "    \n",
    "    # also need SNP positions in GEMMA dataframe\n",
    "    gemma2nfbc_idx = convert(Vector{Int}, indexin(gemma_snps, nfbc.snp_info.snpid))\n",
    "    gemma_snp_pos = Vector{Int}(undef, size(gemma_df, 1))\n",
    "    for i in 1:size(gemma_df, 1)\n",
    "        gemma_snp_pos[i] = nfbc.snp_info.position[gemma2nfbc_idx[i]]\n",
    "    end\n",
    "    insertcols!(gemma_df, size(gemma_df, 2) + 1, :pos => gemma_snp_pos)\n",
    "    \n",
    "    return gemma_df, causal_snp_idx, iht_snps_idx\n",
    "end\n",
    "\n",
    "\"\"\"\n",
    "Imports gemma p-values, causal SNPs, and IHT selected SNP, and plot manhattan plot using MendelPlots.jl\n",
    "\"\"\"\n",
    "function plot_gemma_manhattan(sim::Int)\n",
    "    # GEMMA causal SNPs\n",
    "    gemma_df, causal_snps, iht_snps = get_gemma_causal_snp_pos(sim)\n",
    "    rename!(gemma_df, [:p_wald => :pval, :rs => :snpid])\n",
    "    gemma_df[findall(x -> x < 1e-50, gemma_df[!, :pval]), :pval] .= 1e-50\n",
    "    empty_col = [\"\" for i in 1:size(gemma_df, 1)]\n",
    "    insertcols!(gemma_df, size(gemma_df, 2) + 1, :empty_col => empty_col)\n",
    "\n",
    "    manhattan(gemma_df, outfile = \"NFBCsim/manhattan_gemma_sim$sim.png\",\n",
    "        annotateinds = causal_snps, annotateinds2 = iht_snps,\n",
    "        annotatevar=:empty_col, titles=\"GEMMA simulation $sim\")\n",
    "    display(\"image/png\", read(\"NFBCsim/manhattan_gemma_sim$sim.png\"))\n",
    "end\n",
    "\n",
    "\"\"\"\n",
    "Imports mvPLINK p-values, causal SNPs, and IHT selected SNP, and plot manhattan plot using MendelPlots.jl\n",
    "\"\"\"\n",
    "function plot_mvPLINK_manhattan(sim::Int)\n",
    "    # mvPLINK\n",
    "    filename = \"NFBCsim/sim$sim/plink.mqfam.total\"\n",
    "    mvplink_df = CSV.read(filename, DataFrame, delim=' ', ignorerepeated=true)\n",
    "    if eltype(mvplink_df[!, :P]) == Float64\n",
    "        pval = mvplink_df[!, :P]\n",
    "    else\n",
    "        mvplink_df[findall(x -> x == \"NA\", mvplink_df[!, :P]), :P] .= \"1.0\"\n",
    "        pval = parse.(Float64, mvplink_df[!, :P])\n",
    "    end\n",
    "    pval[findall(x -> x < 1e-50, pval)] .= 1e-50 # limit smallest pvalues\n",
    "\n",
    "    # causal SNPs\n",
    "    causal_snps = get_true_SNPs(sim)\n",
    "    \n",
    "    # IHT SNPs\n",
    "    iht_snps = get_IHT_SNPs(sim)\n",
    "\n",
    "    # make dataframe to input into MendelPlots\n",
    "    snpdata = SnpData(\"/Users/biona001/Benjamin_Folder/UCLA/research/stampeed/imputed_with_0/NFBC_imputed_with_0\")\n",
    "    rename!(snpdata.snp_info, [:chr, :snpid, :genetic_distance, :pos, :allele1, :allele2])\n",
    "    insertcols!(snpdata.snp_info, size(snpdata.snp_info, 2) + 1, :pval => pval)\n",
    "    empty_col = [\"\" for i in 1:size(snpdata.snp_info, 1)]\n",
    "    insertcols!(snpdata.snp_info, size(snpdata.snp_info, 2) + 1, :empty_col => empty_col)\n",
    "\n",
    "    # plot\n",
    "    manhattan(snpdata.snp_info, outfile = \"NFBCsim/manhattan_mvPLINK_sim$sim.png\",\n",
    "        annotateinds = causal_snps, annotateinds2 = iht_snps, \n",
    "        annotatevar=:empty_col, titles=\"mvPLINK simulation $sim\")\n",
    "    display(\"image/png\", read(\"NFBCsim/manhattan_mvPLINK_sim$sim.png\"))\n",
    "end\n",
    "\n",
    "\"\"\"\n",
    "Imports mvPLINK p-values and plot QQ plot using MendelPlots.jl\n",
    "\"\"\"\n",
    "function plot_mvPLINK_QQ(sim::Int)\n",
    "    filename = \"NFBCsim/sim$sim/plink.mqfam.total\"\n",
    "    mvplink_df = CSV.read(filename, DataFrame, delim=' ', ignorerepeated=true)\n",
    "    if eltype(mvplink_df[!, :P]) == Float64\n",
    "        pval = mvplink_df[!, :P]\n",
    "    else\n",
    "        mvplink_df[findall(x -> x == \"NA\", mvplink_df[!, :P]), :P] .= \"1.0\"\n",
    "        pval = parse.(Float64, mvplink_df[!, :P])\n",
    "    end\n",
    "    pval[findall(x -> x < 1e-50, pval)] .= 1e-50 # limit smallest pvalues\n",
    "    qq(pval, outfile = \"NFBCsim/QQ_mvPLINK_sim$sim.png\",\n",
    "        ylabel=\"mvPLINK observed -log10(p)\", titles=\"mvPLINK simulation $sim\")\n",
    "    display(\"image/png\", read(\"NFBCsim/QQ_mvPLINK_sim$sim.png\"))\n",
    "end\n",
    "\n",
    "\"\"\"\n",
    "Imports gemma p-values and plot QQ plot using MendelPlots.jl\n",
    "\"\"\"\n",
    "function plot_gemma_QQ(sim::Int)\n",
    "    filename = \"NFBCsim/sim$sim/gemma.sim$sim.assoc.txt\"\n",
    "    gemma_df = CSV.read(filename, DataFrame)\n",
    "    pval_wald = gemma_df[!, :p_wald]\n",
    "    pval_wald[findall(x -> x < 1e-50, pval_wald)] .= 1e-50 # limit smallest pvalues\n",
    "    qq(pval_wald, outfile = \"NFBCsim/QQ_gemma_sim$sim.png\",\n",
    "        ylabel=\"GEMMA observed -log10(p)\", titles=\"GEMMA simulation $sim\")\n",
    "    display(\"image/png\", read(\"NFBCsim/QQ_gemma_sim$sim.png\"))\n",
    "end\n",
    "\n",
    "# \"\"\"\n",
    "# Summarize all simulations for IHT, mvPLINK, GEMMA in computation time, true positives,\n",
    "# false positives, and false positive rates. \n",
    "# \"\"\"\n",
    "# function summarize_repeats()\n",
    "#     model = \"NFBCsim\"\n",
    "#     n, p = 5340, 24523\n",
    "#     sims = 1:50 # k = 10, r = 2, βoverlap=2, polygenic model\n",
    "#     nfbc = SnpData(\"/Users/biona001/Benjamin_Folder/UCLA/research/stampeed/imputed_with_0/NFBC_imputed_with_0\")\n",
    "#     snp_rsID = nfbc.snp_info.snpid\n",
    "\n",
    "#     iht_time, iht_power, iht_FP, iht_FPR = Float64[], Float64[], Float64[], Float64[]\n",
    "#     mvPLINK_time, mvPLINK_power, mvPLINK_FP, mvPLINK_FPR = Float64[], Float64[], Float64[], Float64[]\n",
    "#     gemma_time, gemma_power, gemma_FP, gemma_FPR = Float64[], Float64[], Float64[], Float64[]\n",
    "\n",
    "#     for sim in sims\n",
    "#         dir = \"NFBCsim/sim$sim/\"\n",
    "#         try\n",
    "#             # correct SNPs\n",
    "#             trueB = readdlm(dir * \"trueb.txt\")\n",
    "#             causal_snps = unique([x[1] for x in findall(!iszero, trueB)])\n",
    "#             causal_snps_rsID = snp_rsID[causal_snps]\n",
    "\n",
    "#             # IHT\n",
    "#             iht_β1 = vec(readdlm(dir * \"iht_beta1.txt\"))\n",
    "#             iht_β2 = vec(readdlm(dir * \"iht_beta2.txt\"))\n",
    "#             detected_snps = findall(!iszero, iht_β1) ∪ findall(!iszero, iht_β2)\n",
    "#             ihtpower, ihtFP, ihtFPR = power_and_fpr(p, causal_snps, detected_snps)\n",
    "\n",
    "#             # MVPLINK\n",
    "#             plinkpower, plinkFP, plinkFPR = process_mvPLINK(dir * \"plink.mqfam.total\", causal_snps)\n",
    "\n",
    "#             # GEMMA \n",
    "#             gemmapower, gemmaFP, gemmaFPR = process_gemma_result(dir * \"gemma.sim$sim.assoc.txt\", causal_snps_rsID)\n",
    "            \n",
    "#             push!(iht_power, ihtpower); push!(iht_FP, ihtFP); push!(iht_FPR, ihtFPR); \n",
    "#             push!(mvPLINK_power, plinkpower); push!(mvPLINK_FP, plinkFP); push!(mvPLINK_FPR, plinkFPR); \n",
    "#             push!(gemma_power, gemmapower); push!(gemma_FP, gemmaFP); push!(gemma_FPR, gemmaFPR);\n",
    "#         catch\n",
    "#             println(\"simulation $sim failed!\")\n",
    "#         end\n",
    "#     end\n",
    "\n",
    "#     return iht_time, iht_power, iht_FP, iht_FPR,\n",
    "#         mvPLINK_time, mvPLINK_power, mvPLINK_FP, mvPLINK_FPR,\n",
    "#         gemma_time, gemma_power, gemma_FP, gemma_FPR\n",
    "# end\n",
    "\n",
    "\"\"\"\n",
    "For each simulation set, after performing n simulations using `run_repeats`,\n",
    "this function reads the summary files for each simulation and summarizes the result. \n",
    "\"\"\"\n",
    "function read_summary(; verbose=true)    \n",
    "    mIHT_time, mIHT_plei_power, mIHT_indp_power, mIHT_FP, mIHT_FPR = \n",
    "        Float64[], Float64[], Float64[], Float64[], Float64[]\n",
    "    uIHT_time, uIHT_plei_power, uIHT_indp_power, uIHT_FP, uIHT_FPR = \n",
    "        Float64[], Float64[], Float64[], Float64[], Float64[]\n",
    "    mvPLINK_time, mvPLINK_plei_power, mvPLINK_indp_power, mvPLINK_FP, \n",
    "        mvPLINK_FPR, mvPLINK_λ = Float64[], Float64[], Float64[], Float64[], Float64[], Float64[]\n",
    "    gemma_time, gemma_plei_power, gemma_indp_power, gemma_FP, gemma_FPR, gemma_λ = \n",
    "        Float64[], Float64[], Float64[], Float64[], Float64[], Float64[]\n",
    "\n",
    "    regex = r\"= (\\d+\\.\\d+) seconds, pleiotropic power = (.+), independent power = (\\d+\\.\\d+), FP = (\\d+), FPR = (\\d\\.\\d+e?-?\\d*), λ = (.+)\"\n",
    "\n",
    "\n",
    "    # polygenic beta = sign * Uniform{0.05, …, 0.5}\n",
    "    # max condition number = 10\n",
    "    # q = 5, iterates ≥5 times, init_beta=true, debias=false)\n",
    "    # sim set 1 are for k = 10, r = 2, βoverlap = 3, path = 5:5:50 (then search around best k)\n",
    "    # sim set 2 are for k = 20, r = 2, βoverlap = 5, path = 5:5:50 (then search around best k)\n",
    "    # sim set 3 are for k = 30, r = 2, βoverlap = 7, path = 5:5:50 (then search around best k)\n",
    "    # sim set 4 are for k = 10, r = 3, βoverlap = 2, path = 5:5:50 (then search around best k)\n",
    "    # sim set 5 are for k = 20, r = 3, βoverlap = 5, path = 5:5:50 (then search around best k)\n",
    "    # sim set 6 are for k = 30, r = 3, βoverlap = 7, path = 5:5:50 (then search around best k)\n",
    "    # sim set 7 are for k = 10, r = 4, βoverlap = 2, path = 5:5:50 (then search around best k)\n",
    "    # sim set 8 are for k = 20, r = 4, βoverlap = 3, path = 5:5:50 (then search around best k)\n",
    "    # sim set 9 are for k = 30, r = 4, βoverlap = 5, path = 5:5:50 (then search around best k)\n",
    "\n",
    "    # compute summary statistics\n",
    "    open(\"summary.txt\", \"w\") do summary_io\n",
    "        for set in 1:6\n",
    "            successes = 0\n",
    "            empty!(mIHT_time); empty!(mIHT_plei_power); empty!(mIHT_indp_power); empty!(mIHT_FP); empty!(mIHT_FPR)\n",
    "            empty!(uIHT_time); empty!(uIHT_plei_power); empty!(uIHT_indp_power); empty!(uIHT_FP); empty!(uIHT_FPR)\n",
    "            empty!(mvPLINK_time); empty!(mvPLINK_plei_power); empty!(mvPLINK_indp_power); empty!(mvPLINK_FP); empty!(mvPLINK_FPR); empty!(mvPLINK_λ)\n",
    "            empty!(gemma_time); empty!(gemma_plei_power); empty!(gemma_indp_power); empty!(gemma_FP); empty!(gemma_FPR); empty!(gemma_λ)\n",
    "\n",
    "            # read each simulation's result\n",
    "            for sim in 1:100\n",
    "                if !isfile(\"set$set/sim$sim/summary.txt\")\n",
    "                    continue\n",
    "                end\n",
    "                try\n",
    "                    open(\"set$set/sim$sim/summary.txt\", \"r\") do io\n",
    "                        readline(io); readline(io); readline(io); readline(io); readline(io)\n",
    "\n",
    "                        # parse mIHT result\n",
    "                        mIHT = match(regex, readline(io))\n",
    "                        push!(mIHT_time, parse(Float64, mIHT[1]))\n",
    "                        push!(mIHT_plei_power, parse(Float64, mIHT[2]))\n",
    "                        push!(mIHT_indp_power, parse(Float64, mIHT[3]))\n",
    "                        push!(mIHT_FP, parse(Float64, mIHT[4]))\n",
    "                        push!(mIHT_FPR, parse(Float64, mIHT[5]))\n",
    "\n",
    "                        # parse uIHT result\n",
    "                        uIHT = match(regex, readline(io))\n",
    "                        push!(uIHT_time, parse(Float64, uIHT[1]))\n",
    "                        push!(uIHT_plei_power, parse(Float64, uIHT[2]))\n",
    "                        push!(uIHT_indp_power, parse(Float64, uIHT[3]))\n",
    "                        push!(uIHT_FP, parse(Float64, uIHT[4]))\n",
    "                        push!(uIHT_FPR, parse(Float64, uIHT[5]))\n",
    "\n",
    "                        # parse mvPLINK result\n",
    "                        mvPLINK = match(regex, readline(io))\n",
    "                        push!(mvPLINK_time, parse(Float64, mvPLINK[1]))\n",
    "                        push!(mvPLINK_plei_power, parse(Float64, mvPLINK[2]))\n",
    "                        push!(mvPLINK_indp_power, parse(Float64, mvPLINK[3]))\n",
    "                        push!(mvPLINK_FP, parse(Float64, mvPLINK[4]))\n",
    "                        push!(mvPLINK_FPR, parse(Float64, mvPLINK[5]))\n",
    "                        push!(mvPLINK_λ, parse(Float64, mvPLINK[6]))\n",
    "\n",
    "                        # parse mvPLINK result\n",
    "                        gemma = match(regex, readline(io))\n",
    "                        push!(gemma_time, parse(Float64, gemma[1]))\n",
    "                        push!(gemma_plei_power, parse(Float64, gemma[2]))\n",
    "                        push!(gemma_indp_power, parse(Float64, gemma[3]))\n",
    "                        push!(gemma_FP, parse(Float64, gemma[4]))\n",
    "                        push!(gemma_FPR, parse(Float64, gemma[5]))\n",
    "                        push!(gemma_λ, parse(Float64, gemma[6]))\n",
    "                    end\n",
    "                    successes += 1\n",
    "                catch\n",
    "                    continue\n",
    "                end\n",
    "            end\n",
    "            \n",
    "            # summary statistics\n",
    "            mIHT_time_mean, mIHT_time_std = round(mean(mIHT_time), digits=1), round(std(mIHT_time), digits=1)\n",
    "            mIHT_plei_TP_mean, mIHT_plei_TP_std = round(mean(mIHT_plei_power), digits=2), round(std(mIHT_plei_power), digits=2)\n",
    "            mIHT_indp_TP_mean, mIHT_indp_TP_std = round(mean(mIHT_indp_power), digits=2), round(std(mIHT_indp_power), digits=2)\n",
    "            mIHT_FP_mean, mIHT_FP_std = round(mean(mIHT_FP), digits=1), round(std(mIHT_FP), digits=1)\n",
    "            \n",
    "            uIHT_time_mean, uIHT_time_std = round(mean(uIHT_time), digits=1), round(std(uIHT_time), digits=1)\n",
    "            uIHT_plei_TP_mean, uIHT_plei_TP_std = round(mean(uIHT_plei_power), digits=2), round(std(uIHT_plei_power), digits=2)\n",
    "            uIHT_indp_TP_mean, uIHT_indp_TP_std = round(mean(uIHT_indp_power), digits=2), round(std(uIHT_indp_power), digits=2)\n",
    "            uIHT_FP_mean, uIHT_FP_std = round(mean(uIHT_FP), digits=1), round(std(uIHT_FP), digits=1)\n",
    "            \n",
    "            mvPLINK_time_mean, mvPLINK_time_std = round(mean(mvPLINK_time), digits=1), round(std(mvPLINK_time), digits=1)\n",
    "            mvPLINK_plei_TP_mean, mvPLINK_plei_TP_std = round(mean(mvPLINK_plei_power), digits=2), round(std(mvPLINK_plei_power), digits=2)\n",
    "            mvPLINK_indp_TP_mean, mvPLINK_indp_TP_std = round(mean(mvPLINK_indp_power), digits=2), round(std(mvPLINK_indp_power), digits=2)\n",
    "            mvPLINK_FP_mean, mvPLINK_FP_std = round(mean(mvPLINK_FP), digits=1), round(std(mvPLINK_FP), digits=1)\n",
    "            mvPLINK_λ_mean, mvPLINK_λ_std = round(mean(mvPLINK_λ), digits=1), round(std(mvPLINK_λ), digits=1)\n",
    "            \n",
    "            gemma_time_mean, gemma_time_std = round(mean(gemma_time), digits=1), round(std(gemma_time), digits=1)\n",
    "            gemma_plei_TP_mean, gemma_plei_TP_std = round(mean(gemma_plei_power), digits=2), round(std(gemma_plei_power), digits=2)\n",
    "            gemma_indp_TP_mean, gemma_indp_TP_std = round(mean(gemma_indp_power), digits=2), round(std(gemma_indp_power), digits=2)\n",
    "            gemma_FP_mean, gemma_FP_std = round(mean(gemma_FP), digits=1), round(std(gemma_FP), digits=1)\n",
    "            gemma_λ_mean, gemma_λ_std = round(mean(gemma_λ), digits=1), round(std(gemma_λ), digits=1)\n",
    "            \n",
    "            println(summary_io, \"set $set summary (successful run = $successes):\")\n",
    "            println(summary_io, \"mIHT time = $mIHT_time_mean ± $mIHT_time_std, plei TP = $mIHT_plei_TP_mean ± $mIHT_plei_TP_std, indep TP = $mIHT_indp_TP_mean ± $mIHT_indp_TP_std, FP = $mIHT_FP_mean ± $mIHT_FP_std\")\n",
    "            println(summary_io, \"uIHT time = $uIHT_time_mean ± $uIHT_time_std, plei TP = $uIHT_plei_TP_mean ± $uIHT_plei_TP_std, indep TP = $uIHT_indp_TP_mean ± $uIHT_indp_TP_std, FP = $uIHT_FP_mean ± $uIHT_FP_std\")\n",
    "            println(summary_io, \"mvPLINK time = $mvPLINK_time_mean ± $mvPLINK_time_std, plei TP = $mvPLINK_plei_TP_mean ± $mvPLINK_plei_TP_std, indep TP = $mvPLINK_indp_TP_mean ± $mvPLINK_indp_TP_std, FP = $mvPLINK_FP_mean ± $mvPLINK_FP_std\")\n",
    "            println(summary_io, \"gemma time = $gemma_time_mean ± $gemma_time_std, plei TP = $gemma_plei_TP_mean ± $gemma_plei_TP_std, indep TP = $gemma_indp_TP_mean ± $gemma_indp_TP_std, FP = $gemma_FP_mean ± $gemma_FP_std \\n\")\n",
    "            \n",
    "            if verbose\n",
    "                println(\"set $set summary (successful run = $successes):\")\n",
    "                println(\"mIHT time = $mIHT_time_mean ± $mIHT_time_std, plei TP = $mIHT_plei_TP_mean ± $mIHT_plei_TP_std, indep TP = $mIHT_indp_TP_mean ± $mIHT_indp_TP_std, FP = $mIHT_FP_mean ± $mIHT_FP_std\")\n",
    "                println(\"uIHT time = $uIHT_time_mean ± $uIHT_time_std, plei TP = $uIHT_plei_TP_mean ± $uIHT_plei_TP_std, indep TP = $uIHT_indp_TP_mean ± $uIHT_indp_TP_std, FP = $uIHT_FP_mean ± $uIHT_FP_std\")\n",
    "                println(\"mvPLINK time = $mvPLINK_time_mean ± $mvPLINK_time_std, plei TP = $mvPLINK_plei_TP_mean ± $mvPLINK_plei_TP_std, indep TP = $mvPLINK_indp_TP_mean ± $mvPLINK_indp_TP_std, FP = $mvPLINK_FP_mean ± $mvPLINK_FP_std\")\n",
    "                println(\"gemma time = $gemma_time_mean ± $gemma_time_std, plei TP = $gemma_plei_TP_mean ± $gemma_plei_TP_std, indep TP = $gemma_indp_TP_mean ± $gemma_indp_TP_std, FP = $gemma_FP_mean ± $gemma_FP_std \\n\")\n",
    "            end\n",
    "            \n",
    "            # for latex table\n",
    "#             if verbose\n",
    "#                 println(\"set $set summary (successful run = $successes):\")\n",
    "#                 println(\"\\\\texttt{mIHT} & \\$$mIHT_time_mean \\\\pm $mIHT_time_std\\$ & \\$$mIHT_plei_TP_mean \\\\pm $mIHT_plei_TP_std\\$ & \\$$mIHT_indp_TP_mean \\\\pm $mIHT_indp_TP_std\\$ & \\$$mIHT_FP_mean \\\\pm $mIHT_FP_std\\$\\\\\\\\\")\n",
    "#                 println(\"\\\\texttt{uIHT} & \\$$uIHT_time_mean \\\\pm $uIHT_time_std\\$ & \\$$uIHT_plei_TP_mean \\\\pm $uIHT_plei_TP_std\\$ & \\$$uIHT_indp_TP_mean \\\\pm $uIHT_indp_TP_std\\$ & \\$$uIHT_FP_mean \\\\pm $uIHT_FP_std\\$\\\\\\\\\")\n",
    "#                 println(\"\\\\texttt{CCA} & \\$$mvPLINK_time_mean \\\\pm $mvPLINK_time_std\\$ & \\$$mvPLINK_plei_TP_mean \\\\pm $mvPLINK_plei_TP_std\\$ & \\$$mvPLINK_indp_TP_mean \\\\pm $mvPLINK_indp_TP_std\\$ & \\$$mvPLINK_FP_mean \\\\pm $mvPLINK_FP_std\\$\\\\\\\\\")\n",
    "#                 println(\"\\\\texttt{mvLMM} & \\$$gemma_time_mean \\\\pm $gemma_time_std\\$ & \\$$gemma_plei_TP_mean \\\\pm $gemma_plei_TP_std\\$ & \\$$gemma_indp_TP_mean \\\\pm $gemma_indp_TP_std\\$ & \\$$gemma_FP_mean \\\\pm $gemma_FP_std\\$\\\\\\\\ \\n\")\n",
    "#             end\n",
    "        end\n",
    "    end\n",
    "    \n",
    "    return nothing\n",
    "end\n",
    "read_summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "@webio": {
   "lastCommId": null,
   "lastKernelId": null
  },
  "kernelspec": {
   "display_name": "Julia 1.6.0",
   "language": "julia",
   "name": "julia-1.6"
  },
  "language_info": {
   "file_extension": ".jl",
   "mimetype": "application/julia",
   "name": "julia",
   "version": "1.6.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
