var documenterSearchIndex = {"docs":
[{"location":"#MendelIHT.jl-Documentation-1","page":"Home","title":"MendelIHT.jl Documentation","text":"","category":"section"},{"location":"#","page":"Home","title":"Home","text":"A modern approach to analyze data from a Genome Wide Association Studies (GWAS)","category":"page"},{"location":"#","page":"Home","title":"Home","text":"note: Note\nPlease see the OpenMendel paper for a review of current statistical methods for GWAS and why we chose Julia as the programming language for this project. Pay attention to sections Handling SNP data and Iterative hard thresholding because they are especially relevant in this package.","category":"page"},{"location":"#Package-Feature-1","page":"Home","title":"Package Feature","text":"","category":"section"},{"location":"#","page":"Home","title":"Home","text":"Analyze large GWAS datasets intuitively.\nAutomatically distributed (parallel) q-fold cross-validation routines.\nAbility to fit a large range of generalized linear models with any choice of link function.\nComputation directly on raw genotype files.\nAbility to include non-genetic covariates.\nOptional acceleration step to dramatically improve speed by setting debias = true.\nRare variant discovery: ability to incorporate prior weight information.\nRare variant discovery: ability to handle within and between group sparsity. \nEasily adjustable convergence criteria, max iteration count, and max backtracking.","category":"page"}]
}
