{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "# Normal: IHT vs LASSO vs Marginal regression: comparing false positive/negatives\n",
    "\n",
    "Marginal regression is the most prevalent method used to analyze GWAS data. LASSO is currently the de facto penalized least squares method for [feature selection](https://en.wikipedia.org/wiki/Feature_selection). Here we compare the performance (in terms of the number of false positives/negative) of LASSO and marginal regression with IHT for GWAS data. We use the `glmnet` implementation of cyclic coordinate descent for LASSO, and `MendelGWAS.jl` module of [OpenMendel](https://openmendel.github.io/MendelGWAS.jl/) for marginal analysis. Since the focus here is not scalability, we test our sample problems on moderately sized genotype matrces of 1000 samples with 10,000 SNPs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "11"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "using Distributed\n",
    "addprocs(10)\n",
    "nprocs()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "┌ Info: Loading DataFrames support into Gadfly.jl\n",
      "└ @ Gadfly /u/home/b/biona001/.julia/packages/Gadfly/09PWZ/src/mapping.jl:228\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall /u/home/b/biona001/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: Loading required package: Matrix\n",
      "│ Loading required package: foreach\n",
      "│ Loaded glmnet 2.0-16\n",
      "│ \n",
      "└ @ RCall /u/home/b/biona001/.julia/packages/RCall/ffM0W/src/io.jl:113\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "RObject{StrSxp}\n",
       " [1] \"glmnet\"    \"foreach\"   \"Matrix\"    \"stats\"     \"graphics\"  \"grDevices\"\n",
       " [7] \"utils\"     \"datasets\"  \"methods\"   \"base\"     \n"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "using MendelIHT\n",
    "using SnpArrays\n",
    "using DataFrames\n",
    "using Distributions\n",
    "using Random\n",
    "using LinearAlgebra\n",
    "using DelimitedFiles\n",
    "using GLM\n",
    "using MendelGWAS\n",
    "using CSV\n",
    "using RCall\n",
    "R\"library(glmnet)\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "# How we ran LASSO\n",
    "\n",
    "We use the R library `glmnet` (implemented in Fortran) to run LASSO. [Documentation of Rcall.jl](http://juliainterop.github.io/RCall.jl/stable/gettingstarted.html) teaches us to to transfer variables between R and Julia. Since glmnet does not operate on genotype file, we need to convert `x` to Float64, and then run `glmnet`. We chose to run IHT across 50 different model sizes $\\{1,2,...,50\\}$ because the best model selected by LASSO averages to about 25 non-zero predictors, so we doubled that number. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "iht_lasso_marginal (generic function with 1 method)"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "function iht_lasso_marginal(n::Int64, p::Int64, k::Int64, d::UnionAll, l::Link)\n",
    "    #construct snpmatrix, covariate files, and true model b\n",
    "    x, = simulate_random_snparray(n, p, \"tmp.bed\")\n",
    "    xbm = SnpBitMatrix{Float64}(x, model=ADDITIVE_MODEL, center=true, scale=true); \n",
    "    z = ones(n, 1) # the intercept\n",
    "    x_float = [convert(Matrix{Float64}, x, center=true, scale=true) z] #Float64 version of x\n",
    "\n",
    "    # simulate response, true model b, and the correct non-0 positions of b\n",
    "    y, true_b, correct_position = simulate_random_response(x, xbm, k, d, l)\n",
    " \n",
    "    #specify path and folds\n",
    "    num_folds = 3\n",
    "    folds = rand(1:num_folds, size(x, 1));\n",
    "\n",
    "    #run glmnet via Rcall\n",
    "    @rput x_float y folds num_folds #make variables visible to R\n",
    "    R\"lasso_cv_result = cv.glmnet(x_float, y, nfolds = num_folds, foldid = folds, family='gaussian')\"\n",
    "    R\"lasso_beta_tmp = glmnet(x_float, y, lambda=lasso_cv_result$lambda.min, family='gaussian')$beta\"\n",
    "    R\"lasso_beta = as.vector(lasso_beta_tmp)\"\n",
    "    @rget lasso_cv_result lasso_beta #pull result from R to Julia\n",
    "    lasso_k_est = count(!iszero, lasso_beta)\n",
    "    \n",
    "    #find non-zero entries returned by best lasso model as largest k estimate\n",
    "    path = collect(1:50);\n",
    "    \n",
    "    #run IHT's cross validation routine \n",
    "    mses = cv_iht(d(), l, x, z, y, 1, path, folds, num_folds, use_maf=false, debias=false, showinfo=false, parallel=true);\n",
    "    iht_k_est = argmin(mses)\n",
    "    iht_result = L0_reg(x, xbm, z, y, 1, iht_k_est, d(), l, debias=false, init=false, use_maf=false)\n",
    "    iht_beta = iht_result.beta\n",
    "    \n",
    "    #Now run MendelGWAS\n",
    "    make_bim_fam_files(x, y, \"tmp\") #create .bim and .bed files for MendelGWAS\n",
    "\n",
    "    #create a control file for MendelGWAS\n",
    "    open(\"tmp_control.txt\", \"w\") do f\n",
    "        write(f, \"plink_input_basename = tmp \\n\")\n",
    "        write(f, \"plink_field_separator = '\\t' \\n\")\n",
    "        write(f, \"output_table = tmp_table.txt \\n\")\n",
    "        write(f, \"regression = linear \\n\")\n",
    "        write(f, \"regression_formula = Trait ~ \\n\")\n",
    "    end\n",
    "\n",
    "    #run marginal analysis\n",
    "    GWAS(\"tmp_control.txt\");\n",
    "\n",
    "    # calculate false positive/negatives based on bonferroni correction\n",
    "    p_values = CSV.read(\"tmp_table.txt\", delim=',', header=true)[:Pvalue]\n",
    "    significance = 0.05 / p\n",
    "    passing_snps_position = findall(p_values .<= significance)\n",
    "    marginal_found = [correct_position[snp] in passing_snps_position for snp in 1:k]\n",
    "\n",
    "    #clean up\n",
    "    rm(\"tmp.bed\", force=true)\n",
    "    rm(\"tmp.bim\", force=true)\n",
    "    rm(\"tmp.fam\", force=true)\n",
    "    rm(\"tmp_table.txt\", force=true)\n",
    "    rm(\"tmp_control.txt\", force=true)\n",
    "    rm(\"Mendel_Output.txt\", force=true)\n",
    "        \n",
    "    #show lasso and IHT's reconstruction result\n",
    "    compare_model = DataFrame(\n",
    "        true_β  = true_b[correct_position], \n",
    "        IHT_β   = iht_beta[correct_position],\n",
    "        lasso_β = lasso_beta[correct_position],\n",
    "        marginal_found = marginal_found)\n",
    "    @show compare_model\n",
    "    \n",
    "    #compute true/false positives/negatives for IHT and lasso\n",
    "    iht_tp = count(!iszero, iht_beta[correct_position])\n",
    "    iht_fp = iht_k_est - iht_tp\n",
    "    iht_fn = k - iht_tp\n",
    "    lasso_tp = count(!iszero, lasso_beta[correct_position])\n",
    "    lasso_fp = lasso_k_est - lasso_tp\n",
    "    lasso_fn = k - lasso_tp\n",
    "    marginal_tp = count(!iszero, true_b[passing_snps_position])\n",
    "    marginal_fp = length(passing_snps_position) - marginal_tp\n",
    "    marginal_fn = k - marginal_tp\n",
    "    \n",
    "    println(\"IHT false positives = $iht_fp\")\n",
    "    println(\"IHT false negatives = $iht_fn\")\n",
    "    println(\"LASSO false positives = $lasso_fp\")\n",
    "    println(\"LASSO false negatives = $lasso_fn\")\n",
    "    println(\"marginal false positives = $marginal_fp\")\n",
    "    println(\"marginal false negatives = $marginal_fn\" * \"\\n\")\n",
    "    \n",
    "    return iht_fp, iht_fn, lasso_fp, lasso_fn, marginal_fp, marginal_fn\n",
    "end"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "# Normal response"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true,
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "run (generic function with 1 method)"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "function run()\n",
    "    #simulat data with k true predictors, from distribution d and with link l.\n",
    "    n = 1000\n",
    "    p = 10000\n",
    "    k = 10\n",
    "    d = Normal\n",
    "    l = canonicallink(d())\n",
    "\n",
    "    #run function above, saving results in 4 vectors\n",
    "    total_runs = 50\n",
    "    iht_false_positives = zeros(total_runs)\n",
    "    iht_false_negatives = zeros(total_runs)\n",
    "    lasso_false_positives = zeros(total_runs)\n",
    "    lasso_false_negatives = zeros(total_runs)\n",
    "    marginal_false_positives = zeros(total_runs)\n",
    "    marginal_false_negatives = zeros(total_runs)\n",
    "    for i in 1:total_runs\n",
    "        println(\"current run = $i\")\n",
    "        \n",
    "        #set random seed\n",
    "        Random.seed!(i)\n",
    "        \n",
    "        iht_fp, iht_fn, lasso_fp, lasso_fn, m_fp, m_fn = iht_lasso_marginal(n, p, k, d, l)\n",
    "        iht_false_positives[i] = iht_fp\n",
    "        iht_false_negatives[i] = iht_fn\n",
    "        lasso_false_positives[i] = lasso_fp\n",
    "        lasso_false_negatives[i] = lasso_fn\n",
    "        marginal_false_positives[i] = m_fp\n",
    "        marginal_false_negatives[i] = m_fn\n",
    "    end\n",
    "    \n",
    "    return iht_false_positives, iht_false_negatives, lasso_false_positives, \n",
    "            lasso_false_negatives, marginal_false_positives, marginal_false_negatives\n",
    "end"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "# Compute average number of false positives/negatives"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "current run = 1\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β    │ IHT_β     │ lasso_β   │ marginal_found │\n",
      "│     │ Float64   │ Float64   │ Float64   │ Bool           │\n",
      "├─────┼───────────┼───────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.607538  │ 0.546901  │ 0.450247  │ false          │\n",
      "│ 2   │ -1.53004  │ -1.58398  │ -1.49073  │ false          │\n",
      "│ 3   │ 1.44416   │ 1.45971   │ 1.37278   │ false          │\n",
      "│ 4   │ 1.75303   │ 1.73809   │ 1.64483   │ false          │\n",
      "│ 5   │ 0.69867   │ 0.704586  │ 0.593622  │ false          │\n",
      "│ 6   │ -1.69162  │ -1.6952   │ -1.5926   │ false          │\n",
      "│ 7   │ -0.211858 │ -0.226112 │ -0.130548 │ false          │\n",
      "│ 8   │ 0.092561  │ 0.0       │ 0.0       │ false          │\n",
      "│ 9   │ 0.611886  │ 0.63118   │ 0.52574   │ false          │\n",
      "│ 10  │ -1.52095  │ -1.5026   │ -1.40507  │ false          │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 1\n",
      "LASSO false positives = 26\n",
      "LASSO false negatives = 1\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 10\n",
      "\n",
      "current run = 2\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β    │ IHT_β     │ lasso_β   │ marginal_found │\n",
      "│     │ Float64   │ Float64   │ Float64   │ Bool           │\n",
      "├─────┼───────────┼───────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.749759  │ 0.802508  │ 0.698892  │ true           │\n",
      "│ 2   │ 0.677934  │ 0.694198  │ 0.598272  │ true           │\n",
      "│ 3   │ 0.0720974 │ 0.0       │ 0.0       │ false          │\n",
      "│ 4   │ -1.13432  │ -1.11165  │ -1.0057   │ true           │\n",
      "│ 5   │ -0.827478 │ -0.795373 │ -0.706908 │ true           │\n",
      "│ 6   │ 1.13273   │ 1.05656   │ 0.966889  │ true           │\n",
      "│ 7   │ -0.898177 │ -0.88663  │ -0.802203 │ true           │\n",
      "│ 8   │ -2.55085  │ -2.60053  │ -2.51343  │ true           │\n",
      "│ 9   │ 1.90046   │ 1.86493   │ 1.77983   │ false          │\n",
      "│ 10  │ 0.668923  │ 0.660692  │ 0.559003  │ true           │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 1\n",
      "LASSO false positives = 14\n",
      "LASSO false negatives = 1\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 2\n",
      "\n",
      "current run = 3\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β    │ IHT_β     │ lasso_β   │ marginal_found │\n",
      "│     │ Float64   │ Float64   │ Float64   │ Bool           │\n",
      "├─────┼───────────┼───────────┼───────────┼────────────────┤\n",
      "│ 1   │ -1.29395  │ -1.31156  │ -1.2234   │ false          │\n",
      "│ 2   │ 0.786253  │ 0.778146  │ 0.677479  │ false          │\n",
      "│ 3   │ -2.06493  │ -2.12481  │ -2.01415  │ false          │\n",
      "│ 4   │ 0.540411  │ 0.587876  │ 0.48026   │ false          │\n",
      "│ 5   │ 1.49302   │ 1.50456   │ 1.40013   │ false          │\n",
      "│ 6   │ -0.640079 │ -0.629336 │ -0.547408 │ false          │\n",
      "│ 7   │ 0.0457283 │ 0.0       │ 0.0       │ false          │\n",
      "│ 8   │ -1.5777   │ -1.60712  │ -1.51259  │ false          │\n",
      "│ 9   │ -1.56178  │ -1.58028  │ -1.48088  │ false          │\n",
      "│ 10  │ 0.127488  │ 0.0       │ 0.0408414 │ false          │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 2\n",
      "LASSO false positives = 41\n",
      "LASSO false negatives = 1\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 10\n",
      "\n",
      "current run = 4\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β     │ IHT_β     │ lasso_β    │ marginal_found │\n",
      "│     │ Float64    │ Float64   │ Float64    │ Bool           │\n",
      "├─────┼────────────┼───────────┼────────────┼────────────────┤\n",
      "│ 1   │ 1.01155    │ 1.02739   │ 0.913529   │ true           │\n",
      "│ 2   │ -0.375936  │ -0.3984   │ -0.295008  │ false          │\n",
      "│ 3   │ -0.909578  │ -0.907387 │ -0.800775  │ true           │\n",
      "│ 4   │ -2.04195   │ -2.02569  │ -1.93188   │ true           │\n",
      "│ 5   │ -1.24787   │ -1.24457  │ -1.15789   │ true           │\n",
      "│ 6   │ -0.219267  │ 0.0       │ -0.0322982 │ false          │\n",
      "│ 7   │ -0.0620693 │ 0.0       │ 0.0        │ false          │\n",
      "│ 8   │ -0.493204  │ -0.448463 │ -0.364609  │ true           │\n",
      "│ 9   │ 0.0735544  │ 0.0       │ 0.0        │ false          │\n",
      "│ 10  │ -0.40849   │ -0.432987 │ -0.321045  │ false          │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 3\n",
      "LASSO false positives = 16\n",
      "LASSO false negatives = 2\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 5\n",
      "\n",
      "current run = 5\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β     │ IHT_β     │ lasso_β   │ marginal_found │\n",
      "│     │ Float64    │ Float64   │ Float64   │ Bool           │\n",
      "├─────┼────────────┼───────────┼───────────┼────────────────┤\n",
      "│ 1   │ 1.49393    │ 1.48671   │ 1.39252   │ true           │\n",
      "│ 2   │ 0.268446   │ 0.217996  │ 0.112064  │ false          │\n",
      "│ 3   │ -1.57681   │ -1.59947  │ -1.50845  │ true           │\n",
      "│ 4   │ -0.0665982 │ 0.0       │ 0.0       │ false          │\n",
      "│ 5   │ 0.564564   │ 0.572764  │ 0.485618  │ true           │\n",
      "│ 6   │ -0.0723161 │ 0.0       │ 0.0       │ false          │\n",
      "│ 7   │ -0.012129  │ 0.0       │ 0.0       │ false          │\n",
      "│ 8   │ 1.43517    │ 1.39943   │ 1.31278   │ true           │\n",
      "│ 9   │ -0.432575  │ -0.37851  │ -0.2861   │ false          │\n",
      "│ 10  │ -0.656486  │ -0.680305 │ -0.580108 │ true           │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 3\n",
      "LASSO false positives = 45\n",
      "LASSO false negatives = 3\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 5\n",
      "\n",
      "current run = 6\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β     │ IHT_β     │ lasso_β    │ marginal_found │\n",
      "│     │ Float64    │ Float64   │ Float64    │ Bool           │\n",
      "├─────┼────────────┼───────────┼────────────┼────────────────┤\n",
      "│ 1   │ 0.557855   │ 0.503238  │ 0.39887    │ true           │\n",
      "│ 2   │ 0.852578   │ 0.814239  │ 0.716678   │ true           │\n",
      "│ 3   │ -0.157019  │ 0.0       │ -0.0243698 │ false          │\n",
      "│ 4   │ -0.167593  │ -0.219871 │ -0.124482  │ false          │\n",
      "│ 5   │ 1.43279    │ 1.37108   │ 1.26639    │ true           │\n",
      "│ 6   │ 0.00814292 │ 0.0       │ 0.0        │ false          │\n",
      "│ 7   │ -0.311245  │ -0.349964 │ -0.235133  │ false          │\n",
      "│ 8   │ 1.14164    │ 1.14145   │ 1.02704    │ true           │\n",
      "│ 9   │ 0.296891   │ 0.319944  │ 0.206203   │ false          │\n",
      "│ 10  │ -0.407489  │ -0.417213 │ -0.306936  │ true           │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 2\n",
      "LASSO false positives = 13\n",
      "LASSO false negatives = 1\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 5\n",
      "\n",
      "current run = 7\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β     │ IHT_β     │ lasso_β   │ marginal_found │\n",
      "│     │ Float64    │ Float64   │ Float64   │ Bool           │\n",
      "├─────┼────────────┼───────────┼───────────┼────────────────┤\n",
      "│ 1   │ -1.11915   │ -1.09187  │ -0.987788 │ true           │\n",
      "│ 2   │ 0.61236    │ 0.582924  │ 0.463789  │ true           │\n",
      "│ 3   │ 0.559179   │ 0.557581  │ 0.437407  │ true           │\n",
      "│ 4   │ 0.460021   │ 0.46291   │ 0.345071  │ true           │\n",
      "│ 5   │ -0.0420588 │ 0.0       │ 0.0       │ false          │\n",
      "│ 6   │ -0.595666  │ -0.489398 │ -0.394499 │ true           │\n",
      "│ 7   │ 0.54228    │ 0.555662  │ 0.437753  │ true           │\n",
      "│ 8   │ 0.933922   │ 0.925222  │ 0.821359  │ true           │\n",
      "│ 9   │ 0.685041   │ 0.690302  │ 0.572829  │ true           │\n",
      "│ 10  │ 1.15986    │ 1.16186   │ 1.04076   │ true           │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 1\n",
      "LASSO false positives = 21\n",
      "LASSO false negatives = 1\n",
      "marginal false positives = 1\n",
      "marginal false negatives = 1\n",
      "\n",
      "current run = 8\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β    │ IHT_β     │ lasso_β   │ marginal_found │\n",
      "│     │ Float64   │ Float64   │ Float64   │ Bool           │\n",
      "├─────┼───────────┼───────────┼───────────┼────────────────┤\n",
      "│ 1   │ -0.318516 │ -0.293541 │ -0.190439 │ false          │\n",
      "│ 2   │ 0.80228   │ 0.755705  │ 0.644118  │ true           │\n",
      "│ 3   │ 0.678794  │ 0.682713  │ 0.577202  │ true           │\n",
      "│ 4   │ -0.550278 │ -0.539721 │ -0.453756 │ true           │\n",
      "│ 5   │ -1.039    │ -1.06408  │ -0.968181 │ true           │\n",
      "│ 6   │ -1.10893  │ -1.12409  │ -1.02399  │ true           │\n",
      "│ 7   │ -2.11847  │ -2.10469  │ -2.01802  │ true           │\n",
      "│ 8   │ 0.539736  │ 0.520486  │ 0.410471  │ true           │\n",
      "│ 9   │ 0.0513002 │ 0.0       │ 0.0       │ false          │\n",
      "│ 10  │ 1.44312   │ 1.44748   │ 1.36078   │ true           │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 1\n",
      "LASSO false positives = 19\n",
      "LASSO false negatives = 1\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 2\n",
      "\n",
      "current run = 9\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β    │ IHT_β     │ lasso_β   │ marginal_found │\n",
      "│     │ Float64   │ Float64   │ Float64   │ Bool           │\n",
      "├─────┼───────────┼───────────┼───────────┼────────────────┤\n",
      "│ 1   │ -1.20375  │ -1.21128  │ -1.10972  │ true           │\n",
      "│ 2   │ -0.334385 │ -0.292532 │ -0.201894 │ false          │\n",
      "│ 3   │ -1.82832  │ -1.79428  │ -1.69684  │ true           │\n",
      "│ 4   │ 1.3708    │ 1.3428    │ 1.22508   │ true           │\n",
      "│ 5   │ 1.20444   │ 1.22552   │ 1.1146    │ true           │\n",
      "│ 6   │ -0.508497 │ -0.500277 │ -0.387289 │ false          │\n",
      "│ 7   │ -1.09875  │ -1.13042  │ -1.02629  │ true           │\n",
      "│ 8   │ 0.486961  │ 0.438299  │ 0.322484  │ false          │\n",
      "│ 9   │ 0.251382  │ 0.248484  │ 0.150615  │ false          │\n",
      "│ 10  │ -0.460604 │ -0.464603 │ -0.351561 │ false          │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 0\n",
      "LASSO false positives = 36\n",
      "LASSO false negatives = 0\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 5\n",
      "\n",
      "current run = 10\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β   │ IHT_β     │ lasso_β   │ marginal_found │\n",
      "│     │ Float64  │ Float64   │ Float64   │ Bool           │\n",
      "├─────┼──────────┼───────────┼───────────┼────────────────┤\n",
      "│ 1   │ 1.17878  │ 1.19405   │ 1.08371   │ true           │\n",
      "│ 2   │ -0.44546 │ -0.405607 │ -0.299839 │ false          │\n",
      "│ 3   │ 0.249749 │ 0.2387    │ 0.123247  │ false          │\n",
      "│ 4   │ 1.47017  │ 1.5306    │ 1.41042   │ true           │\n",
      "│ 5   │ 0.655446 │ 0.656183  │ 0.556901  │ false          │\n",
      "│ 6   │ 2.48177  │ 2.41907   │ 2.31565   │ true           │\n",
      "│ 7   │ -1.38405 │ -1.39172  │ -1.29867  │ true           │\n",
      "│ 8   │ 1.2645   │ 1.24568   │ 1.13037   │ true           │\n",
      "│ 9   │ 1.70905  │ 1.70875   │ 1.61156   │ true           │\n",
      "│ 10  │ 0.494919 │ 0.493086  │ 0.398115  │ true           │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 0\n",
      "LASSO false positives = 26\n",
      "LASSO false negatives = 0\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 3\n",
      "\n",
      "current run = 11\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β    │ IHT_β     │ lasso_β    │ marginal_found │\n",
      "│     │ Float64   │ Float64   │ Float64    │ Bool           │\n",
      "├─────┼───────────┼───────────┼────────────┼────────────────┤\n",
      "│ 1   │ -2.1401   │ -2.12378  │ -2.04098   │ true           │\n",
      "│ 2   │ 0.597831  │ 0.556601  │ 0.470517   │ true           │\n",
      "│ 3   │ -0.068048 │ 0.0       │ -0.0127408 │ false          │\n",
      "│ 4   │ 0.379909  │ 0.434963  │ 0.339117   │ true           │\n",
      "│ 5   │ -1.59939  │ -1.6117   │ -1.52684   │ true           │\n",
      "│ 6   │ -0.519577 │ -0.513549 │ -0.432209  │ true           │\n",
      "│ 7   │ 0.33893   │ 0.346509  │ 0.260918   │ false          │\n",
      "│ 8   │ 0.822734  │ 0.840362  │ 0.752439   │ true           │\n",
      "│ 9   │ -0.437776 │ -0.46378  │ -0.362638  │ false          │\n",
      "│ 10  │ 0.454821  │ 0.445969  │ 0.355061   │ true           │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 1\n",
      "LASSO false positives = 36\n",
      "LASSO false negatives = 0\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 3\n",
      "\n",
      "current run = 12\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β     │ IHT_β    │ lasso_β  │ marginal_found │\n",
      "│     │ Float64    │ Float64  │ Float64  │ Bool           │\n",
      "├─────┼────────────┼──────────┼──────────┼────────────────┤\n",
      "│ 1   │ 0.381684   │ 0.362566 │ 0.266142 │ false          │\n",
      "│ 2   │ -2.25632   │ -2.23578 │ -2.1444  │ true           │\n",
      "│ 3   │ -0.0285488 │ 0.0      │ 0.0      │ false          │\n",
      "│ 4   │ -1.20846   │ -1.21765 │ -1.10221 │ true           │\n",
      "│ 5   │ -1.1725    │ -1.11352 │ -1.00995 │ true           │\n",
      "│ 6   │ 0.666317   │ 0.646822 │ 0.556127 │ true           │\n",
      "│ 7   │ 1.33832    │ 1.35671  │ 1.25494  │ true           │\n",
      "│ 8   │ 0.695851   │ 0.654519 │ 0.555448 │ true           │\n",
      "│ 9   │ 0.0762416  │ 0.0      │ 0.0      │ false          │\n",
      "│ 10  │ 2.51051    │ 2.51597  │ 2.4295   │ true           │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 2\n",
      "LASSO false positives = 35\n",
      "LASSO false negatives = 2\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 3\n",
      "\n",
      "current run = 13\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β     │ IHT_β     │ lasso_β    │ marginal_found │\n",
      "│     │ Float64    │ Float64   │ Float64    │ Bool           │\n",
      "├─────┼────────────┼───────────┼────────────┼────────────────┤\n",
      "│ 1   │ -0.191228  │ -0.154721 │ -0.0563355 │ false          │\n",
      "│ 2   │ -0.530065  │ -0.555367 │ -0.455259  │ true           │\n",
      "│ 3   │ 0.761358   │ 0.844438  │ 0.749028   │ true           │\n",
      "│ 4   │ -0.295163  │ -0.28258  │ -0.18877   │ false          │\n",
      "│ 5   │ 1.14278    │ 1.18463   │ 1.10328    │ true           │\n",
      "│ 6   │ 1.42384    │ 1.41518   │ 1.32586    │ true           │\n",
      "│ 7   │ -0.415621  │ -0.361901 │ -0.279569  │ true           │\n",
      "│ 8   │ 0.815625   │ 0.813288  │ 0.709017   │ true           │\n",
      "│ 9   │ -0.0729403 │ 0.0       │ 0.0        │ false          │\n",
      "│ 10  │ 0.28586    │ 0.273342  │ 0.161847   │ false          │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 1\n",
      "LASSO false positives = 37\n",
      "LASSO false negatives = 1\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 4\n",
      "\n",
      "current run = 14\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β    │ IHT_β     │ lasso_β   │ marginal_found │\n",
      "│     │ Float64   │ Float64   │ Float64   │ Bool           │\n",
      "├─────┼───────────┼───────────┼───────────┼────────────────┤\n",
      "│ 1   │ -2.00106  │ -2.01739  │ -1.9007   │ true           │\n",
      "│ 2   │ 0.837229  │ 0.846887  │ 0.73332   │ true           │\n",
      "│ 3   │ 0.636888  │ 0.615318  │ 0.500977  │ false          │\n",
      "│ 4   │ 0.515326  │ 0.503971  │ 0.399379  │ false          │\n",
      "│ 5   │ 0.331752  │ 0.364085  │ 0.2499    │ false          │\n",
      "│ 6   │ -1.36196  │ -1.3438   │ -1.23662  │ true           │\n",
      "│ 7   │ -0.859086 │ -0.873312 │ -0.768546 │ true           │\n",
      "│ 8   │ -1.66215  │ -1.64069  │ -1.54226  │ true           │\n",
      "│ 9   │ 1.06366   │ 1.10384   │ 1.0001    │ true           │\n",
      "│ 10  │ 1.2262    │ 1.19539   │ 1.08123   │ true           │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 0\n",
      "LASSO false positives = 27\n",
      "LASSO false negatives = 0\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 3\n",
      "\n",
      "current run = 15\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β    │ IHT_β     │ lasso_β   │ marginal_found │\n",
      "│     │ Float64   │ Float64   │ Float64   │ Bool           │\n",
      "├─────┼───────────┼───────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.390517  │ 0.357157  │ 0.254096  │ false          │\n",
      "│ 2   │ 1.13247   │ 1.12183   │ 1.01956   │ false          │\n",
      "│ 3   │ -0.477164 │ -0.515393 │ -0.422042 │ false          │\n",
      "│ 4   │ -0.332879 │ -0.307288 │ -0.214262 │ false          │\n",
      "│ 5   │ 0.601342  │ 0.620502  │ 0.524477  │ false          │\n",
      "│ 6   │ 0.776711  │ 0.766932  │ 0.683351  │ false          │\n",
      "│ 7   │ -0.596155 │ -0.611178 │ -0.520108 │ false          │\n",
      "│ 8   │ -0.305982 │ -0.212148 │ -0.121665 │ false          │\n",
      "│ 9   │ -1.07709  │ -1.06177  │ -0.955763 │ false          │\n",
      "│ 10  │ 0.157683  │ 0.0       │ 0.0       │ false          │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 1\n",
      "LASSO false positives = 34\n",
      "LASSO false negatives = 1\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 10\n",
      "\n",
      "current run = 16\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β     │ IHT_β     │ lasso_β   │ marginal_found │\n",
      "│     │ Float64    │ Float64   │ Float64   │ Bool           │\n",
      "├─────┼────────────┼───────────┼───────────┼────────────────┤\n",
      "│ 1   │ 1.42842    │ 1.52143   │ 1.42197   │ true           │\n",
      "│ 2   │ 1.51933    │ 1.5647    │ 1.44377   │ true           │\n",
      "│ 3   │ 1.72402    │ 1.76965   │ 1.68716   │ true           │\n",
      "│ 4   │ -0.0571394 │ 0.0       │ 0.0       │ false          │\n",
      "│ 5   │ 0.31877    │ 0.281265  │ 0.175161  │ false          │\n",
      "│ 6   │ 0.19792    │ 0.178593  │ 0.0858754 │ false          │\n",
      "│ 7   │ -1.53487   │ -1.54388  │ -1.45439  │ true           │\n",
      "│ 8   │ -0.621343  │ -0.657032 │ -0.564882 │ true           │\n",
      "│ 9   │ -0.804211  │ -0.790684 │ -0.677368 │ true           │\n",
      "│ 10  │ -0.0669473 │ 0.0       │ 0.0       │ false          │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 2\n",
      "LASSO false positives = 34\n",
      "LASSO false negatives = 2\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 4\n",
      "\n",
      "current run = 17\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β    │ IHT_β     │ lasso_β   │ marginal_found │\n",
      "│     │ Float64   │ Float64   │ Float64   │ Bool           │\n",
      "├─────┼───────────┼───────────┼───────────┼────────────────┤\n",
      "│ 1   │ -0.631517 │ -0.671148 │ -0.57059  │ true           │\n",
      "│ 2   │ -0.284542 │ -0.301388 │ -0.209035 │ false          │\n",
      "│ 3   │ -0.90586  │ -0.883717 │ -0.801114 │ true           │\n",
      "│ 4   │ 2.13207   │ 2.14394   │ 2.05782   │ true           │\n",
      "│ 5   │ 0.179385  │ 0.0       │ 0.0556751 │ false          │\n",
      "│ 6   │ -1.47059  │ -1.45064  │ -1.3617   │ true           │\n",
      "│ 7   │ 0.345312  │ 0.362785  │ 0.275209  │ false          │\n",
      "│ 8   │ -0.606927 │ -0.552734 │ -0.447242 │ false          │\n",
      "│ 9   │ 0.685558  │ 0.723834  │ 0.64581   │ true           │\n",
      "│ 10  │ 0.0588637 │ 0.0       │ 0.0       │ false          │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 2\n",
      "LASSO false positives = 31\n",
      "LASSO false negatives = 1\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 5\n",
      "\n",
      "current run = 18\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β      │ IHT_β     │ lasso_β    │ marginal_found │\n",
      "│     │ Float64     │ Float64   │ Float64    │ Bool           │\n",
      "├─────┼─────────────┼───────────┼────────────┼────────────────┤\n",
      "│ 1   │ 0.687324    │ 0.661474  │ 0.538256   │ true           │\n",
      "│ 2   │ 0.578401    │ 0.569775  │ 0.457851   │ true           │\n",
      "│ 3   │ 0.834403    │ 0.839095  │ 0.7386     │ true           │\n",
      "│ 4   │ -0.00685474 │ 0.0       │ 0.0        │ false          │\n",
      "│ 5   │ -0.976679   │ -0.930395 │ -0.824203  │ true           │\n",
      "│ 6   │ -0.108428   │ 0.0       │ -0.0138268 │ false          │\n",
      "│ 7   │ -0.377253   │ -0.418161 │ -0.306445  │ true           │\n",
      "│ 8   │ 0.788596    │ 0.782608  │ 0.681827   │ true           │\n",
      "│ 9   │ -0.155683   │ 0.0       │ -0.0360011 │ false          │\n",
      "│ 10  │ 0.496849    │ 0.509086  │ 0.408224   │ true           │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 3\n",
      "LASSO false positives = 10\n",
      "LASSO false negatives = 1\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 3\n",
      "\n",
      "current run = 19\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β      │ IHT_β     │ lasso_β   │ marginal_found │\n",
      "│     │ Float64     │ Float64   │ Float64   │ Bool           │\n",
      "├─────┼─────────────┼───────────┼───────────┼────────────────┤\n",
      "│ 1   │ -0.419297   │ -0.428871 │ -0.335645 │ true           │\n",
      "│ 2   │ -0.223578   │ -0.240846 │ -0.123378 │ false          │\n",
      "│ 3   │ -0.283353   │ -0.275379 │ -0.175179 │ false          │\n",
      "│ 4   │ 0.294738    │ 0.291887  │ 0.174954  │ false          │\n",
      "│ 5   │ -0.00546738 │ 0.0       │ 0.0       │ false          │\n",
      "│ 6   │ 0.854599    │ 0.808109  │ 0.696778  │ true           │\n",
      "│ 7   │ 1.72256     │ 1.66067   │ 1.54475   │ true           │\n",
      "│ 8   │ -0.0586545  │ 0.0       │ 0.0       │ false          │\n",
      "│ 9   │ 1.02908     │ 1.00852   │ 0.915182  │ true           │\n",
      "│ 10  │ -1.65961    │ -1.63943  │ -1.53901  │ true           │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 2\n",
      "LASSO false positives = 12\n",
      "LASSO false negatives = 2\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 5\n",
      "\n",
      "current run = 20\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β    │ IHT_β     │ lasso_β    │ marginal_found │\n",
      "│     │ Float64   │ Float64   │ Float64    │ Bool           │\n",
      "├─────┼───────────┼───────────┼────────────┼────────────────┤\n",
      "│ 1   │ 0.978725  │ 0.97718   │ 0.873715   │ true           │\n",
      "│ 2   │ 0.135377  │ 0.147285  │ 0.0550684  │ false          │\n",
      "│ 3   │ 0.151616  │ 0.0       │ 0.0461726  │ false          │\n",
      "│ 4   │ -0.169305 │ -0.175268 │ -0.0864801 │ false          │\n",
      "│ 5   │ -0.232999 │ -0.304517 │ -0.211985  │ false          │\n",
      "│ 6   │ 0.181031  │ 0.190307  │ 0.0781171  │ false          │\n",
      "│ 7   │ -0.196465 │ 0.0       │ -0.0393363 │ false          │\n",
      "│ 8   │ 0.763341  │ 0.767269  │ 0.653497   │ true           │\n",
      "│ 9   │ 1.2331    │ 1.19464   │ 1.09757    │ true           │\n",
      "│ 10  │ -2.21915  │ -2.20726  │ -2.10989   │ true           │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 2\n",
      "LASSO false positives = 14\n",
      "LASSO false negatives = 0\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 6\n",
      "\n",
      "current run = 21\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β     │ IHT_β     │ lasso_β   │ marginal_found │\n",
      "│     │ Float64    │ Float64   │ Float64   │ Bool           │\n",
      "├─────┼────────────┼───────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.0729166  │ 0.0       │ 0.0       │ false          │\n",
      "│ 2   │ 1.37107    │ 1.3541    │ 1.25616   │ false          │\n",
      "│ 3   │ -0.329875  │ -0.303463 │ -0.203233 │ false          │\n",
      "│ 4   │ -0.287766  │ -0.324975 │ -0.236145 │ false          │\n",
      "│ 5   │ -0.784903  │ -0.798174 │ -0.705495 │ false          │\n",
      "│ 6   │ -0.0281104 │ 0.0       │ 0.0       │ false          │\n",
      "│ 7   │ 1.76022    │ 1.74482   │ 1.65598   │ false          │\n",
      "│ 8   │ 0.591337   │ 0.602732  │ 0.509522  │ false          │\n",
      "│ 9   │ 0.0898311  │ 0.0       │ 0.0       │ false          │\n",
      "│ 10  │ 1.20634    │ 1.23707   │ 1.16158   │ false          │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 3\n",
      "LASSO false positives = 28\n",
      "LASSO false negatives = 3\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 10\n",
      "\n",
      "current run = 22\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β    │ IHT_β     │ lasso_β    │ marginal_found │\n",
      "│     │ Float64   │ Float64   │ Float64    │ Bool           │\n",
      "├─────┼───────────┼───────────┼────────────┼────────────────┤\n",
      "│ 1   │ -1.43685  │ -1.51324  │ -1.42593   │ true           │\n",
      "│ 2   │ 1.50608   │ 1.56325   │ 1.46432    │ true           │\n",
      "│ 3   │ -0.319684 │ -0.324291 │ -0.22385   │ false          │\n",
      "│ 4   │ -0.122129 │ 0.0       │ -0.0352307 │ false          │\n",
      "│ 5   │ -0.708054 │ -0.632638 │ -0.527489  │ false          │\n",
      "│ 6   │ 2.58886   │ 2.58406   │ 2.49916    │ true           │\n",
      "│ 7   │ 1.56787   │ 1.49121   │ 1.40644    │ true           │\n",
      "│ 8   │ 0.548074  │ 0.564389  │ 0.46333    │ true           │\n",
      "│ 9   │ 1.81003   │ 1.79765   │ 1.69409    │ true           │\n",
      "│ 10  │ -0.681014 │ -0.669926 │ -0.574352  │ true           │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 1\n",
      "LASSO false positives = 24\n",
      "LASSO false negatives = 0\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 3\n",
      "\n",
      "current run = 23\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β    │ IHT_β     │ lasso_β   │ marginal_found │\n",
      "│     │ Float64   │ Float64   │ Float64   │ Bool           │\n",
      "├─────┼───────────┼───────────┼───────────┼────────────────┤\n",
      "│ 1   │ 2.99854   │ 2.99658   │ 2.88545   │ true           │\n",
      "│ 2   │ -0.822753 │ -0.843544 │ -0.736151 │ true           │\n",
      "│ 3   │ -0.696875 │ -0.727511 │ -0.622953 │ true           │\n",
      "│ 4   │ 0.892771  │ 0.901685  │ 0.784014  │ true           │\n",
      "│ 5   │ -0.846611 │ -0.85132  │ -0.753094 │ true           │\n",
      "│ 6   │ -1.82717  │ -1.86586  │ -1.75629  │ true           │\n",
      "│ 7   │ -0.492482 │ -0.478125 │ -0.371577 │ false          │\n",
      "│ 8   │ -1.30192  │ -1.2791   │ -1.17359  │ true           │\n",
      "│ 9   │ -0.528306 │ -0.489776 │ -0.398398 │ true           │\n",
      "│ 10  │ -0.394907 │ -0.415417 │ -0.299823 │ false          │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 0\n",
      "LASSO false positives = 19\n",
      "LASSO false negatives = 0\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 2\n",
      "\n",
      "current run = 24\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β     │ IHT_β     │ lasso_β   │ marginal_found │\n",
      "│     │ Float64    │ Float64   │ Float64   │ Bool           │\n",
      "├─────┼────────────┼───────────┼───────────┼────────────────┤\n",
      "│ 1   │ 2.19342    │ 2.23308   │ 2.14535   │ true           │\n",
      "│ 2   │ 1.61872    │ 1.69845   │ 1.61101   │ true           │\n",
      "│ 3   │ 1.16179    │ 1.15857   │ 1.07149   │ true           │\n",
      "│ 4   │ -1.66043   │ -1.61596  │ -1.53192  │ true           │\n",
      "│ 5   │ -0.0119048 │ 0.0       │ 0.0       │ false          │\n",
      "│ 6   │ -1.61177   │ -1.63537  │ -1.5484   │ true           │\n",
      "│ 7   │ 1.25813    │ 1.28675   │ 1.19779   │ true           │\n",
      "│ 8   │ -0.440263  │ -0.467164 │ -0.375078 │ false          │\n",
      "│ 9   │ 0.833073   │ 0.781349  │ 0.683935  │ true           │\n",
      "│ 10  │ -0.570173  │ -0.551504 │ -0.446835 │ false          │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 1\n",
      "LASSO false positives = 17\n",
      "LASSO false negatives = 1\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 3\n",
      "\n",
      "current run = 25\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β    │ IHT_β     │ lasso_β   │ marginal_found │\n",
      "│     │ Float64   │ Float64   │ Float64   │ Bool           │\n",
      "├─────┼───────────┼───────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.618107  │ 0.590093  │ 0.503062  │ true           │\n",
      "│ 2   │ -0.555343 │ -0.526535 │ -0.439432 │ true           │\n",
      "│ 3   │ -0.855731 │ -0.861544 │ -0.761358 │ true           │\n",
      "│ 4   │ 0.48719   │ 0.4687    │ 0.366725  │ true           │\n",
      "│ 5   │ -0.960793 │ -0.935068 │ -0.829013 │ true           │\n",
      "│ 6   │ 0.459022  │ 0.453786  │ 0.353904  │ true           │\n",
      "│ 7   │ -0.653155 │ -0.668511 │ -0.567084 │ true           │\n",
      "│ 8   │ 1.00266   │ 1.04799   │ 0.949673  │ true           │\n",
      "│ 9   │ 0.0263126 │ 0.0       │ 0.0       │ false          │\n",
      "│ 10  │ 1.07986   │ 1.06834   │ 0.977075  │ true           │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 1\n",
      "LASSO false positives = 30\n",
      "LASSO false negatives = 1\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 1\n",
      "\n",
      "current run = 26\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β    │ IHT_β     │ lasso_β   │ marginal_found │\n",
      "│     │ Float64   │ Float64   │ Float64   │ Bool           │\n",
      "├─────┼───────────┼───────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.648939  │ 0.696082  │ 0.605049  │ true           │\n",
      "│ 2   │ -0.304904 │ -0.325023 │ -0.240285 │ true           │\n",
      "│ 3   │ 0.404686  │ 0.400234  │ 0.321592  │ true           │\n",
      "│ 4   │ -0.536819 │ -0.507576 │ -0.415069 │ true           │\n",
      "│ 5   │ -0.32004  │ -0.323093 │ -0.218425 │ false          │\n",
      "│ 6   │ -1.00312  │ -1.03128  │ -0.946216 │ true           │\n",
      "│ 7   │ -0.585353 │ -0.585813 │ -0.492241 │ true           │\n",
      "│ 8   │ -0.439567 │ -0.452603 │ -0.375007 │ false          │\n",
      "│ 9   │ 1.01234   │ 1.06945   │ 0.966252  │ true           │\n",
      "│ 10  │ -0.294426 │ -0.2771   │ -0.187182 │ false          │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 0\n",
      "LASSO false positives = 41\n",
      "LASSO false negatives = 0\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 3\n",
      "\n",
      "current run = 27\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β    │ IHT_β     │ lasso_β   │ marginal_found │\n",
      "│     │ Float64   │ Float64   │ Float64   │ Bool           │\n",
      "├─────┼───────────┼───────────┼───────────┼────────────────┤\n",
      "│ 1   │ -0.710093 │ -0.695839 │ -0.586026 │ true           │\n",
      "│ 2   │ 2.44475   │ 2.50575   │ 2.3996    │ true           │\n",
      "│ 3   │ 2.3993    │ 2.41334   │ 2.32544   │ true           │\n",
      "│ 4   │ -0.56819  │ -0.570894 │ -0.484099 │ true           │\n",
      "│ 5   │ 1.44143   │ 1.40178   │ 1.304     │ true           │\n",
      "│ 6   │ 1.06554   │ 1.03576   │ 0.947561  │ true           │\n",
      "│ 7   │ 1.10821   │ 1.02366   │ 0.915232  │ true           │\n",
      "│ 8   │ -1.0537   │ -1.12777  │ -1.04664  │ true           │\n",
      "│ 9   │ -0.192688 │ -0.164833 │ -0.074494 │ false          │\n",
      "│ 10  │ 0.671208  │ 0.648915  │ 0.532099  │ false          │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 0\n",
      "LASSO false positives = 18\n",
      "LASSO false negatives = 0\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 2\n",
      "\n",
      "current run = 28\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β    │ IHT_β     │ lasso_β   │ marginal_found │\n",
      "│     │ Float64   │ Float64   │ Float64   │ Bool           │\n",
      "├─────┼───────────┼───────────┼───────────┼────────────────┤\n",
      "│ 1   │ -0.821304 │ -0.800914 │ -0.711673 │ true           │\n",
      "│ 2   │ 0.53879   │ 0.539952  │ 0.450046  │ true           │\n",
      "│ 3   │ 0.386925  │ 0.319112  │ 0.237306  │ true           │\n",
      "│ 4   │ 0.205322  │ 0.267985  │ 0.17174   │ false          │\n",
      "│ 5   │ -0.513565 │ -0.528299 │ -0.42246  │ true           │\n",
      "│ 6   │ 0.879618  │ 0.831067  │ 0.734295  │ true           │\n",
      "│ 7   │ -1.26338  │ -1.28443  │ -1.17872  │ true           │\n",
      "│ 8   │ 0.541092  │ 0.507219  │ 0.408768  │ true           │\n",
      "│ 9   │ -0.919751 │ -0.912427 │ -0.819618 │ true           │\n",
      "│ 10  │ 1.35169   │ 1.35358   │ 1.25312   │ true           │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 0\n",
      "LASSO false positives = 43\n",
      "LASSO false negatives = 0\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 1\n",
      "\n",
      "current run = 29\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β    │ IHT_β     │ lasso_β   │ marginal_found │\n",
      "│     │ Float64   │ Float64   │ Float64   │ Bool           │\n",
      "├─────┼───────────┼───────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.647175  │ 0.660088  │ 0.566585  │ true           │\n",
      "│ 2   │ -0.288778 │ -0.323801 │ -0.227701 │ false          │\n",
      "│ 3   │ -1.2357   │ -1.30154  │ -1.1908   │ true           │\n",
      "│ 4   │ -1.0022   │ -1.07398  │ -0.982937 │ true           │\n",
      "│ 5   │ 0.865008  │ 0.861816  │ 0.783354  │ true           │\n",
      "│ 6   │ 0.104902  │ 0.0       │ 0.0       │ false          │\n",
      "│ 7   │ -1.74413  │ -1.7342   │ -1.64652  │ true           │\n",
      "│ 8   │ -0.755781 │ -0.736398 │ -0.639375 │ true           │\n",
      "│ 9   │ -0.556828 │ -0.577514 │ -0.469179 │ false          │\n",
      "│ 10  │ -0.817378 │ -0.858851 │ -0.764819 │ true           │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 1\n",
      "LASSO false positives = 64\n",
      "LASSO false negatives = 1\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 3\n",
      "\n",
      "current run = 30\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β    │ IHT_β     │ lasso_β    │ marginal_found │\n",
      "│     │ Float64   │ Float64   │ Float64    │ Bool           │\n",
      "├─────┼───────────┼───────────┼────────────┼────────────────┤\n",
      "│ 1   │ 0.10572   │ 0.0       │ 0.0312442  │ false          │\n",
      "│ 2   │ -1.0601   │ -1.05338  │ -0.971861  │ true           │\n",
      "│ 3   │ 0.281754  │ 0.278822  │ 0.18539    │ false          │\n",
      "│ 4   │ 0.702759  │ 0.68602   │ 0.594949   │ true           │\n",
      "│ 5   │ 0.678062  │ 0.666578  │ 0.569036   │ true           │\n",
      "│ 6   │ 0.444067  │ 0.48826   │ 0.390041   │ true           │\n",
      "│ 7   │ -0.32708  │ -0.299329 │ -0.200932  │ false          │\n",
      "│ 8   │ -0.151695 │ 0.0       │ -0.0254815 │ false          │\n",
      "│ 9   │ -0.867023 │ -0.945166 │ -0.84528   │ true           │\n",
      "│ 10  │ 1.57627   │ 1.51016   │ 1.41253    │ true           │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 2\n",
      "LASSO false positives = 44\n",
      "LASSO false negatives = 0\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 4\n",
      "\n",
      "current run = 31\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β     │ IHT_β    │ lasso_β    │ marginal_found │\n",
      "│     │ Float64    │ Float64  │ Float64    │ Bool           │\n",
      "├─────┼────────────┼──────────┼────────────┼────────────────┤\n",
      "│ 1   │ 0.716913   │ 0.764058 │ 0.648225   │ true           │\n",
      "│ 2   │ -0.129531  │ 0.0      │ -0.0462017 │ false          │\n",
      "│ 3   │ 0.0684458  │ 0.0      │ 0.0        │ false          │\n",
      "│ 4   │ 0.012461   │ 0.0      │ 0.0        │ false          │\n",
      "│ 5   │ -1.25199   │ -1.30375 │ -1.18498   │ true           │\n",
      "│ 6   │ 0.468408   │ 0.471502 │ 0.358728   │ false          │\n",
      "│ 7   │ -2.15673   │ -2.13804 │ -2.03593   │ true           │\n",
      "│ 8   │ 0.812352   │ 0.807568 │ 0.696936   │ true           │\n",
      "│ 9   │ 0.492849   │ 0.517933 │ 0.410205   │ true           │\n",
      "│ 10  │ -0.0782995 │ 0.0      │ 0.0        │ false          │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 4\n",
      "LASSO false positives = 14\n",
      "LASSO false negatives = 3\n",
      "marginal false positives = 1\n",
      "marginal false negatives = 5\n",
      "\n",
      "current run = 32\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β    │ IHT_β     │ lasso_β    │ marginal_found │\n",
      "│     │ Float64   │ Float64   │ Float64    │ Bool           │\n",
      "├─────┼───────────┼───────────┼────────────┼────────────────┤\n",
      "│ 1   │ -0.131514 │ 0.0       │ -0.0208083 │ false          │\n",
      "│ 2   │ -0.239057 │ -0.254197 │ -0.133168  │ false          │\n",
      "│ 3   │ 0.739708  │ 0.728436  │ 0.625183   │ true           │\n",
      "│ 4   │ -0.40069  │ -0.435349 │ -0.321515  │ false          │\n",
      "│ 5   │ -0.206226 │ -0.222187 │ -0.133734  │ false          │\n",
      "│ 6   │ -1.06698  │ -1.09773  │ -0.998224  │ true           │\n",
      "│ 7   │ -0.527112 │ -0.582044 │ -0.47763   │ true           │\n",
      "│ 8   │ -1.9998   │ -2.01714  │ -1.91913   │ true           │\n",
      "│ 9   │ -0.793674 │ -0.7929   │ -0.69089   │ true           │\n",
      "│ 10  │ -2.087    │ -2.08219  │ -1.97146   │ true           │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 1\n",
      "LASSO false positives = 19\n",
      "LASSO false negatives = 0\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 4\n",
      "\n",
      "current run = 33\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β     │ IHT_β     │ lasso_β   │ marginal_found │\n",
      "│     │ Float64    │ Float64   │ Float64   │ Bool           │\n",
      "├─────┼────────────┼───────────┼───────────┼────────────────┤\n",
      "│ 1   │ -0.833228  │ -0.832405 │ -0.724359 │ true           │\n",
      "│ 2   │ 0.676049   │ 0.660355  │ 0.548099  │ true           │\n",
      "│ 3   │ -0.0524694 │ 0.0       │ 0.0       │ false          │\n",
      "│ 4   │ 1.21151    │ 1.20423   │ 1.09371   │ true           │\n",
      "│ 5   │ 0.0947078  │ 0.0       │ 0.0144032 │ false          │\n",
      "│ 6   │ 1.50923    │ 1.53035   │ 1.41428   │ true           │\n",
      "│ 7   │ -0.332496  │ -0.380072 │ -0.277982 │ false          │\n",
      "│ 8   │ 0.129556   │ 0.0       │ 0.0442901 │ false          │\n",
      "│ 9   │ 0.0771111  │ 0.0       │ 0.0       │ false          │\n",
      "│ 10  │ -1.36522   │ -1.36449  │ -1.2472   │ true           │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 4\n",
      "LASSO false positives = 19\n",
      "LASSO false negatives = 2\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 5\n",
      "\n",
      "current run = 34\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β     │ IHT_β     │ lasso_β   │ marginal_found │\n",
      "│     │ Float64    │ Float64   │ Float64   │ Bool           │\n",
      "├─────┼────────────┼───────────┼───────────┼────────────────┤\n",
      "│ 1   │ -1.38126   │ -1.35298  │ -1.25364  │ true           │\n",
      "│ 2   │ -0.840492  │ -0.871085 │ -0.770694 │ true           │\n",
      "│ 3   │ 0.438593   │ 0.452572  │ 0.35577   │ true           │\n",
      "│ 4   │ -1.03938   │ -1.05631  │ -0.939835 │ true           │\n",
      "│ 5   │ -0.0858827 │ 0.0       │ 0.0       │ false          │\n",
      "│ 6   │ 0.626138   │ 0.581691  │ 0.481834  │ true           │\n",
      "│ 7   │ -0.896747  │ -0.922546 │ -0.818232 │ true           │\n",
      "│ 8   │ 0.285343   │ 0.28179   │ 0.178834  │ false          │\n",
      "│ 9   │ 0.350203   │ 0.399617  │ 0.29469   │ false          │\n",
      "│ 10  │ 1.12765    │ 1.18863   │ 1.09319   │ true           │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 1\n",
      "LASSO false positives = 20\n",
      "LASSO false negatives = 1\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 3\n",
      "\n",
      "current run = 35\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β    │ IHT_β     │ lasso_β    │ marginal_found │\n",
      "│     │ Float64   │ Float64   │ Float64    │ Bool           │\n",
      "├─────┼───────────┼───────────┼────────────┼────────────────┤\n",
      "│ 1   │ 0.402152  │ 0.415812  │ 0.307026   │ false          │\n",
      "│ 2   │ -0.114413 │ 0.0       │ -0.0155359 │ false          │\n",
      "│ 3   │ -1.07648  │ -1.0901   │ -0.961862  │ true           │\n",
      "│ 4   │ -0.202484 │ -0.216681 │ -0.0999947 │ false          │\n",
      "│ 5   │ 0.245664  │ 0.240545  │ 0.132115   │ false          │\n",
      "│ 6   │ 0.345895  │ 0.325211  │ 0.232508   │ true           │\n",
      "│ 7   │ 1.54896   │ 1.58377   │ 1.46869    │ true           │\n",
      "│ 8   │ 0.601544  │ 0.552843  │ 0.450195   │ true           │\n",
      "│ 9   │ 0.755934  │ 0.79992   │ 0.697685   │ true           │\n",
      "│ 10  │ -0.450207 │ -0.438255 │ -0.341649  │ true           │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 1\n",
      "LASSO false positives = 24\n",
      "LASSO false negatives = 0\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 4\n",
      "\n",
      "current run = 36\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β     │ IHT_β     │ lasso_β   │ marginal_found │\n",
      "│     │ Float64    │ Float64   │ Float64   │ Bool           │\n",
      "├─────┼────────────┼───────────┼───────────┼────────────────┤\n",
      "│ 1   │ -1.07994   │ -1.08149  │ -0.994434 │ true           │\n",
      "│ 2   │ 1.13449    │ 1.19305   │ 1.10661   │ true           │\n",
      "│ 3   │ -0.830169  │ -0.823331 │ -0.730471 │ true           │\n",
      "│ 4   │ -0.498433  │ -0.455547 │ -0.360449 │ true           │\n",
      "│ 5   │ 0.747752   │ 0.726894  │ 0.632679  │ true           │\n",
      "│ 6   │ 0.993698   │ 0.966205  │ 0.879467  │ true           │\n",
      "│ 7   │ 1.02778    │ 1.02856   │ 0.940831  │ true           │\n",
      "│ 8   │ -0.0267846 │ 0.0       │ 0.0       │ false          │\n",
      "│ 9   │ 0.373882   │ 0.412793  │ 0.319527  │ true           │\n",
      "│ 10  │ 0.112726   │ 0.0       │ 0.0584978 │ false          │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 2\n",
      "LASSO false positives = 38\n",
      "LASSO false negatives = 1\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 2\n",
      "\n",
      "current run = 37\n",
      "\u001b[31mDid not converge!!!!! The run time for IHT was 14.027657985687256 seconds and model size was9\u001b[39m\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β    │ IHT_β     │ lasso_β   │ marginal_found │\n",
      "│     │ Float64   │ Float64   │ Float64   │ Bool           │\n",
      "├─────┼───────────┼───────────┼───────────┼────────────────┤\n",
      "│ 1   │ -1.15965  │ -1.1887   │ -1.09724  │ true           │\n",
      "│ 2   │ 2.23898   │ 2.63856   │ 2.11803   │ false          │\n",
      "│ 3   │ 0.0114716 │ 0.0       │ 0.0       │ false          │\n",
      "│ 4   │ 1.50934   │ 1.52057   │ 1.42916   │ true           │\n",
      "│ 5   │ 1.35203   │ 1.35865   │ 1.24999   │ true           │\n",
      "│ 6   │ -0.236083 │ -0.272139 │ -0.179025 │ false          │\n",
      "│ 7   │ -1.43073  │ -1.41551  │ -1.31865  │ true           │\n",
      "│ 8   │ 0.216307  │ 0.221785  │ 0.122856  │ false          │\n",
      "│ 9   │ 0.340348  │ 0.281738  │ 0.187361  │ false          │\n",
      "│ 10  │ 0.366631  │ 0.418762  │ 0.320466  │ false          │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 1\n",
      "LASSO false positives = 9\n",
      "LASSO false negatives = 1\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 6\n",
      "\n",
      "current run = 38\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β    │ IHT_β     │ lasso_β   │ marginal_found │\n",
      "│     │ Float64   │ Float64   │ Float64   │ Bool           │\n",
      "├─────┼───────────┼───────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.978836  │ 1.00652   │ 0.903545  │ true           │\n",
      "│ 2   │ 0.550849  │ 0.565556  │ 0.45086   │ true           │\n",
      "│ 3   │ -0.307564 │ -0.332191 │ -0.216066 │ false          │\n",
      "│ 4   │ 1.60565   │ 1.6283    │ 1.54464   │ true           │\n",
      "│ 5   │ 0.546421  │ 0.507149  │ 0.421066  │ true           │\n",
      "│ 6   │ -1.22867  │ -1.20582  │ -1.1105   │ true           │\n",
      "│ 7   │ 1.38029   │ 1.35707   │ 1.27499   │ true           │\n",
      "│ 8   │ -0.317914 │ -0.343961 │ -0.262911 │ true           │\n",
      "│ 9   │ 0.276958  │ 0.229415  │ 0.145557  │ false          │\n",
      "│ 10  │ 0.229313  │ 0.285963  │ 0.157963  │ false          │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 0\n",
      "LASSO false positives = 35\n",
      "LASSO false negatives = 0\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 3\n",
      "\n",
      "current run = 39\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β    │ IHT_β     │ lasso_β   │ marginal_found │\n",
      "│     │ Float64   │ Float64   │ Float64   │ Bool           │\n",
      "├─────┼───────────┼───────────┼───────────┼────────────────┤\n",
      "│ 1   │ -0.128251 │ 0.0       │ 0.0       │ false          │\n",
      "│ 2   │ 0.919801  │ 0.903501  │ 0.797372  │ true           │\n",
      "│ 3   │ -1.28463  │ -1.27504  │ -1.16129  │ true           │\n",
      "│ 4   │ 0.101823  │ 0.0       │ 0.0270947 │ false          │\n",
      "│ 5   │ 1.71113   │ 1.74653   │ 1.62265   │ true           │\n",
      "│ 6   │ 0.177542  │ 0.227357  │ 0.100073  │ false          │\n",
      "│ 7   │ -2.13612  │ -2.15925  │ -2.07031  │ true           │\n",
      "│ 8   │ -1.4868   │ -1.47657  │ -1.36174  │ true           │\n",
      "│ 9   │ -0.418656 │ -0.420799 │ -0.302736 │ false          │\n",
      "│ 10  │ -0.853007 │ -0.83551  │ -0.735002 │ true           │\n",
      "IHT false positives = 1\n",
      "IHT false negatives = 2\n",
      "LASSO false positives = 6\n",
      "LASSO false negatives = 1\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 4\n",
      "\n",
      "current run = 40\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β    │ IHT_β     │ lasso_β   │ marginal_found │\n",
      "│     │ Float64   │ Float64   │ Float64   │ Bool           │\n",
      "├─────┼───────────┼───────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.497199  │ 0.429229  │ 0.317647  │ true           │\n",
      "│ 2   │ -2.38356  │ -2.47021  │ -2.3706   │ true           │\n",
      "│ 3   │ -0.228254 │ -0.243433 │ -0.141708 │ false          │\n",
      "│ 4   │ 0.294025  │ 0.251284  │ 0.129493  │ false          │\n",
      "│ 5   │ 0.302579  │ 0.267993  │ 0.145058  │ false          │\n",
      "│ 6   │ 0.525926  │ 0.538935  │ 0.445087  │ true           │\n",
      "│ 7   │ 0.0246983 │ 0.0       │ 0.0       │ false          │\n",
      "│ 8   │ 1.37478   │ 1.41021   │ 1.29876   │ true           │\n",
      "│ 9   │ -0.678537 │ -0.667006 │ -0.556766 │ true           │\n",
      "│ 10  │ 0.0118296 │ 0.0       │ 0.0       │ false          │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 2\n",
      "LASSO false positives = 14\n",
      "LASSO false negatives = 2\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 5\n",
      "\n",
      "current run = 41\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β     │ IHT_β     │ lasso_β   │ marginal_found │\n",
      "│     │ Float64    │ Float64   │ Float64   │ Bool           │\n",
      "├─────┼────────────┼───────────┼───────────┼────────────────┤\n",
      "│ 1   │ -0.762556  │ -0.760873 │ -0.635671 │ true           │\n",
      "│ 2   │ -1.37821   │ -1.3826   │ -1.25707  │ true           │\n",
      "│ 3   │ 1.37773    │ 1.378     │ 1.26193   │ true           │\n",
      "│ 4   │ -0.0696107 │ 0.0       │ 0.0       │ false          │\n",
      "│ 5   │ 0.116563   │ 0.0       │ 0.0       │ false          │\n",
      "│ 6   │ -0.989843  │ -1.01507  │ -0.897106 │ true           │\n",
      "│ 7   │ -0.564463  │ -0.601766 │ -0.498344 │ true           │\n",
      "│ 8   │ -0.334853  │ -0.322917 │ -0.214101 │ false          │\n",
      "│ 9   │ 0.548853   │ 0.57052   │ 0.455885  │ true           │\n",
      "│ 10  │ -0.707457  │ -0.727106 │ -0.608103 │ true           │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 2\n",
      "LASSO false positives = 6\n",
      "LASSO false negatives = 2\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 3\n",
      "\n",
      "current run = 42\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β    │ IHT_β     │ lasso_β    │ marginal_found │\n",
      "│     │ Float64   │ Float64   │ Float64    │ Bool           │\n",
      "├─────┼───────────┼───────────┼────────────┼────────────────┤\n",
      "│ 1   │ 1.64416   │ 1.6763    │ 1.5671     │ true           │\n",
      "│ 2   │ -0.984903 │ -0.998582 │ -0.901849  │ true           │\n",
      "│ 3   │ -2.06212  │ -2.12263  │ -2.03538   │ true           │\n",
      "│ 4   │ -0.100535 │ 0.0       │ -0.0253309 │ false          │\n",
      "│ 5   │ 0.467878  │ 0.451508  │ 0.343717   │ false          │\n",
      "│ 6   │ 0.247928  │ 0.263626  │ 0.161845   │ false          │\n",
      "│ 7   │ 0.501326  │ 0.486349  │ 0.38658    │ false          │\n",
      "│ 8   │ 1.6264    │ 1.61545   │ 1.51164    │ true           │\n",
      "│ 9   │ -1.29208  │ -1.35658  │ -1.24858   │ true           │\n",
      "│ 10  │ -0.112418 │ -0.182911 │ -0.0704007 │ false          │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 1\n",
      "LASSO false positives = 23\n",
      "LASSO false negatives = 0\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 5\n",
      "\n",
      "current run = 43\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β    │ IHT_β     │ lasso_β   │ marginal_found │\n",
      "│     │ Float64   │ Float64   │ Float64   │ Bool           │\n",
      "├─────┼───────────┼───────────┼───────────┼────────────────┤\n",
      "│ 1   │ -0.773332 │ -0.771208 │ -0.67552  │ true           │\n",
      "│ 2   │ 0.665313  │ 0.634218  │ 0.527532  │ true           │\n",
      "│ 3   │ 0.293619  │ 0.29056   │ 0.198352  │ false          │\n",
      "│ 4   │ -1.86182  │ -1.86145  │ -1.75971  │ true           │\n",
      "│ 5   │ -1.30875  │ -1.28344  │ -1.17243  │ true           │\n",
      "│ 6   │ 0.64743   │ 0.646705  │ 0.536906  │ true           │\n",
      "│ 7   │ 0.066356  │ 0.0       │ 0.0       │ false          │\n",
      "│ 8   │ 0.381098  │ 0.446398  │ 0.343011  │ false          │\n",
      "│ 9   │ 1.08931   │ 1.15544   │ 1.0555    │ true           │\n",
      "│ 10  │ 0.124671  │ 0.0       │ 0.0544846 │ false          │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 2\n",
      "LASSO false positives = 12\n",
      "LASSO false negatives = 1\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 4\n",
      "\n",
      "current run = 44\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β    │ IHT_β     │ lasso_β   │ marginal_found │\n",
      "│     │ Float64   │ Float64   │ Float64   │ Bool           │\n",
      "├─────┼───────────┼───────────┼───────────┼────────────────┤\n",
      "│ 1   │ -0.566488 │ -0.544225 │ -0.445223 │ true           │\n",
      "│ 2   │ -0.592059 │ -0.58848  │ -0.505459 │ true           │\n",
      "│ 3   │ 0.358378  │ 0.379552  │ 0.286032  │ true           │\n",
      "│ 4   │ -0.177563 │ -0.270488 │ -0.188196 │ false          │\n",
      "│ 5   │ 1.32632   │ 1.33904   │ 1.25273   │ true           │\n",
      "│ 6   │ 1.44183   │ 1.41294   │ 1.34083   │ true           │\n",
      "│ 7   │ -0.389083 │ -0.354141 │ -0.258456 │ false          │\n",
      "│ 8   │ 0.527468  │ 0.471261  │ 0.393624  │ true           │\n",
      "│ 9   │ 1.16854   │ 1.14464   │ 1.05956   │ true           │\n",
      "│ 10  │ 1.03984   │ 1.05942   │ 0.956351  │ true           │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 0\n",
      "LASSO false positives = 38\n",
      "LASSO false negatives = 0\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 2\n",
      "\n",
      "current run = 45\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β    │ IHT_β     │ lasso_β   │ marginal_found │\n",
      "│     │ Float64   │ Float64   │ Float64   │ Bool           │\n",
      "├─────┼───────────┼───────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.574367  │ 0.558226  │ 0.453866  │ true           │\n",
      "│ 2   │ 1.39576   │ 1.36906   │ 1.25429   │ true           │\n",
      "│ 3   │ -0.791119 │ -0.786836 │ -0.681104 │ true           │\n",
      "│ 4   │ -0.708916 │ -0.721463 │ -0.622731 │ true           │\n",
      "│ 5   │ 0.0573325 │ 0.0       │ 0.0       │ false          │\n",
      "│ 6   │ 0.948541  │ 0.936847  │ 0.834434  │ false          │\n",
      "│ 7   │ 0.89135   │ 0.912335  │ 0.804562  │ true           │\n",
      "│ 8   │ 1.36034   │ 1.32073   │ 1.21231   │ true           │\n",
      "│ 9   │ -0.708206 │ -0.684663 │ -0.601392 │ true           │\n",
      "│ 10  │ -0.574003 │ -0.571642 │ -0.479813 │ true           │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 1\n",
      "LASSO false positives = 37\n",
      "LASSO false negatives = 1\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 2\n",
      "\n",
      "current run = 46\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β    │ IHT_β     │ lasso_β   │ marginal_found │\n",
      "│     │ Float64   │ Float64   │ Float64   │ Bool           │\n",
      "├─────┼───────────┼───────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.0579335 │ 0.0       │ 0.0       │ false          │\n",
      "│ 2   │ -0.307419 │ -0.221732 │ -0.127744 │ false          │\n",
      "│ 3   │ -0.11986  │ 0.0       │ 0.0       │ false          │\n",
      "│ 4   │ 1.318     │ 1.32141   │ 1.22079   │ true           │\n",
      "│ 5   │ 1.1655    │ 1.13696   │ 1.03754   │ true           │\n",
      "│ 6   │ 0.308425  │ 0.290792  │ 0.195504  │ false          │\n",
      "│ 7   │ 0.656646  │ 0.569714  │ 0.470835  │ true           │\n",
      "│ 8   │ 0.1531    │ 0.184554  │ 0.0813537 │ false          │\n",
      "│ 9   │ 0.34867   │ 0.384834  │ 0.288042  │ true           │\n",
      "│ 10  │ 1.30663   │ 1.27056   │ 1.15876   │ true           │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 2\n",
      "LASSO false positives = 16\n",
      "LASSO false negatives = 2\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 5\n",
      "\n",
      "current run = 47\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β     │ IHT_β     │ lasso_β   │ marginal_found │\n",
      "│     │ Float64    │ Float64   │ Float64   │ Bool           │\n",
      "├─────┼────────────┼───────────┼───────────┼────────────────┤\n",
      "│ 1   │ 1.65733    │ 1.64265   │ 1.54732   │ true           │\n",
      "│ 2   │ 0.893532   │ 0.898929  │ 0.798805  │ true           │\n",
      "│ 3   │ -0.838084  │ -0.840385 │ -0.727741 │ true           │\n",
      "│ 4   │ 0.213567   │ 0.201273  │ 0.0957702 │ false          │\n",
      "│ 5   │ 0.628961   │ 0.643756  │ 0.54346   │ true           │\n",
      "│ 6   │ -0.0400687 │ 0.0       │ 0.0       │ false          │\n",
      "│ 7   │ 0.360368   │ 0.372952  │ 0.267099  │ false          │\n",
      "│ 8   │ -1.29039   │ -1.27235  │ -1.16364  │ true           │\n",
      "│ 9   │ -1.5451    │ -1.55029  │ -1.44091  │ true           │\n",
      "│ 10  │ 0.794447   │ 0.763424  │ 0.655904  │ true           │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 1\n",
      "LASSO false positives = 21\n",
      "LASSO false negatives = 1\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 3\n",
      "\n",
      "current run = 48\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β    │ IHT_β     │ lasso_β   │ marginal_found │\n",
      "│     │ Float64   │ Float64   │ Float64   │ Bool           │\n",
      "├─────┼───────────┼───────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.23845   │ 0.187939  │ 0.0848846 │ false          │\n",
      "│ 2   │ -1.58552  │ -1.57008  │ -1.47342  │ true           │\n",
      "│ 3   │ -0.923084 │ -0.907151 │ -0.811096 │ true           │\n",
      "│ 4   │ 0.563645  │ 0.585465  │ 0.474382  │ true           │\n",
      "│ 5   │ 0.480607  │ 0.438691  │ 0.334303  │ false          │\n",
      "│ 6   │ -0.671179 │ -0.669614 │ -0.564642 │ true           │\n",
      "│ 7   │ 0.331685  │ 0.298315  │ 0.191323  │ false          │\n",
      "│ 8   │ 1.07535   │ 1.08422   │ 0.992625  │ true           │\n",
      "│ 9   │ -1.03137  │ -1.03815  │ -0.934388 │ true           │\n",
      "│ 10  │ 1.26969   │ 1.28335   │ 1.20572   │ true           │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 0\n",
      "LASSO false positives = 23\n",
      "LASSO false negatives = 0\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 3\n",
      "\n",
      "current run = 49\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β    │ IHT_β     │ lasso_β   │ marginal_found │\n",
      "│     │ Float64   │ Float64   │ Float64   │ Bool           │\n",
      "├─────┼───────────┼───────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.407838  │ 0.387648  │ 0.295258  │ true           │\n",
      "│ 2   │ -1.99473  │ -1.97399  │ -1.87425  │ true           │\n",
      "│ 3   │ -1.78315  │ -1.75527  │ -1.64783  │ true           │\n",
      "│ 4   │ 0.326483  │ 0.34165   │ 0.260201  │ false          │\n",
      "│ 5   │ 0.358946  │ 0.348451  │ 0.255663  │ false          │\n",
      "│ 6   │ -0.177961 │ 0.0       │ -0.102792 │ false          │\n",
      "│ 7   │ -0.385924 │ -0.324594 │ -0.21807  │ false          │\n",
      "│ 8   │ 0.490134  │ 0.460413  │ 0.383079  │ true           │\n",
      "│ 9   │ 1.66554   │ 1.67046   │ 1.55264   │ true           │\n",
      "│ 10  │ 1.31769   │ 1.3236    │ 1.22505   │ true           │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 1\n",
      "LASSO false positives = 37\n",
      "LASSO false negatives = 0\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 4\n",
      "\n",
      "current run = 50\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β      │ IHT_β     │ lasso_β   │ marginal_found │\n",
      "│     │ Float64     │ Float64   │ Float64   │ Bool           │\n",
      "├─────┼─────────────┼───────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.0671508   │ 0.0       │ 0.0       │ false          │\n",
      "│ 2   │ -2.24678    │ -2.24029  │ -2.13896  │ true           │\n",
      "│ 3   │ -0.619753   │ -0.668244 │ -0.562582 │ true           │\n",
      "│ 4   │ 0.697065    │ 0.686334  │ 0.57232   │ true           │\n",
      "│ 5   │ -0.00615383 │ 0.0       │ 0.0       │ false          │\n",
      "│ 6   │ -1.6053     │ -1.62491  │ -1.51535  │ true           │\n",
      "│ 7   │ 1.02524     │ 1.00996   │ 0.899862  │ true           │\n",
      "│ 8   │ 0.429733    │ 0.416425  │ 0.310121  │ false          │\n",
      "│ 9   │ -0.628775   │ -0.649596 │ -0.543887 │ true           │\n",
      "│ 10  │ -0.629107   │ -0.646053 │ -0.53442  │ true           │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 2\n",
      "LASSO false positives = 5\n",
      "LASSO false negatives = 2\n",
      "marginal false positives = 1\n",
      "marginal false negatives = 3\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "([0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0  …  0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0], [1.0, 1.0, 2.0, 3.0, 3.0, 2.0, 1.0, 1.0, 0.0, 0.0  …  2.0, 1.0, 2.0, 0.0, 1.0, 2.0, 1.0, 0.0, 1.0, 2.0], [26.0, 14.0, 41.0, 16.0, 45.0, 13.0, 21.0, 19.0, 36.0, 26.0  …  6.0, 23.0, 12.0, 38.0, 37.0, 16.0, 21.0, 23.0, 37.0, 5.0], [1.0, 1.0, 1.0, 2.0, 3.0, 1.0, 1.0, 1.0, 0.0, 0.0  …  2.0, 0.0, 1.0, 0.0, 1.0, 2.0, 1.0, 0.0, 0.0, 2.0], [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0  …  0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0], [10.0, 2.0, 10.0, 5.0, 5.0, 5.0, 1.0, 2.0, 5.0, 3.0  …  3.0, 5.0, 4.0, 2.0, 2.0, 5.0, 3.0, 3.0, 4.0, 3.0])"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "iht_false_positives, iht_false_negatives, lasso_false_positives, \n",
    "    lasso_false_negatives, marginal_false_positives, marginal_false_negatives = run()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true,
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "6-element Array{Float64,1}:\n",
       "  0.02\n",
       "  1.38\n",
       " 25.42\n",
       "  0.94\n",
       "  0.06\n",
       "  4.02"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "normal_iht_false_positives = sum(iht_false_positives) / 50\n",
    "normal_iht_false_negatives = sum(iht_false_negatives) / 50\n",
    "normal_lasso_false_positives = sum(lasso_false_positives) / 50\n",
    "normal_lasso_false_negatives = sum(lasso_false_negatives) / 50\n",
    "normal_marginal_false_positives = sum(marginal_false_positives) / 50\n",
    "normal_marginal_false_negatives = sum(marginal_false_negatives) / 50\n",
    "\n",
    "result = [normal_iht_false_positives; normal_iht_false_negatives; normal_lasso_false_positives; \n",
    "    normal_lasso_false_negatives; normal_marginal_false_positives; normal_marginal_false_negatives]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Julia 1.0.3",
   "language": "julia",
   "name": "julia-1.0"
  },
  "language_info": {
   "file_extension": ".jl",
   "mimetype": "application/julia",
   "name": "julia",
   "version": "1.0.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
