{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Normal: IHT vs LASSO vs Marginal regression: comparing false positive/negatives\n",
    "\n",
    "Marginal regression is the most prevalent method used to analyze GWAS data. LASSO is currently the de facto penalized least squares method for [feature selection](https://en.wikipedia.org/wiki/Feature_selection). Here we compare the performance (in terms of the number of false positives/negative) of LASSO and marginal regression with IHT for GWAS data. We use the `glmnet` implementation of cyclic coordinate descent for LASSO, and `MendelGWAS.jl` module of [OpenMendel](https://openmendel.github.io/MendelGWAS.jl/) for marginal analysis. Since the focus here is not scalability, we test our sample problems on moderately sized genotype matrces of 1000 samples with 10,000 SNPs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "31"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "using Distributed\n",
    "addprocs(30)\n",
    "nprocs()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "┌ Info: Loading DataFrames support into Gadfly.jl\n",
      "└ @ Gadfly /u/home/b/biona001/.julia/packages/Gadfly/09PWZ/src/mapping.jl:228\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall /u/home/b/biona001/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: Loading required package: Matrix\n",
      "│ Loading required package: foreach\n",
      "│ Loaded glmnet 2.0-16\n",
      "│ \n",
      "└ @ RCall /u/home/b/biona001/.julia/packages/RCall/ffM0W/src/io.jl:113\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "RObject{StrSxp}\n",
       " [1] \"glmnet\"    \"foreach\"   \"Matrix\"    \"stats\"     \"graphics\"  \"grDevices\"\n",
       " [7] \"utils\"     \"datasets\"  \"methods\"   \"base\"     \n"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "using MendelIHT\n",
    "using SnpArrays\n",
    "using DataFrames\n",
    "using Distributions\n",
    "using Random\n",
    "using LinearAlgebra\n",
    "using DelimitedFiles\n",
    "using GLM\n",
    "using MendelGWAS\n",
    "using CSV\n",
    "using RCall\n",
    "R\"library(glmnet)\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Normal response"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "iht_lasso_marginal (generic function with 1 method)"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "function iht_lasso_marginal(n::Int64, p::Int64, d::UnionAll, l::Link)\n",
    "    #construct snpmatrix, covariate files, and true model b\n",
    "    x, = simulate_random_snparray(n, p, \"tmp.bed\")\n",
    "    xbm = SnpBitMatrix{Float64}(x, model=ADDITIVE_MODEL, center=true, scale=true); \n",
    "    z = ones(n, 1) # the intercept\n",
    "    x_float = [convert(Matrix{Float64}, x, center=true, scale=true) z] #Float64 version of x\n",
    "\n",
    "    # simulate response, true model b, and the correct non-0 positions of b\n",
    "    true_b = zeros(p)\n",
    "    true_b[1:10] .= collect(0.1:0.1:1.0)\n",
    "    k = 10\n",
    "    shuffle!(true_b)\n",
    "    correct_position = findall(x -> x != 0, true_b)\n",
    "\n",
    "    #simulate phenotypes (e.g. vector y)\n",
    "    if d == Normal || d == Poisson || d == Bernoulli\n",
    "        prob = linkinv.(l, xbm * true_b)\n",
    "        clamp!(prob, -20, 20)\n",
    "        y = [rand(d(i)) for i in prob]\n",
    "    elseif d == NegativeBinomial\n",
    "        μ = linkinv.(l, xbm * true_b)\n",
    "        prob = 1 ./ (1 .+ μ ./ nn)\n",
    "        y = [rand(d(nn, i)) for i in prob] #number of failtures before nn success occurs\n",
    "    elseif d == Gamma\n",
    "        μ = linkinv.(l, xbm * true_b)\n",
    "        β = 1 ./ μ # here β is the rate parameter for gamma distribution\n",
    "        y = [rand(d(α, i)) for i in β] # α is the shape parameter for gamma\n",
    "    end\n",
    "    y = Float64.(y)\n",
    "\n",
    "    #specify path and folds\n",
    "    num_folds = 3\n",
    "    folds = rand(1:num_folds, size(x, 1));\n",
    "\n",
    "    #run glmnet via Rcall\n",
    "    @rput x_float y folds num_folds #make variables visible to R\n",
    "    R\"lasso_cv_result = cv.glmnet(x_float, y, nfolds = num_folds, foldid = folds, family='gaussian')\"\n",
    "    R\"lasso_beta_tmp = glmnet(x_float, y, lambda=lasso_cv_result$lambda.min, family='gaussian')$beta\"\n",
    "    R\"lasso_beta = as.vector(lasso_beta_tmp)\"\n",
    "    @rget lasso_cv_result lasso_beta #pull result from R to Julia\n",
    "    lasso_k_est = count(!iszero, lasso_beta)\n",
    "    \n",
    "    #find non-zero entries returned by best lasso model as largest k estimate\n",
    "    path = collect(1:50);\n",
    "    \n",
    "    #run IHT's cross validation routine \n",
    "    mses = cv_iht(d(), l, x, z, y, 1, path, num_folds, folds=folds, use_maf=false, debias=false, showinfo=false, parallel=true);\n",
    "    iht_k_est = argmin(mses)\n",
    "    iht_result = L0_reg(x, xbm, z, y, 1, iht_k_est, d(), l, debias=false, init=false, use_maf=false)\n",
    "    iht_beta = iht_result.beta\n",
    "    \n",
    "    #Now run MendelGWAS\n",
    "    make_bim_fam_files(x, y, \"tmp\") #create .bim and .bed files for MendelGWAS\n",
    "\n",
    "    #create a control file for MendelGWAS\n",
    "    open(\"tmp_control.txt\", \"w\") do f\n",
    "        write(f, \"plink_input_basename = tmp \\n\")\n",
    "        write(f, \"plink_field_separator = '\\t' \\n\")\n",
    "        write(f, \"output_table = tmp_table.txt \\n\")\n",
    "        write(f, \"regression = linear \\n\")\n",
    "        write(f, \"regression_formula = Trait ~ \\n\")\n",
    "    end\n",
    "\n",
    "    #run marginal analysis\n",
    "    GWAS(\"tmp_control.txt\");\n",
    "\n",
    "    # calculate false positive/negatives based on bonferroni correction\n",
    "    p_values = CSV.read(\"tmp_table.txt\", delim=',', header=true)[:Pvalue]\n",
    "    significance = 0.05 / Float64(p)\n",
    "    passing_snps_position = findall(p_values .<= significance)\n",
    "    marginal_found = [correct_position[snp] in passing_snps_position for snp in 1:k]\n",
    "\n",
    "    #clean up\n",
    "    rm(\"tmp.bed\", force=true)\n",
    "    rm(\"tmp.bim\", force=true)\n",
    "    rm(\"tmp.fam\", force=true)\n",
    "    rm(\"tmp_table.txt\", force=true)\n",
    "    rm(\"tmp_control.txt\", force=true)\n",
    "    rm(\"Mendel_Output.txt\", force=true)\n",
    "\n",
    "    #show lasso and IHT's reconstruction result\n",
    "    compare_model = DataFrame(\n",
    "        true_β  = true_b[correct_position], \n",
    "        IHT_β   = iht_beta[correct_position],\n",
    "        lasso_β = lasso_beta[correct_position],\n",
    "        marginal_found = marginal_found)\n",
    "    @show compare_model\n",
    "    \n",
    "    #compute true/false positives/negatives for IHT and lasso\n",
    "    iht_tp = count(!iszero, iht_beta[correct_position])\n",
    "    iht_fp = iht_k_est - iht_tp\n",
    "    iht_fn = k - iht_tp\n",
    "    lasso_tp = count(!iszero, lasso_beta[correct_position])\n",
    "    lasso_fp = lasso_k_est - lasso_tp\n",
    "    lasso_fn = k - lasso_tp\n",
    "    marginal_tp = count(!iszero, true_b[passing_snps_position])\n",
    "    marginal_fp = length(passing_snps_position) - marginal_tp\n",
    "    marginal_fn = k - marginal_tp\n",
    "    \n",
    "    println(\"IHT true positives = $iht_tp\")\n",
    "    println(\"IHT false positives = $iht_fp\")\n",
    "    println(\"IHT false negatives = $iht_fn\")\n",
    "    println(\"LASSO true positives = $lasso_tp\")\n",
    "    println(\"LASSO false positives = $lasso_fp\")\n",
    "    println(\"LASSO false negatives = $lasso_fn\")\n",
    "    println(\"marginal true positives = $marginal_tp\")\n",
    "    println(\"marginal false positives = $marginal_fp\")\n",
    "    println(\"marginal false negatives = $marginal_fn\" * \"\\n\")\n",
    "    \n",
    "    return iht_tp, iht_fp, iht_fn, lasso_tp, lasso_fp, lasso_fn, marginal_tp, marginal_fp, marginal_fn\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "run (generic function with 1 method)"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "function run()\n",
    "    #simulat data with k true predictors, from distribution d and with link l.\n",
    "    n = 1000\n",
    "    p = 10000\n",
    "    d = Normal\n",
    "    l = canonicallink(d())\n",
    "\n",
    "    #run function above, saving results in 4 vectors\n",
    "    total_runs = 50\n",
    "    iht_true_positives = zeros(total_runs)\n",
    "    iht_false_positives = zeros(total_runs)\n",
    "    iht_false_negatives = zeros(total_runs)\n",
    "    lasso_true_positives = zeros(total_runs)\n",
    "    lasso_false_positives = zeros(total_runs)\n",
    "    lasso_false_negatives = zeros(total_runs)\n",
    "    marginal_true_positives = zeros(total_runs)\n",
    "    marginal_false_positives = zeros(total_runs)\n",
    "    marginal_false_negatives = zeros(total_runs)\n",
    "    for i in 1:total_runs\n",
    "        println(\"current run = $i\")\n",
    "        \n",
    "        #set random seed\n",
    "        Random.seed!(i)\n",
    "        \n",
    "        iht_tp, iht_fp, iht_fn, lasso_tp, lasso_fp, lasso_fn, m_tp, m_fp, m_fn = iht_lasso_marginal(n, p, d, l)\n",
    "        iht_true_positives[i] = iht_tp\n",
    "        iht_false_positives[i] = iht_fp\n",
    "        iht_false_negatives[i] = iht_fn\n",
    "        lasso_true_positives[i] = lasso_tp\n",
    "        lasso_false_positives[i] = lasso_fp\n",
    "        lasso_false_negatives[i] = lasso_fn\n",
    "        marginal_true_positives[i] = m_tp\n",
    "        marginal_false_positives[i] = m_fp\n",
    "        marginal_false_negatives[i] = m_fn\n",
    "    end\n",
    "    \n",
    "    return iht_true_positives, iht_false_positives, iht_false_negatives, \n",
    "            lasso_true_positives, lasso_false_positives, lasso_false_negatives, \n",
    "            marginal_true_positives, marginal_false_positives, marginal_false_negatives\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "current run = 1\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/normal\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.5     │ 0.523246 │ 0.427391  │ true           │\n",
      "│ 2   │ 1.0     │ 1.00562  │ 0.922618  │ true           │\n",
      "│ 3   │ 0.6     │ 0.61045  │ 0.523953  │ true           │\n",
      "│ 4   │ 0.1     │ 0.0      │ 0.0501758 │ false          │\n",
      "│ 5   │ 0.8     │ 0.7508   │ 0.670865  │ true           │\n",
      "│ 6   │ 0.2     │ 0.207868 │ 0.124228  │ false          │\n",
      "│ 7   │ 0.9     │ 0.881481 │ 0.799268  │ true           │\n",
      "│ 8   │ 0.7     │ 0.70816  │ 0.618471  │ true           │\n",
      "│ 9   │ 0.3     │ 0.295969 │ 0.21511   │ true           │\n",
      "│ 10  │ 0.4     │ 0.36466  │ 0.267788  │ false          │\n",
      "IHT true positives = 9\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 1\n",
      "LASSO true positives = 10\n",
      "LASSO false positives = 78\n",
      "LASSO false negatives = 0\n",
      "marginal true positives = 7\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 3\n",
      "\n",
      "current run = 2\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/normal\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β  │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64  │ Bool           │\n",
      "├─────┼─────────┼──────────┼──────────┼────────────────┤\n",
      "│ 1   │ 0.7     │ 0.680898 │ 0.593978 │ true           │\n",
      "│ 2   │ 0.2     │ 0.25598  │ 0.159202 │ false          │\n",
      "│ 3   │ 0.5     │ 0.487022 │ 0.38533  │ true           │\n",
      "│ 4   │ 1.0     │ 1.04705  │ 0.941458 │ true           │\n",
      "│ 5   │ 0.6     │ 0.618693 │ 0.512273 │ true           │\n",
      "│ 6   │ 0.9     │ 0.932099 │ 0.833615 │ true           │\n",
      "│ 7   │ 0.4     │ 0.351988 │ 0.256657 │ true           │\n",
      "│ 8   │ 0.3     │ 0.26372  │ 0.162427 │ false          │\n",
      "│ 9   │ 0.1     │ 0.0      │ 0.0      │ false          │\n",
      "│ 10  │ 0.8     │ 0.768702 │ 0.662836 │ true           │\n",
      "IHT true positives = 9\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 1\n",
      "LASSO true positives = 9\n",
      "LASSO false positives = 43\n",
      "LASSO false negatives = 1\n",
      "marginal true positives = 7\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 3\n",
      "\n",
      "current run = 3\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/normal\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.8     │ 0.788904 │ 0.688299  │ true           │\n",
      "│ 2   │ 0.6     │ 0.611703 │ 0.514125  │ true           │\n",
      "│ 3   │ 0.9     │ 0.944016 │ 0.844192  │ true           │\n",
      "│ 4   │ 0.5     │ 0.5103   │ 0.421252  │ true           │\n",
      "│ 5   │ 0.2     │ 0.0      │ 0.0660998 │ false          │\n",
      "│ 6   │ 0.4     │ 0.404111 │ 0.303684  │ true           │\n",
      "│ 7   │ 0.7     │ 0.652407 │ 0.545671  │ true           │\n",
      "│ 8   │ 0.1     │ 0.0      │ 0.0403752 │ false          │\n",
      "│ 9   │ 1.0     │ 1.00443  │ 0.906576  │ true           │\n",
      "│ 10  │ 0.3     │ 0.317482 │ 0.211856  │ false          │\n",
      "IHT true positives = 8\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 2\n",
      "LASSO true positives = 10\n",
      "LASSO false positives = 22\n",
      "LASSO false negatives = 0\n",
      "marginal true positives = 7\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 3\n",
      "\n",
      "current run = 4\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/normal\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β    │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64    │ Bool           │\n",
      "├─────┼─────────┼──────────┼────────────┼────────────────┤\n",
      "│ 1   │ 0.4     │ 0.444044 │ 0.352629   │ true           │\n",
      "│ 2   │ 0.3     │ 0.306494 │ 0.206661   │ false          │\n",
      "│ 3   │ 0.8     │ 0.856252 │ 0.767386   │ true           │\n",
      "│ 4   │ 0.1     │ 0.0      │ 0.00785733 │ false          │\n",
      "│ 5   │ 0.7     │ 0.656644 │ 0.575013   │ true           │\n",
      "│ 6   │ 0.9     │ 0.891091 │ 0.804033   │ true           │\n",
      "│ 7   │ 0.5     │ 0.475095 │ 0.384286   │ true           │\n",
      "│ 8   │ 1.0     │ 0.967892 │ 0.878444   │ true           │\n",
      "│ 9   │ 0.6     │ 0.602452 │ 0.516212   │ true           │\n",
      "│ 10  │ 0.2     │ 0.229362 │ 0.138805   │ false          │\n",
      "IHT true positives = 9\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 1\n",
      "LASSO true positives = 10\n",
      "LASSO false positives = 52\n",
      "LASSO false negatives = 0\n",
      "marginal true positives = 7\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 3\n",
      "\n",
      "current run = 5\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/normal\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β  │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64  │ Bool           │\n",
      "├─────┼─────────┼──────────┼──────────┼────────────────┤\n",
      "│ 1   │ 0.2     │ 0.213268 │ 0.105645 │ false          │\n",
      "│ 2   │ 0.6     │ 0.617482 │ 0.515536 │ false          │\n",
      "│ 3   │ 0.9     │ 0.874935 │ 0.770388 │ true           │\n",
      "│ 4   │ 1.0     │ 1.03915  │ 0.95003  │ true           │\n",
      "│ 5   │ 0.3     │ 0.328119 │ 0.211984 │ false          │\n",
      "│ 6   │ 0.4     │ 0.404972 │ 0.293202 │ true           │\n",
      "│ 7   │ 0.5     │ 0.549056 │ 0.439205 │ true           │\n",
      "│ 8   │ 0.1     │ 0.0      │ 0.0      │ false          │\n",
      "│ 9   │ 0.8     │ 0.801971 │ 0.69862  │ true           │\n",
      "│ 10  │ 0.7     │ 0.681155 │ 0.583266 │ true           │\n",
      "IHT true positives = 9\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 1\n",
      "LASSO true positives = 9\n",
      "LASSO false positives = 9\n",
      "LASSO false negatives = 1\n",
      "marginal true positives = 6\n",
      "marginal false positives = 1\n",
      "marginal false negatives = 4\n",
      "\n",
      "current run = 6\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/normal\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β  │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64  │ Bool           │\n",
      "├─────┼─────────┼──────────┼──────────┼────────────────┤\n",
      "│ 1   │ 0.2     │ 0.268399 │ 0.171906 │ false          │\n",
      "│ 2   │ 1.0     │ 0.997609 │ 0.892802 │ true           │\n",
      "│ 3   │ 0.7     │ 0.696039 │ 0.5962   │ true           │\n",
      "│ 4   │ 0.1     │ 0.0      │ 0.0      │ false          │\n",
      "│ 5   │ 0.5     │ 0.521322 │ 0.437446 │ true           │\n",
      "│ 6   │ 0.8     │ 0.800031 │ 0.706385 │ true           │\n",
      "│ 7   │ 0.3     │ 0.246277 │ 0.138299 │ false          │\n",
      "│ 8   │ 0.9     │ 0.902674 │ 0.8078   │ true           │\n",
      "│ 9   │ 0.4     │ 0.346064 │ 0.24069  │ true           │\n",
      "│ 10  │ 0.6     │ 0.621808 │ 0.523346 │ true           │\n",
      "IHT true positives = 9\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 1\n",
      "LASSO true positives = 9\n",
      "LASSO false positives = 56\n",
      "LASSO false negatives = 1\n",
      "marginal true positives = 7\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 3\n",
      "\n",
      "current run = 7\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/normal\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β  │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64  │ Bool           │\n",
      "├─────┼─────────┼──────────┼──────────┼────────────────┤\n",
      "│ 1   │ 0.5     │ 0.549299 │ 0.442888 │ true           │\n",
      "│ 2   │ 1.0     │ 0.987726 │ 0.860501 │ true           │\n",
      "│ 3   │ 0.1     │ 0.0      │ 0.0      │ false          │\n",
      "│ 4   │ 0.9     │ 0.927741 │ 0.814378 │ true           │\n",
      "│ 5   │ 0.8     │ 0.784661 │ 0.68145  │ true           │\n",
      "│ 6   │ 0.7     │ 0.638093 │ 0.508683 │ true           │\n",
      "│ 7   │ 0.2     │ 0.230365 │ 0.132271 │ false          │\n",
      "│ 8   │ 0.6     │ 0.62669  │ 0.512195 │ true           │\n",
      "│ 9   │ 0.3     │ 0.268052 │ 0.169373 │ false          │\n",
      "│ 10  │ 0.4     │ 0.471883 │ 0.357593 │ true           │\n",
      "IHT true positives = 9\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 1\n",
      "LASSO true positives = 9\n",
      "LASSO false positives = 14\n",
      "LASSO false negatives = 1\n",
      "marginal true positives = 7\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 3\n",
      "\n",
      "current run = 8\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/normal\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.8     │ 0.780934 │ 0.666228  │ true           │\n",
      "│ 2   │ 0.1     │ 0.0      │ 0.0       │ false          │\n",
      "│ 3   │ 0.2     │ 0.195072 │ 0.0914901 │ false          │\n",
      "│ 4   │ 0.4     │ 0.445603 │ 0.345152  │ true           │\n",
      "│ 5   │ 0.5     │ 0.585928 │ 0.495711  │ true           │\n",
      "│ 6   │ 0.7     │ 0.758122 │ 0.645737  │ true           │\n",
      "│ 7   │ 0.9     │ 0.84424  │ 0.748957  │ true           │\n",
      "│ 8   │ 0.6     │ 0.644523 │ 0.536723  │ true           │\n",
      "│ 9   │ 1.0     │ 1.0027   │ 0.885608  │ true           │\n",
      "│ 10  │ 0.3     │ 0.298629 │ 0.19004   │ false          │\n",
      "IHT true positives = 9\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 1\n",
      "LASSO true positives = 9\n",
      "LASSO false positives = 14\n",
      "LASSO false negatives = 1\n",
      "marginal true positives = 7\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 3\n",
      "\n",
      "current run = 9\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/normal\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.7     │ 0.719835 │ 0.61145   │ true           │\n",
      "│ 2   │ 0.4     │ 0.473302 │ 0.382601  │ true           │\n",
      "│ 3   │ 0.3     │ 0.368625 │ 0.268633  │ true           │\n",
      "│ 4   │ 0.9     │ 0.941759 │ 0.845276  │ true           │\n",
      "│ 5   │ 0.6     │ 0.650322 │ 0.567492  │ true           │\n",
      "│ 6   │ 1.0     │ 0.984157 │ 0.896008  │ true           │\n",
      "│ 7   │ 0.8     │ 0.797303 │ 0.696505  │ true           │\n",
      "│ 8   │ 0.5     │ 0.497797 │ 0.40366   │ true           │\n",
      "│ 9   │ 0.2     │ 0.0      │ 0.0622135 │ false          │\n",
      "│ 10  │ 0.1     │ 0.0      │ 0.0358477 │ false          │\n",
      "IHT true positives = 8\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 2\n",
      "LASSO true positives = 10\n",
      "LASSO false positives = 25\n",
      "LASSO false negatives = 0\n",
      "marginal true positives = 8\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 2\n",
      "\n",
      "current run = 10\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/normal\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 1.0     │ 1.02116  │ 0.924434  │ true           │\n",
      "│ 2   │ 0.4     │ 0.396885 │ 0.303367  │ true           │\n",
      "│ 3   │ 0.6     │ 0.592406 │ 0.493701  │ true           │\n",
      "│ 4   │ 0.7     │ 0.751654 │ 0.63619   │ true           │\n",
      "│ 5   │ 0.3     │ 0.305386 │ 0.228994  │ true           │\n",
      "│ 6   │ 0.5     │ 0.475798 │ 0.38202   │ true           │\n",
      "│ 7   │ 0.1     │ 0.0      │ 0.0463664 │ false          │\n",
      "│ 8   │ 0.8     │ 0.763354 │ 0.679809  │ true           │\n",
      "│ 9   │ 0.9     │ 0.896451 │ 0.804549  │ true           │\n",
      "│ 10  │ 0.2     │ 0.18118  │ 0.11048   │ true           │\n",
      "IHT true positives = 9\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 1\n",
      "LASSO true positives = 10\n",
      "LASSO false positives = 32\n",
      "LASSO false negatives = 0\n",
      "marginal true positives = 9\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 1\n",
      "\n",
      "current run = 11\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/normal\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.1     │ 0.0      │ 0.0196305 │ false          │\n",
      "│ 2   │ 0.7     │ 0.670399 │ 0.578975  │ true           │\n",
      "│ 3   │ 0.6     │ 0.586725 │ 0.502842  │ true           │\n",
      "│ 4   │ 0.5     │ 0.455424 │ 0.364268  │ true           │\n",
      "│ 5   │ 0.3     │ 0.255528 │ 0.15264   │ false          │\n",
      "│ 6   │ 0.4     │ 0.456677 │ 0.361555  │ true           │\n",
      "│ 7   │ 0.9     │ 0.876898 │ 0.770045  │ true           │\n",
      "│ 8   │ 0.8     │ 0.777438 │ 0.683919  │ true           │\n",
      "│ 9   │ 1.0     │ 1.05087  │ 0.952399  │ true           │\n",
      "│ 10  │ 0.2     │ 0.216549 │ 0.109464  │ false          │\n",
      "IHT true positives = 9\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 1\n",
      "LASSO true positives = 10\n",
      "LASSO false positives = 41\n",
      "LASSO false negatives = 0\n",
      "marginal true positives = 7\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 3\n",
      "\n",
      "current run = 12\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/normal\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β  │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64  │ Bool           │\n",
      "├─────┼─────────┼──────────┼──────────┼────────────────┤\n",
      "│ 1   │ 0.3     │ 0.307535 │ 0.210547 │ true           │\n",
      "│ 2   │ 0.1     │ 0.0      │ 0.0      │ false          │\n",
      "│ 3   │ 0.7     │ 0.715491 │ 0.613356 │ true           │\n",
      "│ 4   │ 0.9     │ 0.854021 │ 0.745753 │ true           │\n",
      "│ 5   │ 0.6     │ 0.590277 │ 0.4746   │ true           │\n",
      "│ 6   │ 0.2     │ 0.209433 │ 0.113626 │ false          │\n",
      "│ 7   │ 0.4     │ 0.390062 │ 0.289572 │ true           │\n",
      "│ 8   │ 0.8     │ 0.783819 │ 0.666988 │ true           │\n",
      "│ 9   │ 0.5     │ 0.485507 │ 0.385209 │ true           │\n",
      "│ 10  │ 1.0     │ 0.983307 │ 0.861039 │ true           │\n",
      "IHT true positives = 9\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 1\n",
      "LASSO true positives = 9\n",
      "LASSO false positives = 19\n",
      "LASSO false negatives = 1\n",
      "marginal true positives = 8\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 2\n",
      "\n",
      "current run = 13\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/normal\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β    │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64    │ Bool           │\n",
      "├─────┼─────────┼──────────┼────────────┼────────────────┤\n",
      "│ 1   │ 0.2     │ 0.191112 │ 0.0826153  │ false          │\n",
      "│ 2   │ 0.9     │ 0.949167 │ 0.859407   │ true           │\n",
      "│ 3   │ 1.0     │ 0.984536 │ 0.875255   │ true           │\n",
      "│ 4   │ 0.6     │ 0.624846 │ 0.523803   │ true           │\n",
      "│ 5   │ 0.4     │ 0.369486 │ 0.255172   │ true           │\n",
      "│ 6   │ 0.8     │ 0.798174 │ 0.685852   │ true           │\n",
      "│ 7   │ 0.1     │ 0.0      │ 0.00821563 │ false          │\n",
      "│ 8   │ 0.7     │ 0.726317 │ 0.623336   │ true           │\n",
      "│ 9   │ 0.5     │ 0.487078 │ 0.371644   │ true           │\n",
      "│ 10  │ 0.3     │ 0.295505 │ 0.173134   │ false          │\n",
      "IHT true positives = 9\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 1\n",
      "LASSO true positives = 10\n",
      "LASSO false positives = 14\n",
      "LASSO false negatives = 0\n",
      "marginal true positives = 7\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 3\n",
      "\n",
      "current run = 14\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/normal\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.1     │ 0.0      │ 0.0525801 │ false          │\n",
      "│ 2   │ 0.2     │ 0.217363 │ 0.128857  │ false          │\n",
      "│ 3   │ 0.8     │ 0.822197 │ 0.728297  │ true           │\n",
      "│ 4   │ 0.3     │ 0.276995 │ 0.182593  │ false          │\n",
      "│ 5   │ 0.6     │ 0.641389 │ 0.538155  │ true           │\n",
      "│ 6   │ 0.5     │ 0.436678 │ 0.339432  │ true           │\n",
      "│ 7   │ 1.0     │ 1.04816  │ 0.955467  │ true           │\n",
      "│ 8   │ 0.9     │ 0.857552 │ 0.76884   │ true           │\n",
      "│ 9   │ 0.4     │ 0.458445 │ 0.366203  │ true           │\n",
      "│ 10  │ 0.7     │ 0.714745 │ 0.63954   │ true           │\n",
      "IHT true positives = 9\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 1\n",
      "LASSO true positives = 10\n",
      "LASSO false positives = 79\n",
      "LASSO false negatives = 0\n",
      "marginal true positives = 7\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 3\n",
      "\n",
      "current run = 15\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/normal\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β  │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64  │ Bool           │\n",
      "├─────┼─────────┼──────────┼──────────┼────────────────┤\n",
      "│ 1   │ 0.9     │ 0.910775 │ 0.808774 │ true           │\n",
      "│ 2   │ 0.7     │ 0.723419 │ 0.621949 │ true           │\n",
      "│ 3   │ 0.1     │ 0.0      │ 0.0      │ false          │\n",
      "│ 4   │ 0.8     │ 0.813141 │ 0.720209 │ true           │\n",
      "│ 5   │ 1.0     │ 0.942277 │ 0.833869 │ true           │\n",
      "│ 6   │ 0.6     │ 0.609677 │ 0.519798 │ true           │\n",
      "│ 7   │ 0.5     │ 0.536393 │ 0.444106 │ true           │\n",
      "│ 8   │ 0.3     │ 0.360881 │ 0.245239 │ false          │\n",
      "│ 9   │ 0.2     │ 0.212177 │ 0.126827 │ false          │\n",
      "│ 10  │ 0.4     │ 0.3951   │ 0.317739 │ true           │\n",
      "IHT true positives = 9\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 1\n",
      "LASSO true positives = 9\n",
      "LASSO false positives = 20\n",
      "LASSO false negatives = 1\n",
      "marginal true positives = 7\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 3\n",
      "\n",
      "current run = 16\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/normal\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.1     │ 0.0      │ 0.0       │ false          │\n",
      "│ 2   │ 0.6     │ 0.568579 │ 0.477418  │ true           │\n",
      "│ 3   │ 0.3     │ 0.306932 │ 0.21132   │ true           │\n",
      "│ 4   │ 0.2     │ 0.193372 │ 0.0978518 │ false          │\n",
      "│ 5   │ 0.5     │ 0.430198 │ 0.335994  │ true           │\n",
      "│ 6   │ 1.0     │ 0.954571 │ 0.867011  │ true           │\n",
      "│ 7   │ 0.8     │ 0.830369 │ 0.749071  │ true           │\n",
      "│ 8   │ 0.4     │ 0.379819 │ 0.311734  │ true           │\n",
      "│ 9   │ 0.9     │ 0.896174 │ 0.802954  │ true           │\n",
      "│ 10  │ 0.7     │ 0.687594 │ 0.597497  │ true           │\n",
      "IHT true positives = 9\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 1\n",
      "LASSO true positives = 9\n",
      "LASSO false positives = 58\n",
      "LASSO false negatives = 1\n",
      "marginal true positives = 8\n",
      "marginal false positives = 1\n",
      "marginal false negatives = 2\n",
      "\n",
      "current run = 17\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/normal\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β    │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64    │ Bool           │\n",
      "├─────┼─────────┼──────────┼────────────┼────────────────┤\n",
      "│ 1   │ 0.2     │ 0.241303 │ 0.142015   │ false          │\n",
      "│ 2   │ 0.4     │ 0.399598 │ 0.301135   │ true           │\n",
      "│ 3   │ 0.1     │ 0.0      │ 0.00487623 │ false          │\n",
      "│ 4   │ 1.0     │ 0.978307 │ 0.895785   │ true           │\n",
      "│ 5   │ 0.7     │ 0.659998 │ 0.557858   │ true           │\n",
      "│ 6   │ 0.8     │ 0.795415 │ 0.688634   │ true           │\n",
      "│ 7   │ 0.5     │ 0.496556 │ 0.402376   │ true           │\n",
      "│ 8   │ 0.6     │ 0.614661 │ 0.51359    │ true           │\n",
      "│ 9   │ 0.3     │ 0.306255 │ 0.194295   │ false          │\n",
      "│ 10  │ 0.9     │ 0.901026 │ 0.819835   │ true           │\n",
      "IHT true positives = 9\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 1\n",
      "LASSO true positives = 10\n",
      "LASSO false positives = 23\n",
      "LASSO false negatives = 0\n",
      "marginal true positives = 7\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 3\n",
      "\n",
      "current run = 18\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/normal\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.6     │ 0.580341 │ 0.446117  │ true           │\n",
      "│ 2   │ 0.2     │ 0.206607 │ 0.0789551 │ false          │\n",
      "│ 3   │ 1.0     │ 0.975359 │ 0.850371  │ true           │\n",
      "│ 4   │ 0.1     │ 0.0      │ 0.0       │ false          │\n",
      "│ 5   │ 0.4     │ 0.395557 │ 0.270666  │ true           │\n",
      "│ 6   │ 0.7     │ 0.708705 │ 0.572822  │ true           │\n",
      "│ 7   │ 0.3     │ 0.259533 │ 0.143188  │ false          │\n",
      "│ 8   │ 0.9     │ 0.87225  │ 0.753836  │ true           │\n",
      "│ 9   │ 0.5     │ 0.491602 │ 0.371576  │ true           │\n",
      "│ 10  │ 0.8     │ 0.826046 │ 0.712151  │ true           │\n",
      "IHT true positives = 9\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 1\n",
      "LASSO true positives = 9\n",
      "LASSO false positives = 8\n",
      "LASSO false negatives = 1\n",
      "marginal true positives = 7\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 3\n",
      "\n",
      "current run = 19\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/normal\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.7     │ 0.710084 │ 0.608129  │ true           │\n",
      "│ 2   │ 0.9     │ 0.895858 │ 0.791678  │ true           │\n",
      "│ 3   │ 0.3     │ 0.259685 │ 0.154275  │ false          │\n",
      "│ 4   │ 0.2     │ 0.184785 │ 0.0771018 │ false          │\n",
      "│ 5   │ 0.1     │ 0.0      │ 0.0       │ false          │\n",
      "│ 6   │ 0.8     │ 0.88636  │ 0.783557  │ true           │\n",
      "│ 7   │ 0.4     │ 0.393048 │ 0.293712  │ true           │\n",
      "│ 8   │ 1.0     │ 0.952785 │ 0.856913  │ true           │\n",
      "│ 9   │ 0.6     │ 0.585691 │ 0.480445  │ true           │\n",
      "│ 10  │ 0.5     │ 0.490722 │ 0.391361  │ true           │\n",
      "IHT true positives = 9\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 1\n",
      "LASSO true positives = 9\n",
      "LASSO false positives = 27\n",
      "LASSO false negatives = 1\n",
      "marginal true positives = 7\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 3\n",
      "\n",
      "current run = 20\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/normal\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.5     │ 0.487471 │ 0.394749  │ true           │\n",
      "│ 2   │ 0.2     │ 0.0      │ 0.0930917 │ false          │\n",
      "│ 3   │ 1.0     │ 1.00952  │ 0.902194  │ true           │\n",
      "│ 4   │ 0.3     │ 0.36585  │ 0.26687   │ true           │\n",
      "│ 5   │ 0.6     │ 0.540074 │ 0.451917  │ true           │\n",
      "│ 6   │ 0.1     │ 0.0      │ 0.0624194 │ false          │\n",
      "│ 7   │ 0.4     │ 0.422234 │ 0.325231  │ true           │\n",
      "│ 8   │ 0.9     │ 0.9375   │ 0.851427  │ true           │\n",
      "│ 9   │ 0.7     │ 0.661437 │ 0.568286  │ true           │\n",
      "│ 10  │ 0.8     │ 0.796533 │ 0.709175  │ true           │\n",
      "IHT true positives = 8\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 2\n",
      "LASSO true positives = 10\n",
      "LASSO false positives = 53\n",
      "LASSO false negatives = 0\n",
      "marginal true positives = 8\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 2\n",
      "\n",
      "current run = 21\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/normal\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β  │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64  │ Bool           │\n",
      "├─────┼─────────┼──────────┼──────────┼────────────────┤\n",
      "│ 1   │ 0.2     │ 0.233921 │ 0.126087 │ false          │\n",
      "│ 2   │ 0.1     │ 0.0      │ 0.0      │ false          │\n",
      "│ 3   │ 0.9     │ 0.925928 │ 0.830966 │ true           │\n",
      "│ 4   │ 0.5     │ 0.520143 │ 0.426166 │ true           │\n",
      "│ 5   │ 0.3     │ 0.306711 │ 0.189001 │ false          │\n",
      "│ 6   │ 0.8     │ 0.838767 │ 0.730064 │ true           │\n",
      "│ 7   │ 0.6     │ 0.637485 │ 0.528126 │ true           │\n",
      "│ 8   │ 0.4     │ 0.464144 │ 0.363616 │ true           │\n",
      "│ 9   │ 1.0     │ 0.973335 │ 0.874472 │ true           │\n",
      "│ 10  │ 0.7     │ 0.748753 │ 0.651229 │ true           │\n",
      "IHT true positives = 9\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 1\n",
      "LASSO true positives = 9\n",
      "LASSO false positives = 12\n",
      "LASSO false negatives = 1\n",
      "marginal true positives = 7\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 3\n",
      "\n",
      "current run = 22\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/normal\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.1     │ 0.0      │ 0.0215375 │ false          │\n",
      "│ 2   │ 0.2     │ 0.225387 │ 0.121977  │ false          │\n",
      "│ 3   │ 0.3     │ 0.314126 │ 0.199367  │ false          │\n",
      "│ 4   │ 0.6     │ 0.578773 │ 0.482254  │ true           │\n",
      "│ 5   │ 0.8     │ 0.833964 │ 0.73506   │ true           │\n",
      "│ 6   │ 0.7     │ 0.710207 │ 0.614451  │ true           │\n",
      "│ 7   │ 0.5     │ 0.420027 │ 0.332869  │ true           │\n",
      "│ 8   │ 0.4     │ 0.418965 │ 0.313829  │ true           │\n",
      "│ 9   │ 1.0     │ 1.03904  │ 0.945548  │ true           │\n",
      "│ 10  │ 0.9     │ 0.882863 │ 0.773152  │ true           │\n",
      "IHT true positives = 9\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 1\n",
      "LASSO true positives = 10\n",
      "LASSO false positives = 16\n",
      "LASSO false negatives = 0\n",
      "marginal true positives = 7\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 3\n",
      "\n",
      "current run = 23\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/normal\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.5     │ 0.492233 │ 0.394938  │ true           │\n",
      "│ 2   │ 0.4     │ 0.431413 │ 0.315521  │ true           │\n",
      "│ 3   │ 0.8     │ 0.799112 │ 0.68992   │ true           │\n",
      "│ 4   │ 0.9     │ 0.909182 │ 0.81297   │ true           │\n",
      "│ 5   │ 0.3     │ 0.338239 │ 0.223021  │ false          │\n",
      "│ 6   │ 0.2     │ 0.221241 │ 0.137507  │ true           │\n",
      "│ 7   │ 0.6     │ 0.606    │ 0.504946  │ true           │\n",
      "│ 8   │ 1.0     │ 1.06707  │ 0.962384  │ true           │\n",
      "│ 9   │ 0.1     │ 0.0      │ 0.0170908 │ false          │\n",
      "│ 10  │ 0.7     │ 0.693205 │ 0.593955  │ true           │\n",
      "IHT true positives = 9\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 1\n",
      "LASSO true positives = 10\n",
      "LASSO false positives = 11\n",
      "LASSO false negatives = 0\n",
      "marginal true positives = 8\n",
      "marginal false positives = 1\n",
      "marginal false negatives = 2\n",
      "\n",
      "current run = 24\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/normal\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.2     │ 0.185998 │ 0.0778577 │ false          │\n",
      "│ 2   │ 0.4     │ 0.418031 │ 0.328585  │ true           │\n",
      "│ 3   │ 0.3     │ 0.29893  │ 0.174358  │ false          │\n",
      "│ 4   │ 0.8     │ 0.781799 │ 0.672044  │ true           │\n",
      "│ 5   │ 0.5     │ 0.494897 │ 0.408638  │ true           │\n",
      "│ 6   │ 1.0     │ 0.980789 │ 0.898212  │ true           │\n",
      "│ 7   │ 0.1     │ 0.0      │ 0.0       │ false          │\n",
      "│ 8   │ 0.9     │ 0.908401 │ 0.804241  │ true           │\n",
      "│ 9   │ 0.7     │ 0.655513 │ 0.538747  │ true           │\n",
      "│ 10  │ 0.6     │ 0.619967 │ 0.527524  │ true           │\n",
      "IHT true positives = 9\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 1\n",
      "LASSO true positives = 9\n",
      "LASSO false positives = 20\n",
      "LASSO false negatives = 1\n",
      "marginal true positives = 7\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 3\n",
      "\n",
      "current run = 25\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/normal\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.8     │ 0.853136 │ 0.745917  │ true           │\n",
      "│ 2   │ 0.1     │ 0.0      │ 0.0       │ false          │\n",
      "│ 3   │ 0.2     │ 0.213732 │ 0.0827807 │ false          │\n",
      "│ 4   │ 0.9     │ 0.971833 │ 0.837691  │ true           │\n",
      "│ 5   │ 0.3     │ 0.364148 │ 0.237103  │ true           │\n",
      "│ 6   │ 0.5     │ 0.502073 │ 0.389167  │ true           │\n",
      "│ 7   │ 1.0     │ 0.971774 │ 0.866392  │ true           │\n",
      "│ 8   │ 0.6     │ 0.613905 │ 0.485412  │ true           │\n",
      "│ 9   │ 0.7     │ 0.689371 │ 0.575093  │ true           │\n",
      "│ 10  │ 0.4     │ 0.392718 │ 0.297217  │ true           │\n",
      "IHT true positives = 9\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 1\n",
      "LASSO true positives = 9\n",
      "LASSO false positives = 5\n",
      "LASSO false negatives = 1\n",
      "marginal true positives = 8\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 2\n",
      "\n",
      "current run = 26\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/normal\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β  │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64  │ Bool           │\n",
      "├─────┼─────────┼──────────┼──────────┼────────────────┤\n",
      "│ 1   │ 0.9     │ 0.923466 │ 0.812147 │ true           │\n",
      "│ 2   │ 0.3     │ 0.26042  │ 0.148969 │ false          │\n",
      "│ 3   │ 1.0     │ 0.974251 │ 0.866509 │ true           │\n",
      "│ 4   │ 0.2     │ 0.242642 │ 0.128684 │ false          │\n",
      "│ 5   │ 0.4     │ 0.39896  │ 0.294288 │ true           │\n",
      "│ 6   │ 0.7     │ 0.694671 │ 0.59298  │ true           │\n",
      "│ 7   │ 0.6     │ 0.58857  │ 0.470707 │ true           │\n",
      "│ 8   │ 0.5     │ 0.505255 │ 0.413017 │ true           │\n",
      "│ 9   │ 0.8     │ 0.82907  │ 0.716552 │ true           │\n",
      "│ 10  │ 0.1     │ 0.0      │ 0.0      │ false          │\n",
      "IHT true positives = 9\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 1\n",
      "LASSO true positives = 9\n",
      "LASSO false positives = 10\n",
      "LASSO false negatives = 1\n",
      "marginal true positives = 7\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 3\n",
      "\n",
      "current run = 27\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/normal\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.7     │ 0.706739 │ 0.609341  │ true           │\n",
      "│ 2   │ 0.8     │ 0.819112 │ 0.734508  │ true           │\n",
      "│ 3   │ 0.4     │ 0.364684 │ 0.263402  │ true           │\n",
      "│ 4   │ 0.5     │ 0.508235 │ 0.415977  │ true           │\n",
      "│ 5   │ 0.1     │ 0.0      │ 0.0132165 │ false          │\n",
      "│ 6   │ 0.2     │ 0.195882 │ 0.119593  │ false          │\n",
      "│ 7   │ 0.9     │ 0.855339 │ 0.739189  │ true           │\n",
      "│ 8   │ 0.3     │ 0.294662 │ 0.19874   │ false          │\n",
      "│ 9   │ 1.0     │ 1.0124   │ 0.897439  │ true           │\n",
      "│ 10  │ 0.6     │ 0.584561 │ 0.481618  │ true           │\n",
      "IHT true positives = 9\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 1\n",
      "LASSO true positives = 10\n",
      "LASSO false positives = 41\n",
      "LASSO false negatives = 0\n",
      "marginal true positives = 7\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 3\n",
      "\n",
      "current run = 28\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/normal\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.7     │ 0.665811 │ 0.571605  │ true           │\n",
      "│ 2   │ 0.1     │ 0.0      │ 0.0154934 │ false          │\n",
      "│ 3   │ 1.0     │ 0.936117 │ 0.829922  │ false          │\n",
      "│ 4   │ 0.2     │ 0.0      │ 0.060363  │ false          │\n",
      "│ 5   │ 0.4     │ 0.391603 │ 0.291024  │ true           │\n",
      "│ 6   │ 0.5     │ 0.519534 │ 0.435177  │ true           │\n",
      "│ 7   │ 0.3     │ 0.266292 │ 0.159436  │ false          │\n",
      "│ 8   │ 0.8     │ 0.818293 │ 0.728134  │ true           │\n",
      "│ 9   │ 0.9     │ 0.838015 │ 0.737787  │ true           │\n",
      "│ 10  │ 0.6     │ 0.593233 │ 0.503108  │ true           │\n",
      "IHT true positives = 8\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 2\n",
      "LASSO true positives = 10\n",
      "LASSO false positives = 33\n",
      "LASSO false negatives = 0\n",
      "marginal true positives = 6\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 4\n",
      "\n",
      "current run = 29\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/normal\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β  │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64  │ Bool           │\n",
      "├─────┼─────────┼──────────┼──────────┼────────────────┤\n",
      "│ 1   │ 0.2     │ 0.212435 │ 0.121545 │ false          │\n",
      "│ 2   │ 0.4     │ 0.454293 │ 0.367931 │ true           │\n",
      "│ 3   │ 1.0     │ 1.02829  │ 0.940466 │ true           │\n",
      "│ 4   │ 0.8     │ 0.750381 │ 0.655813 │ true           │\n",
      "│ 5   │ 0.1     │ 0.0      │ 0.0      │ false          │\n",
      "│ 6   │ 0.3     │ 0.230499 │ 0.154548 │ true           │\n",
      "│ 7   │ 0.6     │ 0.577755 │ 0.500202 │ true           │\n",
      "│ 8   │ 0.5     │ 0.513155 │ 0.434116 │ true           │\n",
      "│ 9   │ 0.7     │ 0.680683 │ 0.599869 │ true           │\n",
      "│ 10  │ 0.9     │ 0.917679 │ 0.846429 │ true           │\n",
      "IHT true positives = 9\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 1\n",
      "LASSO true positives = 9\n",
      "LASSO false positives = 73\n",
      "LASSO false negatives = 1\n",
      "marginal true positives = 8\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 2\n",
      "\n",
      "current run = 30\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/normal\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β    │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64    │ Bool           │\n",
      "├─────┼─────────┼──────────┼────────────┼────────────────┤\n",
      "│ 1   │ 0.5     │ 0.476148 │ 0.386004   │ true           │\n",
      "│ 2   │ 0.4     │ 0.378951 │ 0.292393   │ true           │\n",
      "│ 3   │ 0.3     │ 0.313417 │ 0.198583   │ false          │\n",
      "│ 4   │ 0.9     │ 0.888555 │ 0.790875   │ true           │\n",
      "│ 5   │ 0.2     │ 0.210367 │ 0.102983   │ false          │\n",
      "│ 6   │ 1.0     │ 0.968393 │ 0.862064   │ true           │\n",
      "│ 7   │ 0.1     │ 0.0      │ 0.00490368 │ false          │\n",
      "│ 8   │ 0.8     │ 0.778637 │ 0.685018   │ true           │\n",
      "│ 9   │ 0.6     │ 0.604849 │ 0.506264   │ true           │\n",
      "│ 10  │ 0.7     │ 0.692246 │ 0.599376   │ true           │\n",
      "IHT true positives = 9\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 1\n",
      "LASSO true positives = 10\n",
      "LASSO false positives = 19\n",
      "LASSO false negatives = 0\n",
      "marginal true positives = 7\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 3\n",
      "\n",
      "current run = 31\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/normal\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β    │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64    │ Bool           │\n",
      "├─────┼─────────┼──────────┼────────────┼────────────────┤\n",
      "│ 1   │ 0.2     │ 0.22979  │ 0.142679   │ false          │\n",
      "│ 2   │ 1.0     │ 0.998589 │ 0.89644    │ true           │\n",
      "│ 3   │ 0.4     │ 0.439257 │ 0.346886   │ true           │\n",
      "│ 4   │ 0.9     │ 0.864655 │ 0.770095   │ true           │\n",
      "│ 5   │ 0.1     │ 0.0      │ 9.93207e-5 │ false          │\n",
      "│ 6   │ 0.5     │ 0.554236 │ 0.448512   │ true           │\n",
      "│ 7   │ 0.7     │ 0.789114 │ 0.697961   │ true           │\n",
      "│ 8   │ 0.6     │ 0.590045 │ 0.487651   │ true           │\n",
      "│ 9   │ 0.8     │ 0.812859 │ 0.720606   │ false          │\n",
      "│ 10  │ 0.3     │ 0.368792 │ 0.288721   │ true           │\n",
      "IHT true positives = 9\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 1\n",
      "LASSO true positives = 10\n",
      "LASSO false positives = 45\n",
      "LASSO false negatives = 0\n",
      "marginal true positives = 7\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 3\n",
      "\n",
      "current run = 32\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/normal\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 1.0     │ 0.992321 │ 0.89205   │ true           │\n",
      "│ 2   │ 0.1     │ 0.0      │ 0.0       │ false          │\n",
      "│ 3   │ 0.8     │ 0.825821 │ 0.723129  │ true           │\n",
      "│ 4   │ 0.7     │ 0.686916 │ 0.557176  │ true           │\n",
      "│ 5   │ 0.6     │ 0.590754 │ 0.499132  │ true           │\n",
      "│ 6   │ 0.9     │ 0.902174 │ 0.803317  │ true           │\n",
      "│ 7   │ 0.4     │ 0.382608 │ 0.282964  │ true           │\n",
      "│ 8   │ 0.2     │ 0.19643  │ 0.0886421 │ false          │\n",
      "│ 9   │ 0.5     │ 0.489614 │ 0.386885  │ true           │\n",
      "│ 10  │ 0.3     │ 0.308694 │ 0.199631  │ false          │\n",
      "IHT true positives = 9\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 1\n",
      "LASSO true positives = 9\n",
      "LASSO false positives = 15\n",
      "LASSO false negatives = 1\n",
      "marginal true positives = 7\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 3\n",
      "\n",
      "current run = 33\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/normal\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β  │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64  │ Bool           │\n",
      "├─────┼─────────┼──────────┼──────────┼────────────────┤\n",
      "│ 1   │ 0.7     │ 0.698534 │ 0.600013 │ true           │\n",
      "│ 2   │ 0.9     │ 0.889541 │ 0.793035 │ true           │\n",
      "│ 3   │ 0.8     │ 0.795629 │ 0.700077 │ true           │\n",
      "│ 4   │ 0.4     │ 0.426792 │ 0.316337 │ true           │\n",
      "│ 5   │ 1.0     │ 1.05722  │ 0.943401 │ true           │\n",
      "│ 6   │ 0.2     │ 0.229568 │ 0.124262 │ false          │\n",
      "│ 7   │ 0.6     │ 0.619457 │ 0.526164 │ true           │\n",
      "│ 8   │ 0.5     │ 0.491633 │ 0.386633 │ true           │\n",
      "│ 9   │ 0.1     │ 0.0      │ 0.0      │ false          │\n",
      "│ 10  │ 0.3     │ 0.268022 │ 0.175608 │ true           │\n",
      "IHT true positives = 9\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 1\n",
      "LASSO true positives = 9\n",
      "LASSO false positives = 28\n",
      "LASSO false negatives = 1\n",
      "marginal true positives = 8\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 2\n",
      "\n",
      "current run = 34\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/normal\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.7     │ 0.681299 │ 0.574283  │ true           │\n",
      "│ 2   │ 0.6     │ 0.604837 │ 0.494133  │ true           │\n",
      "│ 3   │ 0.1     │ 0.0      │ 0.0486065 │ false          │\n",
      "│ 4   │ 0.3     │ 0.287819 │ 0.188631  │ false          │\n",
      "│ 5   │ 0.8     │ 0.811679 │ 0.69422   │ true           │\n",
      "│ 6   │ 0.4     │ 0.361844 │ 0.271991  │ false          │\n",
      "│ 7   │ 0.9     │ 0.892001 │ 0.808838  │ true           │\n",
      "│ 8   │ 0.2     │ 0.0      │ 0.0707312 │ false          │\n",
      "│ 9   │ 1.0     │ 1.05814  │ 0.945626  │ true           │\n",
      "│ 10  │ 0.5     │ 0.489613 │ 0.377046  │ true           │\n",
      "IHT true positives = 8\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 2\n",
      "LASSO true positives = 10\n",
      "LASSO false positives = 56\n",
      "LASSO false negatives = 0\n",
      "marginal true positives = 6\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 4\n",
      "\n",
      "current run = 35\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/normal\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 1.0     │ 0.971236 │ 0.868499  │ true           │\n",
      "│ 2   │ 0.7     │ 0.683412 │ 0.596445  │ true           │\n",
      "│ 3   │ 0.3     │ 0.316105 │ 0.203951  │ false          │\n",
      "│ 4   │ 0.9     │ 0.924282 │ 0.8131    │ true           │\n",
      "│ 5   │ 0.4     │ 0.370781 │ 0.258785  │ false          │\n",
      "│ 6   │ 0.6     │ 0.591005 │ 0.486575  │ true           │\n",
      "│ 7   │ 0.1     │ 0.0      │ 0.0125776 │ false          │\n",
      "│ 8   │ 0.8     │ 0.840121 │ 0.728454  │ true           │\n",
      "│ 9   │ 0.5     │ 0.524436 │ 0.41158   │ false          │\n",
      "│ 10  │ 0.2     │ 0.196014 │ 0.0968806 │ false          │\n",
      "IHT true positives = 9\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 1\n",
      "LASSO true positives = 10\n",
      "LASSO false positives = 44\n",
      "LASSO false negatives = 0\n",
      "marginal true positives = 5\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 5\n",
      "\n",
      "current run = 36\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/normal\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β  │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64  │ Bool           │\n",
      "├─────┼─────────┼──────────┼──────────┼────────────────┤\n",
      "│ 1   │ 0.7     │ 0.741885 │ 0.666147 │ true           │\n",
      "│ 2   │ 0.9     │ 0.89867  │ 0.802411 │ true           │\n",
      "│ 3   │ 1.0     │ 1.04701  │ 0.94943  │ true           │\n",
      "│ 4   │ 0.1     │ 0.0      │ 0.0      │ false          │\n",
      "│ 5   │ 0.4     │ 0.387616 │ 0.291556 │ true           │\n",
      "│ 6   │ 0.5     │ 0.418301 │ 0.30802  │ true           │\n",
      "│ 7   │ 0.3     │ 0.258186 │ 0.165457 │ false          │\n",
      "│ 8   │ 0.6     │ 0.603456 │ 0.502148 │ true           │\n",
      "│ 9   │ 0.2     │ 0.213915 │ 0.121049 │ false          │\n",
      "│ 10  │ 0.8     │ 0.773949 │ 0.668597 │ true           │\n",
      "IHT true positives = 9\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 1\n",
      "LASSO true positives = 9\n",
      "LASSO false positives = 16\n",
      "LASSO false negatives = 1\n",
      "marginal true positives = 7\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 3\n",
      "\n",
      "current run = 37\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/normal\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.8     │ 0.798715 │ 0.706296  │ true           │\n",
      "│ 2   │ 0.6     │ 0.610429 │ 0.533687  │ true           │\n",
      "│ 3   │ 0.4     │ 0.43229  │ 0.337445  │ true           │\n",
      "│ 4   │ 0.7     │ 0.759907 │ 0.665784  │ true           │\n",
      "│ 5   │ 1.0     │ 0.996026 │ 0.912671  │ true           │\n",
      "│ 6   │ 0.2     │ 0.18657  │ 0.100275  │ false          │\n",
      "│ 7   │ 0.5     │ 0.530104 │ 0.445586  │ true           │\n",
      "│ 8   │ 0.3     │ 0.294011 │ 0.198444  │ false          │\n",
      "│ 9   │ 0.9     │ 0.919389 │ 0.833857  │ true           │\n",
      "│ 10  │ 0.1     │ 0.0      │ 0.0165732 │ false          │\n",
      "IHT true positives = 9\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 1\n",
      "LASSO true positives = 10\n",
      "LASSO false positives = 45\n",
      "LASSO false negatives = 0\n",
      "marginal true positives = 7\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 3\n",
      "\n",
      "current run = 38\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/normal\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 1.0     │ 0.987413 │ 0.888431  │ true           │\n",
      "│ 2   │ 0.6     │ 0.625659 │ 0.518968  │ true           │\n",
      "│ 3   │ 0.7     │ 0.703604 │ 0.614828  │ true           │\n",
      "│ 4   │ 0.8     │ 0.791006 │ 0.696426  │ true           │\n",
      "│ 5   │ 0.2     │ 0.0      │ 0.0867785 │ false          │\n",
      "│ 6   │ 0.5     │ 0.516826 │ 0.415984  │ true           │\n",
      "│ 7   │ 0.3     │ 0.290102 │ 0.183699  │ false          │\n",
      "│ 8   │ 0.9     │ 0.861042 │ 0.762933  │ true           │\n",
      "│ 9   │ 0.4     │ 0.40574  │ 0.311582  │ true           │\n",
      "│ 10  │ 0.1     │ 0.0      │ 0.0113337 │ false          │\n",
      "IHT true positives = 8\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 2\n",
      "LASSO true positives = 10\n",
      "LASSO false positives = 69\n",
      "LASSO false negatives = 0\n",
      "marginal true positives = 7\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 3\n",
      "\n",
      "current run = 39\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/normal\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.8     │ 0.750837 │ 0.658466  │ true           │\n",
      "│ 2   │ 1.0     │ 0.986024 │ 0.884926  │ true           │\n",
      "│ 3   │ 0.6     │ 0.613811 │ 0.502464  │ true           │\n",
      "│ 4   │ 0.4     │ 0.383326 │ 0.274643  │ true           │\n",
      "│ 5   │ 0.2     │ 0.19028  │ 0.0710588 │ false          │\n",
      "│ 6   │ 0.7     │ 0.682272 │ 0.576661  │ true           │\n",
      "│ 7   │ 0.3     │ 0.361921 │ 0.2626    │ true           │\n",
      "│ 8   │ 0.5     │ 0.500991 │ 0.398658  │ true           │\n",
      "│ 9   │ 0.1     │ 0.0      │ 0.0       │ false          │\n",
      "│ 10  │ 0.9     │ 0.965593 │ 0.851302  │ true           │\n",
      "IHT true positives = 9\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 1\n",
      "LASSO true positives = 9\n",
      "LASSO false positives = 19\n",
      "LASSO false negatives = 1\n",
      "marginal true positives = 8\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 2\n",
      "\n",
      "current run = 40\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/normal\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.1     │ 0.0      │ 0.032711  │ false          │\n",
      "│ 2   │ 0.4     │ 0.39391  │ 0.285267  │ true           │\n",
      "│ 3   │ 0.3     │ 0.375042 │ 0.262446  │ true           │\n",
      "│ 4   │ 1.0     │ 1.02671  │ 0.920738  │ true           │\n",
      "│ 5   │ 0.2     │ 0.198185 │ 0.0890666 │ false          │\n",
      "│ 6   │ 0.7     │ 0.684069 │ 0.572314  │ true           │\n",
      "│ 7   │ 0.9     │ 0.942335 │ 0.825545  │ true           │\n",
      "│ 8   │ 0.5     │ 0.499006 │ 0.392809  │ true           │\n",
      "│ 9   │ 0.6     │ 0.596714 │ 0.507644  │ true           │\n",
      "│ 10  │ 0.8     │ 0.735556 │ 0.632589  │ true           │\n",
      "IHT true positives = 9\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 1\n",
      "LASSO true positives = 10\n",
      "LASSO false positives = 30\n",
      "LASSO false negatives = 0\n",
      "marginal true positives = 8\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 2\n",
      "\n",
      "current run = 41\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/normal\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.4     │ 0.408516 │ 0.312805  │ true           │\n",
      "│ 2   │ 0.6     │ 0.625267 │ 0.510849  │ true           │\n",
      "│ 3   │ 0.5     │ 0.513979 │ 0.391755  │ true           │\n",
      "│ 4   │ 0.9     │ 0.942429 │ 0.839742  │ true           │\n",
      "│ 5   │ 0.7     │ 0.685866 │ 0.571048  │ true           │\n",
      "│ 6   │ 0.2     │ 0.186541 │ 0.0941924 │ false          │\n",
      "│ 7   │ 0.8     │ 0.790081 │ 0.699067  │ true           │\n",
      "│ 8   │ 0.1     │ 0.0      │ 0.0       │ false          │\n",
      "│ 9   │ 1.0     │ 1.00946  │ 0.899457  │ true           │\n",
      "│ 10  │ 0.3     │ 0.286588 │ 0.17931   │ false          │\n",
      "IHT true positives = 9\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 1\n",
      "LASSO true positives = 9\n",
      "LASSO false positives = 21\n",
      "LASSO false negatives = 1\n",
      "marginal true positives = 7\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 3\n",
      "\n",
      "current run = 42\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/normal\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.6     │ 0.663667 │ 0.567187  │ true           │\n",
      "│ 2   │ 0.3     │ 0.266526 │ 0.160497  │ false          │\n",
      "│ 3   │ 0.5     │ 0.514088 │ 0.42643   │ true           │\n",
      "│ 4   │ 0.2     │ 0.168839 │ 0.0784316 │ false          │\n",
      "│ 5   │ 0.8     │ 0.781841 │ 0.687973  │ true           │\n",
      "│ 6   │ 0.7     │ 0.647893 │ 0.552181  │ true           │\n",
      "│ 7   │ 1.0     │ 1.02528  │ 0.92623   │ true           │\n",
      "│ 8   │ 0.1     │ 0.0      │ 0.0300934 │ false          │\n",
      "│ 9   │ 0.4     │ 0.453422 │ 0.35636   │ true           │\n",
      "│ 10  │ 0.9     │ 0.913772 │ 0.812559  │ true           │\n",
      "IHT true positives = 9\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 1\n",
      "LASSO true positives = 10\n",
      "LASSO false positives = 45\n",
      "LASSO false negatives = 0\n",
      "marginal true positives = 7\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 3\n",
      "\n",
      "current run = 43\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/normal\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.7     │ 0.693298 │ 0.606067  │ true           │\n",
      "│ 2   │ 0.4     │ 0.305879 │ 0.232756  │ true           │\n",
      "│ 3   │ 0.6     │ 0.537123 │ 0.438114  │ true           │\n",
      "│ 4   │ 0.1     │ 0.0      │ 0.0       │ false          │\n",
      "│ 5   │ 0.2     │ 0.186123 │ 0.0802885 │ false          │\n",
      "│ 6   │ 0.8     │ 0.84261  │ 0.75461   │ true           │\n",
      "│ 7   │ 0.5     │ 0.497894 │ 0.411904  │ true           │\n",
      "│ 8   │ 0.9     │ 0.945079 │ 0.862736  │ true           │\n",
      "│ 9   │ 0.3     │ 0.28099  │ 0.179452  │ false          │\n",
      "│ 10  │ 1.0     │ 1.00761  │ 0.914926  │ true           │\n",
      "IHT true positives = 9\n",
      "IHT false positives = 1\n",
      "IHT false negatives = 1\n",
      "LASSO true positives = 9\n",
      "LASSO false positives = 31\n",
      "LASSO false negatives = 1\n",
      "marginal true positives = 7\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 3\n",
      "\n",
      "current run = 44\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/normal\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.2     │ 0.187555 │ 0.0879071 │ false          │\n",
      "│ 2   │ 0.1     │ 0.0      │ 0.0317713 │ false          │\n",
      "│ 3   │ 0.8     │ 0.803771 │ 0.694163  │ true           │\n",
      "│ 4   │ 0.3     │ 0.328177 │ 0.23794   │ true           │\n",
      "│ 5   │ 0.4     │ 0.436513 │ 0.331885  │ true           │\n",
      "│ 6   │ 0.6     │ 0.586669 │ 0.492389  │ true           │\n",
      "│ 7   │ 0.9     │ 0.889714 │ 0.788972  │ true           │\n",
      "│ 8   │ 1.0     │ 0.970861 │ 0.871064  │ true           │\n",
      "│ 9   │ 0.7     │ 0.675821 │ 0.568489  │ true           │\n",
      "│ 10  │ 0.5     │ 0.499657 │ 0.387731  │ true           │\n",
      "IHT true positives = 9\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 1\n",
      "LASSO true positives = 10\n",
      "LASSO false positives = 12\n",
      "LASSO false negatives = 0\n",
      "marginal true positives = 8\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 2\n",
      "\n",
      "current run = 45\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/normal\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.3     │ 0.315839 │ 0.208484  │ false          │\n",
      "│ 2   │ 0.7     │ 0.696231 │ 0.597408  │ true           │\n",
      "│ 3   │ 0.2     │ 0.0      │ 0.0391626 │ false          │\n",
      "│ 4   │ 0.1     │ 0.0      │ 0.0       │ false          │\n",
      "│ 5   │ 0.8     │ 0.81253  │ 0.713301  │ true           │\n",
      "│ 6   │ 0.9     │ 0.842234 │ 0.744436  │ true           │\n",
      "│ 7   │ 1.0     │ 0.975655 │ 0.862303  │ true           │\n",
      "│ 8   │ 0.4     │ 0.438833 │ 0.32399   │ true           │\n",
      "│ 9   │ 0.5     │ 0.499711 │ 0.390205  │ true           │\n",
      "│ 10  │ 0.6     │ 0.62255  │ 0.534369  │ true           │\n",
      "IHT true positives = 8\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 2\n",
      "LASSO true positives = 9\n",
      "LASSO false positives = 29\n",
      "LASSO false negatives = 1\n",
      "marginal true positives = 7\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 3\n",
      "\n",
      "current run = 46\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/normal\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.5     │ 0.520575 │ 0.392631  │ true           │\n",
      "│ 2   │ 0.8     │ 0.804254 │ 0.692295  │ true           │\n",
      "│ 3   │ 0.4     │ 0.407901 │ 0.27917   │ false          │\n",
      "│ 4   │ 0.6     │ 0.551048 │ 0.455497  │ true           │\n",
      "│ 5   │ 0.9     │ 0.933981 │ 0.828293  │ true           │\n",
      "│ 6   │ 0.7     │ 0.744722 │ 0.652842  │ true           │\n",
      "│ 7   │ 0.3     │ 0.258966 │ 0.159181  │ false          │\n",
      "│ 8   │ 0.2     │ 0.242154 │ 0.130238  │ false          │\n",
      "│ 9   │ 1.0     │ 0.947836 │ 0.835509  │ true           │\n",
      "│ 10  │ 0.1     │ 0.0      │ 0.0280612 │ false          │\n",
      "IHT true positives = 9\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 1\n",
      "LASSO true positives = 10\n",
      "LASSO false positives = 10\n",
      "LASSO false negatives = 0\n",
      "marginal true positives = 6\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 4\n",
      "\n",
      "current run = 47\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/normal\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β  │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64  │ Bool           │\n",
      "├─────┼─────────┼──────────┼──────────┼────────────────┤\n",
      "│ 1   │ 0.8     │ 0.77699  │ 0.665737 │ true           │\n",
      "│ 2   │ 0.1     │ 0.0      │ 0.0      │ false          │\n",
      "│ 3   │ 1.0     │ 1.04797  │ 0.947422 │ true           │\n",
      "│ 4   │ 0.7     │ 0.697914 │ 0.580139 │ true           │\n",
      "│ 5   │ 0.2     │ 0.185692 │ 0.062303 │ false          │\n",
      "│ 6   │ 0.3     │ 0.286661 │ 0.187764 │ false          │\n",
      "│ 7   │ 0.9     │ 0.895297 │ 0.791939 │ true           │\n",
      "│ 8   │ 0.5     │ 0.468526 │ 0.357422 │ true           │\n",
      "│ 9   │ 0.6     │ 0.618352 │ 0.509248 │ true           │\n",
      "│ 10  │ 0.4     │ 0.417636 │ 0.314356 │ true           │\n",
      "IHT true positives = 9\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 1\n",
      "LASSO true positives = 9\n",
      "LASSO false positives = 14\n",
      "LASSO false negatives = 1\n",
      "marginal true positives = 7\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 3\n",
      "\n",
      "current run = 48\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/normal\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β    │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64    │ Bool           │\n",
      "├─────┼─────────┼──────────┼────────────┼────────────────┤\n",
      "│ 1   │ 1.0     │ 0.938885 │ 0.839775   │ true           │\n",
      "│ 2   │ 0.8     │ 0.814647 │ 0.725302   │ true           │\n",
      "│ 3   │ 0.9     │ 0.938501 │ 0.841124   │ true           │\n",
      "│ 4   │ 0.7     │ 0.694914 │ 0.612694   │ true           │\n",
      "│ 5   │ 0.5     │ 0.55721  │ 0.451961   │ true           │\n",
      "│ 6   │ 0.6     │ 0.611804 │ 0.511173   │ true           │\n",
      "│ 7   │ 0.1     │ 0.0      │ 0.00449656 │ false          │\n",
      "│ 8   │ 0.2     │ 0.224186 │ 0.121623   │ false          │\n",
      "│ 9   │ 0.4     │ 0.467046 │ 0.362878   │ true           │\n",
      "│ 10  │ 0.3     │ 0.317608 │ 0.228681   │ true           │\n",
      "IHT true positives = 9\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 1\n",
      "LASSO true positives = 10\n",
      "LASSO false positives = 20\n",
      "LASSO false negatives = 0\n",
      "marginal true positives = 8\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 2\n",
      "\n",
      "current run = 49\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/normal\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.3     │ 0.295427 │ 0.193751  │ false          │\n",
      "│ 2   │ 0.2     │ 0.0      │ 0.0361561 │ false          │\n",
      "│ 3   │ 0.7     │ 0.715679 │ 0.614123  │ true           │\n",
      "│ 4   │ 0.9     │ 0.896384 │ 0.773986  │ true           │\n",
      "│ 5   │ 0.1     │ 0.0      │ 0.0       │ false          │\n",
      "│ 6   │ 0.5     │ 0.540259 │ 0.43765   │ true           │\n",
      "│ 7   │ 0.8     │ 0.777264 │ 0.676039  │ true           │\n",
      "│ 8   │ 0.4     │ 0.371245 │ 0.261932  │ true           │\n",
      "│ 9   │ 1.0     │ 1.04377  │ 0.932418  │ true           │\n",
      "│ 10  │ 0.6     │ 0.592516 │ 0.491815  │ true           │\n",
      "IHT true positives = 8\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 2\n",
      "LASSO true positives = 9\n",
      "LASSO false positives = 41\n",
      "LASSO false negatives = 1\n",
      "marginal true positives = 7\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 3\n",
      "\n",
      "current run = 50\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/normal\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = linear\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.1     │ 0.0      │ 0.0123745 │ false          │\n",
      "│ 2   │ 0.5     │ 0.561262 │ 0.458125  │ true           │\n",
      "│ 3   │ 0.2     │ 0.221235 │ 0.130474  │ false          │\n",
      "│ 4   │ 0.9     │ 0.890755 │ 0.809934  │ true           │\n",
      "│ 5   │ 0.7     │ 0.689708 │ 0.595975  │ true           │\n",
      "│ 6   │ 1.0     │ 1.00988  │ 0.925323  │ true           │\n",
      "│ 7   │ 0.6     │ 0.625588 │ 0.529473  │ true           │\n",
      "│ 8   │ 0.4     │ 0.396481 │ 0.299968  │ true           │\n",
      "│ 9   │ 0.3     │ 0.317546 │ 0.224709  │ true           │\n",
      "│ 10  │ 0.8     │ 0.776182 │ 0.685802  │ true           │\n",
      "IHT true positives = 9\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 1\n",
      "LASSO true positives = 10\n",
      "LASSO false positives = 46\n",
      "LASSO false negatives = 0\n",
      "marginal true positives = 8\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 2\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "([9.0, 9.0, 8.0, 9.0, 9.0, 9.0, 9.0, 9.0, 8.0, 9.0  …  9.0, 9.0, 9.0, 9.0, 8.0, 9.0, 9.0, 9.0, 8.0, 9.0], [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0  …  0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0], [1.0, 1.0, 2.0, 1.0, 1.0, 1.0, 1.0, 1.0, 2.0, 1.0  …  1.0, 1.0, 1.0, 1.0, 2.0, 1.0, 1.0, 1.0, 2.0, 1.0], [10.0, 9.0, 10.0, 10.0, 9.0, 9.0, 9.0, 9.0, 10.0, 10.0  …  9.0, 10.0, 9.0, 10.0, 9.0, 10.0, 9.0, 10.0, 9.0, 10.0], [78.0, 43.0, 22.0, 52.0, 9.0, 56.0, 14.0, 14.0, 25.0, 32.0  …  21.0, 45.0, 31.0, 12.0, 29.0, 10.0, 14.0, 20.0, 41.0, 46.0], [0.0, 1.0, 0.0, 0.0, 1.0, 1.0, 1.0, 1.0, 0.0, 0.0  …  1.0, 0.0, 1.0, 0.0, 1.0, 0.0, 1.0, 0.0, 1.0, 0.0], [7.0, 7.0, 7.0, 7.0, 6.0, 7.0, 7.0, 7.0, 8.0, 9.0  …  7.0, 7.0, 7.0, 8.0, 7.0, 6.0, 7.0, 8.0, 7.0, 8.0], [0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0  …  0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0], [3.0, 3.0, 3.0, 3.0, 4.0, 3.0, 3.0, 3.0, 2.0, 1.0  …  3.0, 3.0, 3.0, 2.0, 3.0, 4.0, 3.0, 2.0, 3.0, 2.0])"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "iht_true_positives, iht_false_positives, iht_false_negatives, \n",
    "    lasso_true_positives, lasso_false_positives, lasso_false_negatives, \n",
    "    marginal_true_positives, marginal_false_positives, marginal_false_negatives = run()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3×3 Array{Float64,2}:\n",
       " 8.84   0.02  1.16\n",
       " 9.52  31.26  0.48\n",
       " 7.18   0.06  2.82"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "normal_iht_true_positives = sum(iht_true_positives) / 50\n",
    "normal_iht_false_positives = sum(iht_false_positives) / 50\n",
    "normal_iht_false_negatives = sum(iht_false_negatives) / 50\n",
    "normal_lasso_true_positives = sum(lasso_true_positives) / 50\n",
    "normal_lasso_false_positives = sum(lasso_false_positives) / 50\n",
    "normal_lasso_false_negatives = sum(lasso_false_negatives) / 50\n",
    "normal_marginal_true_positives = sum(marginal_true_positives) / 50\n",
    "normal_marginal_false_positives = sum(marginal_false_positives) / 50\n",
    "normal_marginal_false_negatives = sum(marginal_false_negatives) / 50\n",
    "\n",
    "result = [normal_iht_true_positives normal_iht_false_positives normal_iht_false_negatives; \n",
    "        normal_lasso_true_positives normal_lasso_false_positives normal_lasso_false_negatives;\n",
    "        normal_marginal_true_positives normal_marginal_false_positives normal_marginal_false_negatives]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "8.84"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "normal_iht_true_positives"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1.16"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "normal_iht_false_negatives"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Julia 1.0.3",
   "language": "julia",
   "name": "julia-1.0"
  },
  "language_info": {
   "file_extension": ".jl",
   "mimetype": "application/julia",
   "name": "julia",
   "version": "1.0.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
