{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "# Bernoulli: IHT vs LASSO vs Marginal regression: comparing false positive/negatives\n",
    "\n",
    "Marginal regression is the most prevalent method used to analyze GWAS data. LASSO is currently the de facto penalized least squares method for [feature selection](https://en.wikipedia.org/wiki/Feature_selection). Here we compare the performance (in terms of the number of false positives/negative) of LASSO and marginal regression with IHT for GWAS data. We use the `glmnet` implementation of cyclic coordinate descent for LASSO, and `MendelGWAS.jl` module of [OpenMendel](https://openmendel.github.io/MendelGWAS.jl/) for marginal analysis. Since the focus here is not scalability, we test our sample problems on moderately sized genotype matrces of 1000 samples with 10,000 SNPs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "26"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "using Distributed\n",
    "addprocs(25)\n",
    "nprocs()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "┌ Info: Loading DataFrames support into Gadfly.jl\n",
      "└ @ Gadfly /u/home/b/biona001/.julia/packages/Gadfly/09PWZ/src/mapping.jl:228\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall /u/home/b/biona001/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: Loading required package: Matrix\n",
      "│ Loading required package: foreach\n",
      "│ Loaded glmnet 2.0-16\n",
      "│ \n",
      "└ @ RCall /u/home/b/biona001/.julia/packages/RCall/ffM0W/src/io.jl:113\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "RObject{StrSxp}\n",
       " [1] \"glmnet\"    \"foreach\"   \"Matrix\"    \"stats\"     \"graphics\"  \"grDevices\"\n",
       " [7] \"utils\"     \"datasets\"  \"methods\"   \"base\"     \n"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "using MendelIHT\n",
    "using SnpArrays\n",
    "using DataFrames\n",
    "using Distributions\n",
    "using Random\n",
    "using LinearAlgebra\n",
    "using DelimitedFiles\n",
    "using GLM\n",
    "using MendelGWAS\n",
    "using CSV\n",
    "using RCall\n",
    "R\"library(glmnet)\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "# Bernoulli response (i.e. logistic regression)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "iht_lasso_marginal (generic function with 1 method)"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "function iht_lasso_marginal(n::Int64, p::Int64, k::Int64, d::UnionAll, l::Link)\n",
    "    #construct snpmatrix, covariate files, and true model b\n",
    "    x, = simulate_random_snparray(n, p, \"tmp.bed\")\n",
    "    xbm = SnpBitMatrix{Float64}(x, model=ADDITIVE_MODEL, center=true, scale=true); \n",
    "    z = ones(n, 1) # the intercept\n",
    "    x_float = [convert(Matrix{Float64}, x, center=true, scale=true) z] #Float64 version of x\n",
    "\n",
    "    # simulate response, true model b, and the correct non-0 positions of b\n",
    "    y, true_b, correct_position = simulate_random_response(x, xbm, k, d, l)\n",
    " \n",
    "    #specify path and folds\n",
    "    num_folds = 3\n",
    "    folds = rand(1:num_folds, size(x, 1));\n",
    "\n",
    "    #run glmnet via Rcall\n",
    "    @rput x_float y folds num_folds #make variables visible to R\n",
    "    R\"lasso_cv_result = cv.glmnet(x_float, y, nfolds = num_folds, foldid = folds, family='binomial')\"\n",
    "    R\"lasso_beta_tmp = glmnet(x_float, y, lambda=lasso_cv_result$lambda.min, family='binomial')$beta\"\n",
    "    R\"lasso_beta = as.vector(lasso_beta_tmp)\"\n",
    "    @rget lasso_cv_result lasso_beta #pull result from R to Julia\n",
    "    lasso_k_est = count(!iszero, lasso_beta)\n",
    "    \n",
    "    #find non-zero entries returned by best lasso model as largest k estimate\n",
    "    path = collect(1:50);\n",
    "    \n",
    "    #run IHT's cross validation routine \n",
    "    mses = cv_iht(d(), l, x, z, y, 1, path, folds, num_folds, use_maf=false, debias=false, showinfo=false, parallel=true);\n",
    "    iht_k_est = argmin(mses)\n",
    "    iht_result = L0_reg(x, xbm, z, y, 1, iht_k_est, d(), l, debias=false, init=false, use_maf=false)\n",
    "    iht_beta = iht_result.beta\n",
    "    \n",
    "    #Now run MendelGWAS\n",
    "    make_bim_fam_files(x, y, \"tmp\") #create .bim and .bed files for MendelGWAS\n",
    "\n",
    "    #create a control file for MendelGWAS\n",
    "    open(\"tmp_control.txt\", \"w\") do f\n",
    "        write(f, \"plink_input_basename = tmp \\n\")\n",
    "        write(f, \"plink_field_separator = '\\t' \\n\")\n",
    "        write(f, \"output_table = tmp_table.txt \\n\")\n",
    "        write(f, \"regression = logistic \\n\")\n",
    "        write(f, \"affected_designator = 1.0 \\n\")\n",
    "        write(f, \"regression_formula = Trait ~ \\n\")\n",
    "    end\n",
    "\n",
    "    #run marginal analysis\n",
    "    GWAS(\"tmp_control.txt\");\n",
    "\n",
    "    # calculate false positive/negatives based on bonferroni correction\n",
    "    p_values = CSV.read(\"tmp_table.txt\", delim=',', header=true)[:Pvalue]\n",
    "    significance = 0.05 / Float64(p)\n",
    "    passing_snps_position = findall(p_values .<= significance)\n",
    "    marginal_found = [correct_position[snp] in passing_snps_position for snp in 1:k]\n",
    "\n",
    "    #clean up\n",
    "    rm(\"tmp.bed\", force=true)\n",
    "    rm(\"tmp.bim\", force=true)\n",
    "    rm(\"tmp.fam\", force=true)\n",
    "    rm(\"tmp_table.txt\", force=true)\n",
    "    rm(\"tmp_control.txt\", force=true)\n",
    "    rm(\"Mendel_Output.txt\", force=true)\n",
    "        \n",
    "    #show lasso and IHT's reconstruction result\n",
    "    compare_model = DataFrame(\n",
    "        true_β  = true_b[correct_position], \n",
    "        IHT_β   = iht_beta[correct_position],\n",
    "        lasso_β = lasso_beta[correct_position],\n",
    "        marginal_found = marginal_found)\n",
    "    @show compare_model\n",
    "    \n",
    "    #compute true/false positives/negatives for IHT and lasso\n",
    "    iht_tp = count(!iszero, iht_beta[correct_position])\n",
    "    iht_fp = iht_k_est - iht_tp\n",
    "    iht_fn = k - iht_tp\n",
    "    lasso_tp = count(!iszero, lasso_beta[correct_position])\n",
    "    lasso_fp = lasso_k_est - lasso_tp\n",
    "    lasso_fn = k - lasso_tp\n",
    "    marginal_tp = count(!iszero, true_b[passing_snps_position])\n",
    "    marginal_fp = length(passing_snps_position) - marginal_tp\n",
    "    marginal_fn = k - marginal_tp\n",
    "    \n",
    "    println(\"IHT false positives = $iht_fp\")\n",
    "    println(\"IHT false negatives = $iht_fn\")\n",
    "    println(\"LASSO false positives = $lasso_fp\")\n",
    "    println(\"LASSO false negatives = $lasso_fn\")\n",
    "    println(\"marginal false positives = $marginal_fp\")\n",
    "    println(\"marginal false negatives = $marginal_fn\" * \"\\n\")\n",
    "    \n",
    "    return iht_fp, iht_fn, lasso_fp, lasso_fn, marginal_fp, marginal_fn\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "run (generic function with 1 method)"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "function run()\n",
    "    #simulat data with k true predictors, from distribution d and with link l.\n",
    "    n = 1000\n",
    "    p = 10000\n",
    "    k = 10\n",
    "    d = Bernoulli\n",
    "    l = canonicallink(d())\n",
    "\n",
    "    #run function above, saving results in 4 vectors\n",
    "    total_runs = 50\n",
    "    iht_false_positives = zeros(total_runs)\n",
    "    iht_false_negatives = zeros(total_runs)\n",
    "    lasso_false_positives = zeros(total_runs)\n",
    "    lasso_false_negatives = zeros(total_runs)\n",
    "    marginal_false_positives = zeros(total_runs)\n",
    "    marginal_false_negatives = zeros(total_runs)\n",
    "    for i in 1:total_runs\n",
    "        println(\"current run = $i\")\n",
    "        \n",
    "        #set random seed\n",
    "        Random.seed!(i)\n",
    "        \n",
    "        iht_fp, iht_fn, lasso_fp, lasso_fn, m_fp, m_fn = iht_lasso_marginal(n, p, k, d, l)\n",
    "        iht_false_positives[i] = iht_fp\n",
    "        iht_false_negatives[i] = iht_fn\n",
    "        lasso_false_positives[i] = lasso_fp\n",
    "        lasso_false_negatives[i] = lasso_fn\n",
    "        marginal_false_positives[i] = m_fp\n",
    "        marginal_false_negatives[i] = m_fn\n",
    "    end\n",
    "    \n",
    "    return iht_false_positives, iht_false_negatives, lasso_false_positives, \n",
    "            lasso_false_negatives, marginal_false_positives, marginal_false_negatives\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "current run = 1\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β    │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64   │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼───────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.607538  │ 0.0      │ 0.226727  │ false          │\n",
      "│ 2   │ -1.53004  │ -1.68693 │ -0.951184 │ true           │\n",
      "│ 3   │ 1.44416   │ 1.77934  │ 1.04999   │ true           │\n",
      "│ 4   │ 1.75303   │ 1.90754  │ 1.10995   │ true           │\n",
      "│ 5   │ 0.69867   │ 0.889128 │ 0.388388  │ false          │\n",
      "│ 6   │ -1.69162  │ -1.99076 │ -1.15899  │ true           │\n",
      "│ 7   │ -0.211858 │ 0.0      │ 0.0       │ false          │\n",
      "│ 8   │ 0.092561  │ 0.0      │ 0.0       │ false          │\n",
      "│ 9   │ 0.611886  │ 0.828261 │ 0.307231  │ false          │\n",
      "│ 10  │ -1.52095  │ -1.45649 │ -0.751299 │ true           │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 3\n",
      "LASSO false positives = 167\n",
      "LASSO false negatives = 2\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 5\n",
      "\n",
      "current run = 2\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β    │ IHT_β     │ lasso_β   │ marginal_found │\n",
      "│     │ Float64   │ Float64   │ Float64   │ Bool           │\n",
      "├─────┼───────────┼───────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.749759  │ 0.717065  │ 0.272586  │ false          │\n",
      "│ 2   │ 0.677934  │ 0.783906  │ 0.29728   │ false          │\n",
      "│ 3   │ 0.0720974 │ 0.0       │ 0.0       │ false          │\n",
      "│ 4   │ -1.13432  │ -1.22105  │ -0.638395 │ true           │\n",
      "│ 5   │ -0.827478 │ -0.971275 │ -0.46323  │ true           │\n",
      "│ 6   │ 1.13273   │ 1.13115   │ 0.563636  │ true           │\n",
      "│ 7   │ -0.898177 │ -1.01155  │ -0.489227 │ true           │\n",
      "│ 8   │ -2.55085  │ -2.61471  │ -1.62683  │ true           │\n",
      "│ 9   │ 1.90046   │ 0.98474   │ 0.131249  │ false          │\n",
      "│ 10  │ 0.668923  │ 0.0       │ 0.129002  │ false          │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 2\n",
      "LASSO false positives = 85\n",
      "LASSO false negatives = 1\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 5\n",
      "\n",
      "current run = 3\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β    │ IHT_β     │ lasso_β   │ marginal_found │\n",
      "│     │ Float64   │ Float64   │ Float64   │ Bool           │\n",
      "├─────┼───────────┼───────────┼───────────┼────────────────┤\n",
      "│ 1   │ -1.29395  │ -1.35589  │ -0.865015 │ true           │\n",
      "│ 2   │ 0.786253  │ 0.800853  │ 0.424835  │ true           │\n",
      "│ 3   │ -2.06493  │ -2.58402  │ -0.746037 │ true           │\n",
      "│ 4   │ 0.540411  │ 0.0       │ 0.139517  │ false          │\n",
      "│ 5   │ 1.49302   │ 1.65726   │ 1.07272   │ true           │\n",
      "│ 6   │ -0.640079 │ -0.683172 │ -0.283706 │ true           │\n",
      "│ 7   │ 0.0457283 │ 0.0       │ 0.0       │ false          │\n",
      "│ 8   │ -1.5777   │ -1.60591  │ -1.09624  │ true           │\n",
      "│ 9   │ -1.56178  │ -1.47064  │ -0.981208 │ true           │\n",
      "│ 10  │ 0.127488  │ 0.0       │ 0.0       │ false          │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 3\n",
      "LASSO false positives = 221\n",
      "LASSO false negatives = 2\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 3\n",
      "\n",
      "current run = 4\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β     │ IHT_β     │ lasso_β    │ marginal_found │\n",
      "│     │ Float64    │ Float64   │ Float64    │ Bool           │\n",
      "├─────┼────────────┼───────────┼────────────┼────────────────┤\n",
      "│ 1   │ 1.01155    │ 0.907321  │ 0.403845   │ true           │\n",
      "│ 2   │ -0.375936  │ 0.0       │ 0.0        │ false          │\n",
      "│ 3   │ -0.909578  │ -0.740821 │ -0.343297  │ true           │\n",
      "│ 4   │ -2.04195   │ -1.96568  │ -1.31088   │ true           │\n",
      "│ 5   │ -1.24787   │ -1.12528  │ -0.598833  │ true           │\n",
      "│ 6   │ -0.219267  │ 0.0       │ 0.0        │ false          │\n",
      "│ 7   │ -0.0620693 │ 0.0       │ 0.0        │ false          │\n",
      "│ 8   │ -0.493204  │ 0.0       │ -0.0779886 │ true           │\n",
      "│ 9   │ 0.0735544  │ 0.0       │ 0.0        │ false          │\n",
      "│ 10  │ -0.40849   │ 0.0       │ -0.0290676 │ false          │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 6\n",
      "LASSO false positives = 19\n",
      "LASSO false negatives = 4\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 5\n",
      "\n",
      "current run = 5\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β     │ IHT_β     │ lasso_β   │ marginal_found │\n",
      "│     │ Float64    │ Float64   │ Float64   │ Bool           │\n",
      "├─────┼────────────┼───────────┼───────────┼────────────────┤\n",
      "│ 1   │ 1.49393    │ 1.2119    │ 0.649084  │ true           │\n",
      "│ 2   │ 0.268446   │ 0.0       │ 0.0       │ false          │\n",
      "│ 3   │ -1.57681   │ -1.50357  │ -0.91803  │ true           │\n",
      "│ 4   │ -0.0665982 │ 0.0       │ 0.0       │ false          │\n",
      "│ 5   │ 0.564564   │ 0.6325    │ 0.270952  │ true           │\n",
      "│ 6   │ -0.0723161 │ 0.0       │ 0.0       │ false          │\n",
      "│ 7   │ -0.012129  │ 0.0       │ 0.0       │ false          │\n",
      "│ 8   │ 1.43517    │ 1.26495   │ 0.754358  │ true           │\n",
      "│ 9   │ -0.432575  │ 0.0       │ 0.0       │ false          │\n",
      "│ 10  │ -0.656486  │ -0.689057 │ -0.274784 │ true           │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 5\n",
      "LASSO false positives = 26\n",
      "LASSO false negatives = 5\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 5\n",
      "\n",
      "current run = 6\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β     │ IHT_β     │ lasso_β   │ marginal_found │\n",
      "│     │ Float64    │ Float64   │ Float64   │ Bool           │\n",
      "├─────┼────────────┼───────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.557855   │ 0.598211  │ 0.276092  │ true           │\n",
      "│ 2   │ 0.852578   │ 0.657053  │ 0.350499  │ true           │\n",
      "│ 3   │ -0.157019  │ 0.0       │ 0.0       │ false          │\n",
      "│ 4   │ -0.167593  │ 0.0       │ 0.0       │ false          │\n",
      "│ 5   │ 1.43279    │ 1.51991   │ 0.990755  │ true           │\n",
      "│ 6   │ 0.00814292 │ 0.0       │ 0.0       │ false          │\n",
      "│ 7   │ -0.311245  │ 0.0       │ 0.0       │ false          │\n",
      "│ 8   │ 1.14164    │ 1.13185   │ 0.713251  │ true           │\n",
      "│ 9   │ 0.296891   │ 0.0       │ 0.108537  │ false          │\n",
      "│ 10  │ -0.407489  │ -0.503935 │ -0.199605 │ false          │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 5\n",
      "LASSO false positives = 44\n",
      "LASSO false negatives = 4\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 6\n",
      "\n",
      "current run = 7\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β     │ IHT_β     │ lasso_β   │ marginal_found │\n",
      "│     │ Float64    │ Float64   │ Float64   │ Bool           │\n",
      "├─────┼────────────┼───────────┼───────────┼────────────────┤\n",
      "│ 1   │ -1.11915   │ -1.22661  │ -0.649775 │ true           │\n",
      "│ 2   │ 0.61236    │ 0.725363  │ 0.275914  │ true           │\n",
      "│ 3   │ 0.559179   │ 0.500245  │ 0.13032   │ false          │\n",
      "│ 4   │ 0.460021   │ 0.0       │ 0.0107004 │ false          │\n",
      "│ 5   │ -0.0420588 │ 0.0       │ 0.0       │ false          │\n",
      "│ 6   │ -0.595666  │ -0.615084 │ -0.242806 │ true           │\n",
      "│ 7   │ 0.54228    │ 0.556899  │ 0.166434  │ false          │\n",
      "│ 8   │ 0.933922   │ 1.02115   │ 0.498397  │ true           │\n",
      "│ 9   │ 0.685041   │ 0.597861  │ 0.197654  │ true           │\n",
      "│ 10  │ 1.15986    │ 0.9602    │ 0.440085  │ true           │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 2\n",
      "LASSO false positives = 15\n",
      "LASSO false negatives = 1\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 4\n",
      "\n",
      "current run = 8\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β    │ IHT_β     │ lasso_β   │ marginal_found │\n",
      "│     │ Float64   │ Float64   │ Float64   │ Bool           │\n",
      "├─────┼───────────┼───────────┼───────────┼────────────────┤\n",
      "│ 1   │ -0.318516 │ 0.0       │ 0.0       │ false          │\n",
      "│ 2   │ 0.80228   │ 0.851154  │ 0.37411   │ true           │\n",
      "│ 3   │ 0.678794  │ 0.736851  │ 0.310853  │ true           │\n",
      "│ 4   │ -0.550278 │ -0.540062 │ -0.187053 │ true           │\n",
      "│ 5   │ -1.039    │ -1.05465  │ -0.547867 │ true           │\n",
      "│ 6   │ -1.10893  │ -1.21397  │ -0.636564 │ true           │\n",
      "│ 7   │ -2.11847  │ -2.67241  │ -0.355988 │ true           │\n",
      "│ 8   │ 0.539736  │ 0.518204  │ 0.179404  │ false          │\n",
      "│ 9   │ 0.0513002 │ 0.0       │ 0.0       │ false          │\n",
      "│ 10  │ 1.44312   │ 1.49445   │ 0.793217  │ true           │\n",
      "IHT false positives = 1\n",
      "IHT false negatives = 2\n",
      "LASSO false positives = 57\n",
      "LASSO false negatives = 2\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 3\n",
      "\n",
      "current run = 9\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β    │ IHT_β     │ lasso_β    │ marginal_found │\n",
      "│     │ Float64   │ Float64   │ Float64    │ Bool           │\n",
      "├─────┼───────────┼───────────┼────────────┼────────────────┤\n",
      "│ 1   │ -1.20375  │ -1.22896  │ -0.677346  │ true           │\n",
      "│ 2   │ -0.334385 │ 0.0       │ -0.0718554 │ false          │\n",
      "│ 3   │ -1.82832  │ -1.6604   │ -0.83916   │ true           │\n",
      "│ 4   │ 1.3708    │ 1.31173   │ 0.720723   │ true           │\n",
      "│ 5   │ 1.20444   │ 1.15864   │ 0.511059   │ true           │\n",
      "│ 6   │ -0.508497 │ -0.730174 │ -0.314747  │ true           │\n",
      "│ 7   │ -1.09875  │ -1.16447  │ -0.620614  │ true           │\n",
      "│ 8   │ 0.486961  │ 0.49951   │ 0.119164   │ false          │\n",
      "│ 9   │ 0.251382  │ 0.0       │ 0.0149626  │ false          │\n",
      "│ 10  │ -0.460604 │ 0.0       │ -0.0715771 │ false          │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 3\n",
      "LASSO false positives = 79\n",
      "LASSO false negatives = 0\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 4\n",
      "\n",
      "current run = 10\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β   │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64  │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼──────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 1.17878  │ 1.25353  │ 0.680094  │ true           │\n",
      "│ 2   │ -0.44546 │ 0.0      │ -0.209572 │ false          │\n",
      "│ 3   │ 0.249749 │ 0.0      │ 0.0       │ false          │\n",
      "│ 4   │ 1.47017  │ 1.62562  │ 1.03226   │ true           │\n",
      "│ 5   │ 0.655446 │ 0.628649 │ 0.308748  │ false          │\n",
      "│ 6   │ 2.48177  │ 2.74488  │ 1.80644   │ true           │\n",
      "│ 7   │ -1.38405 │ -1.72458 │ -1.03886  │ true           │\n",
      "│ 8   │ 1.2645   │ 1.34732  │ 0.736211  │ true           │\n",
      "│ 9   │ 1.70905  │ 1.98841  │ 1.27148   │ true           │\n",
      "│ 10  │ 0.494919 │ 0.565471 │ 0.132005  │ false          │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 2\n",
      "LASSO false positives = 276\n",
      "LASSO false negatives = 1\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 4\n",
      "\n",
      "current run = 11\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β    │ IHT_β     │ lasso_β    │ marginal_found │\n",
      "│     │ Float64   │ Float64   │ Float64    │ Bool           │\n",
      "├─────┼───────────┼───────────┼────────────┼────────────────┤\n",
      "│ 1   │ -2.1401   │ -1.78949  │ -1.18876   │ true           │\n",
      "│ 2   │ 0.597831  │ 0.726852  │ 0.367054   │ true           │\n",
      "│ 3   │ -0.068048 │ 0.0       │ 0.0        │ false          │\n",
      "│ 4   │ 0.379909  │ 0.0       │ 0.0        │ false          │\n",
      "│ 5   │ -1.59939  │ -1.4977   │ -0.993308  │ true           │\n",
      "│ 6   │ -0.519577 │ -0.556996 │ -0.248097  │ true           │\n",
      "│ 7   │ 0.33893   │ 0.0       │ 0.0        │ false          │\n",
      "│ 8   │ 0.822734  │ 0.631727  │ 0.337895   │ true           │\n",
      "│ 9   │ -0.437776 │ 0.0       │ -0.0280313 │ false          │\n",
      "│ 10  │ 0.454821  │ 0.0       │ 0.132548   │ false          │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 5\n",
      "LASSO false positives = 65\n",
      "LASSO false negatives = 3\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 5\n",
      "\n",
      "current run = 12\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β     │ IHT_β     │ lasso_β   │ marginal_found │\n",
      "│     │ Float64    │ Float64   │ Float64   │ Bool           │\n",
      "├─────┼────────────┼───────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.381684   │ 0.0       │ 0.156361  │ false          │\n",
      "│ 2   │ -2.25632   │ -2.18121  │ -1.0692   │ true           │\n",
      "│ 3   │ -0.0285488 │ 0.0       │ 0.0       │ false          │\n",
      "│ 4   │ -1.20846   │ -1.12498  │ -0.622395 │ true           │\n",
      "│ 5   │ -1.1725    │ -0.994243 │ -0.483532 │ true           │\n",
      "│ 6   │ 0.666317   │ 0.0       │ 0.185598  │ false          │\n",
      "│ 7   │ 1.33832    │ 1.13384   │ 0.575928  │ true           │\n",
      "│ 8   │ 0.695851   │ 1.05227   │ 0.273948  │ false          │\n",
      "│ 9   │ 0.0762416  │ 0.0       │ 0.0       │ false          │\n",
      "│ 10  │ 2.51051    │ 2.3302    │ 1.54248   │ true           │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 4\n",
      "LASSO false positives = 122\n",
      "LASSO false negatives = 2\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 5\n",
      "\n",
      "current run = 13\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β     │ IHT_β     │ lasso_β   │ marginal_found │\n",
      "│     │ Float64    │ Float64   │ Float64   │ Bool           │\n",
      "├─────┼────────────┼───────────┼───────────┼────────────────┤\n",
      "│ 1   │ -0.191228  │ 0.0       │ 0.0       │ false          │\n",
      "│ 2   │ -0.530065  │ -0.692493 │ -0.282947 │ true           │\n",
      "│ 3   │ 0.761358   │ 0.711156  │ 0.335447  │ true           │\n",
      "│ 4   │ -0.295163  │ 0.0       │ 0.0       │ false          │\n",
      "│ 5   │ 1.14278    │ 1.12803   │ 0.630235  │ true           │\n",
      "│ 6   │ 1.42384    │ 1.44798   │ 0.890228  │ true           │\n",
      "│ 7   │ -0.415621  │ -0.407343 │ -0.103756 │ false          │\n",
      "│ 8   │ 0.815625   │ 0.707076  │ 0.323114  │ true           │\n",
      "│ 9   │ -0.0729403 │ 0.0       │ 0.0       │ false          │\n",
      "│ 10  │ 0.28586    │ 0.0       │ 0.0       │ false          │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 4\n",
      "LASSO false positives = 19\n",
      "LASSO false negatives = 4\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 5\n",
      "\n",
      "current run = 14\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β    │ IHT_β     │ lasso_β   │ marginal_found │\n",
      "│     │ Float64   │ Float64   │ Float64   │ Bool           │\n",
      "├─────┼───────────┼───────────┼───────────┼────────────────┤\n",
      "│ 1   │ -2.00106  │ -2.33932  │ -1.47391  │ true           │\n",
      "│ 2   │ 0.837229  │ 0.948353  │ 0.45566   │ true           │\n",
      "│ 3   │ 0.636888  │ 0.62577   │ 0.221858  │ false          │\n",
      "│ 4   │ 0.515326  │ 0.672801  │ 0.307116  │ false          │\n",
      "│ 5   │ 0.331752  │ 0.0       │ 0.0594879 │ false          │\n",
      "│ 6   │ -1.36196  │ -1.48969  │ -0.850554 │ true           │\n",
      "│ 7   │ -0.859086 │ -0.923056 │ -0.485637 │ true           │\n",
      "│ 8   │ -1.66215  │ -1.72575  │ -0.971763 │ true           │\n",
      "│ 9   │ 1.06366   │ 1.23101   │ 0.731172  │ true           │\n",
      "│ 10  │ 1.2262    │ 1.48062   │ 0.8375    │ true           │\n",
      "IHT false positives = 1\n",
      "IHT false negatives = 1\n",
      "LASSO false positives = 240\n",
      "LASSO false negatives = 0\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 3\n",
      "\n",
      "current run = 15\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β    │ IHT_β     │ lasso_β    │ marginal_found │\n",
      "│     │ Float64   │ Float64   │ Float64    │ Bool           │\n",
      "├─────┼───────────┼───────────┼────────────┼────────────────┤\n",
      "│ 1   │ 0.390517  │ 0.0       │ 0.0845016  │ false          │\n",
      "│ 2   │ 1.13247   │ 0.998449  │ 0.718517   │ true           │\n",
      "│ 3   │ -0.477164 │ 0.0       │ -0.121204  │ false          │\n",
      "│ 4   │ -0.332879 │ 0.0       │ -0.0444476 │ false          │\n",
      "│ 5   │ 0.601342  │ 0.626145  │ 0.355606   │ true           │\n",
      "│ 6   │ 0.776711  │ 0.0       │ 0.163717   │ true           │\n",
      "│ 7   │ -0.596155 │ -0.586304 │ -0.348171  │ true           │\n",
      "│ 8   │ -0.305982 │ -0.434374 │ -0.199603  │ true           │\n",
      "│ 9   │ -1.07709  │ -0.973243 │ -0.482601  │ true           │\n",
      "│ 10  │ 0.157683  │ 0.0       │ 0.0        │ false          │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 5\n",
      "LASSO false positives = 100\n",
      "LASSO false negatives = 1\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 4\n",
      "\n",
      "current run = 16\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β     │ IHT_β     │ lasso_β    │ marginal_found │\n",
      "│     │ Float64    │ Float64   │ Float64    │ Bool           │\n",
      "├─────┼────────────┼───────────┼────────────┼────────────────┤\n",
      "│ 1   │ 1.42842    │ 1.49781   │ 0.867734   │ true           │\n",
      "│ 2   │ 1.51933    │ 1.51345   │ 0.736518   │ true           │\n",
      "│ 3   │ 1.72402    │ 1.83167   │ 1.12353    │ true           │\n",
      "│ 4   │ -0.0571394 │ 0.0       │ 0.0        │ false          │\n",
      "│ 5   │ 0.31877    │ 0.0       │ 0.00435351 │ false          │\n",
      "│ 6   │ 0.19792    │ 0.0       │ 0.0        │ false          │\n",
      "│ 7   │ -1.53487   │ -1.64412  │ -0.953097  │ true           │\n",
      "│ 8   │ -0.621343  │ -0.670985 │ -0.271755  │ true           │\n",
      "│ 9   │ -0.804211  │ -0.670566 │ -0.249887  │ false          │\n",
      "│ 10  │ -0.0669473 │ 0.0       │ 0.0        │ false          │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 4\n",
      "LASSO false positives = 96\n",
      "LASSO false negatives = 3\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 5\n",
      "\n",
      "current run = 17\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β    │ IHT_β     │ lasso_β   │ marginal_found │\n",
      "│     │ Float64   │ Float64   │ Float64   │ Bool           │\n",
      "├─────┼───────────┼───────────┼───────────┼────────────────┤\n",
      "│ 1   │ -0.631517 │ -0.454065 │ -0.127846 │ false          │\n",
      "│ 2   │ -0.284542 │ 0.0       │ 0.0       │ false          │\n",
      "│ 3   │ -0.90586  │ -0.912101 │ -0.456829 │ true           │\n",
      "│ 4   │ 2.13207   │ 2.29657   │ 0.355466  │ true           │\n",
      "│ 5   │ 0.179385  │ 0.0       │ 0.0       │ false          │\n",
      "│ 6   │ -1.47059  │ -1.51771  │ -0.950706 │ true           │\n",
      "│ 7   │ 0.345312  │ 0.0       │ 0.0933899 │ false          │\n",
      "│ 8   │ -0.606927 │ -0.634168 │ -0.257087 │ true           │\n",
      "│ 9   │ 0.685558  │ 0.710754  │ 0.346117  │ true           │\n",
      "│ 10  │ 0.0588637 │ 0.0       │ 0.0       │ false          │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 4\n",
      "LASSO false positives = 23\n",
      "LASSO false negatives = 3\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 5\n",
      "\n",
      "current run = 18\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β      │ IHT_β     │ lasso_β    │ marginal_found │\n",
      "│     │ Float64     │ Float64   │ Float64    │ Bool           │\n",
      "├─────┼─────────────┼───────────┼────────────┼────────────────┤\n",
      "│ 1   │ 0.687324    │ 0.65181   │ 0.222953   │ true           │\n",
      "│ 2   │ 0.578401    │ 0.551445  │ 0.24762    │ true           │\n",
      "│ 3   │ 0.834403    │ 0.780656  │ 0.446192   │ true           │\n",
      "│ 4   │ -0.00685474 │ 0.0       │ 0.0        │ false          │\n",
      "│ 5   │ -0.976679   │ -1.10179  │ -0.695502  │ true           │\n",
      "│ 6   │ -0.108428   │ 0.0       │ 0.0        │ false          │\n",
      "│ 7   │ -0.377253   │ -0.382041 │ -0.0942237 │ false          │\n",
      "│ 8   │ 0.788596    │ 1.00421   │ 0.621271   │ true           │\n",
      "│ 9   │ -0.155683   │ 0.0       │ 0.0        │ false          │\n",
      "│ 10  │ 0.496849    │ 0.0       │ 0.068084   │ false          │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 4\n",
      "LASSO false positives = 58\n",
      "LASSO false negatives = 3\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 5\n",
      "\n",
      "current run = 19\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β      │ IHT_β    │ lasso_β    │ marginal_found │\n",
      "│     │ Float64     │ Float64  │ Float64    │ Bool           │\n",
      "├─────┼─────────────┼──────────┼────────────┼────────────────┤\n",
      "│ 1   │ -0.419297   │ 0.0      │ -0.0613105 │ false          │\n",
      "│ 2   │ -0.223578   │ 0.0      │ 0.0        │ false          │\n",
      "│ 3   │ -0.283353   │ 0.0      │ -0.107342  │ false          │\n",
      "│ 4   │ 0.294738    │ 0.0      │ 0.0        │ false          │\n",
      "│ 5   │ -0.00546738 │ 0.0      │ 0.0        │ false          │\n",
      "│ 6   │ 0.854599    │ 0.843814 │ 0.451939   │ true           │\n",
      "│ 7   │ 1.72256     │ 1.58987  │ 1.01831    │ true           │\n",
      "│ 8   │ -0.0586545  │ 0.0      │ 0.0        │ false          │\n",
      "│ 9   │ 1.02908     │ 1.07814  │ 0.619618   │ true           │\n",
      "│ 10  │ -1.65961    │ -1.58727 │ -1.02557   │ true           │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 6\n",
      "LASSO false positives = 77\n",
      "LASSO false negatives = 4\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 6\n",
      "\n",
      "current run = 20\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β    │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64   │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼───────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.978725  │ 1.05328  │ 0.615849  │ true           │\n",
      "│ 2   │ 0.135377  │ 0.0      │ 0.0       │ false          │\n",
      "│ 3   │ 0.151616  │ 0.0      │ 0.0       │ false          │\n",
      "│ 4   │ -0.169305 │ 0.0      │ 0.0       │ false          │\n",
      "│ 5   │ -0.232999 │ 0.0      │ -0.111809 │ false          │\n",
      "│ 6   │ 0.181031  │ 0.0      │ 0.0       │ false          │\n",
      "│ 7   │ -0.196465 │ 0.0      │ 0.0       │ false          │\n",
      "│ 8   │ 0.763341  │ 0.6886   │ 0.323448  │ false          │\n",
      "│ 9   │ 1.2331    │ 1.345    │ 0.860602  │ true           │\n",
      "│ 10  │ -2.21915  │ -2.22388 │ -1.54724  │ true           │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 6\n",
      "LASSO false positives = 95\n",
      "LASSO false negatives = 5\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 7\n",
      "\n",
      "current run = 21\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β     │ IHT_β     │ lasso_β   │ marginal_found │\n",
      "│     │ Float64    │ Float64   │ Float64   │ Bool           │\n",
      "├─────┼────────────┼───────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.0729166  │ 0.0       │ 0.0       │ false          │\n",
      "│ 2   │ 1.37107    │ 1.3342    │ 0.743414  │ true           │\n",
      "│ 3   │ -0.329875  │ 0.0       │ 0.0       │ false          │\n",
      "│ 4   │ -0.287766  │ 0.0       │ 0.0       │ false          │\n",
      "│ 5   │ -0.784903  │ -0.599873 │ -0.177016 │ false          │\n",
      "│ 6   │ -0.0281104 │ 0.0       │ 0.0       │ false          │\n",
      "│ 7   │ 1.76022    │ 1.85366   │ 1.11065   │ true           │\n",
      "│ 8   │ 0.591337   │ 0.472736  │ 0.149207  │ true           │\n",
      "│ 9   │ 0.0898311  │ 0.0       │ 0.0       │ false          │\n",
      "│ 10  │ 1.20634    │ 1.23704   │ 0.682132  │ true           │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 5\n",
      "LASSO false positives = 15\n",
      "LASSO false negatives = 5\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 6\n",
      "\n",
      "current run = 22\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β    │ IHT_β     │ lasso_β   │ marginal_found │\n",
      "│     │ Float64   │ Float64   │ Float64   │ Bool           │\n",
      "├─────┼───────────┼───────────┼───────────┼────────────────┤\n",
      "│ 1   │ -1.43685  │ -1.49274  │ -0.80595  │ true           │\n",
      "│ 2   │ 1.50608   │ 1.48354   │ 0.776634  │ true           │\n",
      "│ 3   │ -0.319684 │ 0.0       │ 0.0       │ false          │\n",
      "│ 4   │ -0.122129 │ 0.0       │ 0.0       │ false          │\n",
      "│ 5   │ -0.708054 │ 0.0       │ -0.125475 │ false          │\n",
      "│ 6   │ 2.58886   │ 2.58349   │ 1.54381   │ true           │\n",
      "│ 7   │ 1.56787   │ 1.56851   │ 0.826073  │ true           │\n",
      "│ 8   │ 0.548074  │ 0.608084  │ 0.20419   │ false          │\n",
      "│ 9   │ 1.81003   │ 1.9737    │ 1.0665    │ true           │\n",
      "│ 10  │ -0.681014 │ -0.736614 │ -0.211744 │ false          │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 3\n",
      "LASSO false positives = 118\n",
      "LASSO false negatives = 2\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 5\n",
      "\n",
      "current run = 23\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β    │ IHT_β     │ lasso_β    │ marginal_found │\n",
      "│     │ Float64   │ Float64   │ Float64    │ Bool           │\n",
      "├─────┼───────────┼───────────┼────────────┼────────────────┤\n",
      "│ 1   │ 2.99854   │ 2.88005   │ 2.15337    │ true           │\n",
      "│ 2   │ -0.822753 │ -0.64847  │ -0.353396  │ false          │\n",
      "│ 3   │ -0.696875 │ -0.609089 │ -0.329432  │ false          │\n",
      "│ 4   │ 0.892771  │ 0.753559  │ 0.401263   │ false          │\n",
      "│ 5   │ -0.846611 │ -0.764656 │ -0.400683  │ false          │\n",
      "│ 6   │ -1.82717  │ -1.68357  │ -1.13612   │ true           │\n",
      "│ 7   │ -0.492482 │ 0.0       │ -0.125177  │ false          │\n",
      "│ 8   │ -1.30192  │ -1.19594  │ -0.813449  │ true           │\n",
      "│ 9   │ -0.528306 │ 0.0       │ -0.213617  │ false          │\n",
      "│ 10  │ -0.394907 │ 0.0       │ -0.0850432 │ false          │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 3\n",
      "LASSO false positives = 247\n",
      "LASSO false negatives = 0\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 7\n",
      "\n",
      "current run = 24\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β     │ IHT_β     │ lasso_β   │ marginal_found │\n",
      "│     │ Float64    │ Float64   │ Float64   │ Bool           │\n",
      "├─────┼────────────┼───────────┼───────────┼────────────────┤\n",
      "│ 1   │ 2.19342    │ 2.2103    │ 1.15313   │ true           │\n",
      "│ 2   │ 1.61872    │ 1.79867   │ 0.827057  │ true           │\n",
      "│ 3   │ 1.16179    │ 1.29599   │ 0.573918  │ true           │\n",
      "│ 4   │ -1.66043   │ -1.58206  │ -0.818499 │ true           │\n",
      "│ 5   │ -0.0119048 │ 0.0       │ 0.0       │ false          │\n",
      "│ 6   │ -1.61177   │ -1.90391  │ -0.906506 │ true           │\n",
      "│ 7   │ 1.25813    │ 1.28359   │ 0.582681  │ true           │\n",
      "│ 8   │ -0.440263  │ -0.579474 │ -0.160204 │ false          │\n",
      "│ 9   │ 0.833073   │ 0.854185  │ 0.371048  │ true           │\n",
      "│ 10  │ -0.570173  │ 0.0       │ -0.107591 │ false          │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 2\n",
      "LASSO false positives = 97\n",
      "LASSO false negatives = 1\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 3\n",
      "\n",
      "current run = 25\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β    │ IHT_β     │ lasso_β   │ marginal_found │\n",
      "│     │ Float64   │ Float64   │ Float64   │ Bool           │\n",
      "├─────┼───────────┼───────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.618107  │ 0.484087  │ 0.208888  │ true           │\n",
      "│ 2   │ -0.555343 │ -0.461303 │ -0.183136 │ true           │\n",
      "│ 3   │ -0.855731 │ -0.748098 │ -0.385824 │ true           │\n",
      "│ 4   │ 0.48719   │ 0.0       │ 0.131292  │ false          │\n",
      "│ 5   │ -0.960793 │ -0.977836 │ -0.53363  │ true           │\n",
      "│ 6   │ 0.459022  │ 0.669947  │ 0.317176  │ true           │\n",
      "│ 7   │ -0.653155 │ -0.527939 │ -0.232687 │ true           │\n",
      "│ 8   │ 1.00266   │ 0.891487  │ 0.518964  │ true           │\n",
      "│ 9   │ 0.0263126 │ 0.0       │ 0.0       │ false          │\n",
      "│ 10  │ 1.07986   │ 1.08957   │ 0.658667  │ true           │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 2\n",
      "LASSO false positives = 45\n",
      "LASSO false negatives = 1\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 2\n",
      "\n",
      "current run = 26\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β    │ IHT_β     │ lasso_β    │ marginal_found │\n",
      "│     │ Float64   │ Float64   │ Float64    │ Bool           │\n",
      "├─────┼───────────┼───────────┼────────────┼────────────────┤\n",
      "│ 1   │ 0.648939  │ 0.516321  │ 0.252903   │ true           │\n",
      "│ 2   │ -0.304904 │ 0.0       │ 0.0        │ false          │\n",
      "│ 3   │ 0.404686  │ 0.0       │ 0.0735054  │ false          │\n",
      "│ 4   │ -0.536819 │ -0.491001 │ -0.197632  │ true           │\n",
      "│ 5   │ -0.32004  │ 0.0       │ 0.0        │ false          │\n",
      "│ 6   │ -1.00312  │ -1.01493  │ -0.619642  │ true           │\n",
      "│ 7   │ -0.585353 │ -0.464533 │ -0.171556  │ true           │\n",
      "│ 8   │ -0.439567 │ 0.0       │ -0.046516  │ false          │\n",
      "│ 9   │ 1.01234   │ 0.972297  │ 0.541725   │ true           │\n",
      "│ 10  │ -0.294426 │ 0.0       │ -0.0520321 │ false          │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 5\n",
      "LASSO false positives = 10\n",
      "LASSO false negatives = 2\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 5\n",
      "\n",
      "current run = 27\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β    │ IHT_β     │ lasso_β   │ marginal_found │\n",
      "│     │ Float64   │ Float64   │ Float64   │ Bool           │\n",
      "├─────┼───────────┼───────────┼───────────┼────────────────┤\n",
      "│ 1   │ -0.710093 │ 0.0       │ -0.226087 │ false          │\n",
      "│ 2   │ 2.44475   │ 2.05237   │ 1.40073   │ true           │\n",
      "│ 3   │ 2.3993    │ 2.05322   │ 1.20937   │ true           │\n",
      "│ 4   │ -0.56819  │ 0.0       │ -0.163899 │ false          │\n",
      "│ 5   │ 1.44143   │ 1.16699   │ 0.77438   │ true           │\n",
      "│ 6   │ 1.06554   │ 0.898414  │ 0.491548  │ true           │\n",
      "│ 7   │ 1.10821   │ 0.910588  │ 0.529519  │ true           │\n",
      "│ 8   │ -1.0537   │ -0.902737 │ -0.461429 │ true           │\n",
      "│ 9   │ -0.192688 │ 0.0       │ 0.0       │ false          │\n",
      "│ 10  │ 0.671208  │ 0.0       │ 0.119712  │ false          │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 4\n",
      "LASSO false positives = 186\n",
      "LASSO false negatives = 1\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 4\n",
      "\n",
      "current run = 28\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β    │ IHT_β     │ lasso_β    │ marginal_found │\n",
      "│     │ Float64   │ Float64   │ Float64    │ Bool           │\n",
      "├─────┼───────────┼───────────┼────────────┼────────────────┤\n",
      "│ 1   │ -0.821304 │ -0.795102 │ -0.400808  │ true           │\n",
      "│ 2   │ 0.53879   │ 0.439317  │ 0.149475   │ false          │\n",
      "│ 3   │ 0.386925  │ 0.0       │ 0.058714   │ false          │\n",
      "│ 4   │ 0.205322  │ 0.0       │ 0.083144   │ false          │\n",
      "│ 5   │ -0.513565 │ 0.0       │ -0.0974104 │ false          │\n",
      "│ 6   │ 0.879618  │ 0.874753  │ 0.460458   │ true           │\n",
      "│ 7   │ -1.26338  │ -1.23176  │ -0.687345  │ true           │\n",
      "│ 8   │ 0.541092  │ 0.595176  │ 0.174868   │ false          │\n",
      "│ 9   │ -0.919751 │ -0.998866 │ -0.516211  │ true           │\n",
      "│ 10  │ 1.35169   │ 1.3211    │ 0.745012   │ true           │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 3\n",
      "LASSO false positives = 27\n",
      "LASSO false negatives = 0\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 5\n",
      "\n",
      "current run = 29\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β    │ IHT_β     │ lasso_β   │ marginal_found │\n",
      "│     │ Float64   │ Float64   │ Float64   │ Bool           │\n",
      "├─────┼───────────┼───────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.647175  │ 0.646397  │ 0.306233  │ true           │\n",
      "│ 2   │ -0.288778 │ 0.0       │ 0.0       │ false          │\n",
      "│ 3   │ -1.2357   │ -1.07273  │ -0.567856 │ true           │\n",
      "│ 4   │ -1.0022   │ -0.914245 │ -0.476269 │ true           │\n",
      "│ 5   │ 0.865008  │ 0.710763  │ 0.359861  │ true           │\n",
      "│ 6   │ 0.104902  │ 0.0       │ 0.0       │ false          │\n",
      "│ 7   │ -1.74413  │ -1.78702  │ -1.14665  │ true           │\n",
      "│ 8   │ -0.755781 │ -0.833578 │ -0.383372 │ true           │\n",
      "│ 9   │ -0.556828 │ -0.570334 │ -0.209011 │ false          │\n",
      "│ 10  │ -0.817378 │ -0.98965  │ -0.508371 │ true           │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 2\n",
      "LASSO false positives = 114\n",
      "LASSO false negatives = 2\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 3\n",
      "\n",
      "current run = 30\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β    │ IHT_β     │ lasso_β    │ marginal_found │\n",
      "│     │ Float64   │ Float64   │ Float64    │ Bool           │\n",
      "├─────┼───────────┼───────────┼────────────┼────────────────┤\n",
      "│ 1   │ 0.10572   │ 0.0       │ 0.0        │ false          │\n",
      "│ 2   │ -1.0601   │ -0.961874 │ -0.588594  │ true           │\n",
      "│ 3   │ 0.281754  │ 0.0       │ 0.0126592  │ false          │\n",
      "│ 4   │ 0.702759  │ 0.714604  │ 0.397703   │ true           │\n",
      "│ 5   │ 0.678062  │ 0.607046  │ 0.307615   │ true           │\n",
      "│ 6   │ 0.444067  │ 0.0       │ 0.14931    │ false          │\n",
      "│ 7   │ -0.32708  │ 0.0       │ -0.0182684 │ false          │\n",
      "│ 8   │ -0.151695 │ 0.0       │ 0.0        │ false          │\n",
      "│ 9   │ -0.867023 │ -0.732544 │ -0.400367  │ true           │\n",
      "│ 10  │ 1.57627   │ 1.37311   │ 0.655792   │ true           │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 5\n",
      "LASSO false positives = 59\n",
      "LASSO false negatives = 2\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 5\n",
      "\n",
      "current run = 31\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β     │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64    │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼────────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.716913   │ 0.730665 │ 0.340687  │ true           │\n",
      "│ 2   │ -0.129531  │ 0.0      │ 0.0       │ false          │\n",
      "│ 3   │ 0.0684458  │ 0.0      │ 0.0       │ false          │\n",
      "│ 4   │ 0.012461   │ 0.0      │ 0.0       │ false          │\n",
      "│ 5   │ -1.25199   │ -1.1071  │ -0.683577 │ true           │\n",
      "│ 6   │ 0.468408   │ 0.0      │ 0.0698842 │ false          │\n",
      "│ 7   │ -2.15673   │ -1.883   │ -1.31788  │ true           │\n",
      "│ 8   │ 0.812352   │ 0.625793 │ 0.288022  │ true           │\n",
      "│ 9   │ 0.492849   │ 0.0      │ 0.142097  │ false          │\n",
      "│ 10  │ -0.0782995 │ 0.0      │ 0.0       │ false          │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 6\n",
      "LASSO false positives = 56\n",
      "LASSO false negatives = 4\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 6\n",
      "\n",
      "current run = 32\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β    │ IHT_β     │ lasso_β   │ marginal_found │\n",
      "│     │ Float64   │ Float64   │ Float64   │ Bool           │\n",
      "├─────┼───────────┼───────────┼───────────┼────────────────┤\n",
      "│ 1   │ -0.131514 │ 0.0       │ 0.0       │ false          │\n",
      "│ 2   │ -0.239057 │ 0.0       │ 0.0       │ false          │\n",
      "│ 3   │ 0.739708  │ 0.786217  │ 0.303005  │ true           │\n",
      "│ 4   │ -0.40069  │ -0.514678 │ -0.132088 │ false          │\n",
      "│ 5   │ -0.206226 │ 0.0       │ 0.0       │ false          │\n",
      "│ 6   │ -1.06698  │ -1.03244  │ -0.519745 │ true           │\n",
      "│ 7   │ -0.527112 │ -0.606825 │ -0.158486 │ false          │\n",
      "│ 8   │ -1.9998   │ -2.00012  │ -0.944636 │ true           │\n",
      "│ 9   │ -0.793674 │ -0.909659 │ -0.407194 │ true           │\n",
      "│ 10  │ -2.087    │ -2.07286  │ -1.15161  │ true           │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 3\n",
      "LASSO false positives = 37\n",
      "LASSO false negatives = 3\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 5\n",
      "\n",
      "current run = 33\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β     │ IHT_β     │ lasso_β   │ marginal_found │\n",
      "│     │ Float64    │ Float64   │ Float64   │ Bool           │\n",
      "├─────┼────────────┼───────────┼───────────┼────────────────┤\n",
      "│ 1   │ -0.833228  │ -0.839651 │ -0.427538 │ true           │\n",
      "│ 2   │ 0.676049   │ 0.861334  │ 0.407599  │ true           │\n",
      "│ 3   │ -0.0524694 │ 0.0       │ 0.0       │ false          │\n",
      "│ 4   │ 1.21151    │ 1.23627   │ 0.691798  │ true           │\n",
      "│ 5   │ 0.0947078  │ 0.0       │ 0.0       │ false          │\n",
      "│ 6   │ 1.50923    │ 1.55995   │ 0.76368   │ true           │\n",
      "│ 7   │ -0.332496  │ 0.0       │ -0.107759 │ false          │\n",
      "│ 8   │ 0.129556   │ 0.0       │ 0.0       │ false          │\n",
      "│ 9   │ 0.0771111  │ 0.0       │ 0.0       │ false          │\n",
      "│ 10  │ -1.36522   │ -1.47003  │ -0.85008  │ true           │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 5\n",
      "LASSO false positives = 22\n",
      "LASSO false negatives = 4\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 5\n",
      "\n",
      "current run = 34\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β     │ IHT_β     │ lasso_β   │ marginal_found │\n",
      "│     │ Float64    │ Float64   │ Float64   │ Bool           │\n",
      "├─────┼────────────┼───────────┼───────────┼────────────────┤\n",
      "│ 1   │ -1.38126   │ -1.24039  │ -0.759051 │ true           │\n",
      "│ 2   │ -0.840492  │ -1.11408  │ -0.410707 │ true           │\n",
      "│ 3   │ 0.438593   │ 0.0       │ 0.116544  │ false          │\n",
      "│ 4   │ -1.03938   │ -1.07217  │ -0.594152 │ true           │\n",
      "│ 5   │ -0.0858827 │ 0.0       │ 0.0       │ false          │\n",
      "│ 6   │ 0.626138   │ 0.68347   │ 0.227531  │ true           │\n",
      "│ 7   │ -0.896747  │ -0.937799 │ -0.528478 │ true           │\n",
      "│ 8   │ 0.285343   │ 0.0       │ 0.0       │ false          │\n",
      "│ 9   │ 0.350203   │ 0.0       │ 0.0       │ false          │\n",
      "│ 10  │ 1.12765    │ 1.13302   │ 0.620963  │ true           │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 4\n",
      "LASSO false positives = 34\n",
      "LASSO false negatives = 3\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 4\n",
      "\n",
      "current run = 35\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β    │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64   │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼───────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.402152  │ 0.40475  │ 0.142282  │ false          │\n",
      "│ 2   │ -0.114413 │ 0.0      │ 0.0       │ false          │\n",
      "│ 3   │ -1.07648  │ -1.0582  │ -0.630806 │ true           │\n",
      "│ 4   │ -0.202484 │ 0.0      │ 0.0       │ false          │\n",
      "│ 5   │ 0.245664  │ 0.0      │ 0.070011  │ false          │\n",
      "│ 6   │ 0.345895  │ 0.490015 │ 0.227271  │ true           │\n",
      "│ 7   │ 1.54896   │ 1.55771  │ 1.05454   │ true           │\n",
      "│ 8   │ 0.601544  │ 0.461855 │ 0.183825  │ false          │\n",
      "│ 9   │ 0.755934  │ 0.695707 │ 0.417708  │ true           │\n",
      "│ 10  │ -0.450207 │ 0.0      │ -0.106434 │ false          │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 4\n",
      "LASSO false positives = 125\n",
      "LASSO false negatives = 2\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 6\n",
      "\n",
      "current run = 36\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β     │ IHT_β     │ lasso_β   │ marginal_found │\n",
      "│     │ Float64    │ Float64   │ Float64   │ Bool           │\n",
      "├─────┼────────────┼───────────┼───────────┼────────────────┤\n",
      "│ 1   │ -1.07994   │ -1.24045  │ -0.672641 │ true           │\n",
      "│ 2   │ 1.13449    │ 1.29856   │ 0.715275  │ true           │\n",
      "│ 3   │ -0.830169  │ -1.00354  │ -0.465722 │ true           │\n",
      "│ 4   │ -0.498433  │ -0.572156 │ -0.195041 │ true           │\n",
      "│ 5   │ 0.747752   │ 0.74678   │ 0.314625  │ true           │\n",
      "│ 6   │ 0.993698   │ 1.13797   │ 0.56684   │ true           │\n",
      "│ 7   │ 1.02778    │ 1.16165   │ 0.579419  │ true           │\n",
      "│ 8   │ -0.0267846 │ 0.0       │ 0.0       │ false          │\n",
      "│ 9   │ 0.373882   │ 0.0       │ 0.0       │ false          │\n",
      "│ 10  │ 0.112726   │ 0.0       │ 0.0       │ false          │\n",
      "IHT false positives = 1\n",
      "IHT false negatives = 3\n",
      "LASSO false positives = 40\n",
      "LASSO false negatives = 3\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 3\n",
      "\n",
      "current run = 37\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β    │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64   │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼───────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ -1.15965  │ -1.04207 │ -0.543425 │ true           │\n",
      "│ 2   │ 2.23898   │ 0.0      │ 0.0       │ false          │\n",
      "│ 3   │ 0.0114716 │ 0.0      │ 0.0       │ false          │\n",
      "│ 4   │ 1.50934   │ 1.58251  │ 1.11456   │ true           │\n",
      "│ 5   │ 1.35203   │ 1.43567  │ 0.788611  │ true           │\n",
      "│ 6   │ -0.236083 │ 0.0      │ 0.0       │ false          │\n",
      "│ 7   │ -1.43073  │ -1.49295 │ -1.03436  │ true           │\n",
      "│ 8   │ 0.216307  │ 0.0      │ 0.0       │ false          │\n",
      "│ 9   │ 0.340348  │ 0.0      │ 0.100317  │ false          │\n",
      "│ 10  │ 0.366631  │ 0.449918 │ 0.182651  │ false          │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 5\n",
      "LASSO false positives = 164\n",
      "LASSO false negatives = 4\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 6\n",
      "\n",
      "current run = 38\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β    │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64   │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼───────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.978836  │ 0.960692 │ 0.592603  │ true           │\n",
      "│ 2   │ 0.550849  │ 0.575345 │ 0.265881  │ false          │\n",
      "│ 3   │ -0.307564 │ 0.0      │ 0.0       │ false          │\n",
      "│ 4   │ 1.60565   │ 1.51013  │ 1.00524   │ true           │\n",
      "│ 5   │ 0.546421  │ 0.0      │ 0.0947224 │ false          │\n",
      "│ 6   │ -1.22867  │ -1.28278 │ -0.828981 │ true           │\n",
      "│ 7   │ 1.38029   │ 1.15946  │ 0.298458  │ true           │\n",
      "│ 8   │ -0.317914 │ 0.0      │ 0.0       │ false          │\n",
      "│ 9   │ 0.276958  │ 0.0      │ 0.0729742 │ false          │\n",
      "│ 10  │ 0.229313  │ 0.0      │ 0.0       │ false          │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 5\n",
      "LASSO false positives = 68\n",
      "LASSO false negatives = 3\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 6\n",
      "\n",
      "current run = 39\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β    │ IHT_β    │ lasso_β    │ marginal_found │\n",
      "│     │ Float64   │ Float64  │ Float64    │ Bool           │\n",
      "├─────┼───────────┼──────────┼────────────┼────────────────┤\n",
      "│ 1   │ -0.128251 │ 0.0      │ -0.0489092 │ false          │\n",
      "│ 2   │ 0.919801  │ 1.00532  │ 0.583379   │ true           │\n",
      "│ 3   │ -1.28463  │ -1.28083 │ -0.75971   │ true           │\n",
      "│ 4   │ 0.101823  │ 0.0      │ 0.0        │ false          │\n",
      "│ 5   │ 1.71113   │ 1.66499  │ 0.999364   │ true           │\n",
      "│ 6   │ 0.177542  │ 0.0      │ 0.0        │ false          │\n",
      "│ 7   │ -2.13612  │ -2.20729 │ -1.51551   │ true           │\n",
      "│ 8   │ -1.4868   │ -1.26926 │ -0.751219  │ true           │\n",
      "│ 9   │ -0.418656 │ 0.0      │ -0.0878921 │ false          │\n",
      "│ 10  │ -0.853007 │ -0.72796 │ -0.323656  │ false          │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 4\n",
      "LASSO false positives = 185\n",
      "LASSO false negatives = 2\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 5\n",
      "\n",
      "current run = 40\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β    │ IHT_β     │ lasso_β   │ marginal_found │\n",
      "│     │ Float64   │ Float64   │ Float64   │ Bool           │\n",
      "├─────┼───────────┼───────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.497199  │ 0.43668   │ 0.105798  │ false          │\n",
      "│ 2   │ -2.38356  │ -2.25872  │ -1.52439  │ true           │\n",
      "│ 3   │ -0.228254 │ 0.0       │ 0.0       │ false          │\n",
      "│ 4   │ 0.294025  │ 0.0       │ 0.0       │ false          │\n",
      "│ 5   │ 0.302579  │ 0.0       │ 0.0       │ false          │\n",
      "│ 6   │ 0.525926  │ 0.678118  │ 0.17963   │ true           │\n",
      "│ 7   │ 0.0246983 │ 0.0       │ 0.0       │ false          │\n",
      "│ 8   │ 1.37478   │ 1.42816   │ 0.474576  │ true           │\n",
      "│ 9   │ -0.678537 │ -0.640995 │ -0.248247 │ true           │\n",
      "│ 10  │ 0.0118296 │ 0.0       │ 0.0       │ false          │\n",
      "IHT false positives = 1\n",
      "IHT false negatives = 5\n",
      "LASSO false positives = 5\n",
      "LASSO false negatives = 5\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 6\n",
      "\n",
      "current run = 41\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β     │ IHT_β     │ lasso_β    │ marginal_found │\n",
      "│     │ Float64    │ Float64   │ Float64    │ Bool           │\n",
      "├─────┼────────────┼───────────┼────────────┼────────────────┤\n",
      "│ 1   │ -0.762556  │ -0.85662  │ -0.36877   │ true           │\n",
      "│ 2   │ -1.37821   │ -1.37789  │ -0.748552  │ true           │\n",
      "│ 3   │ 1.37773    │ 1.64538   │ 0.938508   │ true           │\n",
      "│ 4   │ -0.0696107 │ 0.0       │ 0.0        │ false          │\n",
      "│ 5   │ 0.116563   │ 0.0       │ 0.0        │ false          │\n",
      "│ 6   │ -0.989843  │ -1.1038   │ -0.564057  │ true           │\n",
      "│ 7   │ -0.564463  │ -0.61622  │ -0.245349  │ true           │\n",
      "│ 8   │ -0.334853  │ 0.0       │ -0.0586704 │ false          │\n",
      "│ 9   │ 0.548853   │ 0.738772  │ 0.276241   │ false          │\n",
      "│ 10  │ -0.707457  │ -0.805624 │ -0.382965  │ true           │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 3\n",
      "LASSO false positives = 57\n",
      "LASSO false negatives = 2\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 4\n",
      "\n",
      "current run = 42\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β    │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64   │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼───────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 1.64416   │ 1.32667  │ 0.705941  │ true           │\n",
      "│ 2   │ -0.984903 │ -1.01698 │ -0.555603 │ true           │\n",
      "│ 3   │ -2.06212  │ -1.92669 │ -1.23404  │ true           │\n",
      "│ 4   │ -0.100535 │ 0.0      │ 0.0       │ false          │\n",
      "│ 5   │ 0.467878  │ 0.0      │ 0.182284  │ false          │\n",
      "│ 6   │ 0.247928  │ 0.0      │ 0.0160573 │ false          │\n",
      "│ 7   │ 0.501326  │ 0.0      │ 0.098012  │ false          │\n",
      "│ 8   │ 1.6264    │ 1.47369  │ 0.80778   │ true           │\n",
      "│ 9   │ -1.29208  │ -1.24915 │ -0.701926 │ true           │\n",
      "│ 10  │ -0.112418 │ 0.0      │ 0.0       │ false          │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 5\n",
      "LASSO false positives = 106\n",
      "LASSO false negatives = 2\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 5\n",
      "\n",
      "current run = 43\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β    │ IHT_β     │ lasso_β   │ marginal_found │\n",
      "│     │ Float64   │ Float64   │ Float64   │ Bool           │\n",
      "├─────┼───────────┼───────────┼───────────┼────────────────┤\n",
      "│ 1   │ -0.773332 │ -0.694933 │ -0.325084 │ true           │\n",
      "│ 2   │ 0.665313  │ 0.562918  │ 0.175279  │ false          │\n",
      "│ 3   │ 0.293619  │ 0.0       │ 0.0       │ false          │\n",
      "│ 4   │ -1.86182  │ -1.71732  │ -1.08446  │ true           │\n",
      "│ 5   │ -1.30875  │ -1.06739  │ -0.60336  │ true           │\n",
      "│ 6   │ 0.64743   │ 0.0       │ 0.0897505 │ false          │\n",
      "│ 7   │ 0.066356  │ 0.0       │ 0.0       │ false          │\n",
      "│ 8   │ 0.381098  │ 0.0       │ 0.0       │ false          │\n",
      "│ 9   │ 1.08931   │ 1.01072   │ 0.545334  │ true           │\n",
      "│ 10  │ 0.124671  │ 0.0       │ 0.0       │ false          │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 5\n",
      "LASSO false positives = 22\n",
      "LASSO false negatives = 4\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 6\n",
      "\n",
      "current run = 44\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β    │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64   │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼───────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ -0.566488 │ 0.0      │ -0.143765 │ false          │\n",
      "│ 2   │ -0.592059 │ -0.77728 │ -0.398561 │ true           │\n",
      "│ 3   │ 0.358378  │ 0.0      │ 0.0755532 │ false          │\n",
      "│ 4   │ -0.177563 │ 0.0      │ 0.0       │ false          │\n",
      "│ 5   │ 1.32632   │ 1.29903  │ 0.788833  │ true           │\n",
      "│ 6   │ 1.44183   │ 1.47903  │ 0.840103  │ true           │\n",
      "│ 7   │ -0.389083 │ 0.0      │ -0.126715 │ false          │\n",
      "│ 8   │ 0.527468  │ 0.565748 │ 0.253032  │ true           │\n",
      "│ 9   │ 1.16854   │ 1.1434   │ 0.628817  │ true           │\n",
      "│ 10  │ 1.03984   │ 0.923035 │ 0.483142  │ true           │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 4\n",
      "LASSO false positives = 62\n",
      "LASSO false negatives = 1\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 4\n",
      "\n",
      "current run = 45\n",
      "\u001b[31mDid not converge!!!!! The run time for IHT was 14.030655860900879 seconds and model size was9\u001b[39m\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β    │ IHT_β     │ lasso_β   │ marginal_found │\n",
      "│     │ Float64   │ Float64   │ Float64   │ Bool           │\n",
      "├─────┼───────────┼───────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.574367  │ 0.64419   │ 0.237107  │ false          │\n",
      "│ 2   │ 1.39576   │ 1.42267   │ 0.817152  │ true           │\n",
      "│ 3   │ -0.791119 │ -0.790308 │ -0.355921 │ true           │\n",
      "│ 4   │ -0.708916 │ -0.674372 │ -0.299415 │ true           │\n",
      "│ 5   │ 0.0573325 │ 0.0       │ 0.0       │ false          │\n",
      "│ 6   │ 0.948541  │ 1.09207   │ 0.0984813 │ false          │\n",
      "│ 7   │ 0.89135   │ 0.752353  │ 0.336892  │ true           │\n",
      "│ 8   │ 1.36034   │ 1.30087   │ 0.706616  │ true           │\n",
      "│ 9   │ -0.708206 │ -0.617912 │ -0.267072 │ true           │\n",
      "│ 10  │ -0.574003 │ -0.799962 │ -0.357226 │ true           │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 1\n",
      "LASSO false positives = 85\n",
      "LASSO false negatives = 1\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 3\n",
      "\n",
      "current run = 46\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β    │ IHT_β   │ lasso_β    │ marginal_found │\n",
      "│     │ Float64   │ Float64 │ Float64    │ Bool           │\n",
      "├─────┼───────────┼─────────┼────────────┼────────────────┤\n",
      "│ 1   │ 0.0579335 │ 0.0     │ 0.0        │ false          │\n",
      "│ 2   │ -0.307419 │ 0.0     │ -0.0125462 │ false          │\n",
      "│ 3   │ -0.11986  │ 0.0     │ 0.0        │ false          │\n",
      "│ 4   │ 1.318     │ 1.26452 │ 0.734604   │ true           │\n",
      "│ 5   │ 1.1655    │ 1.05663 │ 0.702654   │ true           │\n",
      "│ 6   │ 0.308425  │ 0.0     │ 0.0579675  │ false          │\n",
      "│ 7   │ 0.656646  │ 0.0     │ 0.155508   │ false          │\n",
      "│ 8   │ 0.1531    │ 0.0     │ 0.0        │ false          │\n",
      "│ 9   │ 0.34867   │ 0.0     │ 0.134663   │ false          │\n",
      "│ 10  │ 1.30663   │ 1.11078 │ 0.753688   │ true           │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 7\n",
      "LASSO false positives = 31\n",
      "LASSO false negatives = 3\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 7\n",
      "\n",
      "current run = 47\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β     │ IHT_β     │ lasso_β   │ marginal_found │\n",
      "│     │ Float64    │ Float64   │ Float64   │ Bool           │\n",
      "├─────┼────────────┼───────────┼───────────┼────────────────┤\n",
      "│ 1   │ 1.65733    │ 1.57434   │ 0.976762  │ true           │\n",
      "│ 2   │ 0.893532   │ 1.04444   │ 0.546474  │ true           │\n",
      "│ 3   │ -0.838084  │ -0.830393 │ -0.415405 │ true           │\n",
      "│ 4   │ 0.213567   │ 0.0       │ 0.0333913 │ false          │\n",
      "│ 5   │ 0.628961   │ 0.652329  │ 0.245186  │ false          │\n",
      "│ 6   │ -0.0400687 │ 0.0       │ 0.0       │ false          │\n",
      "│ 7   │ 0.360368   │ 0.0       │ 0.147262  │ false          │\n",
      "│ 8   │ -1.29039   │ -1.1376   │ -0.666352 │ true           │\n",
      "│ 9   │ -1.5451    │ -1.44992  │ -0.899682 │ true           │\n",
      "│ 10  │ 0.794447   │ 0.751434  │ 0.380203  │ true           │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 3\n",
      "LASSO false positives = 143\n",
      "LASSO false negatives = 1\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 4\n",
      "\n",
      "current run = 48\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β    │ IHT_β     │ lasso_β   │ marginal_found │\n",
      "│     │ Float64   │ Float64   │ Float64   │ Bool           │\n",
      "├─────┼───────────┼───────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.23845   │ 0.0       │ 0.0       │ false          │\n",
      "│ 2   │ -1.58552  │ -1.65082  │ -1.02843  │ true           │\n",
      "│ 3   │ -0.923084 │ -0.850873 │ -0.445237 │ true           │\n",
      "│ 4   │ 0.563645  │ 0.721225  │ 0.246055  │ false          │\n",
      "│ 5   │ 0.480607  │ 0.710043  │ 0.328647  │ false          │\n",
      "│ 6   │ -0.671179 │ -0.888208 │ -0.430339 │ true           │\n",
      "│ 7   │ 0.331685  │ 0.0       │ 0.0503729 │ false          │\n",
      "│ 8   │ 1.07535   │ 1.11109   │ 0.619289  │ true           │\n",
      "│ 9   │ -1.03137  │ -1.09533  │ -0.640948 │ true           │\n",
      "│ 10  │ 1.26969   │ 1.47817   │ 0.930562  │ true           │\n",
      "IHT false positives = 1\n",
      "IHT false negatives = 2\n",
      "LASSO false positives = 184\n",
      "LASSO false negatives = 1\n",
      "marginal false positives = 1\n",
      "marginal false negatives = 4\n",
      "\n",
      "current run = 49\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β    │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64   │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼───────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.407838  │ 0.0      │ 0.124524  │ false          │\n",
      "│ 2   │ -1.99473  │ -1.96063 │ -1.05566  │ true           │\n",
      "│ 3   │ -1.78315  │ -1.86801 │ -0.844052 │ true           │\n",
      "│ 4   │ 0.326483  │ 0.0      │ 0.122877  │ false          │\n",
      "│ 5   │ 0.358946  │ 0.0      │ 0.0714571 │ false          │\n",
      "│ 6   │ -0.177961 │ 0.0      │ 0.0       │ false          │\n",
      "│ 7   │ -0.385924 │ 0.0      │ 0.0       │ false          │\n",
      "│ 8   │ 0.490134  │ 0.491043 │ 0.23377   │ true           │\n",
      "│ 9   │ 1.66554   │ 1.64923  │ 0.801539  │ true           │\n",
      "│ 10  │ 1.31769   │ 1.40167  │ 0.61041   │ true           │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 5\n",
      "LASSO false positives = 132\n",
      "LASSO false negatives = 2\n",
      "marginal false positives = 1\n",
      "marginal false negatives = 5\n",
      "\n",
      "current run = 50\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β      │ IHT_β     │ lasso_β   │ marginal_found │\n",
      "│     │ Float64     │ Float64   │ Float64   │ Bool           │\n",
      "├─────┼─────────────┼───────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.0671508   │ 0.0       │ 0.0       │ false          │\n",
      "│ 2   │ -2.24678    │ -2.17098  │ -1.352    │ true           │\n",
      "│ 3   │ -0.619753   │ -0.657227 │ -0.246502 │ false          │\n",
      "│ 4   │ 0.697065    │ 0.630009  │ 0.235924  │ false          │\n",
      "│ 5   │ -0.00615383 │ 0.0       │ 0.0       │ false          │\n",
      "│ 6   │ -1.6053     │ -1.58095  │ -0.896052 │ true           │\n",
      "│ 7   │ 1.02524     │ 0.961913  │ 0.459899  │ true           │\n",
      "│ 8   │ 0.429733    │ 0.0       │ 0.0       │ false          │\n",
      "│ 9   │ -0.628775   │ -0.522489 │ -0.171661 │ false          │\n",
      "│ 10  │ -0.629107   │ -0.658617 │ -0.292468 │ true           │\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 3\n",
      "LASSO false positives = 67\n",
      "LASSO false negatives = 3\n",
      "marginal false positives = 1\n",
      "marginal false negatives = 6\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "([0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0  …  0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0], [3.0, 2.0, 3.0, 6.0, 5.0, 5.0, 2.0, 2.0, 3.0, 2.0  …  3.0, 5.0, 5.0, 4.0, 1.0, 7.0, 3.0, 2.0, 5.0, 3.0], [167.0, 85.0, 221.0, 19.0, 26.0, 44.0, 15.0, 57.0, 79.0, 276.0  …  57.0, 106.0, 22.0, 62.0, 85.0, 31.0, 143.0, 184.0, 132.0, 67.0], [2.0, 1.0, 2.0, 4.0, 5.0, 4.0, 1.0, 2.0, 0.0, 1.0  …  2.0, 2.0, 4.0, 1.0, 1.0, 3.0, 1.0, 1.0, 2.0, 3.0], [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0  …  0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 1.0, 1.0], [5.0, 5.0, 3.0, 5.0, 5.0, 6.0, 4.0, 3.0, 4.0, 4.0  …  4.0, 5.0, 6.0, 4.0, 3.0, 7.0, 4.0, 4.0, 5.0, 6.0])"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "iht_false_positives, iht_false_negatives, lasso_false_positives, \n",
    "    lasso_false_negatives, marginal_false_positives, marginal_false_negatives = run()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "6-element Array{Float64,1}:\n",
       "  0.1 \n",
       "  3.84\n",
       " 88.54\n",
       "  2.38\n",
       "  0.06\n",
       "  4.78"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bernoulli_iht_false_positives = sum(iht_false_positives) / 50\n",
    "bernoulli_iht_false_negatives = sum(iht_false_negatives) / 50\n",
    "bernoulli_lasso_false_positives = sum(lasso_false_positives) / 50\n",
    "bernoulli_lasso_false_negatives = sum(lasso_false_negatives) / 50\n",
    "bernoulli_marginal_false_positives = sum(marginal_false_positives) / 50\n",
    "bernoulli_marginal_false_negatives = sum(marginal_false_negatives) / 50\n",
    "\n",
    "result = [bernoulli_iht_false_positives; bernoulli_iht_false_negatives; \n",
    "        bernoulli_lasso_false_positives; bernoulli_lasso_false_negatives;\n",
    "        bernoulli_marginal_false_positives; bernoulli_marginal_false_negatives]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Julia 1.0.3",
   "language": "julia",
   "name": "julia-1.0"
  },
  "language_info": {
   "file_extension": ".jl",
   "mimetype": "application/julia",
   "name": "julia",
   "version": "1.0.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
