{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "# Poisson: compare false positive/negatives with LASSO\n",
    "\n",
    "LASSO is currently the de facto penalized least squares method for [feature selection](https://en.wikipedia.org/wiki/Feature_selection). Here we compare the performance (in terms of the number of false positives/negative) of LASSO with IHT for GWAS data, using the `glmnet` implementation of cyclic coordinate descent for LASSO. Since the focus here is not scalability, we test our sample problems on moderately sized genotype matrces of 1000 samples with 10,000 SNPs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "5"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "using Distributed\n",
    "addprocs(4)\n",
    "nprocs()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "┌ Info: Recompiling stale cache file /Users/biona001/.julia/compiled/v1.0/MendelGWAS/y6a7g.ji for MendelGWAS [c1bad3ca-268b-5bf6-b82e-de8a361f94db]\n",
      "└ @ Base loading.jl:1190\n",
      "┌ Info: Loading DataFrames support into Gadfly.jl\n",
      "└ @ Gadfly /Users/biona001/.julia/packages/Gadfly/09PWZ/src/mapping.jl:228\n",
      "┌ Info: Recompiling stale cache file /Users/biona001/.julia/compiled/v1.0/RCall/8GFyb.ji for RCall [6f49c342-dc21-5d91-9882-a32aef131414]\n",
      "└ @ Base loading.jl:1190\n",
      "┌ Warning: RCall.jl:  要求されたパッケージ Matrix をロード中です \n",
      "│  要求されたパッケージ foreach をロード中です \n",
      "│ Loaded glmnet 2.0-16\n",
      "│ \n",
      "└ @ RCall /Users/biona001/.julia/packages/RCall/ffM0W/src/io.jl:113\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "RObject{StrSxp}\n",
       " [1] \"glmnet\"    \"foreach\"   \"Matrix\"    \"stats\"     \"graphics\"  \"grDevices\"\n",
       " [7] \"utils\"     \"datasets\"  \"methods\"   \"base\"     \n"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "using MendelIHT\n",
    "using SnpArrays\n",
    "using DataFrames\n",
    "using Distributions\n",
    "using Random\n",
    "using LinearAlgebra\n",
    "using DelimitedFiles\n",
    "using GLM\n",
    "using MendelGWAS\n",
    "using CSV\n",
    "using RCall\n",
    "R\"library(glmnet)\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "iht_lasso_marginal (generic function with 1 method)"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "function iht_lasso_marginal(n::Int64, p::Int64, k::Int64, d::UnionAll, l::Link)\n",
    "    #construct snpmatrix, covariate files, and true model b\n",
    "    x, = simulate_random_snparray(n, p, \"poisson.bed\")\n",
    "    xbm = SnpBitMatrix{Float64}(x, model=ADDITIVE_MODEL, center=true, scale=true); \n",
    "    z = ones(n, 1) # the intercept\n",
    "    x_float = [convert(Matrix{Float64}, x, center=true, scale=true) z] #Float64 version of x\n",
    "\n",
    "    # simulate response, true model b, and the correct non-0 positions of b\n",
    "    y, true_b, correct_position = simulate_random_response(x, xbm, k, d, l)\n",
    " \n",
    "    #specify path and folds\n",
    "    num_folds = 3\n",
    "    folds = rand(1:num_folds, size(x, 1));\n",
    "\n",
    "    #run glmnet via Rcall\n",
    "    @rput x_float y folds num_folds #make variables visible to R\n",
    "    R\"lasso_cv_result = cv.glmnet(x_float, y, nfolds = num_folds, foldid = folds, family='poisson')\"\n",
    "    R\"lasso_beta_tmp = glmnet(x_float, y, lambda=lasso_cv_result$lambda.min, family='poisson')$beta\"\n",
    "    R\"lasso_beta = as.vector(lasso_beta_tmp)\"\n",
    "    @rget lasso_cv_result lasso_beta #pull result from R to Julia\n",
    "    lasso_k_est = count(!iszero, lasso_beta)\n",
    "    \n",
    "    #find non-zero entries returned by best lasso model as largest k estimate\n",
    "    path = collect(1:50);\n",
    "    \n",
    "    #run IHT's cross validation routine \n",
    "    mses = cv_iht(d(), l, x, z, y, 1, path, folds, num_folds, use_maf=false, debias=false, showinfo=false, parallel=true);\n",
    "    iht_k_est = argmin(mses)\n",
    "    iht_result = L0_reg(x, xbm, z, y, 1, iht_k_est, d(), l, debias=false, init=false, use_maf=false)\n",
    "    iht_beta = iht_result.beta\n",
    "    \n",
    "    #Now run MendelGWAS\n",
    "    make_bim_fam_files(x, y, \"poisson\") #create .bim and .bed files for MendelGWAS\n",
    "\n",
    "    #create a control file for MendelGWAS\n",
    "    open(\"poisson_control.txt\", \"w\") do f\n",
    "        write(f, \"plink_input_basename = poisson \\n\")\n",
    "        write(f, \"plink_field_separator = '\\t' \\n\")\n",
    "        write(f, \"output_table = poisson_table.txt \\n\")\n",
    "        write(f, \"regression = Poisson \\n\")\n",
    "        write(f, \"regression_formula = Trait ~ \\n\")\n",
    "    end\n",
    "\n",
    "    #run marginal analysis\n",
    "    GWAS(\"poisson_control.txt\");\n",
    "\n",
    "    # calculate false positive/negatives based on bonferroni correction\n",
    "    println(\"reached here\")\n",
    "    p_values = CSV.read(\"poisson_table.txt\", delim=',', header=true)[:Pvalue]\n",
    "    println(\"reached here too\")\n",
    "    significance = 0.05 / Float64(p)\n",
    "    println(\"reached here 3\")\n",
    "    passing_snps_position = findall(p_values .<= significance)\n",
    "    println(\"reached here 4\")\n",
    "    marginal_found = [correct_position[snp] in passing_snps_position for snp in 1:k]\n",
    "    println(\"reached here 5\")\n",
    "\n",
    "    #clean up\n",
    "    rm(\"poisson.bed\", force=true)\n",
    "    rm(\"poisson.bim\", force=true)\n",
    "    rm(\"poisson.fam\", force=true)\n",
    "    rm(\"poisson_table.txt\", force=true)\n",
    "    rm(\"poisson_control.txt\", force=true)\n",
    "    rm(\"Mendel_Output.txt\", force=true)\n",
    "    println(\"reached here 6\")\n",
    "\n",
    "    #show lasso and IHT's reconstruction result\n",
    "    compare_model = DataFrame(\n",
    "        true_β  = true_b[correct_position], \n",
    "        IHT_β   = iht_beta[correct_position],\n",
    "        lasso_β = lasso_beta[correct_position],\n",
    "        marginal_found = marginal_found)\n",
    "    @show compare_model\n",
    "    println(\"reached here 7\")\n",
    "\n",
    "    \n",
    "    #compute true/false positives/negatives for IHT and lasso\n",
    "    iht_tp = count(!iszero, iht_beta[correct_position])\n",
    "    iht_fp = iht_k_est - iht_tp\n",
    "    iht_fn = k - iht_tp\n",
    "    lasso_tp = count(!iszero, lasso_beta[correct_position])\n",
    "    lasso_fp = lasso_k_est - lasso_tp\n",
    "    lasso_fn = k - lasso_tp\n",
    "    marginal_tp = count(!iszero, true_b[passing_snps_position])\n",
    "    marginal_fp = length(passing_snps_position) - marginal_tp\n",
    "    marginal_fn = k - marginal_tp\n",
    "    \n",
    "    println(\"IHT false positives = $iht_fp\")\n",
    "    println(\"IHT false negatives = $iht_fn\")\n",
    "    println(\"LASSO false positives = $lasso_fp\")\n",
    "    println(\"LASSO false negatives = $lasso_fn\")\n",
    "    println(\"marginal false positives = $marginal_fp\")\n",
    "    println(\"marginal false negatives = $marginal_fn\" * \"\\n\")\n",
    "    \n",
    "    return iht_fp, iht_fn, lasso_fp, lasso_fn, marginal_fp, marginal_fn\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true,
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "run (generic function with 1 method)"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "function run()\n",
    "    #simulat data with k true predictors, from distribution d and with link l.\n",
    "    n = 1000\n",
    "    p = 10000\n",
    "    k = 10\n",
    "    d = Poisson\n",
    "    l = canonicallink(d())\n",
    "\n",
    "    #run function above, saving results in 4 vectors\n",
    "    total_runs = 50\n",
    "    iht_false_positives = zeros(total_runs)\n",
    "    iht_false_negatives = zeros(total_runs)\n",
    "    lasso_false_positives = zeros(total_runs)\n",
    "    lasso_false_negatives = zeros(total_runs)\n",
    "    marginal_false_positives = zeros(total_runs)\n",
    "    marginal_false_negatives = zeros(total_runs)\n",
    "    for i in 1:total_runs\n",
    "        println(\"current run = $i\")\n",
    "        \n",
    "        #set random seed\n",
    "        Random.seed!(i)\n",
    "        \n",
    "        iht_fp, iht_fn, lasso_fp, lasso_fn, m_fp, m_fn = iht_lasso_marginal(n, p, k, d, l)\n",
    "        iht_false_positives[i] = iht_fp\n",
    "        iht_false_negatives[i] = iht_fn\n",
    "        lasso_false_positives[i] = lasso_fp\n",
    "        lasso_false_negatives[i] = lasso_fn\n",
    "        marginal_false_positives[i] = m_fp\n",
    "        marginal_false_negatives[i] = m_fn\n",
    "    end\n",
    "    \n",
    "    return iht_false_positives, iht_false_negatives, lasso_false_positives, \n",
    "            lasso_false_negatives, marginal_false_positives, marginal_false_negatives\n",
    "end"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Note: this code was ran on cluster\n",
    "\n",
    "The code above was run via the command below on cluster machine. The result was saved, and will be imported later."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "current run = 1\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/Users/biona001/.julia/dev/MendelIHT/figures/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = poisson_control.txt\n",
      "  output_table = poisson_table.txt\n",
      "  pedigree_file = poisson.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = poisson\n",
      "  regression = Poisson\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = poisson.bed\n",
      "  snpdefinition_file = poisson.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "reached here\n",
      "reached here too\n",
      "reached here 3\n",
      "reached here 4\n",
      "reached here 5\n",
      "reached here 6\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β     │ IHT_β     │ lasso_β   │ marginal_found │\n",
      "│     │ Float64    │ Float64   │ Float64   │ Bool           │\n",
      "├─────┼────────────┼───────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.182262   │ 0.152774  │ 0.0966563 │ true           │\n",
      "│ 2   │ -0.459012  │ -0.454145 │ -0.34263  │ true           │\n",
      "│ 3   │ 0.433249   │ 0.434175  │ 0.358665  │ true           │\n",
      "│ 4   │ 0.525909   │ 0.552334  │ 0.489506  │ true           │\n",
      "│ 5   │ 0.209601   │ 0.202243  │ 0.146661  │ true           │\n",
      "│ 6   │ -0.507486  │ -0.524873 │ -0.366219 │ true           │\n",
      "│ 7   │ -0.0635574 │ 0.0       │ 0.0       │ false          │\n",
      "│ 8   │ 0.0277683  │ 0.0       │ 0.0       │ false          │\n",
      "│ 9   │ 0.183566   │ 0.198759  │ 0.137637  │ true           │\n",
      "│ 10  │ -0.456284  │ -0.463115 │ -0.223589 │ true           │\n",
      "reached here 7\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 2\n",
      "LASSO false positives = 68\n",
      "LASSO false negatives = 2\n",
      "marginal false positives = 200\n",
      "marginal false negatives = 2\n",
      "\n",
      "current run = 2\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "┌ Warning: RCall.jl:  警告:  from glmnet Fortran code (error code -1); Convergence for 1th lambda value not reached after maxit=100000 iterations; solutions for larger lambdas returned\n",
      "│  getcoef(fit, nvars, nx, vnames) で警告がありました: \n",
      "│   an empty model has been returned; probably a convergence issue\n",
      "└ @ RCall /Users/biona001/.julia/packages/RCall/ffM0W/src/io.jl:113\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[31mDid not converge!!!!! The run time for IHT was 10.322705030441284 seconds and model size was13\u001b[39m\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/Users/biona001/.julia/dev/MendelIHT/figures/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = poisson_control.txt\n",
      "  output_table = poisson_table.txt\n",
      "  pedigree_file = poisson.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = poisson\n",
      "  regression = Poisson\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = poisson.bed\n",
      "  snpdefinition_file = poisson.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "reached here\n",
      "reached here too\n",
      "reached here 3\n",
      "reached here 4\n",
      "reached here 5\n",
      "reached here 6\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β    │ IHT_β   │ lasso_β │ marginal_found │\n",
      "│     │ Float64   │ Float64 │ Float64 │ Bool           │\n",
      "├─────┼───────────┼─────────┼─────────┼────────────────┤\n",
      "│ 1   │ 0.224928  │ 0.0     │ 0.0     │ true           │\n",
      "│ 2   │ 0.20338   │ 0.0     │ 0.0     │ true           │\n",
      "│ 3   │ 0.0216292 │ 0.0     │ 0.0     │ true           │\n",
      "│ 4   │ -0.340295 │ 0.0     │ 0.0     │ true           │\n",
      "│ 5   │ -0.248243 │ 0.0     │ 0.0     │ true           │\n",
      "│ 6   │ 0.339818  │ 0.0     │ 0.0     │ true           │\n",
      "│ 7   │ -0.269453 │ 0.0     │ 0.0     │ true           │\n",
      "│ 8   │ -0.765255 │ 0.0     │ 0.0     │ true           │\n",
      "│ 9   │ 0.570137  │ 1.0935  │ 0.0     │ false          │\n",
      "│ 10  │ 0.200677  │ 0.0     │ 0.0     │ true           │\n",
      "reached here 7\n",
      "IHT false positives = 12\n",
      "IHT false negatives = 9\n",
      "LASSO false positives = 0\n",
      "LASSO false negatives = 10\n",
      "marginal false positives = 7871\n",
      "marginal false negatives = 1\n",
      "\n",
      "current run = 3\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/Users/biona001/.julia/dev/MendelIHT/figures/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = poisson_control.txt\n",
      "  output_table = poisson_table.txt\n",
      "  pedigree_file = poisson.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = poisson\n",
      "  regression = Poisson\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = poisson.bed\n",
      "  snpdefinition_file = poisson.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "reached here\n",
      "reached here too\n",
      "reached here 3\n",
      "reached here 4\n",
      "reached here 5\n",
      "reached here 6\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β    │ IHT_β     │ lasso_β    │ marginal_found │\n",
      "│     │ Float64   │ Float64   │ Float64    │ Bool           │\n",
      "├─────┼───────────┼───────────┼────────────┼────────────────┤\n",
      "│ 1   │ -0.388184 │ -0.357407 │ -0.267704  │ true           │\n",
      "│ 2   │ 0.235876  │ 0.216451  │ 0.149398   │ true           │\n",
      "│ 3   │ -0.619478 │ -0.593084 │ -0.230178  │ true           │\n",
      "│ 4   │ 0.162123  │ 0.188745  │ 0.115034   │ true           │\n",
      "│ 5   │ 0.447907  │ 0.43613   │ 0.377016   │ true           │\n",
      "│ 6   │ -0.192024 │ -0.190117 │ -0.0855768 │ true           │\n",
      "│ 7   │ 0.0137185 │ 0.0       │ 0.0        │ false          │\n",
      "│ 8   │ -0.47331  │ -0.454614 │ -0.359827  │ true           │\n",
      "│ 9   │ -0.468533 │ -0.478239 │ -0.375317  │ true           │\n",
      "│ 10  │ 0.0382465 │ 0.0       │ 0.0        │ false          │\n",
      "reached here 7\n",
      "IHT false positives = 1\n",
      "IHT false negatives = 2\n",
      "LASSO false positives = 118\n",
      "LASSO false negatives = 2\n",
      "marginal false positives = 60\n",
      "marginal false negatives = 2\n",
      "\n",
      "current run = 4\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/Users/biona001/.julia/dev/MendelIHT/figures/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = poisson_control.txt\n",
      "  output_table = poisson_table.txt\n",
      "  pedigree_file = poisson.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = poisson\n",
      "  regression = Poisson\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = poisson.bed\n",
      "  snpdefinition_file = poisson.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "reached here\n",
      "reached here too\n",
      "reached here 3\n",
      "reached here 4\n",
      "reached here 5\n",
      "reached here 6\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β     │ IHT_β     │ lasso_β    │ marginal_found │\n",
      "│     │ Float64    │ Float64   │ Float64    │ Bool           │\n",
      "├─────┼────────────┼───────────┼────────────┼────────────────┤\n",
      "│ 1   │ 0.303464   │ 0.305611  │ 0.239574   │ true           │\n",
      "│ 2   │ -0.112781  │ -0.201106 │ -0.0948608 │ true           │\n",
      "│ 3   │ -0.272873  │ -0.349354 │ -0.228456  │ true           │\n",
      "│ 4   │ -0.612586  │ -0.600902 │ -0.429166  │ true           │\n",
      "│ 5   │ -0.374362  │ -0.389764 │ -0.180764  │ true           │\n",
      "│ 6   │ -0.0657801 │ 0.0       │ -0.0222101 │ false          │\n",
      "│ 7   │ -0.0186208 │ 0.0       │ 0.0        │ false          │\n",
      "│ 8   │ -0.147961  │ 0.0       │ -0.0513599 │ true           │\n",
      "│ 9   │ 0.0220663  │ 0.0       │ 0.0        │ false          │\n",
      "│ 10  │ -0.122547  │ 0.0       │ -0.0479465 │ false          │\n",
      "reached here 7\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 5\n",
      "LASSO false positives = 36\n",
      "LASSO false negatives = 2\n",
      "marginal false positives = 19\n",
      "marginal false negatives = 4\n",
      "\n",
      "current run = 5\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/Users/biona001/.julia/dev/MendelIHT/figures/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = poisson_control.txt\n",
      "  output_table = poisson_table.txt\n",
      "  pedigree_file = poisson.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = poisson\n",
      "  regression = Poisson\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = poisson.bed\n",
      "  snpdefinition_file = poisson.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "reached here\n",
      "reached here too\n",
      "reached here 3\n",
      "reached here 4\n",
      "reached here 5\n",
      "reached here 6\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β      │ IHT_β     │ lasso_β    │ marginal_found │\n",
      "│     │ Float64     │ Float64   │ Float64    │ Bool           │\n",
      "├─────┼─────────────┼───────────┼────────────┼────────────────┤\n",
      "│ 1   │ 0.448179    │ 0.425105  │ 0.37962    │ true           │\n",
      "│ 2   │ 0.0805337   │ 0.124094  │ 0.0422894  │ false          │\n",
      "│ 3   │ -0.473043   │ -0.450421 │ -0.359597  │ true           │\n",
      "│ 4   │ -0.0199795  │ 0.0       │ 0.0        │ false          │\n",
      "│ 5   │ 0.169369    │ 0.198913  │ 0.121058   │ true           │\n",
      "│ 6   │ -0.0216948  │ 0.0       │ 0.0        │ false          │\n",
      "│ 7   │ -0.00363871 │ 0.0       │ 0.0        │ false          │\n",
      "│ 8   │ 0.430552    │ 0.422544  │ 0.343507   │ true           │\n",
      "│ 9   │ -0.129773   │ 0.0       │ 0.0        │ false          │\n",
      "│ 10  │ -0.196946   │ -0.191711 │ -0.0832795 │ true           │\n",
      "reached here 7\n",
      "IHT false positives = 1\n",
      "IHT false negatives = 4\n",
      "LASSO false positives = 27\n",
      "LASSO false negatives = 4\n",
      "marginal false positives = 74\n",
      "marginal false negatives = 5\n",
      "\n",
      "current run = 6\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/Users/biona001/.julia/dev/MendelIHT/figures/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = poisson_control.txt\n",
      "  output_table = poisson_table.txt\n",
      "  pedigree_file = poisson.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = poisson\n",
      "  regression = Poisson\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = poisson.bed\n",
      "  snpdefinition_file = poisson.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "reached here\n",
      "reached here too\n",
      "reached here 3\n",
      "reached here 4\n",
      "reached here 5\n",
      "reached here 6\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β     │ IHT_β    │ lasso_β     │ marginal_found │\n",
      "│     │ Float64    │ Float64  │ Float64     │ Bool           │\n",
      "├─────┼────────────┼──────────┼─────────────┼────────────────┤\n",
      "│ 1   │ 0.167356   │ 0.175769 │ 0.093401    │ true           │\n",
      "│ 2   │ 0.255773   │ 0.280751 │ 0.185482    │ true           │\n",
      "│ 3   │ -0.0471057 │ 0.0      │ 0.0         │ false          │\n",
      "│ 4   │ -0.0502778 │ 0.0      │ 0.0         │ false          │\n",
      "│ 5   │ 0.429838   │ 0.43208  │ 0.368692    │ true           │\n",
      "│ 6   │ 0.00244288 │ 0.0      │ 0.0         │ false          │\n",
      "│ 7   │ -0.0933734 │ 0.0      │ -0.00354199 │ false          │\n",
      "│ 8   │ 0.342491   │ 0.359835 │ 0.273954    │ true           │\n",
      "│ 9   │ 0.0890674  │ 0.0      │ 0.0         │ false          │\n",
      "│ 10  │ -0.122247  │ 0.0      │ -0.0158001  │ false          │\n",
      "reached here 7\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 6\n",
      "LASSO false positives = 11\n",
      "LASSO false negatives = 4\n",
      "marginal false positives = 11\n",
      "marginal false negatives = 6\n",
      "\n",
      "current run = 7\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/Users/biona001/.julia/dev/MendelIHT/figures/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = poisson_control.txt\n",
      "  output_table = poisson_table.txt\n",
      "  pedigree_file = poisson.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = poisson\n",
      "  regression = Poisson\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = poisson.bed\n",
      "  snpdefinition_file = poisson.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "reached here\n",
      "reached here too\n",
      "reached here 3\n",
      "reached here 4\n",
      "reached here 5\n",
      "reached here 6\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β     │ IHT_β     │ lasso_β   │ marginal_found │\n",
      "│     │ Float64    │ Float64   │ Float64   │ Bool           │\n",
      "├─────┼────────────┼───────────┼───────────┼────────────────┤\n",
      "│ 1   │ -0.335745  │ -0.29831  │ -0.184569 │ true           │\n",
      "│ 2   │ 0.183708   │ 0.143717  │ 0.0618403 │ true           │\n",
      "│ 3   │ 0.167754   │ 0.16881   │ 0.0797714 │ true           │\n",
      "│ 4   │ 0.138006   │ 0.0       │ 0.0385478 │ false          │\n",
      "│ 5   │ -0.0126176 │ 0.0       │ 0.0       │ false          │\n",
      "│ 6   │ -0.1787    │ -0.204253 │ -0.089445 │ true           │\n",
      "│ 7   │ 0.162684   │ 0.0       │ 0.0154076 │ false          │\n",
      "│ 8   │ 0.280177   │ 0.286006  │ 0.218533  │ true           │\n",
      "│ 9   │ 0.205512   │ 0.1868    │ 0.0848187 │ true           │\n",
      "│ 10  │ 0.347957   │ 0.39341   │ 0.32349   │ true           │\n",
      "reached here 7\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 3\n",
      "LASSO false positives = 36\n",
      "LASSO false negatives = 1\n",
      "marginal false positives = 7\n",
      "marginal false negatives = 3\n",
      "\n",
      "current run = 8\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/Users/biona001/.julia/dev/MendelIHT/figures/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = poisson_control.txt\n",
      "  output_table = poisson_table.txt\n",
      "  pedigree_file = poisson.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = poisson\n",
      "  regression = Poisson\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = poisson.bed\n",
      "  snpdefinition_file = poisson.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "reached here\n",
      "reached here too\n",
      "reached here 3\n",
      "reached here 4\n",
      "reached here 5\n",
      "reached here 6\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β     │ IHT_β     │ lasso_β    │ marginal_found │\n",
      "│     │ Float64    │ Float64   │ Float64    │ Bool           │\n",
      "├─────┼────────────┼───────────┼────────────┼────────────────┤\n",
      "│ 1   │ -0.0955549 │ 0.0       │ 0.0        │ false          │\n",
      "│ 2   │ 0.240684   │ 0.202153  │ 0.101882   │ true           │\n",
      "│ 3   │ 0.203638   │ 0.195086  │ 0.11564    │ true           │\n",
      "│ 4   │ -0.165083  │ -0.171398 │ -0.0836322 │ true           │\n",
      "│ 5   │ -0.3117    │ -0.303144 │ -0.188807  │ true           │\n",
      "│ 6   │ -0.332679  │ -0.320963 │ -0.206888  │ true           │\n",
      "│ 7   │ -0.63554   │ -0.635273 │ -0.112872  │ true           │\n",
      "│ 8   │ 0.161921   │ 0.181138  │ 0.0831243  │ true           │\n",
      "│ 9   │ 0.01539    │ 0.0       │ 0.0        │ false          │\n",
      "│ 10  │ 0.432937   │ 0.396964  │ 0.348276   │ true           │\n",
      "reached here 7\n",
      "IHT false positives = 1\n",
      "IHT false negatives = 2\n",
      "LASSO false positives = 39\n",
      "LASSO false negatives = 2\n",
      "marginal false positives = 13\n",
      "marginal false negatives = 2\n",
      "\n",
      "current run = 9\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/Users/biona001/.julia/dev/MendelIHT/figures/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = poisson_control.txt\n",
      "  output_table = poisson_table.txt\n",
      "  pedigree_file = poisson.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = poisson\n",
      "  regression = Poisson\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = poisson.bed\n",
      "  snpdefinition_file = poisson.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "reached here\n",
      "reached here too\n",
      "reached here 3\n",
      "reached here 4\n",
      "reached here 5\n",
      "reached here 6\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β    │ IHT_β     │ lasso_β    │ marginal_found │\n",
      "│     │ Float64   │ Float64   │ Float64    │ Bool           │\n",
      "├─────┼───────────┼───────────┼────────────┼────────────────┤\n",
      "│ 1   │ -0.361125 │ -0.338679 │ -0.193281  │ true           │\n",
      "│ 2   │ -0.100315 │ 0.0       │ -0.0407174 │ true           │\n",
      "│ 3   │ -0.548496 │ -0.655243 │ -0.287019  │ true           │\n",
      "│ 4   │ 0.41124   │ 0.415382  │ 0.318709   │ true           │\n",
      "│ 5   │ 0.361332  │ 0.361911  │ 0.302868   │ true           │\n",
      "│ 6   │ -0.152549 │ 0.0       │ -0.0539856 │ true           │\n",
      "│ 7   │ -0.329625 │ -0.324368 │ -0.147299  │ true           │\n",
      "│ 8   │ 0.146088  │ 0.0       │ 0.0166714  │ false          │\n",
      "│ 9   │ 0.0754147 │ 0.0       │ 0.0068767  │ false          │\n",
      "│ 10  │ -0.138181 │ 0.0       │ -0.0424442 │ false          │\n",
      "reached here 7\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 5\n",
      "LASSO false positives = 25\n",
      "LASSO false negatives = 0\n",
      "marginal false positives = 39\n",
      "marginal false negatives = 3\n",
      "\n",
      "current run = 10\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "┌ Warning: RCall.jl:  警告:  from glmnet Fortran code (error code -1); Convergence for 1th lambda value not reached after maxit=100000 iterations; solutions for larger lambdas returned\n",
      "│  getcoef(fit, nvars, nx, vnames) で警告がありました: \n",
      "│   an empty model has been returned; probably a convergence issue\n",
      "└ @ RCall /Users/biona001/.julia/packages/RCall/ffM0W/src/io.jl:113\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/Users/biona001/.julia/dev/MendelIHT/figures/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = poisson_control.txt\n",
      "  output_table = poisson_table.txt\n",
      "  pedigree_file = poisson.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = poisson\n",
      "  regression = Poisson\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = poisson.bed\n",
      "  snpdefinition_file = poisson.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "reached here\n",
      "reached here too\n",
      "reached here 3\n",
      "reached here 4\n",
      "reached here 5\n",
      "reached here 6\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β    │ IHT_β     │ lasso_β │ marginal_found │\n",
      "│     │ Float64   │ Float64   │ Float64 │ Bool           │\n",
      "├─────┼───────────┼───────────┼─────────┼────────────────┤\n",
      "│ 1   │ 0.353633  │ 0.339164  │ 0.0     │ true           │\n",
      "│ 2   │ -0.133638 │ 0.0       │ 0.0     │ false          │\n",
      "│ 3   │ 0.0749248 │ 0.0       │ 0.0     │ false          │\n",
      "│ 4   │ 0.441051  │ 0.428356  │ 0.0     │ true           │\n",
      "│ 5   │ 0.196634  │ 0.168849  │ 0.0     │ true           │\n",
      "│ 6   │ 0.744532  │ 0.742226  │ 0.0     │ true           │\n",
      "│ 7   │ -0.415215 │ -0.398329 │ 0.0     │ true           │\n",
      "│ 8   │ 0.37935   │ 0.376216  │ 0.0     │ true           │\n",
      "│ 9   │ 0.512715  │ 0.529859  │ 0.0     │ true           │\n",
      "│ 10  │ 0.148476  │ 0.175869  │ 0.0     │ true           │\n",
      "reached here 7\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 2\n",
      "LASSO false positives = 0\n",
      "LASSO false negatives = 10\n",
      "marginal false positives = 3703\n",
      "marginal false negatives = 2\n",
      "\n",
      "current run = 11\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/Users/biona001/.julia/dev/MendelIHT/figures/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = poisson_control.txt\n",
      "  output_table = poisson_table.txt\n",
      "  pedigree_file = poisson.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = poisson\n",
      "  regression = Poisson\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = poisson.bed\n",
      "  snpdefinition_file = poisson.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "reached here\n",
      "reached here too\n",
      "reached here 3\n",
      "reached here 4\n",
      "reached here 5\n",
      "reached here 6\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β     │ IHT_β     │ lasso_β    │ marginal_found │\n",
      "│     │ Float64    │ Float64   │ Float64    │ Bool           │\n",
      "├─────┼────────────┼───────────┼────────────┼────────────────┤\n",
      "│ 1   │ -0.642031  │ -0.669947 │ -0.562903  │ true           │\n",
      "│ 2   │ 0.179349   │ 0.164514  │ 0.0855058  │ true           │\n",
      "│ 3   │ -0.0204144 │ 0.0       │ 0.0        │ false          │\n",
      "│ 4   │ 0.113973   │ 0.0       │ 0.0116843  │ false          │\n",
      "│ 5   │ -0.479816  │ -0.502808 │ -0.396875  │ true           │\n",
      "│ 6   │ -0.155873  │ -0.184737 │ -0.0879138 │ true           │\n",
      "│ 7   │ 0.101679   │ 0.0       │ 0.0425643  │ false          │\n",
      "│ 8   │ 0.24682    │ 0.224992  │ 0.156243   │ true           │\n",
      "│ 9   │ -0.131333  │ -0.154759 │ -0.0552581 │ true           │\n",
      "│ 10  │ 0.136446   │ 0.0       │ 0.0430659  │ true           │\n",
      "reached here 7\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 4\n",
      "LASSO false positives = 24\n",
      "LASSO false negatives = 1\n",
      "marginal false positives = 32\n",
      "marginal false negatives = 3\n",
      "\n",
      "current run = 12\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/Users/biona001/.julia/dev/MendelIHT/figures/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = poisson_control.txt\n",
      "  output_table = poisson_table.txt\n",
      "  pedigree_file = poisson.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = poisson\n",
      "  regression = Poisson\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = poisson.bed\n",
      "  snpdefinition_file = poisson.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "reached here\n",
      "reached here too\n",
      "reached here 3\n",
      "reached here 4\n",
      "reached here 5\n",
      "reached here 6\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β      │ IHT_β     │ lasso_β   │ marginal_found │\n",
      "│     │ Float64     │ Float64   │ Float64   │ Bool           │\n",
      "├─────┼─────────────┼───────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.114505    │ 0.155821  │ 0.0909721 │ true           │\n",
      "│ 2   │ -0.676895   │ -0.655319 │ -0.394543 │ true           │\n",
      "│ 3   │ -0.00856463 │ 0.0       │ 0.0       │ false          │\n",
      "│ 4   │ -0.362537   │ -0.382875 │ -0.286181 │ true           │\n",
      "│ 5   │ -0.351751   │ -0.38317  │ -0.259195 │ true           │\n",
      "│ 6   │ 0.199895    │ 0.230292  │ 0.151058  │ true           │\n",
      "│ 7   │ 0.401497    │ 0.412936  │ 0.333698  │ true           │\n",
      "│ 8   │ 0.208755    │ 0.22145   │ 0.181322  │ true           │\n",
      "│ 9   │ 0.0228725   │ 0.0       │ 0.0       │ false          │\n",
      "│ 10  │ 0.753154    │ 0.75696   │ 0.678867  │ true           │\n",
      "reached here 7\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 2\n",
      "LASSO false positives = 89\n",
      "LASSO false negatives = 2\n",
      "marginal false positives = 801\n",
      "marginal false negatives = 2\n",
      "\n",
      "current run = 13\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/Users/biona001/.julia/dev/MendelIHT/figures/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = poisson_control.txt\n",
      "  output_table = poisson_table.txt\n",
      "  pedigree_file = poisson.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = poisson\n",
      "  regression = Poisson\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = poisson.bed\n",
      "  snpdefinition_file = poisson.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "reached here\n",
      "reached here too\n",
      "reached here 3\n",
      "reached here 4\n",
      "reached here 5\n",
      "reached here 6\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β     │ IHT_β    │ lasso_β    │ marginal_found │\n",
      "│     │ Float64    │ Float64  │ Float64    │ Bool           │\n",
      "├─────┼────────────┼──────────┼────────────┼────────────────┤\n",
      "│ 1   │ -0.0573683 │ 0.0      │ 0.0        │ false          │\n",
      "│ 2   │ -0.15902   │ 0.0      │ -0.0241296 │ false          │\n",
      "│ 3   │ 0.228407   │ 0.212761 │ 0.133779   │ true           │\n",
      "│ 4   │ -0.0885488 │ 0.0      │ 0.0        │ false          │\n",
      "│ 5   │ 0.342833   │ 0.40365  │ 0.334212   │ true           │\n",
      "│ 6   │ 0.427151   │ 0.433835 │ 0.363324   │ true           │\n",
      "│ 7   │ -0.124686  │ 0.0      │ -0.0292191 │ true           │\n",
      "│ 8   │ 0.244688   │ 0.234343 │ 0.152968   │ true           │\n",
      "│ 9   │ -0.0218821 │ 0.0      │ 0.0        │ false          │\n",
      "│ 10  │ 0.085758   │ 0.0      │ 0.0        │ false          │\n",
      "reached here 7\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 6\n",
      "LASSO false positives = 31\n",
      "LASSO false negatives = 4\n",
      "marginal false positives = 14\n",
      "marginal false negatives = 5\n",
      "\n",
      "current run = 14\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/Users/biona001/.julia/dev/MendelIHT/figures/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = poisson_control.txt\n",
      "  output_table = poisson_table.txt\n",
      "  pedigree_file = poisson.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = poisson\n",
      "  regression = Poisson\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = poisson.bed\n",
      "  snpdefinition_file = poisson.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "reached here\n",
      "reached here too\n",
      "reached here 3\n",
      "reached here 4\n",
      "reached here 5\n",
      "reached here 6\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β    │ IHT_β     │ lasso_β   │ marginal_found │\n",
      "│     │ Float64   │ Float64   │ Float64   │ Bool           │\n",
      "├─────┼───────────┼───────────┼───────────┼────────────────┤\n",
      "│ 1   │ -0.600319 │ -0.604808 │ -0.485907 │ true           │\n",
      "│ 2   │ 0.251169  │ 0.203218  │ 0.140136  │ true           │\n",
      "│ 3   │ 0.191066  │ 0.193754  │ 0.113658  │ true           │\n",
      "│ 4   │ 0.154598  │ 0.194022  │ 0.12531   │ true           │\n",
      "│ 5   │ 0.0995257 │ 0.0       │ 0.0490698 │ false          │\n",
      "│ 6   │ -0.408588 │ -0.342267 │ -0.219161 │ true           │\n",
      "│ 7   │ -0.257726 │ -0.282878 │ -0.173284 │ true           │\n",
      "│ 8   │ -0.498645 │ -0.523031 │ -0.361765 │ true           │\n",
      "│ 9   │ 0.319097  │ 0.291163  │ 0.213867  │ true           │\n",
      "│ 10  │ 0.367859  │ 0.336543  │ 0.240854  │ true           │\n",
      "reached here 7\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 1\n",
      "LASSO false positives = 60\n",
      "LASSO false negatives = 0\n",
      "marginal false positives = 132\n",
      "marginal false negatives = 1\n",
      "\n",
      "current run = 15\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/Users/biona001/.julia/dev/MendelIHT/figures/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = poisson_control.txt\n",
      "  output_table = poisson_table.txt\n",
      "  pedigree_file = poisson.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = poisson\n",
      "  regression = Poisson\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = poisson.bed\n",
      "  snpdefinition_file = poisson.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "reached here\n",
      "reached here too\n",
      "reached here 3\n",
      "reached here 4\n",
      "reached here 5\n",
      "reached here 6\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β     │ IHT_β     │ lasso_β    │ marginal_found │\n",
      "│     │ Float64    │ Float64   │ Float64    │ Bool           │\n",
      "├─────┼────────────┼───────────┼────────────┼────────────────┤\n",
      "│ 1   │ 0.117155   │ 0.0       │ 0.00328414 │ false          │\n",
      "│ 2   │ 0.339741   │ 0.327567  │ 0.22414    │ true           │\n",
      "│ 3   │ -0.143149  │ 0.0       │ 0.0        │ false          │\n",
      "│ 4   │ -0.0998637 │ 0.0       │ 0.0        │ false          │\n",
      "│ 5   │ 0.180403   │ 0.171013  │ 0.0678431  │ true           │\n",
      "│ 6   │ 0.233013   │ 0.222095  │ 0.183828   │ true           │\n",
      "│ 7   │ -0.178846  │ -0.189582 │ -0.0828428 │ true           │\n",
      "│ 8   │ -0.0917945 │ 0.0       │ 0.0        │ false          │\n",
      "│ 9   │ -0.323128  │ -0.334305 │ -0.0984025 │ true           │\n",
      "│ 10  │ 0.047305   │ 0.0       │ 0.0        │ false          │\n",
      "reached here 7\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 5\n",
      "LASSO false positives = 6\n",
      "LASSO false negatives = 4\n",
      "marginal false positives = 3\n",
      "marginal false negatives = 5\n",
      "\n",
      "current run = 16\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/Users/biona001/.julia/dev/MendelIHT/figures/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = poisson_control.txt\n",
      "  output_table = poisson_table.txt\n",
      "  pedigree_file = poisson.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = poisson\n",
      "  regression = Poisson\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = poisson.bed\n",
      "  snpdefinition_file = poisson.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "iht_false_positives, iht_false_negatives, lasso_false_positives, \n",
    "    lasso_false_negatives, marginal_false_positives, marginal_false_negatives = run()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Importing results.... "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "50×1 Array{Float64,2}:\n",
       " 2.0\n",
       " 1.0\n",
       " 2.0\n",
       " 4.0\n",
       " 5.0\n",
       " 6.0\n",
       " 3.0\n",
       " 2.0\n",
       " 3.0\n",
       " 2.0\n",
       " 3.0\n",
       " 2.0\n",
       " 5.0\n",
       " ⋮  \n",
       " 3.0\n",
       " 5.0\n",
       " 3.0\n",
       " 3.0\n",
       " 3.0\n",
       " 2.0\n",
       " 2.0\n",
       " 6.0\n",
       " 3.0\n",
       " 3.0\n",
       " 2.0\n",
       " 2.0"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#this code was ran on cluster\n",
    "using DelimitedFiles\n",
    "poisson_iht_false_positives = readdlm(\"poisson_results/iht_false_positives\")\n",
    "poisson_iht_false_negatives = readdlm(\"poisson_results/lasso_false_negatives\")\n",
    "poisson_lasso_false_positives = readdlm(\"poisson_results/lasso_false_positives\")\n",
    "poisson_lasso_false_negatives = readdlm(\"poisson_results/lasso_false_negatives\")\n",
    "poisson_marginal_false_positives = readdlm(\"poisson_results/marginal_false_positives\")\n",
    "poisson_marginal_false_negatives = readdlm(\"poisson_results/marginal_false_negatives\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "6-element Array{Float64,1}:\n",
       "   0.56\n",
       "   3.64\n",
       "  27.74\n",
       "   3.64\n",
       " 866.68\n",
       "   3.14"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "IHT_dnc = 1\n",
    "LASSO_dnc = 8\n",
    "marginal_dnc = 0\n",
    "lasso_dnc_index = [2; 10; 17; 24; 27; 37; 38; 40; ]\n",
    "\n",
    "result = [sum(poisson_iht_false_positives)/50; sum(poisson_iht_false_negatives)/50; \n",
    "        sum(poisson_lasso_false_positives)/50; sum(poisson_lasso_false_negatives)/50;\n",
    "        sum(poisson_marginal_false_positives)/50; sum(poisson_marginal_false_negatives)/50]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "50×1 Array{Float64,2}:\n",
       " 2.0\n",
       " 1.0\n",
       " 2.0\n",
       " 4.0\n",
       " 5.0\n",
       " 6.0\n",
       " 3.0\n",
       " 2.0\n",
       " 3.0\n",
       " 2.0\n",
       " 3.0\n",
       " 2.0\n",
       " 5.0\n",
       " ⋮  \n",
       " 3.0\n",
       " 5.0\n",
       " 3.0\n",
       " 3.0\n",
       " 3.0\n",
       " 2.0\n",
       " 2.0\n",
       " 6.0\n",
       " 3.0\n",
       " 3.0\n",
       " 2.0\n",
       " 2.0"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "poisson_iht_false_positives = readdlm(\"poisson_results/iht_false_positives\")\n",
    "poisson_iht_false_negatives = readdlm(\"poisson_results/lasso_false_negatives\")\n",
    "poisson_lasso_false_positives = readdlm(\"poisson_results/lasso_false_positives\")\n",
    "poisson_lasso_false_negatives = readdlm(\"poisson_results/lasso_false_negatives\")\n",
    "poisson_marginal_false_positives = readdlm(\"poisson_results/marginal_false_positives\")\n",
    "poisson_marginal_false_negatives = readdlm(\"poisson_results/marginal_false_negatives\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "42-element Array{Float64,1}:\n",
       " 2.0\n",
       " 2.0\n",
       " 4.0\n",
       " 5.0\n",
       " 6.0\n",
       " 3.0\n",
       " 2.0\n",
       " 3.0\n",
       " 3.0\n",
       " 2.0\n",
       " 5.0\n",
       " 1.0\n",
       " 5.0\n",
       " ⋮  \n",
       " 3.0\n",
       " 3.0\n",
       " 3.0\n",
       " 3.0\n",
       " 3.0\n",
       " 2.0\n",
       " 2.0\n",
       " 6.0\n",
       " 3.0\n",
       " 3.0\n",
       " 2.0\n",
       " 2.0"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "iht_false_positive_exclude_nonconvergent = [getindex(poisson_iht_false_positives, i) for i = 1:50 if !(i in lasso_dnc_index)]\n",
    "iht_false_negative_exclude_nonconvergent = [getindex(poisson_iht_false_negatives, i) for i = 1:50 if !(i in lasso_dnc_index)]\n",
    "lasso_false_positive_exclude_nonconvergent = [getindex(poisson_lasso_false_positives, i) for i = 1:50 if !(i in lasso_dnc_index)]\n",
    "lasso_false_negative_exclude_nonconvergent = [getindex(poisson_lasso_false_negatives, i) for i = 1:50 if !(i in lasso_dnc_index)]\n",
    "marginal_false_positive_exclude_nonconvergent = [getindex(poisson_marginal_false_positives, i) for i = 1:50 if !(i in lasso_dnc_index)]\n",
    "marginal_false_negative_exclude_nonconvergent = [getindex(poisson_marginal_false_negatives, i) for i = 1:50 if !(i in lasso_dnc_index)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sum(iht_false_positive_exclude_nonconvergent) / 42 = 0.38095238095238093\n",
      "sum(iht_false_negative_exclude_nonconvergent) / 42 = 2.4285714285714284\n",
      "sum(lasso_false_positive_exclude_nonconvergent) / 42 = 33.023809523809526\n",
      "sum(lasso_false_negative_exclude_nonconvergent) / 42 = 2.4285714285714284\n",
      "sum(marginal_false_positive_exclude_nonconvergent) / 42 = 142.23809523809524\n",
      "sum(marginal_false_negative_exclude_nonconvergent) / 42 = 3.3095238095238093\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "3.3095238095238093"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "@show sum(iht_false_positive_exclude_nonconvergent) / 42\n",
    "@show sum(iht_false_negative_exclude_nonconvergent) / 42\n",
    "@show sum(lasso_false_positive_exclude_nonconvergent) / 42\n",
    "@show sum(lasso_false_negative_exclude_nonconvergent) / 42\n",
    "@show sum(marginal_false_positive_exclude_nonconvergent) / 42 #still pretty bad\n",
    "@show sum(marginal_false_negative_exclude_nonconvergent) / 42 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "8-element Array{Int64,1}:\n",
       "  2\n",
       " 10\n",
       " 17\n",
       " 24\n",
       " 27\n",
       " 37\n",
       " 38\n",
       " 40"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lasso_dnc_index"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 9000+ hits are a consequence of extreme y values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/Users/biona001/.julia/dev/MendelIHT/figures/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = poisson_control.txt\n",
      "  output_table = poisson_table.txt\n",
      "  pedigree_file = poisson.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = poisson\n",
      "  regression = Poisson\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = poisson.bed\n",
      "  snpdefinition_file = poisson.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "10-element Array{Bool,1}:\n",
       "  true\n",
       " false\n",
       "  true\n",
       "  true\n",
       "  true\n",
       "  true\n",
       "  true\n",
       "  true\n",
       "  true\n",
       "  true"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "n = 1000\n",
    "p = 10000\n",
    "k = 10\n",
    "d = Poisson\n",
    "l = canonicallink(d())\n",
    "\n",
    "Random.seed!(37)\n",
    "\n",
    "x, = simulate_random_snparray(n, p, \"poisson.bed\")\n",
    "xbm = SnpBitMatrix{Float64}(x, model=ADDITIVE_MODEL, center=true, scale=true); \n",
    "z = ones(n, 1) # the intercept\n",
    "x_float = [convert(Matrix{Float64}, x, center=true, scale=true) z] #Float64 version of x\n",
    "y, true_b, correct_position = simulate_random_response(x, xbm, k, d, l)\n",
    "\n",
    "make_bim_fam_files(x, y, \"poisson\")\n",
    "\n",
    "#create a control file for MendelGWAS\n",
    "open(\"poisson_control.txt\", \"w\") do f\n",
    "    write(f, \"plink_input_basename = poisson \\n\")\n",
    "    write(f, \"plink_field_separator = '\\t' \\n\")\n",
    "    write(f, \"output_table = poisson_table.txt \\n\")\n",
    "    write(f, \"regression = Poisson \\n\")\n",
    "    write(f, \"regression_formula = Trait ~ \\n\")\n",
    "end\n",
    "\n",
    "GWAS(\"poisson_control.txt\");\n",
    "\n",
    "# calculate false positive/negatives based on bonferroni correction\n",
    "p_values = CSV.read(\"poisson_table.txt\", delim=',', header=true)[:Pvalue]\n",
    "significance = 0.05 / Float64(p)\n",
    "passing_snps_position = findall(p_values .<= significance)\n",
    "marginal_found = [correct_position[snp] in passing_snps_position for snp in 1:k]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "10000-element Array{Union{Missing, Float64},1}:\n",
       " 1.3451240680413934e-75 \n",
       " 0.0                    \n",
       " 0.0                    \n",
       " 0.0                    \n",
       " 0.0                    \n",
       " 0.0                    \n",
       " 0.0                    \n",
       " 0.0                    \n",
       " 0.0                    \n",
       " 0.0                    \n",
       " 0.0                    \n",
       " 0.0                    \n",
       " 0.0                    \n",
       " ⋮                      \n",
       " 0.0                    \n",
       " 0.0                    \n",
       " 1.6597422626641304e-238\n",
       " 0.0                    \n",
       " 0.0008455539546774351  \n",
       " 2.0e-323               \n",
       " 0.0                    \n",
       " 0.0                    \n",
       " 0.0                    \n",
       " 0.0                    \n",
       " 0.0                    \n",
       " 0.0                    "
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "p_values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(24217.0, 980310.9098058061)"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# result = CSV.read(\"poisson_table.txt\", delim=',', header=true)\n",
    "maximum(y), var(y)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "rm(\"poisson.bed\", force=true)\n",
    "rm(\"poisson.bim\", force=true)\n",
    "rm(\"poisson.fam\", force=true)\n",
    "rm(\"poisson_table.txt\", force=true)\n",
    "rm(\"poisson_control.txt\", force=true)\n",
    "rm(\"Mendel_Output.txt\", force=true)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "@webio": {
   "lastCommId": null,
   "lastKernelId": null
  },
  "kernelspec": {
   "display_name": "Julia 1.0.3",
   "language": "julia",
   "name": "julia-1.0"
  },
  "language_info": {
   "file_extension": ".jl",
   "mimetype": "application/julia",
   "name": "julia",
   "version": "1.0.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
