{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "# Bernoulli: IHT vs LASSO vs Marginal regression: comparing false positive/negatives\n",
    "\n",
    "Marginal regression is the most prevalent method used to analyze GWAS data. LASSO is currently the de facto penalized least squares method for [feature selection](https://en.wikipedia.org/wiki/Feature_selection). Here we compare the performance (in terms of the number of false positives/negative) of LASSO and marginal regression with IHT for GWAS data. We use the `glmnet` implementation of cyclic coordinate descent for LASSO, and `MendelGWAS.jl` module of [OpenMendel](https://openmendel.github.io/MendelGWAS.jl/) for marginal analysis. Since the focus here is not scalability, we test our sample problems on moderately sized genotype matrces of 1000 samples with 10,000 SNPs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "31"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "using Distributed\n",
    "addprocs(30)\n",
    "nprocs()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "┌ Info: Loading DataFrames support into Gadfly.jl\n",
      "└ @ Gadfly /u/home/b/biona001/.julia/packages/Gadfly/09PWZ/src/mapping.jl:228\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall /u/home/b/biona001/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: Loading required package: Matrix\n",
      "│ Loading required package: foreach\n",
      "│ Loaded glmnet 2.0-16\n",
      "│ \n",
      "└ @ RCall /u/home/b/biona001/.julia/packages/RCall/ffM0W/src/io.jl:113\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "RObject{StrSxp}\n",
       " [1] \"glmnet\"    \"foreach\"   \"Matrix\"    \"stats\"     \"graphics\"  \"grDevices\"\n",
       " [7] \"utils\"     \"datasets\"  \"methods\"   \"base\"     \n"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "using MendelIHT\n",
    "using SnpArrays\n",
    "using DataFrames\n",
    "using Distributions\n",
    "using Random\n",
    "using LinearAlgebra\n",
    "using DelimitedFiles\n",
    "using GLM\n",
    "using MendelGWAS\n",
    "using CSV\n",
    "using RCall\n",
    "R\"library(glmnet)\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "# Bernoulli response (i.e. logistic regression)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "iht_lasso_marginal (generic function with 1 method)"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "function iht_lasso_marginal(n::Int64, p::Int64, d::UnionAll, l::Link)\n",
    "    #construct snpmatrix, covariate files, and true model b\n",
    "    x, = simulate_random_snparray(n, p, \"tmp.bed\")\n",
    "    xbm = SnpBitMatrix{Float64}(x, model=ADDITIVE_MODEL, center=true, scale=true); \n",
    "    z = ones(n, 1) # the intercept\n",
    "    x_float = [convert(Matrix{Float64}, x, center=true, scale=true) z] #Float64 version of x\n",
    "\n",
    "    # simulate response, true model b, and the correct non-0 positions of b\n",
    "    true_b = zeros(p)\n",
    "    true_b[1:10] .= collect(0.1:0.1:1.0)\n",
    "    k = 10\n",
    "    shuffle!(true_b)\n",
    "    correct_position = findall(x -> x != 0, true_b)\n",
    "\n",
    "    #simulate phenotypes (e.g. vector y)\n",
    "    if d == Normal || d == Poisson || d == Bernoulli\n",
    "        prob = linkinv.(l, xbm * true_b)\n",
    "        clamp!(prob, -20, 20)\n",
    "        y = [rand(d(i)) for i in prob]\n",
    "    elseif d == NegativeBinomial\n",
    "        μ = linkinv.(l, xbm * true_b)\n",
    "        prob = 1 ./ (1 .+ μ ./ nn)\n",
    "        y = [rand(d(nn, i)) for i in prob] #number of failtures before nn success occurs\n",
    "    elseif d == Gamma\n",
    "        μ = linkinv.(l, xbm * true_b)\n",
    "        β = 1 ./ μ # here β is the rate parameter for gamma distribution\n",
    "        y = [rand(d(α, i)) for i in β] # α is the shape parameter for gamma\n",
    "    end\n",
    "    y = Float64.(y)\n",
    "\n",
    "    #specify path and folds\n",
    "    num_folds = 3\n",
    "    folds = rand(1:num_folds, size(x, 1));\n",
    "\n",
    "    #run glmnet via Rcall\n",
    "    @rput x_float y folds num_folds #make variables visible to R\n",
    "    R\"lasso_cv_result = cv.glmnet(x_float, y, nfolds = num_folds, foldid = folds, family='binomial')\"\n",
    "    R\"lasso_beta_tmp = glmnet(x_float, y, lambda=lasso_cv_result$lambda.min, family='binomial')$beta\"\n",
    "    R\"lasso_beta = as.vector(lasso_beta_tmp)\"\n",
    "    @rget lasso_cv_result lasso_beta #pull result from R to Julia\n",
    "    lasso_k_est = count(!iszero, lasso_beta)\n",
    "    \n",
    "    #find non-zero entries returned by best lasso model as largest k estimate\n",
    "    path = collect(1:50);\n",
    "    \n",
    "    #run IHT's cross validation routine \n",
    "    mses = cv_iht(d(), l, x, z, y, 1, path, num_folds, folds=folds, use_maf=false, debias=false, showinfo=false, parallel=true);\n",
    "    iht_k_est = argmin(mses)\n",
    "    iht_result = L0_reg(x, xbm, z, y, 1, iht_k_est, d(), l, debias=false, init=false, use_maf=false)\n",
    "    iht_beta = iht_result.beta\n",
    "    \n",
    "    #Now run MendelGWAS\n",
    "    make_bim_fam_files(x, y, \"tmp\") #create .bim and .bed files for MendelGWAS\n",
    "\n",
    "    #create a control file for MendelGWAS\n",
    "    open(\"tmp_control.txt\", \"w\") do f\n",
    "        write(f, \"plink_input_basename = tmp \\n\")\n",
    "        write(f, \"plink_field_separator = '\\t' \\n\")\n",
    "        write(f, \"output_table = tmp_table.txt \\n\")\n",
    "        write(f, \"regression = logistic \\n\")\n",
    "        write(f, \"affected_designator = 1.0 \\n\")\n",
    "        write(f, \"regression_formula = Trait ~ \\n\")\n",
    "    end\n",
    "\n",
    "    #run marginal analysis\n",
    "    GWAS(\"tmp_control.txt\");\n",
    "\n",
    "    # calculate false positive/negatives based on bonferroni correction\n",
    "    p_values = CSV.read(\"tmp_table.txt\", delim=',', header=true)[:Pvalue]\n",
    "    significance = 0.05 / Float64(p)\n",
    "    passing_snps_position = findall(p_values .<= significance)\n",
    "    marginal_found = [correct_position[snp] in passing_snps_position for snp in 1:k]\n",
    "\n",
    "    #clean up\n",
    "    rm(\"tmp.bed\", force=true)\n",
    "    rm(\"tmp.bim\", force=true)\n",
    "    rm(\"tmp.fam\", force=true)\n",
    "    rm(\"tmp_table.txt\", force=true)\n",
    "    rm(\"tmp_control.txt\", force=true)\n",
    "    rm(\"Mendel_Output.txt\", force=true)\n",
    "\n",
    "    #show lasso and IHT's reconstruction result\n",
    "    compare_model = DataFrame(\n",
    "        true_β  = true_b[correct_position], \n",
    "        IHT_β   = iht_beta[correct_position],\n",
    "        lasso_β = lasso_beta[correct_position],\n",
    "        marginal_found = marginal_found)\n",
    "    @show compare_model\n",
    "    \n",
    "    #compute true/false positives/negatives for IHT and lasso\n",
    "    iht_tp = count(!iszero, iht_beta[correct_position])\n",
    "    iht_fp = iht_k_est - iht_tp\n",
    "    iht_fn = k - iht_tp\n",
    "    lasso_tp = count(!iszero, lasso_beta[correct_position])\n",
    "    lasso_fp = lasso_k_est - lasso_tp\n",
    "    lasso_fn = k - lasso_tp\n",
    "    marginal_tp = count(!iszero, true_b[passing_snps_position])\n",
    "    marginal_fp = length(passing_snps_position) - marginal_tp\n",
    "    marginal_fn = k - marginal_tp\n",
    "    \n",
    "    println(\"IHT true positives = $iht_tp\")\n",
    "    println(\"IHT false positives = $iht_fp\")\n",
    "    println(\"IHT false negatives = $iht_fn\")\n",
    "    println(\"LASSO true positives = $lasso_tp\")\n",
    "    println(\"LASSO false positives = $lasso_fp\")\n",
    "    println(\"LASSO false negatives = $lasso_fn\")\n",
    "    println(\"marginal true positives = $marginal_tp\")\n",
    "    println(\"marginal false positives = $marginal_fp\")\n",
    "    println(\"marginal false negatives = $marginal_fn\" * \"\\n\")\n",
    "    \n",
    "    return iht_tp, iht_fp, iht_fn, lasso_tp, lasso_fp, lasso_fn, marginal_tp, marginal_fp, marginal_fn\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "run (generic function with 1 method)"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "function run()\n",
    "    #simulat data with k true predictors, from distribution d and with link l.\n",
    "    n = 1000\n",
    "    p = 10000\n",
    "    d = Bernoulli\n",
    "    l = canonicallink(d())\n",
    "\n",
    "    #run function above, saving results in 4 vectors\n",
    "    total_runs = 50\n",
    "    iht_true_positives = zeros(total_runs)\n",
    "    iht_false_positives = zeros(total_runs)\n",
    "    iht_false_negatives = zeros(total_runs)\n",
    "    lasso_true_positives = zeros(total_runs)\n",
    "    lasso_false_positives = zeros(total_runs)\n",
    "    lasso_false_negatives = zeros(total_runs)\n",
    "    marginal_true_positives = zeros(total_runs)\n",
    "    marginal_false_positives = zeros(total_runs)\n",
    "    marginal_false_negatives = zeros(total_runs)\n",
    "    for i in 1:total_runs\n",
    "        println(\"current run = $i\")\n",
    "        \n",
    "        #set random seed\n",
    "        Random.seed!(i)\n",
    "        \n",
    "        iht_tp, iht_fp, iht_fn, lasso_tp, lasso_fp, lasso_fn, m_tp, m_fp, m_fn = iht_lasso_marginal(n, p, d, l)\n",
    "        iht_true_positives[i] = iht_tp\n",
    "        iht_false_positives[i] = iht_fp\n",
    "        iht_false_negatives[i] = iht_fn\n",
    "        lasso_true_positives[i] = lasso_tp\n",
    "        lasso_false_positives[i] = lasso_fp\n",
    "        lasso_false_negatives[i] = lasso_fn\n",
    "        marginal_true_positives[i] = m_tp\n",
    "        marginal_false_positives[i] = m_fp\n",
    "        marginal_false_negatives[i] = m_fn\n",
    "    end\n",
    "    \n",
    "    return iht_true_positives, iht_false_positives, iht_false_negatives, \n",
    "            lasso_true_positives, lasso_false_positives, lasso_false_negatives, \n",
    "            marginal_true_positives, marginal_false_positives, marginal_false_negatives\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "current run = 1\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.5     │ 0.464049 │ 0.162386  │ false          │\n",
      "│ 2   │ 1.0     │ 0.965866 │ 0.548849  │ true           │\n",
      "│ 3   │ 0.6     │ 0.492834 │ 0.221417  │ true           │\n",
      "│ 4   │ 0.1     │ 0.0      │ 0.0       │ false          │\n",
      "│ 5   │ 0.8     │ 0.913407 │ 0.385438  │ true           │\n",
      "│ 6   │ 0.2     │ 0.0      │ 0.086002  │ false          │\n",
      "│ 7   │ 0.9     │ 0.804275 │ 0.451149  │ true           │\n",
      "│ 8   │ 0.7     │ 0.593493 │ 0.28826   │ true           │\n",
      "│ 9   │ 0.3     │ 0.374088 │ 0.126692  │ false          │\n",
      "│ 10  │ 0.4     │ 0.0      │ 0.0314002 │ false          │\n",
      "IHT true positives = 7\n",
      "IHT false positives = 1\n",
      "IHT false negatives = 3\n",
      "LASSO true positives = 9\n",
      "LASSO false positives = 40\n",
      "LASSO false negatives = 1\n",
      "marginal true positives = 5\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 5\n",
      "\n",
      "current run = 2\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.7     │ 0.709959 │ 0.394699  │ true           │\n",
      "│ 2   │ 0.2     │ 0.0      │ 0.0851549 │ false          │\n",
      "│ 3   │ 0.5     │ 0.492419 │ 0.188848  │ false          │\n",
      "│ 4   │ 1.0     │ 0.952304 │ 0.623434  │ true           │\n",
      "│ 5   │ 0.6     │ 0.53342  │ 0.272188  │ true           │\n",
      "│ 6   │ 0.9     │ 0.877677 │ 0.552251  │ true           │\n",
      "│ 7   │ 0.4     │ 0.483007 │ 0.224354  │ true           │\n",
      "│ 8   │ 0.3     │ 0.0      │ 0.0564483 │ false          │\n",
      "│ 9   │ 0.1     │ 0.0      │ 0.0       │ false          │\n",
      "│ 10  │ 0.8     │ 0.789484 │ 0.443726  │ true           │\n",
      "IHT true positives = 7\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 3\n",
      "LASSO true positives = 9\n",
      "LASSO false positives = 122\n",
      "LASSO false negatives = 1\n",
      "marginal true positives = 6\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 4\n",
      "\n",
      "current run = 3\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.8     │ 0.675481 │ 0.346857  │ true           │\n",
      "│ 2   │ 0.6     │ 0.510962 │ 0.24107   │ true           │\n",
      "│ 3   │ 0.9     │ 0.730469 │ 0.395465  │ true           │\n",
      "│ 4   │ 0.5     │ 0.496007 │ 0.209115  │ true           │\n",
      "│ 5   │ 0.2     │ 0.0      │ 0.0530344 │ false          │\n",
      "│ 6   │ 0.4     │ 0.359331 │ 0.0938084 │ false          │\n",
      "│ 7   │ 0.7     │ 0.586314 │ 0.269368  │ true           │\n",
      "│ 8   │ 0.1     │ 0.0      │ 0.0       │ false          │\n",
      "│ 9   │ 1.0     │ 0.925745 │ 0.548922  │ true           │\n",
      "│ 10  │ 0.3     │ 0.0      │ 0.0280756 │ false          │\n",
      "IHT true positives = 7\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 3\n",
      "LASSO true positives = 9\n",
      "LASSO false positives = 28\n",
      "LASSO false negatives = 1\n",
      "marginal true positives = 6\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 4\n",
      "\n",
      "current run = 4\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.4     │ 0.380124 │ 0.0985316 │ false          │\n",
      "│ 2   │ 0.3     │ 0.413839 │ 0.11299   │ false          │\n",
      "│ 3   │ 0.8     │ 0.641027 │ 0.30867   │ true           │\n",
      "│ 4   │ 0.1     │ 0.0      │ 0.0       │ false          │\n",
      "│ 5   │ 0.7     │ 0.66378  │ 0.335999  │ true           │\n",
      "│ 6   │ 0.9     │ 0.742957 │ 0.379904  │ true           │\n",
      "│ 7   │ 0.5     │ 0.556487 │ 0.250216  │ true           │\n",
      "│ 8   │ 1.0     │ 0.75107  │ 0.379325  │ true           │\n",
      "│ 9   │ 0.6     │ 0.419473 │ 0.111726  │ true           │\n",
      "│ 10  │ 0.2     │ 0.0      │ 0.0       │ false          │\n",
      "IHT true positives = 8\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 2\n",
      "LASSO true positives = 8\n",
      "LASSO false positives = 7\n",
      "LASSO false negatives = 2\n",
      "marginal true positives = 6\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 4\n",
      "\n",
      "current run = 5\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.2     │ 0.0      │ 0.0       │ false          │\n",
      "│ 2   │ 0.6     │ 0.934131 │ 0.11416   │ false          │\n",
      "│ 3   │ 0.9     │ 0.926292 │ 0.567725  │ true           │\n",
      "│ 4   │ 1.0     │ 1.09125  │ 0.183338  │ true           │\n",
      "│ 5   │ 0.3     │ 0.323604 │ 0.0630213 │ false          │\n",
      "│ 6   │ 0.4     │ 0.444111 │ 0.149171  │ false          │\n",
      "│ 7   │ 0.5     │ 0.46339  │ 0.165112  │ true           │\n",
      "│ 8   │ 0.1     │ 0.0      │ 0.0       │ false          │\n",
      "│ 9   │ 0.8     │ 0.570325 │ 0.248565  │ true           │\n",
      "│ 10  │ 0.7     │ 0.725652 │ 0.416094  │ true           │\n",
      "IHT true positives = 8\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 2\n",
      "LASSO true positives = 8\n",
      "LASSO false positives = 75\n",
      "LASSO false negatives = 2\n",
      "marginal true positives = 5\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 5\n",
      "\n",
      "current run = 6\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β    │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64    │ Bool           │\n",
      "├─────┼─────────┼──────────┼────────────┼────────────────┤\n",
      "│ 1   │ 0.2     │ 0.0      │ 0.00666594 │ false          │\n",
      "│ 2   │ 1.0     │ 0.840699 │ 0.472991   │ true           │\n",
      "│ 3   │ 0.7     │ 0.710429 │ 0.33925    │ true           │\n",
      "│ 4   │ 0.1     │ 0.0      │ 0.0        │ false          │\n",
      "│ 5   │ 0.5     │ 0.431109 │ 0.159108   │ true           │\n",
      "│ 6   │ 0.8     │ 0.720415 │ 0.37568    │ true           │\n",
      "│ 7   │ 0.3     │ 0.0      │ 0.0        │ false          │\n",
      "│ 8   │ 0.9     │ 0.840681 │ 0.483108   │ true           │\n",
      "│ 9   │ 0.4     │ 0.0      │ 0.0141281  │ false          │\n",
      "│ 10  │ 0.6     │ 0.495203 │ 0.19809    │ true           │\n",
      "IHT true positives = 6\n",
      "IHT false positives = 1\n",
      "IHT false negatives = 4\n",
      "LASSO true positives = 8\n",
      "LASSO false positives = 31\n",
      "LASSO false negatives = 2\n",
      "marginal true positives = 6\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 4\n",
      "\n",
      "current run = 7\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.5     │ 0.586402 │ 0.289817  │ true           │\n",
      "│ 2   │ 1.0     │ 0.90865  │ 0.520675  │ true           │\n",
      "│ 3   │ 0.1     │ 0.0      │ 0.0       │ false          │\n",
      "│ 4   │ 0.9     │ 0.82613  │ 0.502882  │ true           │\n",
      "│ 5   │ 0.8     │ 0.64973  │ 0.331599  │ true           │\n",
      "│ 6   │ 0.7     │ 0.653469 │ 0.323028  │ true           │\n",
      "│ 7   │ 0.2     │ 0.0      │ 0.0       │ false          │\n",
      "│ 8   │ 0.6     │ 0.582838 │ 0.296828  │ true           │\n",
      "│ 9   │ 0.3     │ 0.0      │ 0.0216486 │ false          │\n",
      "│ 10  │ 0.4     │ 0.0      │ 0.0935478 │ false          │\n",
      "IHT true positives = 6\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 4\n",
      "LASSO true positives = 8\n",
      "LASSO false positives = 52\n",
      "LASSO false negatives = 2\n",
      "marginal true positives = 6\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 4\n",
      "\n",
      "current run = 8\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.8     │ 0.854902 │ 0.468972  │ true           │\n",
      "│ 2   │ 0.1     │ 0.0      │ 0.0       │ false          │\n",
      "│ 3   │ 0.2     │ 0.0      │ 0.0       │ false          │\n",
      "│ 4   │ 0.4     │ 0.0      │ 0.0700382 │ false          │\n",
      "│ 5   │ 0.5     │ 0.500031 │ 0.131089  │ false          │\n",
      "│ 6   │ 0.7     │ 0.624055 │ 0.323658  │ true           │\n",
      "│ 7   │ 0.9     │ 0.835741 │ 0.485677  │ true           │\n",
      "│ 8   │ 0.6     │ 0.602497 │ 0.258442  │ true           │\n",
      "│ 9   │ 1.0     │ 0.916071 │ 0.538617  │ true           │\n",
      "│ 10  │ 0.3     │ 0.0      │ 0.0197119 │ false          │\n",
      "IHT true positives = 6\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 4\n",
      "LASSO true positives = 8\n",
      "LASSO false positives = 41\n",
      "LASSO false negatives = 2\n",
      "marginal true positives = 5\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 5\n",
      "\n",
      "current run = 9\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.7     │ 0.690458 │ 0.353065  │ true           │\n",
      "│ 2   │ 0.4     │ 0.0      │ 0.092175  │ false          │\n",
      "│ 3   │ 0.3     │ 0.447431 │ 0.181717  │ true           │\n",
      "│ 4   │ 0.9     │ 0.842133 │ 0.475045  │ true           │\n",
      "│ 5   │ 0.6     │ 0.677646 │ 0.338953  │ true           │\n",
      "│ 6   │ 1.0     │ 1.01585  │ 0.400767  │ true           │\n",
      "│ 7   │ 0.8     │ 0.699056 │ 0.350678  │ true           │\n",
      "│ 8   │ 0.5     │ 0.0      │ 0.132693  │ true           │\n",
      "│ 9   │ 0.2     │ 0.0      │ 0.0676959 │ false          │\n",
      "│ 10  │ 0.1     │ 0.0      │ 0.0       │ false          │\n",
      "IHT true positives = 6\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 4\n",
      "LASSO true positives = 9\n",
      "LASSO false positives = 19\n",
      "LASSO false negatives = 1\n",
      "marginal true positives = 7\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 3\n",
      "\n",
      "current run = 10\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 1.0     │ 1.15799  │ 0.682967  │ true           │\n",
      "│ 2   │ 0.4     │ 0.607854 │ 0.236572  │ true           │\n",
      "│ 3   │ 0.6     │ 0.679429 │ 0.292794  │ true           │\n",
      "│ 4   │ 0.7     │ 0.627397 │ 0.214848  │ false          │\n",
      "│ 5   │ 0.3     │ 0.0      │ 0.0677736 │ false          │\n",
      "│ 6   │ 0.5     │ 0.486887 │ 0.191651  │ true           │\n",
      "│ 7   │ 0.1     │ 0.0      │ 0.0       │ false          │\n",
      "│ 8   │ 0.8     │ 0.867599 │ 0.431159  │ true           │\n",
      "│ 9   │ 0.9     │ 0.881403 │ 0.468373  │ true           │\n",
      "│ 10  │ 0.2     │ 0.0      │ 0.0       │ false          │\n",
      "IHT true positives = 7\n",
      "IHT false positives = 1\n",
      "IHT false negatives = 3\n",
      "LASSO true positives = 8\n",
      "LASSO false positives = 30\n",
      "LASSO false negatives = 2\n",
      "marginal true positives = 6\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 4\n",
      "\n",
      "current run = 11\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β  │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64  │ Bool           │\n",
      "├─────┼─────────┼──────────┼──────────┼────────────────┤\n",
      "│ 1   │ 0.1     │ 0.0      │ 0.0      │ false          │\n",
      "│ 2   │ 0.7     │ 0.789529 │ 0.443814 │ true           │\n",
      "│ 3   │ 0.6     │ 0.622341 │ 0.290237 │ true           │\n",
      "│ 4   │ 0.5     │ 0.532164 │ 0.255766 │ true           │\n",
      "│ 5   │ 0.3     │ 0.0      │ 0.0      │ false          │\n",
      "│ 6   │ 0.4     │ 0.0      │ 0.10381  │ false          │\n",
      "│ 7   │ 0.9     │ 0.885982 │ 0.539703 │ true           │\n",
      "│ 8   │ 0.8     │ 0.728428 │ 0.236208 │ true           │\n",
      "│ 9   │ 1.0     │ 0.982007 │ 0.619481 │ true           │\n",
      "│ 10  │ 0.2     │ 0.0      │ 0.0      │ false          │\n",
      "IHT true positives = 6\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 4\n",
      "LASSO true positives = 7\n",
      "LASSO false positives = 61\n",
      "LASSO false negatives = 3\n",
      "marginal true positives = 6\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 4\n",
      "\n",
      "current run = 12\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.3     │ 0.0      │ 7.2942e-5 │ false          │\n",
      "│ 2   │ 0.1     │ 0.0      │ 0.0       │ false          │\n",
      "│ 3   │ 0.7     │ 0.717842 │ 0.424281  │ true           │\n",
      "│ 4   │ 0.9     │ 0.845221 │ 0.277604  │ true           │\n",
      "│ 5   │ 0.6     │ 0.0      │ 0.117126  │ false          │\n",
      "│ 6   │ 0.2     │ 0.0      │ 0.0       │ false          │\n",
      "│ 7   │ 0.4     │ 0.0      │ 0.127203  │ false          │\n",
      "│ 8   │ 0.8     │ 0.646221 │ 0.382995  │ true           │\n",
      "│ 9   │ 0.5     │ 0.562382 │ 0.283411  │ true           │\n",
      "│ 10  │ 1.0     │ 0.80234  │ 0.448923  │ true           │\n",
      "IHT true positives = 5\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 5\n",
      "LASSO true positives = 8\n",
      "LASSO false positives = 58\n",
      "LASSO false negatives = 2\n",
      "marginal true positives = 5\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 5\n",
      "\n",
      "current run = 13\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.2     │ 0.0      │ 0.0       │ false          │\n",
      "│ 2   │ 0.9     │ 0.993235 │ 0.593845  │ true           │\n",
      "│ 3   │ 1.0     │ 0.865435 │ 0.493092  │ true           │\n",
      "│ 4   │ 0.6     │ 0.444414 │ 0.15488   │ true           │\n",
      "│ 5   │ 0.4     │ 0.0      │ 0.0547519 │ false          │\n",
      "│ 6   │ 0.8     │ 0.829484 │ 0.458596  │ true           │\n",
      "│ 7   │ 0.1     │ 0.0      │ 0.0       │ false          │\n",
      "│ 8   │ 0.7     │ 0.715573 │ 0.374642  │ true           │\n",
      "│ 9   │ 0.5     │ 0.475715 │ 0.184372  │ true           │\n",
      "│ 10  │ 0.3     │ 0.0      │ 0.0668804 │ false          │\n",
      "IHT true positives = 6\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 4\n",
      "LASSO true positives = 8\n",
      "LASSO false positives = 37\n",
      "LASSO false negatives = 2\n",
      "marginal true positives = 6\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 4\n",
      "\n",
      "current run = 14\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.1     │ 0.0      │ 0.0       │ false          │\n",
      "│ 2   │ 0.2     │ 0.0      │ 0.0772191 │ false          │\n",
      "│ 3   │ 0.8     │ 0.841813 │ 0.318093  │ true           │\n",
      "│ 4   │ 0.3     │ 0.0      │ 0.0       │ false          │\n",
      "│ 5   │ 0.6     │ 0.556563 │ 0.266232  │ true           │\n",
      "│ 6   │ 0.5     │ 0.0      │ 0.037174  │ false          │\n",
      "│ 7   │ 1.0     │ 0.895696 │ 0.563027  │ true           │\n",
      "│ 8   │ 0.9     │ 0.867979 │ 0.521898  │ true           │\n",
      "│ 9   │ 0.4     │ 0.462315 │ 0.190381  │ true           │\n",
      "│ 10  │ 0.7     │ 0.542367 │ 0.273612  │ true           │\n",
      "IHT true positives = 6\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 4\n",
      "LASSO true positives = 8\n",
      "LASSO false positives = 51\n",
      "LASSO false negatives = 2\n",
      "marginal true positives = 6\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 4\n",
      "\n",
      "current run = 15\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β  │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64  │ Bool           │\n",
      "├─────┼─────────┼──────────┼──────────┼────────────────┤\n",
      "│ 1   │ 0.9     │ 0.999155 │ 0.470987 │ true           │\n",
      "│ 2   │ 0.7     │ 0.714983 │ 0.335195 │ true           │\n",
      "│ 3   │ 0.1     │ 0.0      │ 0.0      │ false          │\n",
      "│ 4   │ 0.8     │ 0.783364 │ 0.428642 │ true           │\n",
      "│ 5   │ 1.0     │ 1.13359  │ 0.629874 │ true           │\n",
      "│ 6   │ 0.6     │ 0.535477 │ 0.219398 │ true           │\n",
      "│ 7   │ 0.5     │ 0.564364 │ 0.231996 │ true           │\n",
      "│ 8   │ 0.3     │ 0.0      │ 0.0      │ false          │\n",
      "│ 9   │ 0.2     │ 0.0      │ 0.0      │ false          │\n",
      "│ 10  │ 0.4     │ 0.0      │ 0.12852  │ true           │\n",
      "IHT true positives = 6\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 4\n",
      "LASSO true positives = 7\n",
      "LASSO false positives = 18\n",
      "LASSO false negatives = 3\n",
      "marginal true positives = 7\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 3\n",
      "\n",
      "current run = 16\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β    │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64    │ Bool           │\n",
      "├─────┼─────────┼──────────┼────────────┼────────────────┤\n",
      "│ 1   │ 0.1     │ 0.0      │ 0.0        │ false          │\n",
      "│ 2   │ 0.6     │ 0.575761 │ 0.264261   │ true           │\n",
      "│ 3   │ 0.3     │ 0.0      │ 0.0594738  │ false          │\n",
      "│ 4   │ 0.2     │ 0.0      │ 0.00985868 │ false          │\n",
      "│ 5   │ 0.5     │ 0.541443 │ 0.242547   │ true           │\n",
      "│ 6   │ 1.0     │ 1.01688  │ 0.595683   │ true           │\n",
      "│ 7   │ 0.8     │ 0.63407  │ 0.301564   │ true           │\n",
      "│ 8   │ 0.4     │ 0.473117 │ 0.170076   │ true           │\n",
      "│ 9   │ 0.9     │ 0.854271 │ 0.463257   │ true           │\n",
      "│ 10  │ 0.7     │ 0.671811 │ 0.33724    │ true           │\n",
      "IHT true positives = 7\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 3\n",
      "LASSO true positives = 9\n",
      "LASSO false positives = 40\n",
      "LASSO false negatives = 1\n",
      "marginal true positives = 7\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 3\n",
      "\n",
      "current run = 17\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.2     │ 0.0      │ 0.0       │ false          │\n",
      "│ 2   │ 0.4     │ 0.0      │ 0.0730643 │ false          │\n",
      "│ 3   │ 0.1     │ 0.0      │ 0.0       │ false          │\n",
      "│ 4   │ 1.0     │ 1.02307  │ 0.573481  │ true           │\n",
      "│ 5   │ 0.7     │ 0.631516 │ 0.312696  │ true           │\n",
      "│ 6   │ 0.8     │ 0.548035 │ 0.22072   │ true           │\n",
      "│ 7   │ 0.5     │ 0.0      │ 0.123706  │ true           │\n",
      "│ 8   │ 0.6     │ 0.557133 │ 0.248199  │ true           │\n",
      "│ 9   │ 0.3     │ 0.0      │ 0.0690182 │ false          │\n",
      "│ 10  │ 0.9     │ 0.879367 │ 0.53104   │ true           │\n",
      "IHT true positives = 5\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 5\n",
      "LASSO true positives = 8\n",
      "LASSO false positives = 13\n",
      "LASSO false negatives = 2\n",
      "marginal true positives = 6\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 4\n",
      "\n",
      "current run = 18\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.6     │ 0.607715 │ 0.242731  │ true           │\n",
      "│ 2   │ 0.2     │ 0.0      │ 0.0       │ false          │\n",
      "│ 3   │ 1.0     │ 1.00223  │ 0.551676  │ true           │\n",
      "│ 4   │ 0.1     │ 0.0      │ 0.0       │ false          │\n",
      "│ 5   │ 0.4     │ 0.0      │ 0.0586618 │ false          │\n",
      "│ 6   │ 0.7     │ 0.776481 │ 0.377501  │ true           │\n",
      "│ 7   │ 0.3     │ 0.0      │ 0.0357733 │ false          │\n",
      "│ 8   │ 0.9     │ 0.868313 │ 0.460552  │ true           │\n",
      "│ 9   │ 0.5     │ 0.574518 │ 0.231936  │ true           │\n",
      "│ 10  │ 0.8     │ 0.801032 │ 0.42098   │ true           │\n",
      "IHT true positives = 6\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 4\n",
      "LASSO true positives = 8\n",
      "LASSO false positives = 22\n",
      "LASSO false negatives = 2\n",
      "marginal true positives = 6\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 4\n",
      "\n",
      "current run = 19\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β  │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64  │ Bool           │\n",
      "├─────┼─────────┼──────────┼──────────┼────────────────┤\n",
      "│ 1   │ 0.7     │ 0.635951 │ 0.313151 │ true           │\n",
      "│ 2   │ 0.9     │ 0.748674 │ 0.432817 │ true           │\n",
      "│ 3   │ 0.3     │ 0.0      │ 0.0      │ false          │\n",
      "│ 4   │ 0.2     │ 0.0      │ 0.0      │ false          │\n",
      "│ 5   │ 0.1     │ 0.0      │ 0.0      │ false          │\n",
      "│ 6   │ 0.8     │ 0.712926 │ 0.381978 │ true           │\n",
      "│ 7   │ 0.4     │ 0.0      │ 0.11997  │ false          │\n",
      "│ 8   │ 1.0     │ 0.935609 │ 0.596371 │ true           │\n",
      "│ 9   │ 0.6     │ 0.444719 │ 0.191115 │ true           │\n",
      "│ 10  │ 0.5     │ 0.381878 │ 0.177894 │ true           │\n",
      "IHT true positives = 6\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 4\n",
      "LASSO true positives = 7\n",
      "LASSO false positives = 39\n",
      "LASSO false negatives = 3\n",
      "marginal true positives = 6\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 4\n",
      "\n",
      "current run = 20\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.5     │ 0.576779 │ 0.249151  │ true           │\n",
      "│ 2   │ 0.2     │ 0.0      │ 0.0       │ false          │\n",
      "│ 3   │ 1.0     │ 0.921558 │ 0.544404  │ true           │\n",
      "│ 4   │ 0.3     │ 0.0      │ 0.0745884 │ false          │\n",
      "│ 5   │ 0.6     │ 0.0      │ 0.106004  │ false          │\n",
      "│ 6   │ 0.1     │ 0.0      │ 0.0       │ false          │\n",
      "│ 7   │ 0.4     │ 0.0      │ 0.0457298 │ false          │\n",
      "│ 8   │ 0.9     │ 0.849792 │ 0.489159  │ true           │\n",
      "│ 9   │ 0.7     │ 0.754696 │ 0.392819  │ true           │\n",
      "│ 10  │ 0.8     │ 0.694103 │ 0.367317  │ true           │\n",
      "IHT true positives = 5\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 5\n",
      "LASSO true positives = 8\n",
      "LASSO false positives = 19\n",
      "LASSO false negatives = 2\n",
      "marginal true positives = 5\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 5\n",
      "\n",
      "current run = 21\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β    │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64    │ Bool           │\n",
      "├─────┼─────────┼──────────┼────────────┼────────────────┤\n",
      "│ 1   │ 0.2     │ 0.0      │ 0.0        │ false          │\n",
      "│ 2   │ 0.1     │ 0.0      │ 0.0        │ false          │\n",
      "│ 3   │ 0.9     │ 0.985468 │ 0.646724   │ true           │\n",
      "│ 4   │ 0.5     │ 0.627475 │ 0.339024   │ true           │\n",
      "│ 5   │ 0.3     │ 0.0      │ 0.00949512 │ false          │\n",
      "│ 6   │ 0.8     │ 0.873836 │ 0.523167   │ true           │\n",
      "│ 7   │ 0.6     │ 0.619914 │ 0.325271   │ true           │\n",
      "│ 8   │ 0.4     │ 0.432067 │ 0.190359   │ true           │\n",
      "│ 9   │ 1.0     │ 0.973689 │ 0.647192   │ true           │\n",
      "│ 10  │ 0.7     │ 0.767594 │ 0.470267   │ true           │\n",
      "IHT true positives = 7\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 3\n",
      "LASSO true positives = 8\n",
      "LASSO false positives = 118\n",
      "LASSO false negatives = 2\n",
      "marginal true positives = 7\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 3\n",
      "\n",
      "current run = 22\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.1     │ 0.0      │ 0.0       │ false          │\n",
      "│ 2   │ 0.2     │ 0.0      │ 0.0860364 │ false          │\n",
      "│ 3   │ 0.3     │ 0.0      │ 0.0265473 │ false          │\n",
      "│ 4   │ 0.6     │ 0.45549  │ 0.225071  │ true           │\n",
      "│ 5   │ 0.8     │ 0.75729  │ 0.48247   │ true           │\n",
      "│ 6   │ 0.7     │ 0.587614 │ 0.306061  │ true           │\n",
      "│ 7   │ 0.5     │ 0.458187 │ 0.201996  │ true           │\n",
      "│ 8   │ 0.4     │ 0.0      │ 0.0567072 │ false          │\n",
      "│ 9   │ 1.0     │ 0.954947 │ 0.655183  │ true           │\n",
      "│ 10  │ 0.9     │ 0.908672 │ 0.597394  │ true           │\n",
      "IHT true positives = 6\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 4\n",
      "LASSO true positives = 9\n",
      "LASSO false positives = 108\n",
      "LASSO false negatives = 1\n",
      "marginal true positives = 6\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 4\n",
      "\n",
      "current run = 23\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.5     │ 0.450723 │ 0.157794  │ true           │\n",
      "│ 2   │ 0.4     │ 0.375798 │ 0.0873432 │ false          │\n",
      "│ 3   │ 0.8     │ 0.717487 │ 0.355001  │ true           │\n",
      "│ 4   │ 0.9     │ 0.760581 │ 0.4051    │ true           │\n",
      "│ 5   │ 0.3     │ 0.0      │ 0.0656473 │ false          │\n",
      "│ 6   │ 0.2     │ 0.0      │ 0.0       │ false          │\n",
      "│ 7   │ 0.6     │ 0.485389 │ 0.208463  │ true           │\n",
      "│ 8   │ 1.0     │ 0.937495 │ 0.533983  │ true           │\n",
      "│ 9   │ 0.1     │ 0.0      │ 0.0       │ false          │\n",
      "│ 10  │ 0.7     │ 0.607043 │ 0.283714  │ true           │\n",
      "IHT true positives = 7\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 3\n",
      "LASSO true positives = 8\n",
      "LASSO false positives = 20\n",
      "LASSO false negatives = 2\n",
      "marginal true positives = 6\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 4\n",
      "\n",
      "current run = 24\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β  │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64  │ Bool           │\n",
      "├─────┼─────────┼──────────┼──────────┼────────────────┤\n",
      "│ 1   │ 0.2     │ 0.0      │ 0.0      │ false          │\n",
      "│ 2   │ 0.4     │ 0.408097 │ 0.154459 │ true           │\n",
      "│ 3   │ 0.3     │ 0.0      │ 0.0      │ false          │\n",
      "│ 4   │ 0.8     │ 1.0246   │ 0.585963 │ true           │\n",
      "│ 5   │ 0.5     │ 0.404177 │ 0.159114 │ true           │\n",
      "│ 6   │ 1.0     │ 0.896991 │ 0.53134  │ true           │\n",
      "│ 7   │ 0.1     │ 0.0      │ 0.0      │ false          │\n",
      "│ 8   │ 0.9     │ 1.01912  │ 0.605509 │ true           │\n",
      "│ 9   │ 0.7     │ 0.632082 │ 0.254428 │ true           │\n",
      "│ 10  │ 0.6     │ 0.714155 │ 0.351545 │ true           │\n",
      "IHT true positives = 7\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 3\n",
      "LASSO true positives = 7\n",
      "LASSO false positives = 48\n",
      "LASSO false negatives = 3\n",
      "marginal true positives = 7\n",
      "marginal false positives = 1\n",
      "marginal false negatives = 3\n",
      "\n",
      "current run = 25\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.8     │ 0.763271 │ 0.447319  │ true           │\n",
      "│ 2   │ 0.1     │ 0.0      │ 0.0       │ false          │\n",
      "│ 3   │ 0.2     │ 0.0      │ 0.0       │ false          │\n",
      "│ 4   │ 0.9     │ 0.870804 │ 0.441068  │ true           │\n",
      "│ 5   │ 0.3     │ 0.0      │ 0.0918773 │ false          │\n",
      "│ 6   │ 0.5     │ 0.430862 │ 0.192195  │ true           │\n",
      "│ 7   │ 1.0     │ 1.08158  │ 0.70825   │ true           │\n",
      "│ 8   │ 0.6     │ 0.652794 │ 0.331765  │ true           │\n",
      "│ 9   │ 0.7     │ 0.551451 │ 0.290984  │ true           │\n",
      "│ 10  │ 0.4     │ 0.0      │ 0.058591  │ false          │\n",
      "IHT true positives = 6\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 4\n",
      "LASSO true positives = 8\n",
      "LASSO false positives = 47\n",
      "LASSO false negatives = 2\n",
      "marginal true positives = 6\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 4\n",
      "\n",
      "current run = 26\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.9     │ 0.890547 │ 0.459162  │ true           │\n",
      "│ 2   │ 0.3     │ 0.0      │ 0.0379161 │ false          │\n",
      "│ 3   │ 1.0     │ 1.07301  │ 0.614577  │ true           │\n",
      "│ 4   │ 0.2     │ 0.0      │ 0.0       │ false          │\n",
      "│ 5   │ 0.4     │ 0.604685 │ 0.272977  │ true           │\n",
      "│ 6   │ 0.7     │ 0.608468 │ 0.282534  │ true           │\n",
      "│ 7   │ 0.6     │ 0.471677 │ 0.158644  │ false          │\n",
      "│ 8   │ 0.5     │ 0.595994 │ 0.262842  │ true           │\n",
      "│ 9   │ 0.8     │ 0.67876  │ 0.280316  │ true           │\n",
      "│ 10  │ 0.1     │ 0.0      │ 0.0       │ false          │\n",
      "IHT true positives = 7\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 3\n",
      "LASSO true positives = 8\n",
      "LASSO false positives = 17\n",
      "LASSO false negatives = 2\n",
      "marginal true positives = 6\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 4\n",
      "\n",
      "current run = 27\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.7     │ 0.726895 │ 0.265548  │ true           │\n",
      "│ 2   │ 0.8     │ 0.910793 │ 0.504557  │ true           │\n",
      "│ 3   │ 0.4     │ 0.0      │ 0.0462442 │ false          │\n",
      "│ 4   │ 0.5     │ 0.448489 │ 0.153084  │ false          │\n",
      "│ 5   │ 0.1     │ 0.0      │ 0.0       │ false          │\n",
      "│ 6   │ 0.2     │ 0.0      │ 0.0469064 │ false          │\n",
      "│ 7   │ 0.9     │ 1.0095   │ 0.575574  │ true           │\n",
      "│ 8   │ 0.3     │ 0.425917 │ 0.150294  │ false          │\n",
      "│ 9   │ 1.0     │ 1.12887  │ 0.624841  │ true           │\n",
      "│ 10  │ 0.6     │ 0.508025 │ 0.181952  │ false          │\n",
      "IHT true positives = 7\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 3\n",
      "LASSO true positives = 9\n",
      "LASSO false positives = 38\n",
      "LASSO false negatives = 1\n",
      "marginal true positives = 4\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 6\n",
      "\n",
      "current run = 28\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.7     │ 0.70203  │ 0.373134  │ true           │\n",
      "│ 2   │ 0.1     │ 0.0      │ 0.0       │ false          │\n",
      "│ 3   │ 1.0     │ 0.0      │ 0.0460271 │ false          │\n",
      "│ 4   │ 0.2     │ 0.0      │ 0.0       │ false          │\n",
      "│ 5   │ 0.4     │ 0.434961 │ 0.135482  │ false          │\n",
      "│ 6   │ 0.5     │ 0.482073 │ 0.178727  │ true           │\n",
      "│ 7   │ 0.3     │ 0.0      │ 0.0184877 │ false          │\n",
      "│ 8   │ 0.8     │ 0.786629 │ 0.435365  │ true           │\n",
      "│ 9   │ 0.9     │ 0.933805 │ 0.552074  │ true           │\n",
      "│ 10  │ 0.6     │ 0.56978  │ 0.260556  │ true           │\n",
      "IHT true positives = 6\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 4\n",
      "LASSO true positives = 8\n",
      "LASSO false positives = 11\n",
      "LASSO false negatives = 2\n",
      "marginal true positives = 5\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 5\n",
      "\n",
      "current run = 29\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β    │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64    │ Bool           │\n",
      "├─────┼─────────┼──────────┼────────────┼────────────────┤\n",
      "│ 1   │ 0.2     │ 0.0      │ 0.00926907 │ false          │\n",
      "│ 2   │ 0.4     │ 0.491116 │ 0.197444   │ true           │\n",
      "│ 3   │ 1.0     │ 0.89322  │ 0.488465   │ true           │\n",
      "│ 4   │ 0.8     │ 0.772526 │ 0.401561   │ true           │\n",
      "│ 5   │ 0.1     │ 0.0      │ 0.0        │ false          │\n",
      "│ 6   │ 0.3     │ 0.0      │ 0.0546797  │ false          │\n",
      "│ 7   │ 0.6     │ 0.597883 │ 0.303942   │ true           │\n",
      "│ 8   │ 0.5     │ 0.505323 │ 0.191641   │ true           │\n",
      "│ 9   │ 0.7     │ 0.66923  │ 0.354049   │ true           │\n",
      "│ 10  │ 0.9     │ 0.915203 │ 0.513553   │ true           │\n",
      "IHT true positives = 7\n",
      "IHT false positives = 1\n",
      "IHT false negatives = 3\n",
      "LASSO true positives = 9\n",
      "LASSO false positives = 46\n",
      "LASSO false negatives = 1\n",
      "marginal true positives = 7\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 3\n",
      "\n",
      "current run = 30\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.5     │ 0.444684 │ 0.15937   │ true           │\n",
      "│ 2   │ 0.4     │ 0.362678 │ 0.108442  │ false          │\n",
      "│ 3   │ 0.3     │ 0.0      │ 0.0972615 │ false          │\n",
      "│ 4   │ 0.9     │ 0.794333 │ 0.436833  │ true           │\n",
      "│ 5   │ 0.2     │ 0.0      │ 0.0843848 │ false          │\n",
      "│ 6   │ 1.0     │ 0.938162 │ 0.517629  │ true           │\n",
      "│ 7   │ 0.1     │ 0.0      │ 0.0       │ false          │\n",
      "│ 8   │ 0.8     │ 0.761634 │ 0.429932  │ true           │\n",
      "│ 9   │ 0.6     │ 0.520029 │ 0.246035  │ true           │\n",
      "│ 10  │ 0.7     │ 0.738181 │ 0.41332   │ true           │\n",
      "IHT true positives = 7\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 3\n",
      "LASSO true positives = 9\n",
      "LASSO false positives = 47\n",
      "LASSO false negatives = 1\n",
      "marginal true positives = 6\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 4\n",
      "\n",
      "current run = 31\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β    │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64    │ Bool           │\n",
      "├─────┼─────────┼──────────┼────────────┼────────────────┤\n",
      "│ 1   │ 0.2     │ 0.0      │ 0.00828067 │ false          │\n",
      "│ 2   │ 1.0     │ 0.830723 │ 0.4994     │ true           │\n",
      "│ 3   │ 0.4     │ 0.0      │ 0.0897489  │ false          │\n",
      "│ 4   │ 0.9     │ 0.692752 │ 0.392689   │ true           │\n",
      "│ 5   │ 0.1     │ 0.0      │ 0.0        │ false          │\n",
      "│ 6   │ 0.5     │ 0.0      │ 0.0910232  │ false          │\n",
      "│ 7   │ 0.7     │ 0.717895 │ 0.400382   │ true           │\n",
      "│ 8   │ 0.6     │ 0.635714 │ 0.326699   │ true           │\n",
      "│ 9   │ 0.8     │ 0.0      │ 0.0608871  │ false          │\n",
      "│ 10  │ 0.3     │ 0.0      │ 0.0412942  │ false          │\n",
      "IHT true positives = 4\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 6\n",
      "LASSO true positives = 9\n",
      "LASSO false positives = 21\n",
      "LASSO false negatives = 1\n",
      "marginal true positives = 4\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 6\n",
      "\n",
      "current run = 32\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 1.0     │ 0.85864  │ 0.521806  │ true           │\n",
      "│ 2   │ 0.1     │ 0.0      │ 0.0180191 │ false          │\n",
      "│ 3   │ 0.8     │ 0.79716  │ 0.358518  │ true           │\n",
      "│ 4   │ 0.7     │ 0.727523 │ 0.397702  │ true           │\n",
      "│ 5   │ 0.6     │ 0.604671 │ 0.333157  │ true           │\n",
      "│ 6   │ 0.9     │ 0.925252 │ 0.579486  │ true           │\n",
      "│ 7   │ 0.4     │ 0.0      │ 0.105055  │ false          │\n",
      "│ 8   │ 0.2     │ 0.0      │ 0.0       │ false          │\n",
      "│ 9   │ 0.5     │ 0.60235  │ 0.25473   │ true           │\n",
      "│ 10  │ 0.3     │ 0.420457 │ 0.16356   │ false          │\n",
      "IHT true positives = 7\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 3\n",
      "LASSO true positives = 9\n",
      "LASSO false positives = 117\n",
      "LASSO false negatives = 1\n",
      "marginal true positives = 6\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 4\n",
      "\n",
      "current run = 33\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.7     │ 0.560691 │ 0.264613  │ true           │\n",
      "│ 2   │ 0.9     │ 0.837837 │ 0.473388  │ true           │\n",
      "│ 3   │ 0.8     │ 0.721231 │ 0.363206  │ true           │\n",
      "│ 4   │ 0.4     │ 0.0      │ 0.0679089 │ false          │\n",
      "│ 5   │ 1.0     │ 0.833116 │ 0.464721  │ true           │\n",
      "│ 6   │ 0.2     │ 0.0      │ 0.0       │ false          │\n",
      "│ 7   │ 0.6     │ 0.687556 │ 0.354763  │ true           │\n",
      "│ 8   │ 0.5     │ 0.575712 │ 0.26969   │ true           │\n",
      "│ 9   │ 0.1     │ 0.0      │ 0.0       │ false          │\n",
      "│ 10  │ 0.3     │ 0.0      │ 0.0632281 │ false          │\n",
      "IHT true positives = 6\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 4\n",
      "LASSO true positives = 8\n",
      "LASSO false positives = 39\n",
      "LASSO false negatives = 2\n",
      "marginal true positives = 6\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 4\n",
      "\n",
      "current run = 34\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.7     │ 0.518452 │ 0.23595   │ false          │\n",
      "│ 2   │ 0.6     │ 0.568934 │ 0.30206   │ true           │\n",
      "│ 3   │ 0.1     │ 0.0      │ 0.0       │ false          │\n",
      "│ 4   │ 0.3     │ 0.0      │ 0.142029  │ false          │\n",
      "│ 5   │ 0.8     │ 0.649641 │ 0.346092  │ true           │\n",
      "│ 6   │ 0.4     │ 0.0      │ 0.0387174 │ false          │\n",
      "│ 7   │ 0.9     │ 0.791095 │ 0.509347  │ true           │\n",
      "│ 8   │ 0.2     │ 0.0      │ 0.0376166 │ false          │\n",
      "│ 9   │ 1.0     │ 0.860793 │ 0.530989  │ true           │\n",
      "│ 10  │ 0.5     │ 0.481192 │ 0.245787  │ true           │\n",
      "IHT true positives = 6\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 4\n",
      "LASSO true positives = 9\n",
      "LASSO false positives = 103\n",
      "LASSO false negatives = 1\n",
      "marginal true positives = 5\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 5\n",
      "\n",
      "current run = 35\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 1.0     │ 1.0716   │ 0.616817  │ true           │\n",
      "│ 2   │ 0.7     │ 0.617388 │ 0.305827  │ true           │\n",
      "│ 3   │ 0.3     │ 0.0      │ 0.0403806 │ false          │\n",
      "│ 4   │ 0.9     │ 0.914056 │ 0.52991   │ true           │\n",
      "│ 5   │ 0.4     │ 0.405515 │ 0.123701  │ false          │\n",
      "│ 6   │ 0.6     │ 0.58388  │ 0.270245  │ true           │\n",
      "│ 7   │ 0.1     │ 0.0      │ 0.0       │ false          │\n",
      "│ 8   │ 0.8     │ 0.896144 │ 0.583712  │ true           │\n",
      "│ 9   │ 0.5     │ 0.516869 │ 0.147351  │ false          │\n",
      "│ 10  │ 0.2     │ 0.0      │ 0.0       │ false          │\n",
      "IHT true positives = 7\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 3\n",
      "LASSO true positives = 8\n",
      "LASSO false positives = 79\n",
      "LASSO false negatives = 2\n",
      "marginal true positives = 5\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 5\n",
      "\n",
      "current run = 36\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.7     │ 0.818948 │ 0.484817  │ true           │\n",
      "│ 2   │ 0.9     │ 0.987382 │ 0.638536  │ true           │\n",
      "│ 3   │ 1.0     │ 0.971542 │ 0.615807  │ true           │\n",
      "│ 4   │ 0.1     │ 0.0      │ 0.0       │ false          │\n",
      "│ 5   │ 0.4     │ 0.0      │ 0.0582323 │ false          │\n",
      "│ 6   │ 0.5     │ 0.405865 │ 0.151072  │ false          │\n",
      "│ 7   │ 0.3     │ 0.0      │ 0.0150599 │ false          │\n",
      "│ 8   │ 0.6     │ 0.61255  │ 0.273084  │ true           │\n",
      "│ 9   │ 0.2     │ 0.0      │ 0.0289527 │ false          │\n",
      "│ 10  │ 0.8     │ 0.665462 │ 0.324654  │ true           │\n",
      "IHT true positives = 6\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 4\n",
      "LASSO true positives = 9\n",
      "LASSO false positives = 72\n",
      "LASSO false negatives = 1\n",
      "marginal true positives = 5\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 5\n",
      "\n",
      "current run = 37\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.8     │ 0.808631 │ 0.384195  │ true           │\n",
      "│ 2   │ 0.6     │ 0.780601 │ 0.407328  │ true           │\n",
      "│ 3   │ 0.4     │ 0.0      │ 0.0443346 │ false          │\n",
      "│ 4   │ 0.7     │ 0.721877 │ 0.337069  │ true           │\n",
      "│ 5   │ 1.0     │ 1.0028   │ 0.573165  │ true           │\n",
      "│ 6   │ 0.2     │ 0.0      │ 0.0       │ false          │\n",
      "│ 7   │ 0.5     │ 0.449095 │ 0.1394    │ false          │\n",
      "│ 8   │ 0.3     │ 0.340997 │ 0.0606009 │ false          │\n",
      "│ 9   │ 0.9     │ 0.799309 │ 0.422671  │ true           │\n",
      "│ 10  │ 0.1     │ 0.0      │ 0.0       │ false          │\n",
      "IHT true positives = 7\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 3\n",
      "LASSO true positives = 8\n",
      "LASSO false positives = 20\n",
      "LASSO false negatives = 2\n",
      "marginal true positives = 5\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 5\n",
      "\n",
      "current run = 38\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 1.0     │ 1.09846  │ 0.668877  │ true           │\n",
      "│ 2   │ 0.6     │ 0.64138  │ 0.305831  │ true           │\n",
      "│ 3   │ 0.7     │ 0.623225 │ 0.293462  │ true           │\n",
      "│ 4   │ 0.8     │ 0.954045 │ 0.556247  │ true           │\n",
      "│ 5   │ 0.2     │ 0.0      │ 0.0       │ false          │\n",
      "│ 6   │ 0.5     │ 0.559322 │ 0.242857  │ true           │\n",
      "│ 7   │ 0.3     │ 0.0      │ 0.0303946 │ false          │\n",
      "│ 8   │ 0.9     │ 0.738578 │ 0.398435  │ true           │\n",
      "│ 9   │ 0.4     │ 0.0      │ 0.117215  │ false          │\n",
      "│ 10  │ 0.1     │ 0.0      │ 0.0       │ false          │\n",
      "IHT true positives = 6\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 4\n",
      "LASSO true positives = 8\n",
      "LASSO false positives = 41\n",
      "LASSO false negatives = 2\n",
      "marginal true positives = 6\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 4\n",
      "\n",
      "current run = 39\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.8     │ 0.723762 │ 0.374709  │ true           │\n",
      "│ 2   │ 1.0     │ 0.830871 │ 0.461295  │ true           │\n",
      "│ 3   │ 0.6     │ 0.47017  │ 0.163009  │ true           │\n",
      "│ 4   │ 0.4     │ 0.0      │ 0.0668645 │ false          │\n",
      "│ 5   │ 0.2     │ 0.0      │ 0.0       │ false          │\n",
      "│ 6   │ 0.7     │ 0.655515 │ 0.293149  │ true           │\n",
      "│ 7   │ 0.3     │ 0.0      │ 0.0       │ false          │\n",
      "│ 8   │ 0.5     │ 0.55116  │ 0.250632  │ true           │\n",
      "│ 9   │ 0.1     │ 0.0      │ 0.0       │ false          │\n",
      "│ 10  │ 0.9     │ 0.805071 │ 0.424746  │ true           │\n",
      "IHT true positives = 6\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 4\n",
      "LASSO true positives = 7\n",
      "LASSO false positives = 28\n",
      "LASSO false negatives = 3\n",
      "marginal true positives = 6\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 4\n",
      "\n",
      "current run = 40\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.1     │ 0.0      │ 0.0       │ false          │\n",
      "│ 2   │ 0.4     │ 0.0      │ 0.065     │ false          │\n",
      "│ 3   │ 0.3     │ 0.0      │ 0.0765434 │ false          │\n",
      "│ 4   │ 1.0     │ 0.873786 │ 0.464308  │ true           │\n",
      "│ 5   │ 0.2     │ 0.0      │ 0.0       │ false          │\n",
      "│ 6   │ 0.7     │ 0.712277 │ 0.365306  │ true           │\n",
      "│ 7   │ 0.9     │ 0.755782 │ 0.420657  │ true           │\n",
      "│ 8   │ 0.5     │ 0.0      │ 0.0922344 │ false          │\n",
      "│ 9   │ 0.6     │ 0.655156 │ 0.327507  │ true           │\n",
      "│ 10  │ 0.8     │ 0.75099  │ 0.424221  │ true           │\n",
      "IHT true positives = 5\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 5\n",
      "LASSO true positives = 8\n",
      "LASSO false positives = 27\n",
      "LASSO false negatives = 2\n",
      "marginal true positives = 5\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 5\n",
      "\n",
      "current run = 41\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β    │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64    │ Bool           │\n",
      "├─────┼─────────┼──────────┼────────────┼────────────────┤\n",
      "│ 1   │ 0.4     │ 0.41818  │ 0.157827   │ false          │\n",
      "│ 2   │ 0.6     │ 0.585098 │ 0.279236   │ true           │\n",
      "│ 3   │ 0.5     │ 0.399243 │ 0.126335   │ false          │\n",
      "│ 4   │ 0.9     │ 0.861845 │ 0.517258   │ true           │\n",
      "│ 5   │ 0.7     │ 0.717022 │ 0.391833   │ true           │\n",
      "│ 6   │ 0.2     │ 0.0      │ 0.00241304 │ false          │\n",
      "│ 7   │ 0.8     │ 0.709602 │ 0.420775   │ true           │\n",
      "│ 8   │ 0.1     │ 0.0      │ 0.0        │ false          │\n",
      "│ 9   │ 1.0     │ 1.02579  │ 0.617811   │ true           │\n",
      "│ 10  │ 0.3     │ 0.0      │ 0.132904   │ false          │\n",
      "IHT true positives = 7\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 3\n",
      "LASSO true positives = 9\n",
      "LASSO false positives = 100\n",
      "LASSO false negatives = 1\n",
      "marginal true positives = 5\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 5\n",
      "\n",
      "current run = 42\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.6     │ 0.403459 │ 0.135112  │ false          │\n",
      "│ 2   │ 0.3     │ 0.0      │ 0.072035  │ false          │\n",
      "│ 3   │ 0.5     │ 0.0      │ 0.0235897 │ false          │\n",
      "│ 4   │ 0.2     │ 0.0      │ 0.0       │ false          │\n",
      "│ 5   │ 0.8     │ 0.772452 │ 0.439127  │ true           │\n",
      "│ 6   │ 0.7     │ 0.50003  │ 0.225048  │ true           │\n",
      "│ 7   │ 1.0     │ 0.864949 │ 0.51498   │ true           │\n",
      "│ 8   │ 0.1     │ 0.0      │ 0.0       │ false          │\n",
      "│ 9   │ 0.4     │ 0.0      │ 0.0788868 │ false          │\n",
      "│ 10  │ 0.9     │ 0.687788 │ 0.325017  │ true           │\n",
      "IHT true positives = 5\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 5\n",
      "LASSO true positives = 8\n",
      "LASSO false positives = 8\n",
      "LASSO false negatives = 2\n",
      "marginal true positives = 4\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 6\n",
      "\n",
      "current run = 43\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β  │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64  │ Bool           │\n",
      "├─────┼─────────┼──────────┼──────────┼────────────────┤\n",
      "│ 1   │ 0.7     │ 0.674637 │ 0.337381 │ true           │\n",
      "│ 2   │ 0.4     │ 0.0      │ 0.140329 │ true           │\n",
      "│ 3   │ 0.6     │ 0.598628 │ 0.254094 │ true           │\n",
      "│ 4   │ 0.1     │ 0.0      │ 0.0      │ false          │\n",
      "│ 5   │ 0.2     │ 0.0      │ 0.0      │ false          │\n",
      "│ 6   │ 0.8     │ 0.660451 │ 0.317312 │ true           │\n",
      "│ 7   │ 0.5     │ 0.418876 │ 0.156516 │ true           │\n",
      "│ 8   │ 0.9     │ 0.672597 │ 0.386514 │ true           │\n",
      "│ 9   │ 0.3     │ 0.0      │ 0.065986 │ false          │\n",
      "│ 10  │ 1.0     │ 0.819398 │ 0.488731 │ true           │\n",
      "IHT true positives = 6\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 4\n",
      "LASSO true positives = 8\n",
      "LASSO false positives = 23\n",
      "LASSO false negatives = 2\n",
      "marginal true positives = 7\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 3\n",
      "\n",
      "current run = 44\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β  │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64  │ Bool           │\n",
      "├─────┼─────────┼──────────┼──────────┼────────────────┤\n",
      "│ 1   │ 0.2     │ 0.0      │ 0.0      │ false          │\n",
      "│ 2   │ 0.1     │ 0.0      │ 0.0      │ false          │\n",
      "│ 3   │ 0.8     │ 0.666172 │ 0.322054 │ true           │\n",
      "│ 4   │ 0.3     │ 0.0      │ 0.0      │ false          │\n",
      "│ 5   │ 0.4     │ 0.0      │ 0.110887 │ false          │\n",
      "│ 6   │ 0.6     │ 0.691079 │ 0.383924 │ true           │\n",
      "│ 7   │ 0.9     │ 0.741535 │ 0.377643 │ true           │\n",
      "│ 8   │ 1.0     │ 1.18199  │ 0.734309 │ true           │\n",
      "│ 9   │ 0.7     │ 0.604375 │ 0.285495 │ true           │\n",
      "│ 10  │ 0.5     │ 0.545101 │ 0.253259 │ true           │\n",
      "IHT true positives = 6\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 4\n",
      "LASSO true positives = 7\n",
      "LASSO false positives = 34\n",
      "LASSO false negatives = 3\n",
      "marginal true positives = 6\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 4\n",
      "\n",
      "current run = 45\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β  │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64  │ Bool           │\n",
      "├─────┼─────────┼──────────┼──────────┼────────────────┤\n",
      "│ 1   │ 0.3     │ 0.0      │ 0.0      │ false          │\n",
      "│ 2   │ 0.7     │ 0.842753 │ 0.442551 │ true           │\n",
      "│ 3   │ 0.2     │ 0.0      │ 0.0      │ false          │\n",
      "│ 4   │ 0.1     │ 0.0      │ 0.0      │ false          │\n",
      "│ 5   │ 0.8     │ 0.729033 │ 0.414261 │ true           │\n",
      "│ 6   │ 0.9     │ 0.727785 │ 0.398587 │ true           │\n",
      "│ 7   │ 1.0     │ 0.853803 │ 0.498159 │ true           │\n",
      "│ 8   │ 0.4     │ 0.0      │ 0.104047 │ false          │\n",
      "│ 9   │ 0.5     │ 0.546907 │ 0.235639 │ true           │\n",
      "│ 10  │ 0.6     │ 0.604048 │ 0.31566  │ true           │\n",
      "IHT true positives = 6\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 4\n",
      "LASSO true positives = 7\n",
      "LASSO false positives = 56\n",
      "LASSO false negatives = 3\n",
      "marginal true positives = 6\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 4\n",
      "\n",
      "current run = 46\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β    │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64    │ Bool           │\n",
      "├─────┼─────────┼──────────┼────────────┼────────────────┤\n",
      "│ 1   │ 0.5     │ 0.463837 │ 0.122883   │ false          │\n",
      "│ 2   │ 0.8     │ 0.759727 │ 0.393282   │ true           │\n",
      "│ 3   │ 0.4     │ 0.521458 │ 0.169805   │ false          │\n",
      "│ 4   │ 0.6     │ 0.736244 │ 0.356339   │ true           │\n",
      "│ 5   │ 0.9     │ 0.782345 │ 0.379864   │ true           │\n",
      "│ 6   │ 0.7     │ 0.692802 │ 0.346316   │ true           │\n",
      "│ 7   │ 0.3     │ 0.0      │ 0.00422443 │ false          │\n",
      "│ 8   │ 0.2     │ 0.352494 │ 0.072355   │ false          │\n",
      "│ 9   │ 1.0     │ 0.882295 │ 0.409935   │ true           │\n",
      "│ 10  │ 0.1     │ 0.0      │ 0.0        │ false          │\n",
      "IHT true positives = 8\n",
      "IHT false positives = 1\n",
      "IHT false negatives = 2\n",
      "LASSO true positives = 9\n",
      "LASSO false positives = 48\n",
      "LASSO false negatives = 1\n",
      "marginal true positives = 5\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 5\n",
      "\n",
      "current run = 47\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.8     │ 0.658326 │ 0.403935  │ true           │\n",
      "│ 2   │ 0.1     │ 0.0      │ 0.0       │ false          │\n",
      "│ 3   │ 1.0     │ 1.07598  │ 0.691063  │ true           │\n",
      "│ 4   │ 0.7     │ 0.670537 │ 0.37657   │ true           │\n",
      "│ 5   │ 0.2     │ 0.0      │ 0.08815   │ false          │\n",
      "│ 6   │ 0.3     │ 0.0      │ 0.0358827 │ false          │\n",
      "│ 7   │ 0.9     │ 0.881036 │ 0.491326  │ true           │\n",
      "│ 8   │ 0.5     │ 0.568065 │ 0.279942  │ true           │\n",
      "│ 9   │ 0.6     │ 0.530217 │ 0.269293  │ true           │\n",
      "│ 10  │ 0.4     │ 0.0      │ 0.11268   │ false          │\n",
      "IHT true positives = 6\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 4\n",
      "LASSO true positives = 9\n",
      "LASSO false positives = 82\n",
      "LASSO false negatives = 1\n",
      "marginal true positives = 6\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 4\n",
      "\n",
      "current run = 48\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 1.0     │ 0.931564 │ 0.596104  │ true           │\n",
      "│ 2   │ 0.8     │ 0.735384 │ 0.395256  │ true           │\n",
      "│ 3   │ 0.9     │ 0.694223 │ 0.373158  │ true           │\n",
      "│ 4   │ 0.7     │ 0.701218 │ 0.419901  │ true           │\n",
      "│ 5   │ 0.5     │ 0.587413 │ 0.280345  │ true           │\n",
      "│ 6   │ 0.6     │ 0.570492 │ 0.271315  │ true           │\n",
      "│ 7   │ 0.1     │ 0.0      │ 0.0       │ false          │\n",
      "│ 8   │ 0.2     │ 0.0      │ 0.0       │ false          │\n",
      "│ 9   │ 0.4     │ 0.0      │ 0.124178  │ false          │\n",
      "│ 10  │ 0.3     │ 0.0      │ 0.0277466 │ false          │\n",
      "IHT true positives = 6\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 4\n",
      "LASSO true positives = 8\n",
      "LASSO false positives = 43\n",
      "LASSO false negatives = 2\n",
      "marginal true positives = 6\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 4\n",
      "\n",
      "current run = 49\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.3     │ 0.0      │ 0.0       │ false          │\n",
      "│ 2   │ 0.2     │ 0.0      │ 0.0       │ false          │\n",
      "│ 3   │ 0.7     │ 0.734939 │ 0.340656  │ true           │\n",
      "│ 4   │ 0.9     │ 0.919406 │ 0.452677  │ true           │\n",
      "│ 5   │ 0.1     │ 0.0      │ 0.0       │ false          │\n",
      "│ 6   │ 0.5     │ 0.632056 │ 0.283654  │ true           │\n",
      "│ 7   │ 0.8     │ 0.825943 │ 0.417897  │ true           │\n",
      "│ 8   │ 0.4     │ 0.0      │ 0.0200044 │ false          │\n",
      "│ 9   │ 1.0     │ 1.00329  │ 0.536308  │ true           │\n",
      "│ 10  │ 0.6     │ 0.470194 │ 0.174087  │ true           │\n",
      "IHT true positives = 6\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 4\n",
      "LASSO true positives = 7\n",
      "LASSO false positives = 18\n",
      "LASSO false negatives = 3\n",
      "marginal true positives = 6\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 4\n",
      "\n",
      "current run = 50\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 1.0\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = logistic\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.1     │ 0.0      │ 0.0       │ false          │\n",
      "│ 2   │ 0.5     │ 0.0      │ 0.0444567 │ false          │\n",
      "│ 3   │ 0.2     │ 0.0      │ 0.0       │ false          │\n",
      "│ 4   │ 0.9     │ 0.849397 │ 0.484971  │ true           │\n",
      "│ 5   │ 0.7     │ 0.680756 │ 0.370854  │ true           │\n",
      "│ 6   │ 1.0     │ 0.947222 │ 0.551212  │ true           │\n",
      "│ 7   │ 0.6     │ 0.54616  │ 0.261105  │ true           │\n",
      "│ 8   │ 0.4     │ 0.0      │ 0.0599635 │ false          │\n",
      "│ 9   │ 0.3     │ 0.0      │ 0.151399  │ true           │\n",
      "│ 10  │ 0.8     │ 0.798082 │ 0.454913  │ true           │\n",
      "IHT true positives = 5\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 5\n",
      "LASSO true positives = 8\n",
      "LASSO false positives = 26\n",
      "LASSO false negatives = 2\n",
      "marginal true positives = 6\n",
      "marginal false positives = 0\n",
      "marginal false negatives = 4\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "([7.0, 7.0, 7.0, 8.0, 8.0, 6.0, 6.0, 6.0, 6.0, 7.0  …  7.0, 5.0, 6.0, 6.0, 6.0, 8.0, 6.0, 6.0, 6.0, 5.0], [1.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 1.0  …  0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0], [3.0, 3.0, 3.0, 2.0, 2.0, 4.0, 4.0, 4.0, 4.0, 3.0  …  3.0, 5.0, 4.0, 4.0, 4.0, 2.0, 4.0, 4.0, 4.0, 5.0], [9.0, 9.0, 9.0, 8.0, 8.0, 8.0, 8.0, 8.0, 9.0, 8.0  …  9.0, 8.0, 8.0, 7.0, 7.0, 9.0, 9.0, 8.0, 7.0, 8.0], [40.0, 122.0, 28.0, 7.0, 75.0, 31.0, 52.0, 41.0, 19.0, 30.0  …  100.0, 8.0, 23.0, 34.0, 56.0, 48.0, 82.0, 43.0, 18.0, 26.0], [1.0, 1.0, 1.0, 2.0, 2.0, 2.0, 2.0, 2.0, 1.0, 2.0  …  1.0, 2.0, 2.0, 3.0, 3.0, 1.0, 1.0, 2.0, 3.0, 2.0], [5.0, 6.0, 6.0, 6.0, 5.0, 6.0, 6.0, 5.0, 7.0, 6.0  …  5.0, 4.0, 7.0, 6.0, 6.0, 5.0, 6.0, 6.0, 6.0, 6.0], [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0  …  0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0], [5.0, 4.0, 4.0, 4.0, 5.0, 4.0, 4.0, 5.0, 3.0, 4.0  …  5.0, 6.0, 3.0, 4.0, 4.0, 5.0, 4.0, 4.0, 4.0, 4.0])"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "iht_true_positives, iht_false_positives, iht_false_negatives, \n",
    "    lasso_true_positives, lasso_false_positives, lasso_false_negatives, \n",
    "    marginal_true_positives, marginal_false_positives, marginal_false_negatives = run()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3×3 Array{Float64,2}:\n",
       " 6.28   0.1   3.72\n",
       " 8.16  45.76  1.84\n",
       " 5.76   0.02  4.24"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "logistic_iht_true_positives = sum(iht_true_positives) / 50\n",
    "logistic_iht_false_positives = sum(iht_false_positives) / 50\n",
    "logistic_iht_false_negatives = sum(iht_false_negatives) / 50\n",
    "logistic_lasso_true_positives = sum(lasso_true_positives) / 50\n",
    "logistic_lasso_false_positives = sum(lasso_false_positives) / 50\n",
    "logistic_lasso_false_negatives = sum(lasso_false_negatives) / 50\n",
    "logistic_marginal_true_positives = sum(marginal_true_positives) / 50\n",
    "logistic_marginal_false_positives = sum(marginal_false_positives) / 50\n",
    "logistic_marginal_false_negatives = sum(marginal_false_negatives) / 50\n",
    "\n",
    "result = [logistic_iht_true_positives logistic_iht_false_positives logistic_iht_false_negatives; \n",
    "        logistic_lasso_true_positives logistic_lasso_false_positives logistic_lasso_false_negatives;\n",
    "        logistic_marginal_true_positives logistic_marginal_false_positives logistic_marginal_false_negatives]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Julia 1.0.3",
   "language": "julia",
   "name": "julia-1.0"
  },
  "language_info": {
   "file_extension": ".jl",
   "mimetype": "application/julia",
   "name": "julia",
   "version": "1.0.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
