{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Poisson: IHT vs LASSO vs Marginal regression: comparing false positive/negatives\n",
    "\n",
    "Marginal regression is the most prevalent method used to analyze GWAS data. LASSO is currently the de facto penalized least squares method for [feature selection](https://en.wikipedia.org/wiki/Feature_selection). Here we compare the performance (in terms of the number of false positives/negative) of LASSO and marginal regression with IHT for GWAS data. We use the `glmnet` implementation of cyclic coordinate descent for LASSO, and `MendelGWAS.jl` module of [OpenMendel](https://openmendel.github.io/MendelGWAS.jl/) for marginal analysis. Since the focus here is not scalability, we test our sample problems on moderately sized genotype matrces of 1000 samples with 10,000 SNPs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "31"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "using Distributed\n",
    "addprocs(30)\n",
    "nprocs()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "┌ Info: Loading DataFrames support into Gadfly.jl\n",
      "└ @ Gadfly /u/home/b/biona001/.julia/packages/Gadfly/09PWZ/src/mapping.jl:228\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "[ Info: Loading DataFrames support into Gadfly.jl\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall /u/home/b/biona001/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: During startup - Warning message:\n",
      "│ Setting LC_CTYPE failed, using \"C\" \n",
      "└ @ RCall ~/.julia/packages/RCall/ffM0W/src/io.jl:113\n",
      "┌ Warning: RCall.jl: Loading required package: Matrix\n",
      "│ Loading required package: foreach\n",
      "│ Loaded glmnet 2.0-16\n",
      "│ \n",
      "└ @ RCall /u/home/b/biona001/.julia/packages/RCall/ffM0W/src/io.jl:113\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "RObject{StrSxp}\n",
       " [1] \"glmnet\"    \"foreach\"   \"Matrix\"    \"stats\"     \"graphics\"  \"grDevices\"\n",
       " [7] \"utils\"     \"datasets\"  \"methods\"   \"base\"     \n"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "using MendelIHT\n",
    "using SnpArrays\n",
    "using DataFrames\n",
    "using Distributions\n",
    "using Random\n",
    "using LinearAlgebra\n",
    "using DelimitedFiles\n",
    "using GLM\n",
    "using MendelGWAS\n",
    "using CSV\n",
    "using RCall\n",
    "R\"library(glmnet)\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Poisson response"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "iht_lasso_marginal (generic function with 1 method)"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "function iht_lasso_marginal(n::Int64, p::Int64, d::UnionAll, l::Link)\n",
    "    #construct snpmatrix, covariate files, and true model b\n",
    "    x, = simulate_random_snparray(n, p, \"tmp.bed\")\n",
    "    xbm = SnpBitMatrix{Float64}(x, model=ADDITIVE_MODEL, center=true, scale=true); \n",
    "    z = ones(n, 1) # the intercept\n",
    "    x_float = [convert(Matrix{Float64}, x, center=true, scale=true) z] #Float64 version of x\n",
    "\n",
    "    # simulate response, true model b, and the correct non-0 positions of b\n",
    "    true_b = zeros(p)\n",
    "    true_b[1:10] .= collect(0.1:0.1:1.0)\n",
    "    k = 10\n",
    "    shuffle!(true_b)\n",
    "    correct_position = findall(x -> x != 0, true_b)\n",
    "\n",
    "    #simulate phenotypes (e.g. vector y)\n",
    "    if d == Normal || d == Poisson || d == Bernoulli\n",
    "        prob = linkinv.(l, xbm * true_b)\n",
    "        clamp!(prob, -20, 20)\n",
    "        y = [rand(d(i)) for i in prob]\n",
    "    elseif d == NegativeBinomial\n",
    "        μ = linkinv.(l, xbm * true_b)\n",
    "        prob = 1 ./ (1 .+ μ ./ nn)\n",
    "        y = [rand(d(nn, i)) for i in prob] #number of failtures before nn success occurs\n",
    "    elseif d == Gamma\n",
    "        μ = linkinv.(l, xbm * true_b)\n",
    "        β = 1 ./ μ # here β is the rate parameter for gamma distribution\n",
    "        y = [rand(d(α, i)) for i in β] # α is the shape parameter for gamma\n",
    "    end\n",
    "    y = Float64.(y)\n",
    "\n",
    "    #specify path and folds\n",
    "    num_folds = 3\n",
    "    folds = rand(1:num_folds, size(x, 1));\n",
    "\n",
    "    #run glmnet via Rcall\n",
    "    @rput x_float y folds num_folds #make variables visible to R\n",
    "    R\"lasso_cv_result = cv.glmnet(x_float, y, nfolds = num_folds, foldid = folds, family='poisson')\"\n",
    "    R\"lasso_beta_tmp = glmnet(x_float, y, lambda=lasso_cv_result$lambda.min, family='poisson')$beta\"\n",
    "    R\"lasso_beta = as.vector(lasso_beta_tmp)\"\n",
    "    @rget lasso_cv_result lasso_beta #pull result from R to Julia\n",
    "    lasso_k_est = count(!iszero, lasso_beta)\n",
    "    \n",
    "    #find non-zero entries returned by best lasso model as largest k estimate\n",
    "    path = collect(1:50);\n",
    "    \n",
    "    #run IHT's cross validation routine \n",
    "    mses = cv_iht(d(), l, x, z, y, 1, path, num_folds, folds=folds, use_maf=false, debias=false, showinfo=false, parallel=true);\n",
    "    iht_k_est = argmin(mses)\n",
    "    iht_result = L0_reg(x, xbm, z, y, 1, iht_k_est, d(), l, debias=false, init=false, use_maf=false)\n",
    "    iht_beta = iht_result.beta\n",
    "    \n",
    "    #Now run MendelGWAS\n",
    "    make_bim_fam_files(x, y, \"tmp\") #create .bim and .bed files for MendelGWAS\n",
    "\n",
    "    #create a control file for MendelGWAS\n",
    "    open(\"tmp_control.txt\", \"w\") do f\n",
    "        write(f, \"plink_input_basename = tmp \\n\")\n",
    "        write(f, \"plink_field_separator = '\\t' \\n\")\n",
    "        write(f, \"output_table = tmp_table.txt \\n\")\n",
    "        write(f, \"regression = poisson \\n\")\n",
    "        write(f, \"regression_formula = Trait ~ \\n\")\n",
    "    end\n",
    "\n",
    "    #run marginal analysis\n",
    "    GWAS(\"tmp_control.txt\");\n",
    "\n",
    "    # calculate false positive/negatives based on bonferroni correction\n",
    "    p_values = CSV.read(\"tmp_table.txt\", delim=',', header=true)[:Pvalue]\n",
    "    significance = 0.05 / Float64(p)\n",
    "    passing_snps_position = findall(p_values .<= significance)\n",
    "    marginal_found = [correct_position[snp] in passing_snps_position for snp in 1:k]\n",
    "\n",
    "    #clean up\n",
    "    rm(\"tmp.bed\", force=true)\n",
    "    rm(\"tmp.bim\", force=true)\n",
    "    rm(\"tmp.fam\", force=true)\n",
    "    rm(\"tmp_table.txt\", force=true)\n",
    "    rm(\"tmp_control.txt\", force=true)\n",
    "    rm(\"Mendel_Output.txt\", force=true)\n",
    "\n",
    "    #show lasso and IHT's reconstruction result\n",
    "    compare_model = DataFrame(\n",
    "        true_β  = true_b[correct_position], \n",
    "        IHT_β   = iht_beta[correct_position],\n",
    "        lasso_β = lasso_beta[correct_position],\n",
    "        marginal_found = marginal_found)\n",
    "    @show compare_model\n",
    "    \n",
    "    #compute true/false positives/negatives for IHT and lasso\n",
    "    iht_tp = count(!iszero, iht_beta[correct_position])\n",
    "    iht_fp = iht_k_est - iht_tp\n",
    "    iht_fn = k - iht_tp\n",
    "    lasso_tp = count(!iszero, lasso_beta[correct_position])\n",
    "    lasso_fp = lasso_k_est - lasso_tp\n",
    "    lasso_fn = k - lasso_tp\n",
    "    marginal_tp = count(!iszero, true_b[passing_snps_position])\n",
    "    marginal_fp = length(passing_snps_position) - marginal_tp\n",
    "    marginal_fn = k - marginal_tp\n",
    "    \n",
    "    println(\"IHT true positives = $iht_tp\")\n",
    "    println(\"IHT false positives = $iht_fp\")\n",
    "    println(\"IHT false negatives = $iht_fn\")\n",
    "    println(\"LASSO true positives = $lasso_tp\")\n",
    "    println(\"LASSO false positives = $lasso_fp\")\n",
    "    println(\"LASSO false negatives = $lasso_fn\")\n",
    "    println(\"marginal true positives = $marginal_tp\")\n",
    "    println(\"marginal false positives = $marginal_fp\")\n",
    "    println(\"marginal false negatives = $marginal_fn\" * \"\\n\")\n",
    "    \n",
    "    return iht_tp, iht_fp, iht_fn, lasso_tp, lasso_fp, lasso_fn, marginal_tp, marginal_fp, marginal_fn\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "run (generic function with 1 method)"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "function run()\n",
    "    #simulat data with k true predictors, from distribution d and with link l.\n",
    "    n = 1000\n",
    "    p = 10000\n",
    "    d = Poisson\n",
    "    l = canonicallink(d())\n",
    "\n",
    "    #run function above, saving results in 4 vectors\n",
    "    total_runs = 50\n",
    "    iht_true_positives = zeros(total_runs)\n",
    "    iht_false_positives = zeros(total_runs)\n",
    "    iht_false_negatives = zeros(total_runs)\n",
    "    lasso_true_positives = zeros(total_runs)\n",
    "    lasso_false_positives = zeros(total_runs)\n",
    "    lasso_false_negatives = zeros(total_runs)\n",
    "    marginal_true_positives = zeros(total_runs)\n",
    "    marginal_false_positives = zeros(total_runs)\n",
    "    marginal_false_negatives = zeros(total_runs)\n",
    "    for i in 1:total_runs\n",
    "        println(\"current run = $i\")\n",
    "        \n",
    "        #set random seed\n",
    "        Random.seed!(i)\n",
    "        \n",
    "        iht_tp, iht_fp, iht_fn, lasso_tp, lasso_fp, lasso_fn, m_tp, m_fp, m_fn = iht_lasso_marginal(n, p, d, l)\n",
    "        iht_true_positives[i] = iht_tp\n",
    "        iht_false_positives[i] = iht_fp\n",
    "        iht_false_negatives[i] = iht_fn\n",
    "        lasso_true_positives[i] = lasso_tp\n",
    "        lasso_false_positives[i] = lasso_fp\n",
    "        lasso_false_negatives[i] = lasso_fn\n",
    "        marginal_true_positives[i] = m_tp\n",
    "        marginal_false_positives[i] = m_fp\n",
    "        marginal_false_negatives[i] = m_fn\n",
    "    end\n",
    "    \n",
    "    return iht_true_positives, iht_false_positives, iht_false_negatives, \n",
    "            lasso_true_positives, lasso_false_positives, lasso_false_negatives, \n",
    "            marginal_true_positives, marginal_false_positives, marginal_false_negatives\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "current run = 1\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/poisson\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = poisson\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β    │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64    │ Bool           │\n",
      "├─────┼─────────┼──────────┼────────────┼────────────────┤\n",
      "│ 1   │ 0.5     │ 0.269235 │ 0.156067   │ true           │\n",
      "│ 2   │ 1.0     │ 0.711308 │ 0.569163   │ true           │\n",
      "│ 3   │ 0.6     │ 0.412619 │ 0.286741   │ true           │\n",
      "│ 4   │ 0.1     │ 0.0      │ 0.00490646 │ false          │\n",
      "│ 5   │ 0.8     │ 0.48799  │ 0.400223   │ true           │\n",
      "│ 6   │ 0.2     │ 0.0      │ 0.0898611  │ true           │\n",
      "│ 7   │ 0.9     │ 0.689053 │ 0.491235   │ true           │\n",
      "│ 8   │ 0.7     │ 0.535089 │ 0.353369   │ true           │\n",
      "│ 9   │ 0.3     │ 0.253753 │ 0.172409   │ true           │\n",
      "│ 10  │ 0.4     │ 0.378347 │ 0.207514   │ true           │\n",
      "IHT true positives = 8\n",
      "IHT false positives = 1\n",
      "IHT false negatives = 2\n",
      "LASSO true positives = 10\n",
      "LASSO false positives = 84\n",
      "LASSO false negatives = 0\n",
      "marginal true positives = 9\n",
      "marginal false positives = 1614\n",
      "marginal false negatives = 1\n",
      "\n",
      "current run = 2\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/poisson\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = poisson\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.7     │ 0.469279 │ 0.367305  │ true           │\n",
      "│ 2   │ 0.2     │ 0.0      │ 0.0907549 │ true           │\n",
      "│ 3   │ 0.5     │ 0.532566 │ 0.337825  │ true           │\n",
      "│ 4   │ 1.0     │ 0.706782 │ 0.594096  │ true           │\n",
      "│ 5   │ 0.6     │ 0.458143 │ 0.252235  │ true           │\n",
      "│ 6   │ 0.9     │ 0.641179 │ 0.516227  │ true           │\n",
      "│ 7   │ 0.4     │ 0.22059  │ 0.149234  │ true           │\n",
      "│ 8   │ 0.3     │ 0.0      │ 0.0714994 │ true           │\n",
      "│ 9   │ 0.1     │ 0.0      │ 0.0       │ true           │\n",
      "│ 10  │ 0.8     │ 0.676611 │ 0.434515  │ true           │\n",
      "IHT true positives = 7\n",
      "IHT false positives = 1\n",
      "IHT false negatives = 3\n",
      "LASSO true positives = 9\n",
      "LASSO false positives = 87\n",
      "LASSO false negatives = 1\n",
      "marginal true positives = 10\n",
      "marginal false positives = 1648\n",
      "marginal false negatives = 0\n",
      "\n",
      "current run = 3\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/poisson\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = poisson\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.8     │ 0.493909 │ 0.381002  │ true           │\n",
      "│ 2   │ 0.6     │ 0.0      │ 0.324946  │ true           │\n",
      "│ 3   │ 0.9     │ 0.661962 │ 0.443846  │ true           │\n",
      "│ 4   │ 0.5     │ 0.0      │ 0.284553  │ true           │\n",
      "│ 5   │ 0.2     │ 0.0      │ 0.0429285 │ true           │\n",
      "│ 6   │ 0.4     │ 0.0      │ 0.148626  │ true           │\n",
      "│ 7   │ 0.7     │ 0.459309 │ 0.333962  │ true           │\n",
      "│ 8   │ 0.1     │ 0.0      │ 0.0       │ false          │\n",
      "│ 9   │ 1.0     │ 0.747522 │ 0.550783  │ true           │\n",
      "│ 10  │ 0.3     │ 0.0      │ 0.0774877 │ true           │\n",
      "IHT true positives = 4\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 6\n",
      "LASSO true positives = 9\n",
      "LASSO false positives = 75\n",
      "LASSO false negatives = 1\n",
      "marginal true positives = 9\n",
      "marginal false positives = 1583\n",
      "marginal false negatives = 1\n",
      "\n",
      "current run = 4\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/poisson\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = poisson\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.4     │ 0.0      │ 0.142098  │ true           │\n",
      "│ 2   │ 0.3     │ 0.297864 │ 0.153144  │ true           │\n",
      "│ 3   │ 0.8     │ 0.535011 │ 0.442055  │ true           │\n",
      "│ 4   │ 0.1     │ 0.0      │ 0.0       │ false          │\n",
      "│ 5   │ 0.7     │ 0.610134 │ 0.453175  │ true           │\n",
      "│ 6   │ 0.9     │ 0.677616 │ 0.517509  │ true           │\n",
      "│ 7   │ 0.5     │ 0.410825 │ 0.268829  │ true           │\n",
      "│ 8   │ 1.0     │ 0.691278 │ 0.549242  │ true           │\n",
      "│ 9   │ 0.6     │ 0.289934 │ 0.268168  │ true           │\n",
      "│ 10  │ 0.2     │ 0.0      │ 0.0847915 │ true           │\n",
      "IHT true positives = 7\n",
      "IHT false positives = 1\n",
      "IHT false negatives = 3\n",
      "LASSO true positives = 9\n",
      "LASSO false positives = 103\n",
      "LASSO false negatives = 1\n",
      "marginal true positives = 9\n",
      "marginal false positives = 1600\n",
      "marginal false negatives = 1\n",
      "\n",
      "current run = 5\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/poisson\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = poisson\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.2     │ 0.0      │ 0.0249618 │ true           │\n",
      "│ 2   │ 0.6     │ 0.25095  │ 0.210307  │ false          │\n",
      "│ 3   │ 0.9     │ 0.540784 │ 0.429466  │ true           │\n",
      "│ 4   │ 1.0     │ 0.271356 │ 0.247027  │ true           │\n",
      "│ 5   │ 0.3     │ 0.0      │ 0.0300392 │ false          │\n",
      "│ 6   │ 0.4     │ 0.0      │ 0.0742434 │ true           │\n",
      "│ 7   │ 0.5     │ 0.256915 │ 0.131358  │ true           │\n",
      "│ 8   │ 0.1     │ 0.0      │ 0.0       │ false          │\n",
      "│ 9   │ 0.8     │ 0.444818 │ 0.416762  │ true           │\n",
      "│ 10  │ 0.7     │ 0.462882 │ 0.323793  │ true           │\n",
      "IHT true positives = 6\n",
      "IHT false positives = 2\n",
      "IHT false negatives = 4\n",
      "LASSO true positives = 9\n",
      "LASSO false positives = 33\n",
      "LASSO false negatives = 1\n",
      "marginal true positives = 7\n",
      "marginal false positives = 1431\n",
      "marginal false negatives = 3\n",
      "\n",
      "current run = 6\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/poisson\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = poisson\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.2     │ 0.151263 │ 0.0786576 │ true           │\n",
      "│ 2   │ 1.0     │ 0.823597 │ 0.635338  │ true           │\n",
      "│ 3   │ 0.7     │ 0.552536 │ 0.391943  │ true           │\n",
      "│ 4   │ 0.1     │ 0.125961 │ 0.0141926 │ false          │\n",
      "│ 5   │ 0.5     │ 0.473984 │ 0.312057  │ true           │\n",
      "│ 6   │ 0.8     │ 0.62267  │ 0.459432  │ true           │\n",
      "│ 7   │ 0.3     │ 0.202743 │ 0.112402  │ true           │\n",
      "│ 8   │ 0.9     │ 0.764291 │ 0.565621  │ true           │\n",
      "│ 9   │ 0.4     │ 0.326002 │ 0.19416   │ true           │\n",
      "│ 10  │ 0.6     │ 0.474368 │ 0.353109  │ true           │\n",
      "IHT true positives = 10\n",
      "IHT false positives = 9\n",
      "IHT false negatives = 0\n",
      "LASSO true positives = 10\n",
      "LASSO false positives = 71\n",
      "LASSO false negatives = 0\n",
      "marginal true positives = 9\n",
      "marginal false positives = 1475\n",
      "marginal false negatives = 1\n",
      "\n",
      "current run = 7\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/poisson\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = poisson\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.5     │ 0.422583 │ 0.356929  │ true           │\n",
      "│ 2   │ 1.0     │ 0.76482  │ 0.631257  │ true           │\n",
      "│ 3   │ 0.1     │ 0.0      │ 0.0       │ false          │\n",
      "│ 4   │ 0.9     │ 0.711835 │ 0.587874  │ true           │\n",
      "│ 5   │ 0.8     │ 0.635858 │ 0.514744  │ true           │\n",
      "│ 6   │ 0.7     │ 0.534104 │ 0.398494  │ true           │\n",
      "│ 7   │ 0.2     │ 0.158263 │ 0.0865238 │ true           │\n",
      "│ 8   │ 0.6     │ 0.463426 │ 0.386594  │ true           │\n",
      "│ 9   │ 0.3     │ 0.251767 │ 0.172832  │ true           │\n",
      "│ 10  │ 0.4     │ 0.347452 │ 0.260131  │ true           │\n",
      "IHT true positives = 9\n",
      "IHT false positives = 21\n",
      "IHT false negatives = 1\n",
      "LASSO true positives = 9\n",
      "LASSO false positives = 147\n",
      "LASSO false negatives = 1\n",
      "marginal true positives = 9\n",
      "marginal false positives = 1492\n",
      "marginal false negatives = 1\n",
      "\n",
      "current run = 8\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/poisson\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = poisson\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.8     │ 0.612578 │ 0.380716  │ true           │\n",
      "│ 2   │ 0.1     │ 0.0      │ 0.0       │ false          │\n",
      "│ 3   │ 0.2     │ 0.0      │ 0.0406051 │ true           │\n",
      "│ 4   │ 0.4     │ 0.25458  │ 0.146019  │ true           │\n",
      "│ 5   │ 0.5     │ 0.289418 │ 0.225666  │ true           │\n",
      "│ 6   │ 0.7     │ 0.45631  │ 0.288043  │ true           │\n",
      "│ 7   │ 0.9     │ 0.617638 │ 0.484434  │ true           │\n",
      "│ 8   │ 0.6     │ 0.419392 │ 0.287905  │ true           │\n",
      "│ 9   │ 1.0     │ 0.771852 │ 0.586727  │ true           │\n",
      "│ 10  │ 0.3     │ 0.227603 │ 0.140207  │ true           │\n",
      "IHT true positives = 8\n",
      "IHT false positives = 10\n",
      "IHT false negatives = 2\n",
      "LASSO true positives = 9\n",
      "LASSO false positives = 99\n",
      "LASSO false negatives = 1\n",
      "marginal true positives = 9\n",
      "marginal false positives = 1557\n",
      "marginal false negatives = 1\n",
      "\n",
      "current run = 9\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/poisson\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = poisson\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.7     │ 0.515368 │ 0.327566  │ true           │\n",
      "│ 2   │ 0.4     │ 0.346846 │ 0.206102  │ true           │\n",
      "│ 3   │ 0.3     │ 0.0      │ 0.0688951 │ true           │\n",
      "│ 4   │ 0.9     │ 0.654586 │ 0.434133  │ true           │\n",
      "│ 5   │ 0.6     │ 0.35727  │ 0.250965  │ true           │\n",
      "│ 6   │ 1.0     │ 0.442354 │ 0.406102  │ true           │\n",
      "│ 7   │ 0.8     │ 0.574458 │ 0.384585  │ true           │\n",
      "│ 8   │ 0.5     │ 0.390613 │ 0.237027  │ true           │\n",
      "│ 9   │ 0.2     │ 0.0      │ 0.0279537 │ true           │\n",
      "│ 10  │ 0.1     │ 0.0      │ 0.0       │ true           │\n",
      "IHT true positives = 7\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 3\n",
      "LASSO true positives = 9\n",
      "LASSO false positives = 50\n",
      "LASSO false negatives = 1\n",
      "marginal true positives = 10\n",
      "marginal false positives = 1656\n",
      "marginal false negatives = 0\n",
      "\n",
      "current run = 10\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/poisson\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = poisson\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 1.0     │ 0.717801 │ 0.602194  │ true           │\n",
      "│ 2   │ 0.4     │ 0.0      │ 0.165112  │ true           │\n",
      "│ 3   │ 0.6     │ 0.417866 │ 0.321252  │ true           │\n",
      "│ 4   │ 0.7     │ 0.44116  │ 0.380033  │ true           │\n",
      "│ 5   │ 0.3     │ 0.0      │ 0.151484  │ true           │\n",
      "│ 6   │ 0.5     │ 0.382725 │ 0.282721  │ true           │\n",
      "│ 7   │ 0.1     │ 0.0      │ 0.0201182 │ true           │\n",
      "│ 8   │ 0.8     │ 0.668027 │ 0.444556  │ true           │\n",
      "│ 9   │ 0.9     │ 0.658832 │ 0.511244  │ true           │\n",
      "│ 10  │ 0.2     │ 0.0      │ 0.02433   │ true           │\n",
      "IHT true positives = 6\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 4\n",
      "LASSO true positives = 10\n",
      "LASSO false positives = 82\n",
      "LASSO false negatives = 0\n",
      "marginal true positives = 10\n",
      "marginal false positives = 1443\n",
      "marginal false negatives = 0\n",
      "\n",
      "current run = 11\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/poisson\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = poisson\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.1     │ 0.0      │ 0.0065292 │ false          │\n",
      "│ 2   │ 0.7     │ 0.467753 │ 0.314573  │ true           │\n",
      "│ 3   │ 0.6     │ 0.38607  │ 0.283733  │ true           │\n",
      "│ 4   │ 0.5     │ 0.417742 │ 0.259965  │ true           │\n",
      "│ 5   │ 0.3     │ 0.0      │ 0.0736968 │ true           │\n",
      "│ 6   │ 0.4     │ 0.285973 │ 0.180492  │ true           │\n",
      "│ 7   │ 0.9     │ 0.780108 │ 0.546522  │ true           │\n",
      "│ 8   │ 0.8     │ 0.329288 │ 0.305488  │ true           │\n",
      "│ 9   │ 1.0     │ 0.789921 │ 0.577311  │ true           │\n",
      "│ 10  │ 0.2     │ 0.0      │ 0.01043   │ false          │\n",
      "IHT true positives = 7\n",
      "IHT false positives = 1\n",
      "IHT false negatives = 3\n",
      "LASSO true positives = 10\n",
      "LASSO false positives = 59\n",
      "LASSO false negatives = 0\n",
      "marginal true positives = 8\n",
      "marginal false positives = 1451\n",
      "marginal false negatives = 2\n",
      "\n",
      "current run = 12\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/poisson\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = poisson\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.3     │ 0.0      │ 0.0777383 │ true           │\n",
      "│ 2   │ 0.1     │ 0.0      │ 0.0148612 │ false          │\n",
      "│ 3   │ 0.7     │ 0.555384 │ 0.411271  │ true           │\n",
      "│ 4   │ 0.9     │ 0.493864 │ 0.432623  │ true           │\n",
      "│ 5   │ 0.6     │ 0.457316 │ 0.313175  │ true           │\n",
      "│ 6   │ 0.2     │ 0.0      │ 0.0159635 │ true           │\n",
      "│ 7   │ 0.4     │ 0.296365 │ 0.187683  │ true           │\n",
      "│ 8   │ 0.8     │ 0.606349 │ 0.433454  │ true           │\n",
      "│ 9   │ 0.5     │ 0.4155   │ 0.266241  │ true           │\n",
      "│ 10  │ 1.0     │ 0.69044  │ 0.565912  │ true           │\n",
      "IHT true positives = 7\n",
      "IHT false positives = 2\n",
      "IHT false negatives = 3\n",
      "LASSO true positives = 10\n",
      "LASSO false positives = 94\n",
      "LASSO false negatives = 0\n",
      "marginal true positives = 9\n",
      "marginal false positives = 1672\n",
      "marginal false negatives = 1\n",
      "\n",
      "current run = 13\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/poisson\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = poisson\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.2     │ 0.0      │ 0.0516345 │ true           │\n",
      "│ 2   │ 0.9     │ 0.707759 │ 0.537284  │ true           │\n",
      "│ 3   │ 1.0     │ 0.690903 │ 0.557977  │ true           │\n",
      "│ 4   │ 0.6     │ 0.38626  │ 0.320345  │ true           │\n",
      "│ 5   │ 0.4     │ 0.263623 │ 0.175128  │ true           │\n",
      "│ 6   │ 0.8     │ 0.675862 │ 0.466043  │ true           │\n",
      "│ 7   │ 0.1     │ 0.0      │ 0.0       │ false          │\n",
      "│ 8   │ 0.7     │ 0.454112 │ 0.397095  │ true           │\n",
      "│ 9   │ 0.5     │ 0.351557 │ 0.237408  │ true           │\n",
      "│ 10  │ 0.3     │ 0.0      │ 0.077272  │ true           │\n",
      "IHT true positives = 7\n",
      "IHT false positives = 1\n",
      "IHT false negatives = 3\n",
      "LASSO true positives = 9\n",
      "LASSO false positives = 78\n",
      "LASSO false negatives = 1\n",
      "marginal true positives = 9\n",
      "marginal false positives = 1681\n",
      "marginal false negatives = 1\n",
      "\n",
      "current run = 14\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/poisson\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = poisson\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.1     │ 0.105633 │ 0.0244388 │ false          │\n",
      "│ 2   │ 0.2     │ 0.116124 │ 0.0451808 │ true           │\n",
      "│ 3   │ 0.8     │ 0.48648  │ 0.419739  │ true           │\n",
      "│ 4   │ 0.3     │ 0.193165 │ 0.141856  │ true           │\n",
      "│ 5   │ 0.6     │ 0.435851 │ 0.324676  │ true           │\n",
      "│ 6   │ 0.5     │ 0.3087   │ 0.216407  │ true           │\n",
      "│ 7   │ 1.0     │ 0.707908 │ 0.574683  │ true           │\n",
      "│ 8   │ 0.9     │ 0.631244 │ 0.542575  │ true           │\n",
      "│ 9   │ 0.4     │ 0.27445  │ 0.188933  │ true           │\n",
      "│ 10  │ 0.7     │ 0.542758 │ 0.457966  │ true           │\n",
      "IHT true positives = 10\n",
      "IHT false positives = 16\n",
      "IHT false negatives = 0\n",
      "LASSO true positives = 10\n",
      "LASSO false positives = 130\n",
      "LASSO false negatives = 0\n",
      "marginal true positives = 9\n",
      "marginal false positives = 1532\n",
      "marginal false negatives = 1\n",
      "\n",
      "current run = 15\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/poisson\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = poisson\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.9     │ 0.559358 │ 0.471592  │ true           │\n",
      "│ 2   │ 0.7     │ 0.44261  │ 0.352483  │ true           │\n",
      "│ 3   │ 0.1     │ 0.0      │ 0.0       │ true           │\n",
      "│ 4   │ 0.8     │ 0.62216  │ 0.486808  │ true           │\n",
      "│ 5   │ 1.0     │ 0.627596 │ 0.507494  │ true           │\n",
      "│ 6   │ 0.6     │ 0.389841 │ 0.314317  │ true           │\n",
      "│ 7   │ 0.5     │ 0.325241 │ 0.228457  │ true           │\n",
      "│ 8   │ 0.3     │ 0.0      │ 0.0293587 │ true           │\n",
      "│ 9   │ 0.2     │ 0.120941 │ 0.0677291 │ true           │\n",
      "│ 10  │ 0.4     │ 0.272885 │ 0.215345  │ true           │\n",
      "IHT true positives = 8\n",
      "IHT false positives = 11\n",
      "IHT false negatives = 2\n",
      "LASSO true positives = 9\n",
      "LASSO false positives = 65\n",
      "LASSO false negatives = 1\n",
      "marginal true positives = 10\n",
      "marginal false positives = 1453\n",
      "marginal false negatives = 0\n",
      "\n",
      "current run = 16\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/poisson\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = poisson\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.1     │ 0.0      │ 0.0196043 │ false          │\n",
      "│ 2   │ 0.6     │ 0.501872 │ 0.328245  │ true           │\n",
      "│ 3   │ 0.3     │ 0.273203 │ 0.165103  │ true           │\n",
      "│ 4   │ 0.2     │ 0.0      │ 0.0696716 │ true           │\n",
      "│ 5   │ 0.5     │ 0.410646 │ 0.246627  │ true           │\n",
      "│ 6   │ 1.0     │ 0.71781  │ 0.570998  │ true           │\n",
      "│ 7   │ 0.8     │ 0.523742 │ 0.375529  │ true           │\n",
      "│ 8   │ 0.4     │ 0.0      │ 0.142187  │ true           │\n",
      "│ 9   │ 0.9     │ 0.56801  │ 0.458301  │ true           │\n",
      "│ 10  │ 0.7     │ 0.400281 │ 0.325275  │ true           │\n",
      "IHT true positives = 7\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 3\n",
      "LASSO true positives = 10\n",
      "LASSO false positives = 66\n",
      "LASSO false negatives = 0\n",
      "marginal true positives = 9\n",
      "marginal false positives = 1539\n",
      "marginal false negatives = 1\n",
      "\n",
      "current run = 17\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/poisson\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = poisson\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.2     │ 0.0      │ 0.0       │ true           │\n",
      "│ 2   │ 0.4     │ 0.0      │ 0.0668857 │ true           │\n",
      "│ 3   │ 0.1     │ 0.0      │ 0.0       │ false          │\n",
      "│ 4   │ 1.0     │ 0.605727 │ 0.450796  │ true           │\n",
      "│ 5   │ 0.7     │ 0.494679 │ 0.310812  │ true           │\n",
      "│ 6   │ 0.8     │ 0.510959 │ 0.350302  │ true           │\n",
      "│ 7   │ 0.5     │ 0.0      │ 0.147652  │ true           │\n",
      "│ 8   │ 0.6     │ 0.0      │ 0.165799  │ true           │\n",
      "│ 9   │ 0.3     │ 0.0      │ 0.0       │ true           │\n",
      "│ 10  │ 0.9     │ 0.666313 │ 0.420411  │ true           │\n",
      "IHT true positives = 4\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 6\n",
      "LASSO true positives = 7\n",
      "LASSO false positives = 14\n",
      "LASSO false negatives = 3\n",
      "marginal true positives = 9\n",
      "marginal false positives = 1699\n",
      "marginal false negatives = 1\n",
      "\n",
      "current run = 18\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/poisson\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = poisson\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.6     │ 0.453214 │ 0.288177  │ true           │\n",
      "│ 2   │ 0.2     │ 0.0      │ 0.0404256 │ true           │\n",
      "│ 3   │ 1.0     │ 0.727693 │ 0.584606  │ true           │\n",
      "│ 4   │ 0.1     │ 0.0      │ 0.0       │ true           │\n",
      "│ 5   │ 0.4     │ 0.365456 │ 0.193203  │ true           │\n",
      "│ 6   │ 0.7     │ 0.568335 │ 0.369517  │ true           │\n",
      "│ 7   │ 0.3     │ 0.251852 │ 0.158881  │ true           │\n",
      "│ 8   │ 0.9     │ 0.590818 │ 0.486395  │ true           │\n",
      "│ 9   │ 0.5     │ 0.387784 │ 0.199511  │ true           │\n",
      "│ 10  │ 0.8     │ 0.620334 │ 0.408984  │ true           │\n",
      "IHT true positives = 8\n",
      "IHT false positives = 2\n",
      "IHT false negatives = 2\n",
      "LASSO true positives = 9\n",
      "LASSO false positives = 72\n",
      "LASSO false negatives = 1\n",
      "marginal true positives = 10\n",
      "marginal false positives = 1383\n",
      "marginal false negatives = 0\n",
      "\n",
      "current run = 19\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/poisson\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = poisson\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β    │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64    │ Bool           │\n",
      "├─────┼─────────┼──────────┼────────────┼────────────────┤\n",
      "│ 1   │ 0.7     │ 0.606568 │ 0.449348   │ true           │\n",
      "│ 2   │ 0.9     │ 0.73001  │ 0.565886   │ true           │\n",
      "│ 3   │ 0.3     │ 0.279539 │ 0.157292   │ true           │\n",
      "│ 4   │ 0.2     │ 0.0      │ 0.00949943 │ true           │\n",
      "│ 5   │ 0.1     │ 0.0      │ 0.0        │ true           │\n",
      "│ 6   │ 0.8     │ 0.692212 │ 0.516946   │ true           │\n",
      "│ 7   │ 0.4     │ 0.191343 │ 0.136242   │ true           │\n",
      "│ 8   │ 1.0     │ 0.777863 │ 0.597024   │ true           │\n",
      "│ 9   │ 0.6     │ 0.455292 │ 0.342435   │ true           │\n",
      "│ 10  │ 0.5     │ 0.410104 │ 0.296233   │ true           │\n",
      "IHT true positives = 8\n",
      "IHT false positives = 4\n",
      "IHT false negatives = 2\n",
      "LASSO true positives = 9\n",
      "LASSO false positives = 113\n",
      "LASSO false negatives = 1\n",
      "marginal true positives = 10\n",
      "marginal false positives = 1422\n",
      "marginal false negatives = 0\n",
      "\n",
      "current run = 20\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/poisson\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = poisson\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β  │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64  │ Bool           │\n",
      "├─────┼─────────┼──────────┼──────────┼────────────────┤\n",
      "│ 1   │ 0.5     │ 0.361748 │ 0.216895 │ true           │\n",
      "│ 2   │ 0.2     │ 0.0      │ 0.0      │ true           │\n",
      "│ 3   │ 1.0     │ 0.724518 │ 0.505215 │ true           │\n",
      "│ 4   │ 0.3     │ 0.206918 │ 0.117583 │ true           │\n",
      "│ 5   │ 0.6     │ 0.378845 │ 0.242565 │ true           │\n",
      "│ 6   │ 0.1     │ 0.0      │ 0.0      │ false          │\n",
      "│ 7   │ 0.4     │ 0.272762 │ 0.193381 │ true           │\n",
      "│ 8   │ 0.9     │ 0.639958 │ 0.436074 │ true           │\n",
      "│ 9   │ 0.7     │ 0.516098 │ 0.335585 │ true           │\n",
      "│ 10  │ 0.8     │ 0.649913 │ 0.452567 │ true           │\n",
      "IHT true positives = 8\n",
      "IHT false positives = 9\n",
      "IHT false negatives = 2\n",
      "LASSO true positives = 8\n",
      "LASSO false positives = 60\n",
      "LASSO false negatives = 2\n",
      "marginal true positives = 9\n",
      "marginal false positives = 1472\n",
      "marginal false negatives = 1\n",
      "\n",
      "current run = 21\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/poisson\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = poisson\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.2     │ 0.0      │ 0.0218715 │ true           │\n",
      "│ 2   │ 0.1     │ 0.0      │ 0.0       │ false          │\n",
      "│ 3   │ 0.9     │ 0.594319 │ 0.458946  │ true           │\n",
      "│ 4   │ 0.5     │ 0.307888 │ 0.215252  │ true           │\n",
      "│ 5   │ 0.3     │ 0.0      │ 0.10255   │ true           │\n",
      "│ 6   │ 0.8     │ 0.524124 │ 0.413399  │ true           │\n",
      "│ 7   │ 0.6     │ 0.349061 │ 0.279373  │ true           │\n",
      "│ 8   │ 0.4     │ 0.260921 │ 0.160585  │ true           │\n",
      "│ 9   │ 1.0     │ 0.718204 │ 0.535005  │ true           │\n",
      "│ 10  │ 0.7     │ 0.533659 │ 0.390506  │ true           │\n",
      "IHT true positives = 7\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 3\n",
      "LASSO true positives = 9\n",
      "LASSO false positives = 44\n",
      "LASSO false negatives = 1\n",
      "marginal true positives = 9\n",
      "marginal false positives = 1487\n",
      "marginal false negatives = 1\n",
      "\n",
      "current run = 22\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/poisson\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = poisson\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.1     │ 0.0      │ 0.0       │ true           │\n",
      "│ 2   │ 0.2     │ 0.0      │ 0.0577124 │ true           │\n",
      "│ 3   │ 0.3     │ 0.256204 │ 0.151342  │ true           │\n",
      "│ 4   │ 0.6     │ 0.46113  │ 0.380418  │ true           │\n",
      "│ 5   │ 0.8     │ 0.668066 │ 0.490831  │ true           │\n",
      "│ 6   │ 0.7     │ 0.537786 │ 0.416647  │ true           │\n",
      "│ 7   │ 0.5     │ 0.336097 │ 0.290861  │ true           │\n",
      "│ 8   │ 0.4     │ 0.232939 │ 0.209487  │ true           │\n",
      "│ 9   │ 1.0     │ 0.761424 │ 0.613745  │ true           │\n",
      "│ 10  │ 0.9     │ 0.655677 │ 0.54642   │ true           │\n",
      "IHT true positives = 8\n",
      "IHT false positives = 1\n",
      "IHT false negatives = 2\n",
      "LASSO true positives = 9\n",
      "LASSO false positives = 125\n",
      "LASSO false negatives = 1\n",
      "marginal true positives = 10\n",
      "marginal false positives = 1387\n",
      "marginal false negatives = 0\n",
      "\n",
      "current run = 23\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/poisson\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = poisson\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.5     │ 0.0      │ 0.250118  │ true           │\n",
      "│ 2   │ 0.4     │ 0.0      │ 0.205242  │ true           │\n",
      "│ 3   │ 0.8     │ 0.0      │ 0.439381  │ true           │\n",
      "│ 4   │ 0.9     │ 0.0      │ 0.573824  │ true           │\n",
      "│ 5   │ 0.3     │ 0.0      │ 0.122374  │ true           │\n",
      "│ 6   │ 0.2     │ 0.0      │ 0.0830359 │ true           │\n",
      "│ 7   │ 0.6     │ 0.0      │ 0.348195  │ true           │\n",
      "│ 8   │ 1.0     │ 0.546512 │ 0.566852  │ true           │\n",
      "│ 9   │ 0.1     │ 0.0      │ 0.0       │ false          │\n",
      "│ 10  │ 0.7     │ 0.0      │ 0.395034  │ true           │\n",
      "IHT true positives = 1\n",
      "IHT false positives = 1\n",
      "IHT false negatives = 9\n",
      "LASSO true positives = 9\n",
      "LASSO false positives = 98\n",
      "LASSO false negatives = 1\n",
      "marginal true positives = 9\n",
      "marginal false positives = 1515\n",
      "marginal false negatives = 1\n",
      "\n",
      "current run = 24\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/poisson\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = poisson\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.2     │ 0.0      │ 0.0846416 │ true           │\n",
      "│ 2   │ 0.4     │ 0.373014 │ 0.291444  │ true           │\n",
      "│ 3   │ 0.3     │ 0.268083 │ 0.140248  │ true           │\n",
      "│ 4   │ 0.8     │ 0.624574 │ 0.506783  │ true           │\n",
      "│ 5   │ 0.5     │ 0.32295  │ 0.286295  │ true           │\n",
      "│ 6   │ 1.0     │ 0.844305 │ 0.683691  │ true           │\n",
      "│ 7   │ 0.1     │ 0.0      │ 0.0       │ false          │\n",
      "│ 8   │ 0.9     │ 0.698806 │ 0.581925  │ true           │\n",
      "│ 9   │ 0.7     │ 0.532298 │ 0.402572  │ true           │\n",
      "│ 10  │ 0.6     │ 0.435351 │ 0.341781  │ true           │\n",
      "IHT true positives = 8\n",
      "IHT false positives = 3\n",
      "IHT false negatives = 2\n",
      "LASSO true positives = 9\n",
      "LASSO false positives = 144\n",
      "LASSO false negatives = 1\n",
      "marginal true positives = 9\n",
      "marginal false positives = 1548\n",
      "marginal false negatives = 1\n",
      "\n",
      "current run = 25\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/poisson\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = poisson\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.8     │ 0.622179 │ 0.503535  │ true           │\n",
      "│ 2   │ 0.1     │ 0.0      │ 0.0551596 │ false          │\n",
      "│ 3   │ 0.2     │ 0.0      │ 0.0531229 │ false          │\n",
      "│ 4   │ 0.9     │ 0.658503 │ 0.542674  │ true           │\n",
      "│ 5   │ 0.3     │ 0.279054 │ 0.140198  │ true           │\n",
      "│ 6   │ 0.5     │ 0.402578 │ 0.282249  │ true           │\n",
      "│ 7   │ 1.0     │ 0.790711 │ 0.675298  │ true           │\n",
      "│ 8   │ 0.6     │ 0.477621 │ 0.385307  │ true           │\n",
      "│ 9   │ 0.7     │ 0.570758 │ 0.4607    │ true           │\n",
      "│ 10  │ 0.4     │ 0.401893 │ 0.307707  │ true           │\n",
      "IHT true positives = 8\n",
      "IHT false positives = 14\n",
      "IHT false negatives = 2\n",
      "LASSO true positives = 10\n",
      "LASSO false positives = 166\n",
      "LASSO false negatives = 0\n",
      "marginal true positives = 8\n",
      "marginal false positives = 1485\n",
      "marginal false negatives = 2\n",
      "\n",
      "current run = 26\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/poisson\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = poisson\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.9     │ 0.68015  │ 0.644592  │ true           │\n",
      "│ 2   │ 0.3     │ 0.0      │ 0.150166  │ true           │\n",
      "│ 3   │ 1.0     │ 0.847213 │ 0.749951  │ true           │\n",
      "│ 4   │ 0.2     │ 0.0      │ 0.0614828 │ true           │\n",
      "│ 5   │ 0.4     │ 0.324263 │ 0.281093  │ true           │\n",
      "│ 6   │ 0.7     │ 0.565769 │ 0.503799  │ true           │\n",
      "│ 7   │ 0.6     │ 0.458713 │ 0.389418  │ true           │\n",
      "│ 8   │ 0.5     │ 0.386467 │ 0.340507  │ true           │\n",
      "│ 9   │ 0.8     │ 0.620018 │ 0.567648  │ true           │\n",
      "│ 10  │ 0.1     │ 0.0      │ 0.0388817 │ false          │\n",
      "IHT true positives = 7\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 3\n",
      "LASSO true positives = 10\n",
      "LASSO false positives = 194\n",
      "LASSO false negatives = 0\n",
      "marginal true positives = 9\n",
      "marginal false positives = 1473\n",
      "marginal false negatives = 1\n",
      "\n",
      "current run = 27\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/poisson\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = poisson\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.7     │ 0.418519 │ 0.304396  │ true           │\n",
      "│ 2   │ 0.8     │ 0.693323 │ 0.347488  │ true           │\n",
      "│ 3   │ 0.4     │ 0.0      │ 0.0450958 │ true           │\n",
      "│ 4   │ 0.5     │ 0.0      │ 0.209273  │ true           │\n",
      "│ 5   │ 0.1     │ 0.0      │ 0.0       │ false          │\n",
      "│ 6   │ 0.2     │ 0.0      │ 0.0       │ true           │\n",
      "│ 7   │ 0.9     │ 0.876706 │ 0.42666   │ true           │\n",
      "│ 8   │ 0.3     │ 0.0      │ 0.128836  │ true           │\n",
      "│ 9   │ 1.0     │ 0.729931 │ 0.503447  │ true           │\n",
      "│ 10  │ 0.6     │ 0.0      │ 0.248917  │ true           │\n",
      "IHT true positives = 4\n",
      "IHT false positives = 1\n",
      "IHT false negatives = 6\n",
      "LASSO true positives = 8\n",
      "LASSO false positives = 19\n",
      "LASSO false negatives = 2\n",
      "marginal true positives = 9\n",
      "marginal false positives = 1565\n",
      "marginal false negatives = 1\n",
      "\n",
      "current run = 28\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/poisson\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = poisson\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.7     │ 0.572804 │ 0.480081  │ true           │\n",
      "│ 2   │ 0.1     │ 0.0      │ 0.0       │ true           │\n",
      "│ 3   │ 1.0     │ 0.269864 │ 0.245492  │ false          │\n",
      "│ 4   │ 0.2     │ 0.0      │ 0.0778733 │ true           │\n",
      "│ 5   │ 0.4     │ 0.21686  │ 0.221367  │ true           │\n",
      "│ 6   │ 0.5     │ 0.265364 │ 0.274884  │ true           │\n",
      "│ 7   │ 0.3     │ 0.267715 │ 0.161722  │ true           │\n",
      "│ 8   │ 0.8     │ 0.671095 │ 0.593273  │ true           │\n",
      "│ 9   │ 0.9     │ 0.751177 │ 0.636419  │ true           │\n",
      "│ 10  │ 0.6     │ 0.471677 │ 0.32163   │ true           │\n",
      "IHT true positives = 8\n",
      "IHT false positives = 2\n",
      "IHT false negatives = 2\n",
      "LASSO true positives = 9\n",
      "LASSO false positives = 129\n",
      "LASSO false negatives = 1\n",
      "marginal true positives = 9\n",
      "marginal false positives = 1521\n",
      "marginal false negatives = 1\n",
      "\n",
      "current run = 29\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/poisson\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = poisson\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β  │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64  │ Bool           │\n",
      "├─────┼─────────┼──────────┼──────────┼────────────────┤\n",
      "│ 1   │ 0.2     │ 0.0      │ 0.067626 │ true           │\n",
      "│ 2   │ 0.4     │ 0.0      │ 0.19041  │ true           │\n",
      "│ 3   │ 1.0     │ 0.786327 │ 0.584587 │ true           │\n",
      "│ 4   │ 0.8     │ 0.580428 │ 0.421714 │ true           │\n",
      "│ 5   │ 0.1     │ 0.0      │ 0.0      │ false          │\n",
      "│ 6   │ 0.3     │ 0.202439 │ 0.123896 │ true           │\n",
      "│ 7   │ 0.6     │ 0.459397 │ 0.317843 │ true           │\n",
      "│ 8   │ 0.5     │ 0.369222 │ 0.280839 │ true           │\n",
      "│ 9   │ 0.7     │ 0.600115 │ 0.432414 │ true           │\n",
      "│ 10  │ 0.9     │ 0.646582 │ 0.554909 │ true           │\n",
      "IHT true positives = 7\n",
      "IHT false positives = 1\n",
      "IHT false negatives = 3\n",
      "LASSO true positives = 9\n",
      "LASSO false positives = 93\n",
      "LASSO false negatives = 1\n",
      "marginal true positives = 9\n",
      "marginal false positives = 1556\n",
      "marginal false negatives = 1\n",
      "\n",
      "current run = 30\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/poisson\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = poisson\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β  │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64  │ Bool           │\n",
      "├─────┼─────────┼──────────┼──────────┼────────────────┤\n",
      "│ 1   │ 0.5     │ 0.421839 │ 0.272819 │ true           │\n",
      "│ 2   │ 0.4     │ 0.182924 │ 0.13988  │ true           │\n",
      "│ 3   │ 0.3     │ 0.0      │ 0.177531 │ true           │\n",
      "│ 4   │ 0.9     │ 0.674225 │ 0.519923 │ true           │\n",
      "│ 5   │ 0.2     │ 0.0      │ 0.058117 │ true           │\n",
      "│ 6   │ 1.0     │ 0.64223  │ 0.577365 │ true           │\n",
      "│ 7   │ 0.1     │ 0.0      │ 0.0      │ false          │\n",
      "│ 8   │ 0.8     │ 0.513579 │ 0.444605 │ true           │\n",
      "│ 9   │ 0.6     │ 0.601149 │ 0.409952 │ true           │\n",
      "│ 10  │ 0.7     │ 0.460467 │ 0.397301 │ true           │\n",
      "IHT true positives = 7\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 3\n",
      "LASSO true positives = 9\n",
      "LASSO false positives = 113\n",
      "LASSO false negatives = 1\n",
      "marginal true positives = 9\n",
      "marginal false positives = 1678\n",
      "marginal false negatives = 1\n",
      "\n",
      "current run = 31\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/poisson\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = poisson\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β  │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64  │ Bool           │\n",
      "├─────┼─────────┼──────────┼──────────┼────────────────┤\n",
      "│ 1   │ 0.2     │ 0.0      │ 0.0      │ true           │\n",
      "│ 2   │ 1.0     │ 0.799724 │ 0.537835 │ true           │\n",
      "│ 3   │ 0.4     │ 0.0      │ 0.168367 │ true           │\n",
      "│ 4   │ 0.9     │ 0.819922 │ 0.524689 │ true           │\n",
      "│ 5   │ 0.1     │ 0.0      │ 0.0      │ false          │\n",
      "│ 6   │ 0.5     │ 0.42136  │ 0.292916 │ true           │\n",
      "│ 7   │ 0.7     │ 0.571733 │ 0.365364 │ true           │\n",
      "│ 8   │ 0.6     │ 0.435664 │ 0.278867 │ true           │\n",
      "│ 9   │ 0.8     │ 0.0      │ 0.161977 │ false          │\n",
      "│ 10  │ 0.3     │ 0.0      │ 0.082529 │ true           │\n",
      "IHT true positives = 5\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 5\n",
      "LASSO true positives = 8\n",
      "LASSO false positives = 6\n",
      "LASSO false negatives = 2\n",
      "marginal true positives = 8\n",
      "marginal false positives = 1298\n",
      "marginal false negatives = 2\n",
      "\n",
      "current run = 32\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/poisson\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = poisson\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β  │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64  │ Bool           │\n",
      "├─────┼─────────┼──────────┼──────────┼────────────────┤\n",
      "│ 1   │ 1.0     │ 0.827497 │ 0.557832 │ true           │\n",
      "│ 2   │ 0.1     │ 0.0      │ 0.0      │ false          │\n",
      "│ 3   │ 0.8     │ 0.4955   │ 0.387911 │ true           │\n",
      "│ 4   │ 0.7     │ 0.520736 │ 0.376951 │ true           │\n",
      "│ 5   │ 0.6     │ 0.554473 │ 0.371029 │ true           │\n",
      "│ 6   │ 0.9     │ 0.576654 │ 0.446079 │ true           │\n",
      "│ 7   │ 0.4     │ 0.0      │ 0.136326 │ true           │\n",
      "│ 8   │ 0.2     │ 0.0      │ 0.0      │ true           │\n",
      "│ 9   │ 0.5     │ 0.271906 │ 0.230238 │ true           │\n",
      "│ 10  │ 0.3     │ 0.0      │ 0.122605 │ true           │\n",
      "IHT true positives = 6\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 4\n",
      "LASSO true positives = 8\n",
      "LASSO false positives = 68\n",
      "LASSO false negatives = 2\n",
      "marginal true positives = 9\n",
      "marginal false positives = 1541\n",
      "marginal false negatives = 1\n",
      "\n",
      "current run = 33\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/poisson\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = poisson\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.7     │ 0.49048  │ 0.38078   │ true           │\n",
      "│ 2   │ 0.9     │ 0.675865 │ 0.544355  │ true           │\n",
      "│ 3   │ 0.8     │ 0.493579 │ 0.409094  │ true           │\n",
      "│ 4   │ 0.4     │ 0.227392 │ 0.151086  │ true           │\n",
      "│ 5   │ 1.0     │ 0.811262 │ 0.61451   │ true           │\n",
      "│ 6   │ 0.2     │ 0.0      │ 0.0792634 │ true           │\n",
      "│ 7   │ 0.6     │ 0.443873 │ 0.31918   │ true           │\n",
      "│ 8   │ 0.5     │ 0.368003 │ 0.24109   │ true           │\n",
      "│ 9   │ 0.1     │ 0.0      │ 0.0192111 │ true           │\n",
      "│ 10  │ 0.3     │ 0.292968 │ 0.16714   │ true           │\n",
      "IHT true positives = 8\n",
      "IHT false positives = 3\n",
      "IHT false negatives = 2\n",
      "LASSO true positives = 10\n",
      "LASSO false positives = 124\n",
      "LASSO false negatives = 0\n",
      "marginal true positives = 10\n",
      "marginal false positives = 1549\n",
      "marginal false negatives = 0\n",
      "\n",
      "current run = 34\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/poisson\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = poisson\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.7     │ 0.385081 │ 0.350361  │ true           │\n",
      "│ 2   │ 0.6     │ 0.493593 │ 0.325054  │ true           │\n",
      "│ 3   │ 0.1     │ 0.0      │ 0.0597605 │ true           │\n",
      "│ 4   │ 0.3     │ 0.0      │ 0.124181  │ true           │\n",
      "│ 5   │ 0.8     │ 0.469903 │ 0.395522  │ true           │\n",
      "│ 6   │ 0.4     │ 0.0      │ 0.105253  │ true           │\n",
      "│ 7   │ 0.9     │ 0.733311 │ 0.530867  │ true           │\n",
      "│ 8   │ 0.2     │ 0.0      │ 0.0206105 │ true           │\n",
      "│ 9   │ 1.0     │ 0.842425 │ 0.580915  │ true           │\n",
      "│ 10  │ 0.5     │ 0.0      │ 0.250217  │ true           │\n",
      "IHT true positives = 5\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 5\n",
      "LASSO true positives = 10\n",
      "LASSO false positives = 108\n",
      "LASSO false negatives = 0\n",
      "marginal true positives = 10\n",
      "marginal false positives = 1369\n",
      "marginal false negatives = 0\n",
      "\n",
      "current run = 35\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/poisson\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = poisson\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β  │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64  │ Bool           │\n",
      "├─────┼─────────┼──────────┼──────────┼────────────────┤\n",
      "│ 1   │ 1.0     │ 0.681633 │ 0.560499 │ true           │\n",
      "│ 2   │ 0.7     │ 0.490532 │ 0.402023 │ true           │\n",
      "│ 3   │ 0.3     │ 0.134958 │ 0.086665 │ true           │\n",
      "│ 4   │ 0.9     │ 0.597327 │ 0.500084 │ true           │\n",
      "│ 5   │ 0.4     │ 0.264683 │ 0.154603 │ true           │\n",
      "│ 6   │ 0.6     │ 0.434242 │ 0.329904 │ true           │\n",
      "│ 7   │ 0.1     │ 0.0      │ 0.0      │ false          │\n",
      "│ 8   │ 0.8     │ 0.565653 │ 0.388727 │ true           │\n",
      "│ 9   │ 0.5     │ 0.251279 │ 0.224147 │ false          │\n",
      "│ 10  │ 0.2     │ 0.0      │ 0.0      │ false          │\n",
      "IHT true positives = 8\n",
      "IHT false positives = 27\n",
      "IHT false negatives = 2\n",
      "LASSO true positives = 8\n",
      "LASSO false positives = 88\n",
      "LASSO false negatives = 2\n",
      "marginal true positives = 7\n",
      "marginal false positives = 1601\n",
      "marginal false negatives = 3\n",
      "\n",
      "current run = 36\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/poisson\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = poisson\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β  │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64  │ Bool           │\n",
      "├─────┼─────────┼──────────┼──────────┼────────────────┤\n",
      "│ 1   │ 0.7     │ 0.401805 │ 0.351608 │ true           │\n",
      "│ 2   │ 0.9     │ 0.605587 │ 0.482873 │ true           │\n",
      "│ 3   │ 1.0     │ 0.666047 │ 0.55033  │ true           │\n",
      "│ 4   │ 0.1     │ 0.0      │ 0.0      │ false          │\n",
      "│ 5   │ 0.4     │ 0.288858 │ 0.164079 │ true           │\n",
      "│ 6   │ 0.5     │ 0.296594 │ 0.203725 │ true           │\n",
      "│ 7   │ 0.3     │ 0.185214 │ 0.110372 │ true           │\n",
      "│ 8   │ 0.6     │ 0.365271 │ 0.313539 │ true           │\n",
      "│ 9   │ 0.2     │ 0.0      │ 0.0      │ true           │\n",
      "│ 10  │ 0.8     │ 0.521659 │ 0.421292 │ true           │\n",
      "IHT true positives = 8\n",
      "IHT false positives = 8\n",
      "IHT false negatives = 2\n",
      "LASSO true positives = 8\n",
      "LASSO false positives = 92\n",
      "LASSO false negatives = 2\n",
      "marginal true positives = 9\n",
      "marginal false positives = 1487\n",
      "marginal false negatives = 1\n",
      "\n",
      "current run = 37\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/poisson\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = poisson\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.8     │ 0.658071 │ 0.465159  │ true           │\n",
      "│ 2   │ 0.6     │ 0.498644 │ 0.372422  │ true           │\n",
      "│ 3   │ 0.4     │ 0.315885 │ 0.222999  │ true           │\n",
      "│ 4   │ 0.7     │ 0.529712 │ 0.39195   │ true           │\n",
      "│ 5   │ 1.0     │ 0.691793 │ 0.538714  │ true           │\n",
      "│ 6   │ 0.2     │ 0.0      │ 0.031009  │ true           │\n",
      "│ 7   │ 0.5     │ 0.44982  │ 0.278749  │ true           │\n",
      "│ 8   │ 0.3     │ 0.16483  │ 0.108927  │ true           │\n",
      "│ 9   │ 0.9     │ 0.649348 │ 0.500679  │ true           │\n",
      "│ 10  │ 0.1     │ 0.0      │ 0.0408969 │ true           │\n",
      "IHT true positives = 8\n",
      "IHT false positives = 7\n",
      "IHT false negatives = 2\n",
      "LASSO true positives = 10\n",
      "LASSO false positives = 84\n",
      "LASSO false negatives = 0\n",
      "marginal true positives = 10\n",
      "marginal false positives = 1537\n",
      "marginal false negatives = 0\n",
      "\n",
      "current run = 38\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/poisson\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = poisson\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 1.0     │ 0.683978 │ 0.64975   │ true           │\n",
      "│ 2   │ 0.6     │ 0.401162 │ 0.345808  │ true           │\n",
      "│ 3   │ 0.7     │ 0.446517 │ 0.399189  │ true           │\n",
      "│ 4   │ 0.8     │ 0.513718 │ 0.490683  │ true           │\n",
      "│ 5   │ 0.2     │ 0.0      │ 0.0744955 │ true           │\n",
      "│ 6   │ 0.5     │ 0.287425 │ 0.240175  │ true           │\n",
      "│ 7   │ 0.3     │ 0.0      │ 0.0605703 │ true           │\n",
      "│ 8   │ 0.9     │ 0.594008 │ 0.530462  │ true           │\n",
      "│ 9   │ 0.4     │ 0.251683 │ 0.213324  │ true           │\n",
      "│ 10  │ 0.1     │ 0.0      │ 0.0       │ false          │\n",
      "IHT true positives = 7\n",
      "IHT false positives = 2\n",
      "IHT false negatives = 3\n",
      "LASSO true positives = 9\n",
      "LASSO false positives = 116\n",
      "LASSO false negatives = 1\n",
      "marginal true positives = 9\n",
      "marginal false positives = 1538\n",
      "marginal false negatives = 1\n",
      "\n",
      "current run = 39\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/poisson\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = poisson\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.8     │ 0.519165 │ 0.420496  │ true           │\n",
      "│ 2   │ 1.0     │ 0.749498 │ 0.561561  │ true           │\n",
      "│ 3   │ 0.6     │ 0.327905 │ 0.293387  │ true           │\n",
      "│ 4   │ 0.4     │ 0.245636 │ 0.157179  │ true           │\n",
      "│ 5   │ 0.2     │ 0.0      │ 0.0579667 │ true           │\n",
      "│ 6   │ 0.7     │ 0.553997 │ 0.411991  │ true           │\n",
      "│ 7   │ 0.3     │ 0.212928 │ 0.140496  │ true           │\n",
      "│ 8   │ 0.5     │ 0.346342 │ 0.238564  │ true           │\n",
      "│ 9   │ 0.1     │ 0.0      │ 0.0       │ false          │\n",
      "│ 10  │ 0.9     │ 0.688371 │ 0.498065  │ true           │\n",
      "IHT true positives = 8\n",
      "IHT false positives = 7\n",
      "IHT false negatives = 2\n",
      "LASSO true positives = 9\n",
      "LASSO false positives = 150\n",
      "LASSO false negatives = 1\n",
      "marginal true positives = 9\n",
      "marginal false positives = 1414\n",
      "marginal false negatives = 1\n",
      "\n",
      "current run = 40\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/poisson\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = poisson\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.1     │ 0.0      │ 0.0390544 │ true           │\n",
      "│ 2   │ 0.4     │ 0.235878 │ 0.169491  │ true           │\n",
      "│ 3   │ 0.3     │ 0.0      │ 0.0840205 │ true           │\n",
      "│ 4   │ 1.0     │ 0.64988  │ 0.54742   │ true           │\n",
      "│ 5   │ 0.2     │ 0.0      │ 0.0442373 │ true           │\n",
      "│ 6   │ 0.7     │ 0.465028 │ 0.34537   │ true           │\n",
      "│ 7   │ 0.9     │ 0.762111 │ 0.50811   │ true           │\n",
      "│ 8   │ 0.5     │ 0.316163 │ 0.22594   │ true           │\n",
      "│ 9   │ 0.6     │ 0.616211 │ 0.397713  │ true           │\n",
      "│ 10  │ 0.8     │ 0.664001 │ 0.472159  │ true           │\n",
      "IHT true positives = 7\n",
      "IHT false positives = 1\n",
      "IHT false negatives = 3\n",
      "LASSO true positives = 10\n",
      "LASSO false positives = 58\n",
      "LASSO false negatives = 0\n",
      "marginal true positives = 10\n",
      "marginal false positives = 1459\n",
      "marginal false negatives = 0\n",
      "\n",
      "current run = 41\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/poisson\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = poisson\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.4     │ 0.340025 │ 0.20775   │ true           │\n",
      "│ 2   │ 0.6     │ 0.455426 │ 0.343987  │ true           │\n",
      "│ 3   │ 0.5     │ 0.378279 │ 0.298161  │ true           │\n",
      "│ 4   │ 0.9     │ 0.539708 │ 0.504963  │ true           │\n",
      "│ 5   │ 0.7     │ 0.621076 │ 0.450508  │ true           │\n",
      "│ 6   │ 0.2     │ 0.0      │ 0.0704789 │ true           │\n",
      "│ 7   │ 0.8     │ 0.703206 │ 0.510953  │ true           │\n",
      "│ 8   │ 0.1     │ 0.0      │ 0.0       │ false          │\n",
      "│ 9   │ 1.0     │ 0.698777 │ 0.627185  │ true           │\n",
      "│ 10  │ 0.3     │ 0.0      │ 0.13988   │ true           │\n",
      "IHT true positives = 7\n",
      "IHT false positives = 0\n",
      "IHT false negatives = 3\n",
      "LASSO true positives = 9\n",
      "LASSO false positives = 123\n",
      "LASSO false negatives = 1\n",
      "marginal true positives = 9\n",
      "marginal false positives = 1435\n",
      "marginal false negatives = 1\n",
      "\n",
      "current run = 42\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/poisson\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = poisson\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.6     │ 0.411144 │ 0.358405  │ true           │\n",
      "│ 2   │ 0.3     │ 0.204098 │ 0.129843  │ true           │\n",
      "│ 3   │ 0.5     │ 0.339368 │ 0.275453  │ true           │\n",
      "│ 4   │ 0.2     │ 0.0      │ 0.0320659 │ true           │\n",
      "│ 5   │ 0.8     │ 0.557145 │ 0.464018  │ true           │\n",
      "│ 6   │ 0.7     │ 0.485551 │ 0.442672  │ true           │\n",
      "│ 7   │ 1.0     │ 0.745199 │ 0.630594  │ true           │\n",
      "│ 8   │ 0.1     │ 0.0      │ 0.0       │ false          │\n",
      "│ 9   │ 0.4     │ 0.245917 │ 0.18207   │ true           │\n",
      "│ 10  │ 0.9     │ 0.616174 │ 0.555088  │ true           │\n",
      "IHT true positives = 8\n",
      "IHT false positives = 8\n",
      "IHT false negatives = 2\n",
      "LASSO true positives = 9\n",
      "LASSO false positives = 137\n",
      "LASSO false negatives = 1\n",
      "marginal true positives = 9\n",
      "marginal false positives = 1614\n",
      "marginal false negatives = 1\n",
      "\n",
      "current run = 43\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/poisson\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = poisson\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.7     │ 0.520003 │ 0.394767  │ true           │\n",
      "│ 2   │ 0.4     │ 0.310165 │ 0.181402  │ true           │\n",
      "│ 3   │ 0.6     │ 0.415884 │ 0.312067  │ true           │\n",
      "│ 4   │ 0.1     │ 0.0      │ 0.0       │ false          │\n",
      "│ 5   │ 0.2     │ 0.171725 │ 0.0610361 │ true           │\n",
      "│ 6   │ 0.8     │ 0.559521 │ 0.44057   │ true           │\n",
      "│ 7   │ 0.5     │ 0.434923 │ 0.308506  │ true           │\n",
      "│ 8   │ 0.9     │ 0.729888 │ 0.560918  │ true           │\n",
      "│ 9   │ 0.3     │ 0.189277 │ 0.0897948 │ true           │\n",
      "│ 10  │ 1.0     │ 0.778393 │ 0.590987  │ true           │\n",
      "IHT true positives = 9\n",
      "IHT false positives = 22\n",
      "IHT false negatives = 1\n",
      "LASSO true positives = 9\n",
      "LASSO false positives = 123\n",
      "LASSO false negatives = 1\n",
      "marginal true positives = 9\n",
      "marginal false positives = 1631\n",
      "marginal false negatives = 1\n",
      "\n",
      "current run = 44\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/poisson\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = poisson\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.2     │ 0.0      │ 0.0601488 │ true           │\n",
      "│ 2   │ 0.1     │ 0.0      │ 0.0       │ true           │\n",
      "│ 3   │ 0.8     │ 0.729818 │ 0.478808  │ true           │\n",
      "│ 4   │ 0.3     │ 0.0      │ 0.0603755 │ true           │\n",
      "│ 5   │ 0.4     │ 0.0      │ 0.140479  │ true           │\n",
      "│ 6   │ 0.6     │ 0.486966 │ 0.322449  │ true           │\n",
      "│ 7   │ 0.9     │ 0.655607 │ 0.473972  │ true           │\n",
      "│ 8   │ 1.0     │ 0.694384 │ 0.515743  │ true           │\n",
      "│ 9   │ 0.7     │ 0.576398 │ 0.377102  │ true           │\n",
      "│ 10  │ 0.5     │ 0.464763 │ 0.277384  │ true           │\n",
      "IHT true positives = 6\n",
      "IHT false positives = 1\n",
      "IHT false negatives = 4\n",
      "LASSO true positives = 9\n",
      "LASSO false positives = 68\n",
      "LASSO false negatives = 1\n",
      "marginal true positives = 10\n",
      "marginal false positives = 1586\n",
      "marginal false negatives = 0\n",
      "\n",
      "current run = 45\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/poisson\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = poisson\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β     │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64   │ Float64   │ Bool           │\n",
      "├─────┼─────────┼───────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.3     │ 0.204419  │ 0.148197  │ true           │\n",
      "│ 2   │ 0.7     │ 0.527842  │ 0.437424  │ true           │\n",
      "│ 3   │ 0.2     │ 0.0946938 │ 0.0578157 │ true           │\n",
      "│ 4   │ 0.1     │ 0.0       │ 0.0       │ false          │\n",
      "│ 5   │ 0.8     │ 0.629514  │ 0.539366  │ true           │\n",
      "│ 6   │ 0.9     │ 0.681887  │ 0.593177  │ true           │\n",
      "│ 7   │ 1.0     │ 0.738926  │ 0.651438  │ true           │\n",
      "│ 8   │ 0.4     │ 0.234206  │ 0.165937  │ true           │\n",
      "│ 9   │ 0.5     │ 0.366306  │ 0.283484  │ true           │\n",
      "│ 10  │ 0.6     │ 0.473224  │ 0.386027  │ true           │\n",
      "IHT true positives = 9\n",
      "IHT false positives = 19\n",
      "IHT false negatives = 1\n",
      "LASSO true positives = 9\n",
      "LASSO false positives = 211\n",
      "LASSO false negatives = 1\n",
      "marginal true positives = 9\n",
      "marginal false positives = 1478\n",
      "marginal false negatives = 1\n",
      "\n",
      "current run = 46\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/poisson\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = poisson\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.5     │ 0.318478 │ 0.187698  │ true           │\n",
      "│ 2   │ 0.8     │ 0.550585 │ 0.421135  │ true           │\n",
      "│ 3   │ 0.4     │ 0.297477 │ 0.184375  │ true           │\n",
      "│ 4   │ 0.6     │ 0.441207 │ 0.307557  │ true           │\n",
      "│ 5   │ 0.9     │ 0.656412 │ 0.480679  │ true           │\n",
      "│ 6   │ 0.7     │ 0.58837  │ 0.407109  │ true           │\n",
      "│ 7   │ 0.3     │ 0.0      │ 0.0664074 │ true           │\n",
      "│ 8   │ 0.2     │ 0.0      │ 0.0412395 │ true           │\n",
      "│ 9   │ 1.0     │ 0.630641 │ 0.523443  │ true           │\n",
      "│ 10  │ 0.1     │ 0.0      │ 0.0       │ false          │\n",
      "IHT true positives = 7\n",
      "IHT false positives = 2\n",
      "IHT false negatives = 3\n",
      "LASSO true positives = 9\n",
      "LASSO false positives = 94\n",
      "LASSO false negatives = 1\n",
      "marginal true positives = 9\n",
      "marginal false positives = 1669\n",
      "marginal false negatives = 1\n",
      "\n",
      "current run = 47\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/poisson\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = poisson\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.8     │ 0.605825 │ 0.429204  │ true           │\n",
      "│ 2   │ 0.1     │ 0.0      │ 0.0       │ true           │\n",
      "│ 3   │ 1.0     │ 0.689334 │ 0.558228  │ true           │\n",
      "│ 4   │ 0.7     │ 0.464966 │ 0.303998  │ true           │\n",
      "│ 5   │ 0.2     │ 0.0      │ 0.0       │ false          │\n",
      "│ 6   │ 0.3     │ 0.174762 │ 0.0811726 │ true           │\n",
      "│ 7   │ 0.9     │ 0.591993 │ 0.451993  │ true           │\n",
      "│ 8   │ 0.5     │ 0.321994 │ 0.251572  │ true           │\n",
      "│ 9   │ 0.6     │ 0.371794 │ 0.270948  │ true           │\n",
      "│ 10  │ 0.4     │ 0.276552 │ 0.195481  │ true           │\n",
      "IHT true positives = 8\n",
      "IHT false positives = 25\n",
      "IHT false negatives = 2\n",
      "LASSO true positives = 8\n",
      "LASSO false positives = 74\n",
      "LASSO false negatives = 2\n",
      "marginal true positives = 9\n",
      "marginal false positives = 1446\n",
      "marginal false negatives = 1\n",
      "\n",
      "current run = 48\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/poisson\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = poisson\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 1.0     │ 0.8148   │ 0.679784  │ true           │\n",
      "│ 2   │ 0.8     │ 0.586736 │ 0.494957  │ true           │\n",
      "│ 3   │ 0.9     │ 0.750834 │ 0.602499  │ true           │\n",
      "│ 4   │ 0.7     │ 0.642758 │ 0.458169  │ true           │\n",
      "│ 5   │ 0.5     │ 0.0      │ 0.24243   │ true           │\n",
      "│ 6   │ 0.6     │ 0.427188 │ 0.376141  │ true           │\n",
      "│ 7   │ 0.1     │ 0.0      │ 0.0       │ false          │\n",
      "│ 8   │ 0.2     │ 0.0      │ 0.0672205 │ true           │\n",
      "│ 9   │ 0.4     │ 0.0      │ 0.225553  │ true           │\n",
      "│ 10  │ 0.3     │ 0.0      │ 0.133884  │ true           │\n",
      "IHT true positives = 5\n",
      "IHT false positives = 1\n",
      "IHT false negatives = 5\n",
      "LASSO true positives = 9\n",
      "LASSO false positives = 188\n",
      "LASSO false negatives = 1\n",
      "marginal true positives = 9\n",
      "marginal false positives = 1601\n",
      "marginal false negatives = 1\n",
      "\n",
      "current run = 49\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/poisson\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = poisson\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.3     │ 0.241303 │ 0.162111  │ true           │\n",
      "│ 2   │ 0.2     │ 0.0      │ 0.0179707 │ false          │\n",
      "│ 3   │ 0.7     │ 0.451966 │ 0.342384  │ true           │\n",
      "│ 4   │ 0.9     │ 0.748155 │ 0.55021   │ true           │\n",
      "│ 5   │ 0.1     │ 0.0      │ 0.0       │ false          │\n",
      "│ 6   │ 0.5     │ 0.365515 │ 0.247688  │ true           │\n",
      "│ 7   │ 0.8     │ 0.665092 │ 0.45864   │ true           │\n",
      "│ 8   │ 0.4     │ 0.26747  │ 0.156894  │ true           │\n",
      "│ 9   │ 1.0     │ 0.821844 │ 0.642205  │ true           │\n",
      "│ 10  │ 0.6     │ 0.504464 │ 0.334313  │ true           │\n",
      "IHT true positives = 8\n",
      "IHT false positives = 3\n",
      "IHT false negatives = 2\n",
      "LASSO true positives = 9\n",
      "LASSO false positives = 79\n",
      "LASSO false negatives = 1\n",
      "marginal true positives = 8\n",
      "marginal false positives = 1439\n",
      "marginal false negatives = 2\n",
      "\n",
      "current run = 50\n",
      " \n",
      " \n",
      "     Welcome to OpenMendel's\n",
      "      GWAS analysis option\n",
      "        version 0.5.0\n",
      " \n",
      " \n",
      "Reading the data.\n",
      "\n",
      "The current working directory is \"/u/home/b/biona001/precision_recall/poisson\".\n",
      "\n",
      "Keywords modified by the user:\n",
      "\n",
      "  affected_designator = 2\n",
      "  control_file = tmp_control.txt\n",
      "  output_table = tmp_table.txt\n",
      "  pedigree_file = tmp.fam\n",
      "  plink_field_separator = \t\n",
      "  plink_input_basename = tmp\n",
      "  regression = poisson\n",
      "  regression_formula = Trait ~\n",
      "  snpdata_file = tmp.bed\n",
      "  snpdefinition_file = tmp.bim\n",
      " \n",
      " \n",
      "Analyzing the data.\n",
      "\n",
      " \n",
      " \n",
      "Mendel's analysis is finished.\n",
      "\n",
      "compare_model = 10×4 DataFrame\n",
      "│ Row │ true_β  │ IHT_β    │ lasso_β   │ marginal_found │\n",
      "│     │ Float64 │ Float64  │ Float64   │ Bool           │\n",
      "├─────┼─────────┼──────────┼───────────┼────────────────┤\n",
      "│ 1   │ 0.1     │ 0.0      │ 0.0       │ false          │\n",
      "│ 2   │ 0.5     │ 0.344381 │ 0.217125  │ true           │\n",
      "│ 3   │ 0.2     │ 0.0      │ 0.0166546 │ true           │\n",
      "│ 4   │ 0.9     │ 0.643612 │ 0.488672  │ true           │\n",
      "│ 5   │ 0.7     │ 0.451139 │ 0.330958  │ true           │\n",
      "│ 6   │ 1.0     │ 0.8001   │ 0.559272  │ true           │\n",
      "│ 7   │ 0.6     │ 0.476192 │ 0.296632  │ true           │\n",
      "│ 8   │ 0.4     │ 0.0      │ 0.16691   │ true           │\n",
      "│ 9   │ 0.3     │ 0.0      │ 0.0871088 │ true           │\n",
      "│ 10  │ 0.8     │ 0.609667 │ 0.420741  │ true           │\n",
      "IHT true positives = 6\n",
      "IHT false positives = 2\n",
      "IHT false negatives = 4\n",
      "LASSO true positives = 9\n",
      "LASSO false positives = 53\n",
      "LASSO false negatives = 1\n",
      "marginal true positives = 9\n",
      "marginal false positives = 1587\n",
      "marginal false negatives = 1\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "([8.0, 7.0, 4.0, 7.0, 6.0, 10.0, 9.0, 8.0, 7.0, 6.0  …  7.0, 8.0, 9.0, 6.0, 9.0, 7.0, 8.0, 5.0, 8.0, 6.0], [1.0, 1.0, 0.0, 1.0, 2.0, 9.0, 21.0, 10.0, 0.0, 0.0  …  0.0, 8.0, 22.0, 1.0, 19.0, 2.0, 25.0, 1.0, 3.0, 2.0], [2.0, 3.0, 6.0, 3.0, 4.0, 0.0, 1.0, 2.0, 3.0, 4.0  …  3.0, 2.0, 1.0, 4.0, 1.0, 3.0, 2.0, 5.0, 2.0, 4.0], [10.0, 9.0, 9.0, 9.0, 9.0, 10.0, 9.0, 9.0, 9.0, 10.0  …  9.0, 9.0, 9.0, 9.0, 9.0, 9.0, 8.0, 9.0, 9.0, 9.0], [84.0, 87.0, 75.0, 103.0, 33.0, 71.0, 147.0, 99.0, 50.0, 82.0  …  123.0, 137.0, 123.0, 68.0, 211.0, 94.0, 74.0, 188.0, 79.0, 53.0], [0.0, 1.0, 1.0, 1.0, 1.0, 0.0, 1.0, 1.0, 1.0, 0.0  …  1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 2.0, 1.0, 1.0, 1.0], [9.0, 10.0, 9.0, 9.0, 7.0, 9.0, 9.0, 9.0, 10.0, 10.0  …  9.0, 9.0, 9.0, 10.0, 9.0, 9.0, 9.0, 9.0, 8.0, 9.0], [1614.0, 1648.0, 1583.0, 1600.0, 1431.0, 1475.0, 1492.0, 1557.0, 1656.0, 1443.0  …  1435.0, 1614.0, 1631.0, 1586.0, 1478.0, 1669.0, 1446.0, 1601.0, 1439.0, 1587.0], [1.0, 0.0, 1.0, 1.0, 3.0, 1.0, 1.0, 1.0, 0.0, 0.0  …  1.0, 1.0, 1.0, 0.0, 1.0, 1.0, 1.0, 1.0, 2.0, 1.0])"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "iht_true_positives, iht_false_positives, iht_false_negatives, \n",
    "    lasso_true_positives, lasso_false_positives, lasso_false_negatives, \n",
    "    marginal_true_positives, marginal_false_positives, marginal_false_negatives = run()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3×3 Array{Float64,2}:\n",
       " 7.08     5.04  2.92\n",
       " 9.08    95.02  0.92\n",
       " 9.08  1525.94  0.92"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "poisson_iht_true_positives = sum(iht_true_positives) / 50\n",
    "poisson_iht_false_positives = sum(iht_false_positives) / 50\n",
    "poisson_iht_false_negatives = sum(iht_false_negatives) / 50\n",
    "poisson_lasso_true_positives = sum(lasso_true_positives) / 50\n",
    "poisson_lasso_false_positives = sum(lasso_false_positives) / 50\n",
    "poisson_lasso_false_negatives = sum(lasso_false_negatives) / 50\n",
    "poisson_marginal_true_positives = sum(marginal_true_positives) / 50\n",
    "poisson_marginal_false_positives = sum(marginal_false_positives) / 50\n",
    "poisson_marginal_false_negatives = sum(marginal_false_negatives) / 50\n",
    "\n",
    "result = [poisson_iht_true_positives poisson_iht_false_positives poisson_iht_false_negatives; \n",
    "        poisson_lasso_true_positives poisson_lasso_false_positives poisson_lasso_false_negatives;\n",
    "        poisson_marginal_true_positives poisson_marginal_false_positives poisson_marginal_false_negatives]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Julia 1.0.1",
   "language": "julia",
   "name": "julia-1.0"
  },
  "language_info": {
   "file_extension": ".jl",
   "mimetype": "application/julia",
   "name": "julia",
   "version": "1.0.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
