<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>API · MendelIHT</title><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.0/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.0/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.0/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.11.1/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit">MendelIHT</span></div><form class="docs-search" action="../../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li><a class="tocitem" href="../getting_started/">Getting Started</a></li><li><a class="tocitem" href="../examples/">Examples</a></li><li><a class="tocitem" href="../math/">Mathematical Details</a></li><li><a class="tocitem" href="../contributing/">Contributing</a></li><li class="is-active"><a class="tocitem" href>API</a><ul class="internal"><li><a class="tocitem" href="#Wrapper-Functions"><span>Wrapper Functions</span></a></li><li><a class="tocitem" href="#Core-Functions"><span>Core Functions</span></a></li><li><a class="tocitem" href="#Specifying-Groups-and-Weights"><span>Specifying Groups and Weights</span></a></li><li><a class="tocitem" href="#Simulation-Utilities"><span>Simulation Utilities</span></a></li><li><a class="tocitem" href="#Other-Useful-Functions"><span>Other Useful Functions</span></a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>API</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>API</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/OpenMendel/MendelIHT.jl/blob/master/docs/src/man/api.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="API"><a class="docs-heading-anchor" href="#API">API</a><a id="API-1"></a><a class="docs-heading-anchor-permalink" href="#API" title="Permalink"></a></h1><h2 id="Wrapper-Functions"><a class="docs-heading-anchor" href="#Wrapper-Functions">Wrapper Functions</a><a id="Wrapper-Functions-1"></a><a class="docs-heading-anchor-permalink" href="#Wrapper-Functions" title="Permalink"></a></h2><p>Most users will use the following wrapper functions, which automatically handles everything. The user only has to specify where the PLINK files (and possibly the phenotype/covariate files) are located. </p><article class="docstring"><header><a class="docstring-binding" id="MendelIHT.iht" href="#MendelIHT.iht"><code>MendelIHT.iht</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">iht(plinkfile, k, kwargs...)</code></pre><p>Runs IHT with sparsity level <code>k</code>. Example:</p><pre><code class="language-julia">result = iht(&quot;plinkfile&quot;, 10)</code></pre><p><strong>Phenotypes and other covariates</strong></p><p>Will use 6th column of <code>.fam</code> file for phenotype values and will automatically include an intercept as the only non-genetic covariate. </p><p><strong>Arguments</strong></p><ul><li><code>plinkfile</code>: A <code>String</code> for input PLINK file name (without <code>.bim/.bed/.fam</code> suffixes)</li><li><code>k</code>: An <code>Int</code> for sparsity parameter = number of none-zero coefficients</li></ul><p><strong>Optional Arguments</strong></p><p>All arguments available in <a href="#StatsBase.fit"><code>fit</code></a></p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/OpenMendel/MendelIHT.jl/blob/84e3705bd851b1d409d0bb6e34f11867f94ddb00/src/wrapper.jl#L1-L20">source</a></section><section><div><pre><code class="language-none">iht(plinkfile, covariates, k, kwargs...)</code></pre><p>Runs IHT with sparsity level <code>k</code>, with additional covariates stored separately. Example:</p><pre><code class="language-julia">result = iht(&quot;plinkfile&quot;, &quot;covariates.txt&quot;, 10)</code></pre><p><strong>Phenotypes</strong></p><p>Will use 6th column of <code>.fam</code> file for phenotype values. </p><p><strong>Other covariates</strong></p><p>Covariates are read using <a href="man/@ref"><code>readdlm</code></a> function in Julia base. We require the covariate file to be comma separated, and not include a header line. Each row should be listed in the same order as in the PLINK. The first column should be all 1s to indicate an intercept. All other columns will be standardized to mean 0 variance 1. </p><p><strong>Arguments</strong></p><ul><li><code>plinkfile</code>: A <code>String</code> for input PLINK file name (without <code>.bim/.bed/.fam</code> suffixes)</li><li><code>covariates</code>: A <code>String</code> for covariate file name</li><li><code>k</code>: An <code>Int</code> for sparsity parameter = number of none-zero coefficients</li></ul><p><strong>Optional Arguments</strong></p><p>All arguments available in <a href="#StatsBase.fit"><code>fit</code></a></p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/OpenMendel/MendelIHT.jl/blob/84e3705bd851b1d409d0bb6e34f11867f94ddb00/src/wrapper.jl#L33-L59">source</a></section><section><div><pre><code class="language-none">iht(phenotypes, plinkfile, covariates, k, kwargs...)</code></pre><p>Runs IHT with sparsity level <code>k</code>, where both phenotypes and additional covariates are stored separately. Example:</p><pre><code class="language-julia">result = iht(&quot;phenotypes.txt&quot;, &quot;plinkfile&quot;, &quot;covariates.txt&quot;, 10)</code></pre><p><strong>Phenotypes</strong></p><p>Phenotypes are read using <a href="man/@ref"><code>readdlm</code></a> function in Julia base. We require each  subject&#39;s phenotype to occupy a different row. The file should not include a header line. Each row should be listed in the same order as in the PLINK. </p><p><strong>Other covariates</strong></p><p>Covariates are read using <a href="man/@ref"><code>readdlm</code></a> function in Julia base. We require the covariate file to be comma separated, and not include a header line. Each row should be listed in the same order as in the PLINK. The first column should be all 1s to indicate an intercept. All other columns will be standardized to mean 0 variance 1. </p><p><strong>Arguments</strong></p><ul><li><code>phenotypes</code>: A <code>String</code> for phenotype file name</li><li><code>plinkfile</code>: A <code>String</code> for input PLINK file name (without <code>.bim/.bed/.fam</code> suffixes)</li><li><code>covariates</code>: A <code>String</code> for covariate file name</li><li><code>k</code>: An <code>Int</code> for sparsity parameter = number of none-zero coefficients</li></ul><p><strong>Optional Arguments</strong></p><p>All arguments available in <a href="#StatsBase.fit"><code>fit</code></a></p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/OpenMendel/MendelIHT.jl/blob/84e3705bd851b1d409d0bb6e34f11867f94ddb00/src/wrapper.jl#L81-L111">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="MendelIHT.cross_validate" href="#MendelIHT.cross_validate"><code>MendelIHT.cross_validate</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">cross_validate(plinkfile, path, kwargs...)</code></pre><p>Runs cross-validation to determinal optimal sparsity level <code>k</code>. Sparsity levels is specified in `path. Example:</p><pre><code class="language-julia">mses = cross_validate(&quot;plinkfile&quot;, 1:20)
mses = cross_validate(&quot;plinkfile&quot;, [1, 2, 3, 4, 5]) # alternative syntax</code></pre><p><strong>Phenotypes and other covariates</strong></p><p>Will use 6th column of <code>.fam</code> file for phenotype values and will automatically include an intercept as the only non-genetic covariate. </p><p><strong>Arguments</strong></p><ul><li><code>plinkfile</code>: A <code>String</code> for input PLINK file name (without <code>.bim/.bed/.fam</code> suffixes)</li><li><code>path</code>: Different sparsity levels. Can be an integer range (default 1:20) or vector of integers. </li></ul><p><strong>Optional Arguments</strong></p><p>All arguments available in <a href="#MendelIHT.cv_iht"><code>cv_iht</code></a></p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/OpenMendel/MendelIHT.jl/blob/84e3705bd851b1d409d0bb6e34f11867f94ddb00/src/wrapper.jl#L137-L158">source</a></section><section><div><pre><code class="language-none">cross_validate(plinkfile, covariates, path, kwargs...)</code></pre><p>Runs cross-validation to determinal optimal sparsity level <code>k</code>. Sparsity levels is specified in `path. Example:</p><pre><code class="language-julia">mses = cross_validate(&quot;plinkfile&quot;, &quot;covariates.txt&quot;, 1:20)
mses = cross_validate(&quot;plinkfile&quot;, &quot;covariates.txt&quot;, [1, 10, 20]) # alternative syntax</code></pre><p><strong>Phenotypes</strong></p><p>Will use 6th column of <code>.fam</code> file for phenotype values. </p><p><strong>Other covariates</strong></p><p>Covariates are read using <a href="man/@ref"><code>readdlm</code></a> function in Julia base. We require the covariate file to be comma separated, and not include a header line. Each row should be listed in the same order as in the PLINK. The first column should be all 1s to indicate an intercept. All other columns will be standardized to mean 0 variance 1. </p><p><strong>Arguments</strong></p><ul><li><code>plinkfile</code>: A <code>String</code> for input PLINK file name (without <code>.bim/.bed/.fam</code> suffixes)</li><li><code>covariates</code>: A <code>String</code> for covariate file name</li><li><code>path</code>: Different sparsity levels. Can be an integer range (default 1:20) or vector of integers. </li></ul><p><strong>Optional Arguments</strong></p><p>All arguments available in <a href="#MendelIHT.cv_iht"><code>cv_iht</code></a></p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/OpenMendel/MendelIHT.jl/blob/84e3705bd851b1d409d0bb6e34f11867f94ddb00/src/wrapper.jl#L170-L198">source</a></section><section><div><pre><code class="language-none">cross_validate(phenotypes, plinkfile, covariates, path, kwargs...)</code></pre><p>Runs cross-validation to determinal optimal sparsity level <code>k</code>. Sparsity levels is specified in `path. Example:</p><pre><code class="language-julia">mses = cross_validate(&quot;phenotypes.txt&quot;, &quot;plinkfile&quot;, &quot;covariates.txt&quot;, 1:20)
mses = cross_validate(&quot;phenotypes.txt&quot;, &quot;plinkfile&quot;, &quot;covariates.txt&quot;, [1, 10, 20]) # alternative syntax</code></pre><p><strong>Phenotypes</strong></p><p>Phenotypes are read using <a href="man/@ref"><code>readdlm</code></a> function in Julia base. We require each  subject&#39;s phenotype to occupy a different row. The file should not include a header line. Each row should be listed in the same order as in the PLINK. </p><p><strong>Other covariates</strong></p><p>Covariates are read using <a href="man/@ref"><code>readdlm</code></a> function in Julia base. We require the covariate file to be comma separated, and not include a header line. Each row should be listed in the same order as in the PLINK. The first column should be all 1s to indicate an intercept. All other columns will be standardized to mean 0 variance 1. </p><p><strong>Arguments</strong></p><ul><li><code>phenotypes</code>: A <code>String</code> for phenotype file name</li><li><code>plinkfile</code>: A <code>String</code> for input PLINK file name (without <code>.bim/.bed/.fam</code> suffixes)</li><li><code>covariates</code>: A <code>String</code> for covariate file name</li><li><code>path</code>: Different sparsity levels. Can be an integer range (default 1:20) or vector of integers. </li></ul><p><strong>Optional Arguments</strong></p><p>All arguments available in <a href="#MendelIHT.cv_iht"><code>cv_iht</code></a></p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/OpenMendel/MendelIHT.jl/blob/84e3705bd851b1d409d0bb6e34f11867f94ddb00/src/wrapper.jl#L219-L250">source</a></section></article><h2 id="Core-Functions"><a class="docs-heading-anchor" href="#Core-Functions">Core Functions</a><a id="Core-Functions-1"></a><a class="docs-heading-anchor-permalink" href="#Core-Functions" title="Permalink"></a></h2><p>For advanced users, one can also run IHT regression or cross-validation directly. For cross validation, we generally recommend using <code>cv_iht</code>. This function cycles through the testing sets sequentially and fits different sparsity models in parallel. For larger problems (e.g. UK Biobank sized), one can instead choose to run <code>cv_iht_distribute_fold</code>. This function fits different sparsity models sequentially but initializes all training/testing model in parallel, which consumes more memory (see below). The later strategy allows one to distribute different sparsity parameters to different computers, achieving greater parallel power. </p><article class="docstring"><header><a class="docstring-binding" id="MendelIHT.cv_iht" href="#MendelIHT.cv_iht"><code>MendelIHT.cv_iht</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">cv_iht(y, x, z; kwargs...)</code></pre><p>For each model specified in <code>path</code>, performs <code>q</code>-fold cross validation and  returns the (averaged) deviance residuals. </p><p>The purpose of this function is to find the best sparsity level <code>k</code>, obtained from selecting the model with the minimum out-of-sample error. Different cross validation folds are cycled through sequentially different <code>paths</code> are fitted in parallel on different CPUs. Currently there are no routines to cross validate different group sizes. </p><p><strong>Warning</strong></p><p>Do not remove files with random file names when you run this function. These are  memory mapped files that will be deleted automatically once they are no longer needed.</p><p><strong>Arguments</strong></p><ul><li><code>y</code>: Response vector (phenotypes), should be an <code>Array{T, 1}</code>.</li><li><code>x</code>: A design matrix (genotypes). Should be a <code>SnpArray</code> or an <code>Array{T, 2}</code>. </li><li><code>z</code>: Matrix of non-genetic covariates of type <code>Array{T, 2}</code> or <code>Array{T, 1}</code>. The first column should be the intercept (i.e. column of 1). </li></ul><p><strong>Optional Arguments:</strong></p><ul><li><code>path</code>: Different model sizes to be tested in cross validation (default 1:20)</li><li><code>q</code>: Number of cross validation folds. (default 5)</li><li><code>d</code>: Distribution of your phenotype. (default Normal)</li><li><code>l</code>: A link function (default IdentityLink)</li><li><code>est_r</code>: Symbol for whether to estimate nuisance parameters. Only supported distribution is negative binomial and choices include :Newton or :MM.</li><li><code>group</code>: vector storing group membership for each predictor</li><li><code>weight</code>: vector storing vector of weights containing prior knowledge on each predictor</li><li><code>folds</code>: Vector that separates the sample into q disjoint subsets</li><li><code>destin</code>: Directory where intermediate files will be generated. Directory name must end with <code>/</code>.</li><li><code>init</code>: Boolean indicating whether we should initialize IHT algorithm at a good starting guess</li><li><code>use_maf</code>: Boolean indicating we should scale the projection step by a weight vector </li><li><code>debias</code>: Boolean indicating whether we should debias at each IHT step</li><li><code>verbose</code>: Whether we want IHT to print meaningful intermediate steps</li><li><code>parallel</code>: Whether we want to run cv_iht using multiple CPUs (highly recommended)</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/OpenMendel/MendelIHT.jl/blob/84e3705bd851b1d409d0bb6e34f11867f94ddb00/src/cross_validation.jl#L1-L37">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="MendelIHT.cv_iht_distribute_fold" href="#MendelIHT.cv_iht_distribute_fold"><code>MendelIHT.cv_iht_distribute_fold</code></a> — <span class="docstring-category">Function</span></header><section><div><p>Performs q-fold cross validation for Iterative hard thresholding to  determine the best model size <code>k</code>. The function is the same as <code>cv_iht</code>  except here each <code>fold</code> is distributed to a different CPU as opposed  to each <code>path</code> to a different CPU. </p><p>This function has the edge over <code>cv_iht</code> because one can fit different  sparsity levels on different computers. But this is assuming you have  enough RAM and disk space to store all training data simultaneously.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/OpenMendel/MendelIHT.jl/blob/84e3705bd851b1d409d0bb6e34f11867f94ddb00/src/cross_validation.jl#L86-L95">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="StatsBase.fit" href="#StatsBase.fit"><code>StatsBase.fit</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">fit(Mod::Type{&lt;:StatisticalModel}, f::FormulaTerm, data, args...; 
    contrasts::Dict{Symbol}, kwargs...)</code></pre><p>Convert tabular data into a numeric response vector and predictor matrix using the formula <code>f</code>, and then <code>fit</code> the specified model type, wrapping the result in a <a href="man/@ref"><code>TableRegressionModel</code></a> or <a href="man/@ref"><code>TableStatisticalModel</code></a> (as appropriate).</p><p>This is intended as a backstop for modeling packages that implement model types that are subtypes of <code>StatsBase.StatisticalModel</code> but do not explicitly support the full StatsModels terms-based interface.  Currently this works by creating a <a href="man/@ref"><code>ModelFrame</code></a> from the formula and data, and then converting this to a <a href="man/@ref"><code>ModelMatrix</code></a>, but this is an internal implementation detail which may change in the near future.</p></div></section></article><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p><strong>Do not</strong> delete intermediate files with random file names created by <code>cv_iht</code> and <code>cv_iht_distribute_fold</code> (windows users will be instructed to manually do so via print statements). These are memory-mapped files necessary for cross validation. For <code>cv_iht</code>, <strong>you must have <code>x</code> GB of free space and RAM on your hard disk</strong> where <code>x</code> is your <code>.bed</code> file size. For <code>cv_iht_distribute_fold</code>, you must have enough RAM and disk space to fit all <code>q</code> training datasets simultaneously, each of which typically requires <code>(q - 1)/q * x</code> GB. </p></div></div><h2 id="Specifying-Groups-and-Weights"><a class="docs-heading-anchor" href="#Specifying-Groups-and-Weights">Specifying Groups and Weights</a><a id="Specifying-Groups-and-Weights-1"></a><a class="docs-heading-anchor-permalink" href="#Specifying-Groups-and-Weights" title="Permalink"></a></h2><p>When you have group and weight information, you input them as optional arguments in <code>fit</code> and <code>cv_iht</code>. The weight vector is a vector of <code>Float64</code>, while the group vector is a vector of <code>Int</code>. For instance,</p><pre><code class="language-Julia">    g = #import group vector
    w = #import weight vector
    ng = length(unique(g)) # specify number of non-zero groups
    result = fit(y, x, z; J=ng, k=10, d=Normal(), l=IdentityLink(), group=g, weight=w)</code></pre><h2 id="Simulation-Utilities"><a class="docs-heading-anchor" href="#Simulation-Utilities">Simulation Utilities</a><a id="Simulation-Utilities-1"></a><a class="docs-heading-anchor-permalink" href="#Simulation-Utilities" title="Permalink"></a></h2><p>MendelIHT provides some simulation utilities that help users explore the function and capabilities of iterative hard thresholding. </p><article class="docstring"><header><a class="docstring-binding" id="MendelIHT.simulate_random_snparray" href="#MendelIHT.simulate_random_snparray"><code>MendelIHT.simulate_random_snparray</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">simulate_random_snparray(s::String, n::Integer, p::Integer; 
    [mafs::Vector{Float64}], [min_ma::Integer])</code></pre><p>Creates a random SnpArray in the current directory without missing value,  where each SNP has ⫺5 (default) minor alleles. </p><p>Note: if supplied minor allele frequency is extremely small, it could take a long time for the simulation to generate samples where at least <code>min_ma</code> (defaults to 5) are present. </p><p><strong>Arguments:</strong></p><ul><li><code>s</code>: name of SnpArray that will be created in the current directory. To not   create file, use <code>undef</code>.</li><li><code>n</code>: number of samples</li><li><code>p</code>: number of SNPs</li></ul><p><strong>Optional Arguments:</strong></p><ul><li><code>mafs</code>: vector of desired minor allele freuqencies (uniform(0,0.5) by default)</li><li><code>min_ma</code>: the minimum number of minor alleles that must be present for each   SNP (defaults to 5)</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/OpenMendel/MendelIHT.jl/blob/84e3705bd851b1d409d0bb6e34f11867f94ddb00/src/simulate_utilities.jl#L1-L22">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="MendelIHT.simulate_correlated_snparray" href="#MendelIHT.simulate_correlated_snparray"><code>MendelIHT.simulate_correlated_snparray</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">simulate_correlated_snparray(s, n, p; block_length, hap, prob)</code></pre><p>Simulates a SnpArray with correlation. SNPs are divided into blocks where each adjacent SNP is the same with probability prob. There are no correlation between blocks.</p><p><strong>Arguments:</strong></p><ul><li><code>n</code>: number of samples</li><li><code>p</code>: number of SNPs</li><li><code>s</code>: name of SnpArray that will be created (memory mapped) in the current directory. To not memory map, use <code>undef</code>.</li></ul><p><strong>Optional arguments:</strong></p><ul><li><code>block_length</code>: length of each LD block</li><li><code>hap</code>: number of haplotypes to simulate for each block</li><li><code>prob</code>: with probability <code>prob</code> an adjacent SNP would be the same. </li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/OpenMendel/MendelIHT.jl/blob/84e3705bd851b1d409d0bb6e34f11867f94ddb00/src/simulate_utilities.jl#L103-L118">source</a></section></article><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p>Simulating a SnpArray with <span>$n$</span> subjects and <span>$p$</span> SNPs requires up to <span>$4np$</span> bits of RAM. Make sure you have enough RAM before simulating very large SnpArrays.</p></div></div><article class="docstring"><header><a class="docstring-binding" id="MendelIHT.simulate_random_response" href="#MendelIHT.simulate_random_response"><code>MendelIHT.simulate_random_response</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">simulate_random_response(x, k, d, l; kwargs...)</code></pre><p>This function simulates a random response (trait) vector <code>y</code>. When the  distribution <code>d</code> is from Poisson, Gamma, or Negative Binomial, we simulate  <code>β ∼ N(0, 0.3)</code> to roughly ensure the mean of response <code>y</code> doesn&#39;t become too large. For other distributions, we choose <code>β ∼ N(0, 1)</code>. </p><p><strong>Arguments</strong></p><ul><li><code>x</code>: Design matrix</li><li><code>k</code>: the true number of predictors. </li><li><code>d</code>: The distribution of the simulated trait (note <code>typeof(d) = UnionAll</code> but <code>typeof(d())</code> is an actual distribution: e.g. Normal)</li><li><code>l</code>: The link function. Input <code>canonicallink(d())</code> if you want to use the canonical link of <code>d</code>.</li></ul><p><strong>Optional arguments</strong></p><ul><li><code>r</code>: The number of success until stopping in negative binomial regression, defaults to 10</li><li><code>α</code>: Shape parameter of the gamma distribution, defaults to 1</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/OpenMendel/MendelIHT.jl/blob/84e3705bd851b1d409d0bb6e34f11867f94ddb00/src/simulate_utilities.jl#L188-L205">source</a></section></article><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p>For negative binomial and gamma, the link function must be LogLink. For Bernoulli, the probit link seems to work better than logitlink when used in <code>cv_iht</code> or <code>L0_reg</code>. </p></div></div><article class="docstring"><header><a class="docstring-binding" id="MendelIHT.make_bim_fam_files" href="#MendelIHT.make_bim_fam_files"><code>MendelIHT.make_bim_fam_files</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">make_bim_fam_files(x::SnpArray, y, name::String)</code></pre><p>Creates .bim and .bed files from a SnpArray. </p><p><strong>Arguments:</strong></p><ul><li><code>x</code>: A SnpArray (i.e. <code>.bed</code> file on the disk) for which you wish to create corresponding <code>.bim</code> and <code>.fam</code> files.</li><li><code>name</code>: string that should match the <code>.bed</code> file (Do not include <code>.bim</code> or <code>.fam</code> extensions in <code>name</code>).</li><li><code>y</code>: Trait vector that will go in to the 6th column of <code>.fam</code> file. </li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/OpenMendel/MendelIHT.jl/blob/84e3705bd851b1d409d0bb6e34f11867f94ddb00/src/simulate_utilities.jl#L265-L274">source</a></section></article><h2 id="Other-Useful-Functions"><a class="docs-heading-anchor" href="#Other-Useful-Functions">Other Useful Functions</a><a id="Other-Useful-Functions-1"></a><a class="docs-heading-anchor-permalink" href="#Other-Useful-Functions" title="Permalink"></a></h2><p>MendelIHT additionally provides useful utilities that may be of interest to a few advanced users. </p><article class="docstring"><header><a class="docstring-binding" id="MendelIHT.iht_run_many_models" href="#MendelIHT.iht_run_many_models"><code>MendelIHT.iht_run_many_models</code></a> — <span class="docstring-category">Function</span></header><section><div><p>Runs IHT across many different model sizes specifed in <code>path</code> using the full design matrix. Same as <code>cv_iht</code> but <strong>DOES NOT</strong> validate in a holdout set, meaning that this will definitely induce overfitting as we increase model size. Use this if you want to quickly estimate a range of feasible model sizes before  engaging in full cross validation. </p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/OpenMendel/MendelIHT.jl/blob/84e3705bd851b1d409d0bb6e34f11867f94ddb00/src/cross_validation.jl#L141-L147">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="MendelIHT.maf_weights" href="#MendelIHT.maf_weights"><code>MendelIHT.maf_weights</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">maf_weights(x::SnpArray; max_weight::T = Inf)</code></pre><p>Calculates the prior weight based on minor allele frequencies. </p><p>Returns an array of weights where <code>w[i] = 1 / (2 * sqrt(p[i] (1 - p[i]))) ∈ (1, ∞).</code> Here <code>p</code> is the minor allele frequency computed by <code>maf()</code> in SnpArrays. </p><ul><li><code>x</code>: A SnpArray </li><li><code>max_weight</code>: Maximum weight for any predictor. Defaults to <code>Inf</code>. </li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/OpenMendel/MendelIHT.jl/blob/84e3705bd851b1d409d0bb6e34f11867f94ddb00/src/utilities.jl#L541-L551">source</a></section></article></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../contributing/">« Contributing</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> on <span class="colophon-date" title="Sunday 3 January 2021 22:45">Sunday 3 January 2021</span>. Using Julia version 1.5.3.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
