{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Details of Parameter Estimation\n",
    "\n",
    "This note is meant to supplement our [paper](https://doi.org/10.1093/gigascience/giaa044). \n",
    "\n",
    "For a review on generalized linear models, the following resources are recommended:\n",
    "+ (3rd edition) Chapter 15.3 of [Applied regression analysis and generalized linear models](https://www.amazon.com/Applied-Regression-Analysis-Generalized-Linear/dp/1452205663/ref=sr_1_2?dchild=1&keywords=Applied+Regression+Analysis+and+Generalized+Linear+Models&qid=1609298891&s=books&sr=1-2) by John Fox\n",
    "+ (3rd edition) Chapter 3-5 of [An introduction to generalized linear models](https://www.amazon.com/Introduction-Generalized-Chapman-Statistical-Science/dp/1138741515/ref=sr_1_2?crid=18BN4MONNYYJH&dchild=1&keywords=an+introduction+to+generalized+linear+models&qid=1609298924&s=books&sprefix=an+introduction+to+ge%2Cstripbooks%2C222&sr=1-2) by Dobson and Barnett\n",
    "\n",
    "For review on projected gradient descent, I recommend\n",
    "+ Chapter 5 of [MM optimization algorithms](https://www.amazon.com/MM-Optimization-Algorithms-Kenneth-Lange/dp/1611974399) by Kenneth Lange (2nd edition is almost out, as of 1/2/2021)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Generalized linear models\n",
    "\n",
    "In `MendelIHT.jl`, phenotypes $(\\bf y)$ are modeled as a [generalized linear model](https://en.wikipedia.org/wiki/Generalized_linear_model):\n",
    "```math\n",
    "\\begin{aligned}\n",
    "    \\mu_i = E(y_i) = g({\\bf x}_i^t {\\boldsymbol \\beta})\n",
    "\\end{aligned}\n",
    "```\n",
    "where $\\bf x$ is sample $i$'s $p$-dimensional vector of *covariates* (genotypes + other fixed effects), $\\boldsymbol \\beta$ is a $p$-dimensional regression coefficients, $g$ is a non-linear *inverse-link* function, $y_i$ is sample $i$'s phenotype value, and $\\mu_i$ is the *average predicted value* of $y_i$ given $\\bf x$. \n",
    "\n",
    "The full design matrix ${\\bf X}_{n \\times p}$ and phenotypes ${\\bf y}_{n \\times 1}$ are observed. The distribution of $\\bf y$ and the inverse link $g$ are chosen before fitting. The regression coefficients $\\boldsymbol \\beta$ are not observed and are estimated by *maximum likelihood* methods, traditionally via iteratively reweighted least squares (IRLS). For high dimensional problems where $n < p$, we substitute iterative hard thresholding in place of IRLS. \n",
    "\n",
    "GLMs offer a natural way to model common non-continuous phenotypes. For instance, logistic regression for binary phenotypes and Poisson regression for integer valued phenotypes are special cases under the GLM framework. When $g(\\alpha) = \\alpha,$ we get standard linear regression used for Gaussian phenotypes. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Loglikelihood, gradient, and expected information\n",
    "\n",
    "In GLM, the distribution of $\\bf y$ is from the exponential family with density\n",
    "```math\n",
    "\\begin{aligned}\n",
    "    f(y \\mid \\theta, \\phi) = \\exp \\left[ \\frac{y \\theta - b(\\theta)}{a(\\phi)} + c(y, \\phi) \\right].\n",
    "\\end{aligned}\n",
    "```\n",
    "Here $\\theta$ is called the *canonical (location) parameter* and under the canonical link, $\\theta = g(\\bf x^t \\bf \\beta)$. $\\phi$ is the *dispersion (scale) parameter*. The functions $a, b, c$ are known functions that vary depending on the distribution of $y$. \n",
    "\n",
    "Given $n$ independent observations, the loglikelihood is:\n",
    "```math\n",
    "\\begin{aligned}\n",
    "    L({\\bf \\theta}, \\phi; {\\bf y}) &= \\sum_{i=1}^n \\frac{y_i\\theta_i - b(\\theta_i)}{a_i(\\phi)} + c(y_i, \\phi).\n",
    "\\end{aligned}\n",
    "```\n",
    "To evaluate the loglikelihood, we evaluate sample $i$'s logpdf using the [logpdf](https://juliastats.org/Distributions.jl/latest/univariate/#Distributions.logpdf-Tuple{Distribution{Univariate,S}%20where%20S%3C:ValueSupport,Real}) function in [Distributions.jl](https://github.com/JuliaStats/Distributions.jl).\n",
    "\n",
    "The perform maximum likelihood estimation, we compute partial derivatives for $\\beta$s. The $j$th score component is (eq 4.18 in Dobson):\n",
    "```math\n",
    "\\begin{aligned}\n",
    "    \\frac{\\partial L}{\\partial \\beta_j} = \\sum_{i=1}^n \\left[\\frac{y_i - \\mu_i}{var(y_i)}x_{ij}\\left(\\frac{\\partial \\mu_i}{\\partial \\eta_i}\\right)\\right].\n",
    "\\end{aligned}\n",
    "```\n",
    "Thus the full *gradient* is\n",
    "```math\n",
    "\\begin{aligned}\n",
    "    \\nabla L&= {\\bf X}^t{\\bf W}({\\bf y} - \\boldsymbol\\mu), \\quad W_{ii} = \\frac{1}{var(y_i)}\\left(\\frac{\\partial \\mu_i}{\\partial \\eta_i}\\right),\n",
    "\\end{aligned}\n",
    "```\n",
    "and similarly, the *expected information* is (eq 4.23 in Dobson):\n",
    "```math\n",
    "\\begin{aligned}\n",
    "    J = {\\bf X^t\\tilde{W}X}, \\quad \\tilde{W}_{ii} = \\frac{1}{var(y_i)}\\left(\\frac{\\partial \\mu_i}{\\partial \\eta_i}\\right)^2.\n",
    "\\end{aligned}\n",
    "```\n",
    "To evaluate $\\nabla L$ and $J$, note ${\\bf y}$ and ${\\bf X}$ are known, so we just need to calculate $\\boldsymbol\\mu, \\frac{\\partial\\mu_i}{\\partial\\eta_i},$ and $var(y_i)$. The first simply uses the inverse link: $\\mu_i = g({\\bf x}_i^t {\\boldsymbol \\beta})$. For the second, note $\\frac{\\partial \\mu_i}{\\partial\\eta_i} = \\frac{\\partial g({\\bf x}_i^t {\\boldsymbol \\beta})}{\\partial{\\bf x}_i^t {\\boldsymbol \\beta}}$ is just the derivative of the inverse link function evaluated at the linear predictor $\\eta_i = {\\bf x}_i^t {\\boldsymbol \\beta}$. This is already implemented for various link functions as [mueta](https://github.com/JuliaStats/GLM.jl/blob/master/src/glmtools.jl#L149) in [GLM.jl](https://github.com/JuliaStats/GLM.jl), which we call internally. To compute $var(y_i)$, we note that the exponential family distributions have variance\n",
    "```math\n",
    "\\begin{aligned}\n",
    "    var(y) &= a(\\phi)b''(\\theta) = a(\\phi)\\frac{\\partial^2b(\\theta)}{\\partial\\theta^2} = a(\\phi) var(\\mu).\n",
    "\\end{aligned}\n",
    "```\n",
    "That is, $var(y_i)$ is a product of 2 terms where the first depends solely on $\\phi$, and the second solely on $\\mu_i = g({\\bf x}_i^t {\\boldsymbol \\beta})$. In our code, we use [glmvar](https://github.com/JuliaStats/GLM.jl/blob/master/src/glmtools.jl#L315) implemented in [GLM.jl](https://github.com/JuliaStats/GLM.jl) to calculate $var(\\mu)$. Because $\\phi$ is unknown, we assume $a(\\phi) = 1$ for all models in computing $W_{ii}$ and $\\tilde{W}_{ii}$, except for the negative binomial model. For negative binomial model, we discuss how to estimate $\\phi$ and $\\boldsymbol\\beta$ using alternate block descent below.  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## Iterative hard thresholding\n",
    "\n",
    "In `MendelIHT.jl`, the loglikelihood is maximized using iterative hard thresholding. This is achieved by repeating the following iteration:\n",
    "```math\n",
    "\\begin{aligned}\n",
    "    \\boldsymbol\\beta_{n+1} = \\overbrace{P_{S_k}}^{(3)}\\big(\\boldsymbol\\beta_n + \\underbrace{s_n}_{(2)} \\overbrace{\\nabla f(\\boldsymbol\\beta_n)}^{(1)}\\big)\n",
    "\\end{aligned}\n",
    "```\n",
    "where $f$ is the loglikelihood to maximize. Step (1) computes the gradient as previously discussed. Step (2) computes the step size $s_k$. Step (3) evaluates the projection operator $P_{S_k}$, which sets all but $k$ largest entries in magnitude to $0$. To perform $P_{S_k}$, we first partially sort the *dense* vector $\\beta_n + s_n \\nabla f(\\beta_n)$, and set the smallest $k+1 ... n$ entries in magnitude to $0$. Note the step size $s_n$ is derived in our paper to be\n",
    "```math\n",
    "\\begin{aligned}\n",
    "    s_n = \\frac{||\\nabla f(\\boldsymbol\\beta_n)||_2^2}{\\nabla f(\\boldsymbol\\beta_n)^t J(\\boldsymbol\\beta_n) \\nabla f(\\boldsymbol\\beta_n)}\n",
    "\\end{aligned}\n",
    "```\n",
    "where $J = {\\bf X^t\\tilde{W}X}$ is the expected information matrix (derived in the previous section) *which should never be explicitly formed*. To evaluate the denominator, observe that \n",
    "```math\n",
    "\\begin{aligned}\n",
    "    \\nabla f(\\boldsymbol\\beta_n)^t J(\\boldsymbol\\beta_n) \\nabla f(\\boldsymbol\\beta_n) = \\left(\\nabla f(\\boldsymbol\\beta_n)^t{\\bf X}^t \\sqrt(\\tilde{W})\\right)\\left(\\sqrt(\\tilde{W}){\\bf X}\\nabla f(\\boldsymbol\\beta_n)\\right).\n",
    "\\end{aligned}\n",
    "```\n",
    "Thus one computes ${\\bf v} = \\sqrt(\\tilde{W}){\\bf X}\\nabla f(\\boldsymbol\\beta_n)$ and calculate its inner product with itself. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## Nuisance parameter estimation\n",
    "\n",
    "Currently `MendelIHT.jl` only estimates nuisance parameter for the Negative Binomial model. Estimation of $\\phi$ and $\\boldsymbol \\beta$ can be achieved with alternating block updates. That is, we run 1 IHT iteration to estimate $\\boldsymbol \\beta_n$, followed by 1 iteration of Newton or MM update to estimate $\\phi_n$. Below we derive the Newton and MM updates. \n",
    "\n",
    "Note 1: This feature is provided by our [2019 Bruins in Genomics](https://qcb.ucla.edu/big-summer/big2019-2/) summer student [Vivian Garcia](https://github.com/viviangarcia) and [Francis Adusei](https://github.com/fadusei). \n",
    "\n",
    "Note 2: for Gaussian response, one can use the sample variance formula to estimate $\\phi$ from the estimated mean $\\hat{\\mu}$. \n",
    "\n",
    "### Parametrization for Negative Binomial model\n",
    "\n",
    "The negative binomial distribution has density\n",
    "```math\n",
    "\\begin{aligned}\n",
    "\tP(Y = y) = \\binom{y+r-1}{y}p^r(1-p)^y\n",
    "\\end{aligned}\n",
    "```\n",
    "where $y$ is the number of failures before the $r$th success and $p$ is the probability of success in each individual trial. Adhering to these definitions, the mean and variance according to [WOLFRAM](https://reference.wolfram.com/language/ref/NegativeBinomialDistribution.html) is \n",
    "```math\n",
    "\\begin{aligned}\n",
    "\t\\mu_i = \\frac{r(1-p_i)}{r}, \\quad\n",
    "\tVar(y_i) = \\frac{r(1-p_i)}{p_i^2}.\n",
    "\\end{aligned}\n",
    "```\n",
    "Note these formula are different than the default on [wikipedia](https://en.wikipedia.org/wiki/Negative_binomial_distribution) because in wiki $y$ is the number of *success* and $r$ is the number of *failure*. \n",
    "Therefore, solving for $p_i$, we have \n",
    "```math\n",
    "\\begin{aligned}\n",
    "\tp_i = \\frac{r}{\\mu_i + r} = \\frac{r}{e^{\\mathbf{x}_i^T\\beta} + r} \\in (0, 1).\n",
    "\\end{aligned}\n",
    "```\n",
    "And indeed this this is how we [parametrize the negative binomial model](https://github.com/OpenMendel/MendelIHT.jl/blob/master/src/utilities.jl#L41). **Importantly, we can interpret $p_i$ as a probability**, since $\\mathbf{x}_i^T\\beta$ can take on any number between $-\\infty$ and $+\\infty$ (since $\\beta$ and $\\mathbf{x}_i$ can have positive and negative entries), so $exp(\\mathbf{x}_i^T\\beta)\\in(0, \\infty)$.\n",
    "\n",
    "We can also try to express $Var(y_i)$ in terms of $\\mu_i$ and $r$ by doing some algebra:\n",
    "```math\n",
    "\\begin{aligned}\n",
    "\tVar(y_i)\n",
    "\t&= \\frac{r(1-p_i)}{p_i^2} = \\frac{r\\left( 1 - \\frac{r}{\\mu_i + r} \\right)}{\\frac{r^2}{(\\mu_i + r)^2}} = \\frac{1}{r}\\left(1 - \\frac{r}{\\mu_i + r}\\right)(\\mu_i + r)^2 \\\\\n",
    "\t&= \\frac{1}{r} \\left[ (\\mu_i + r)^2 - r(\\mu_r + r) \\right] = \\frac{1}{r}(\\mu_i + r)\\mu_i\\\\\n",
    "\t&= \\mu_i \\left( \\frac{\\mu_i}{r} + 1 \\right)\n",
    "\\end{aligned}\n",
    "```\n",
    "You can verify [in GLM.jl](https://github.com/JuliaStats/GLM.jl/blob/ef246bb8fdbfa3f3058435035d0b0cf42abdd06e/src/glmtools.jl#L320) that this is indeed how they compute the variance of a negative binomial distribution. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Estimating nuisance parameter using MM algorithms\n",
    "\n",
    "The MM algorithm is very stable, but converges much slower than Newton's alogorithm below. Thus use MM only if Newton's method fails.\n",
    "\n",
    "The loglikelihood for $n$ independent samples under a Negative Binomial model is \n",
    "```math\n",
    "\\begin{aligned}\n",
    "\tL(p_1, ..., p_m, r)\n",
    "\t&= \\sum_{i=1}^m \\ln \\binom{y_i+r-1}{y_i} + r\\ln(p_i) + y_i\\ln(1-p_i)\\\\\n",
    "\t&= \\sum_{i=1}^m \\left[ \\sum_{j=0}^{y_i - 1} \\ln(r+j) + r\\ln(p_i) - \\ln(y_i!) + y_i\\ln(1-p_i) \\right]\\\\\n",
    "\t&\\geq \\sum_{i=1}^m\\left[ \\sum_{j=0}^{y_i-1}\\frac{r_n}{r_n+j}\\ln(r) + c_n + r\\ln(p_i) - \\ln(y_i!) + y_i \\ln(1-p_i) \\right]\\\\\n",
    "    &\\equiv M(p_1, ..., p_m, r)\n",
    "\\end{aligned}\n",
    "```\n",
    "The last inequality can be seen by applying Jensen's inequality:\n",
    "```math\n",
    "\\begin{aligned}\n",
    "\tf\\left[ \\sum_{i}u_i(\\boldsymbol\\theta)\\right] \\leq \\sum_{i} \\frac{u_i(\\boldsymbol\\theta_n)}{\\sum_j u_j(\\boldsymbol\\theta_n)}f \\left[ \\frac{\\sum_j u_j(\\boldsymbol\\theta_n)}{u_i(\\boldsymbol\\theta_n)} u_i(\\boldsymbol\\theta)\\right]\n",
    "\\end{aligned}\n",
    "```\n",
    "to the function $f(u) = - \\ln(u).$ Maximizing $M$ over $r$ (i.e. differentiating with respect to $r$ and setting equal to zero, then solving for $r$), we have\n",
    "```math\n",
    "\\begin{aligned}\n",
    "    \\frac{d}{dr} M\n",
    "\t&= \\sum_{i=1}^{m} \\left[ \\sum_{j=0}^{y_i-1} \\frac{r_n}{r_n + j} \\frac{1}{r} + \\ln(p_i) \\right] \\\\\n",
    "\t&= \\sum_{i=1}^{m}\\sum_{j=0}^{y_i-1} \\frac{r_n}{r_n + j} \\frac{1}{r} + \\sum_{i=1}^{m}\\ln(p_i)\\\\ \n",
    "\t&\\equiv 0\\\\\n",
    "\t\\iff r_{n+1} &= \\frac{-\\sum_{i=1}^{m}\\sum_{j=0}^{y_i-1} \\frac{r_n}{r_n + j}}{\\sum_{i=1}^{m}\\ln(p_i) } \n",
    "\\end{aligned}\n",
    "```\n",
    "Since $L \\ge M$ (M *minorizes* L), maximizing $M$ will maximize $L$. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Estimating Nuisance parameter using Newton's method\n",
    "\n",
    "\n",
    "Since we are dealing with 1 parameter optimization, Newton's method is likely a better candidate due to its quadratic rate of convergence. To estimate the nuisance parameter ($r$), we use maximum likelihood estimates. By $p_i = r / (\\mu_i + r)$ in above, we have\n",
    "```math\n",
    "\\begin{aligned}\n",
    "\t& L(p_1, ..., p_m, r)\\\\\n",
    "\t=& \\sum_{i=1}^m \\ln \\binom{y_i+r-1}{y_i} + r\\ln(p_i) + y_i\\ln(1-p_i)\\\\\n",
    "\t=& \\sum_{i=1}^m \\left[ \\ln\\left((y_i+r-1)!\\right) - \\ln\\left(y_i!\\right) - \\ln\\left((r-1)!\\right) + r\\ln(r) - r\\ln(\\mu_i+r) + y_i\\ln(\\mu_i) + y_i\\ln(\\mu_i + r)\\right]\\\\\n",
    "\t=& \\sum_{i=1}^m\\left[\\ln\\left((y_i+r-1)!\\right)-\\ln(y_i!) - \\ln\\left((r-1)\\right) + r\\ln(r) - (r+y_i)\\ln(\\mu_i + r) + y_i\\ln(\\mu_i)\\right]\n",
    "\\end{aligned}\n",
    "```\n",
    "Recalling the definition of [digamma and trigamma functions](https://en.wikipedia.org/wiki/Digamma_function), the first and second derivative of our last expression with respect to $r$ is:\n",
    "```math\n",
    "\\begin{aligned}\n",
    "\t\\frac{d}{dr} L(p_1, ..., p_m, r) = & \\sum_{i=1}^m \\left[ \\operatorname{digamma}(y_i+r) - \\operatorname{digamma}(r) + 1 + \\ln(r) - \\frac{r+y_i}{\\mu_i+r} - \\ln(\\mu_i + r) \\right]\\\\\n",
    "\t\\frac{d^2}{dr^2} L(p_1, ..., p_m, r) =&\\sum_{i=1}^m \\left[ \\operatorname{trigamma}(y_i+r) - \\operatorname{trigamma}(r) + \\frac{1}{r} - \\frac{2}{\\mu_i + r} + \\frac{r+y_i}{(\\mu_i + r)^2} \\right]\n",
    "\\end{aligned}\n",
    "```\n",
    "So the iteration to use is:\n",
    "```math\n",
    "\\begin{aligned}\n",
    "\tr_{n+1} = r_n - \\frac{\\frac{d}{dr}L(p_1,...,p_m,r)}{\\frac{d^2}{dr^2}L(p_1,...,p_m,r)}.\n",
    "\\end{aligned}\n",
    "```\n",
    "For stability, we set the denominator equal to $1$ if it is less than 0. That is, we use gradient descent if the current iteration has non-positive definite Hessian matrices. "
   ]
  }
 ],
 "metadata": {
  "@webio": {
   "lastCommId": null,
   "lastKernelId": null
  },
  "kernelspec": {
   "display_name": "Julia 1.5.0",
   "language": "julia",
   "name": "julia-1.5"
  },
  "language_info": {
   "file_extension": ".jl",
   "mimetype": "application/julia",
   "name": "julia",
   "version": "1.5.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
